<!DOCTYPE html><html lang="en | zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1"><title>CUT&amp;Tag Analysis | Jocelyn's Blog</title><meta name="keywords" content="CUT&amp;Tag"><meta name="author" content="yilingzhang"><meta name="copyright" content="yilingzhang"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="有一个隐藏的节点可以用，ssh 192.168.1.23 (有24个线程) jobs命令，查看后台有没有任务在运行 IGV可以save session回到之前的界面  分析流程参考： https:&#x2F;&#x2F;yezhengstat.github.io&#x2F;CUTTag_tutorial数据来源： https:&#x2F;&#x2F;www.nature.com&#x2F;articles&#x2F;s41467-019-09982-5  主要步骤">
<meta property="og:type" content="article">
<meta property="og:title" content="CUT&amp;Tag Analysis">
<meta property="og:url" content="http://yilingzhang.gitee.io/2021/03/25/CUT-Tag-Analysis/index.html">
<meta property="og:site_name" content="Jocelyn&#39;s Blog">
<meta property="og:description" content="有一个隐藏的节点可以用，ssh 192.168.1.23 (有24个线程) jobs命令，查看后台有没有任务在运行 IGV可以save session回到之前的界面  分析流程参考： https:&#x2F;&#x2F;yezhengstat.github.io&#x2F;CUTTag_tutorial数据来源： https:&#x2F;&#x2F;www.nature.com&#x2F;articles&#x2F;s41467-019-09982-5  主要步骤">
<meta property="og:locale">
<meta property="og:image" content="http://yilingzhang.gitee.io/img/cover/CUT&Tag.png">
<meta property="article:published_time" content="2021-03-25T09:37:16.000Z">
<meta property="article:modified_time" content="2021-03-30T08:40:41.079Z">
<meta property="article:author" content="yilingzhang">
<meta property="article:tag" content="CUT&amp;Tag">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://yilingzhang.gitee.io/img/cover/CUT&Tag.png"><link rel="shortcut icon" href="/img/love_32px.png"><link rel="canonical" href="http://yilingzhang.gitee.io/2021/03/25/CUT-Tag-Analysis/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    jQuery: 'https://cdn.jsdelivr.net/npm/jquery@latest/dist/jquery.min.js',
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/js/jquery.justifiedGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/css/justifiedGallery.min.css'
    },
    fancybox: {
      js: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.js',
      css: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isanchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = { 
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2021-03-30 16:40:41'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    })(window)</script><meta name="generator" content="Hexo 5.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="author-avatar"><img class="avatar-img" src="/img/selfie.JPG" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data"><div class="data-item is-center"><div class="data-item-link"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">37</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">15</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">14</div></a></div></div></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/img/cover/CUT&amp;Tag.png')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Jocelyn's Blog</a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">CUT&amp;Tag Analysis</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2021-03-25T09:37:16.000Z" title="Created 2021-03-25 17:37:16">2021-03-25</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2021-03-30T08:40:41.079Z" title="Updated 2021-03-30 16:40:41">2021-03-30</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/CUT-Tag/">CUT&amp;Tag</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">Word count:</span><span class="word-count">2.4k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">Reading time:</span><span>12min</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><ul>
<li>有一个隐藏的节点可以用，ssh 192.168.1.23 (有24个线程)</li>
<li>jobs命令，查看后台有没有任务在运行</li>
<li>IGV可以save session回到之前的界面</li>
</ul>
<h1 id="分析流程"><a href="#分析流程" class="headerlink" title="分析流程"></a>分析流程</h1><p>参考： <a target="_blank" rel="noopener" href="https://yezhengstat.github.io/CUTTag_tutorial">https://yezhengstat.github.io/CUTTag_tutorial</a><br>数据来源： <a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41467-019-09982-5">https://www.nature.com/articles/s41467-019-09982-5</a> </p>
<h2 id="主要步骤："><a href="#主要步骤：" class="headerlink" title="主要步骤："></a>主要步骤：</h2><ol>
<li>质量监控（fastqc，multiqc）（略）</li>
<li>读段比对（bowtie2）（略）及利用spike-in计算scale factor</li>
<li>sam过滤及格式转换（samtools、bedtools）及归一化</li>
<li>Peak calling（SEACR、MACS2）</li>
<li>motif识别（MEME）</li>
<li>IGV可视化</li>
<li>deeptools可视化及homer注释（可选）</li>
</ol>
<ul>
<li>BlackList</li>
</ul>
<p>被设为黑名单后，不能当做是peak<br>call peak之后，用Black List过滤一下</p>
<h3 id="Bam文件"><a href="#Bam文件" class="headerlink" title="Bam文件"></a>Bam文件</h3><p><a target="_blank" rel="noopener" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">http://genome.ucsc.edu/FAQ/FAQformat.html#format1</a></p>
<p>相当于压缩文件，要用samtools view看</p>
<h3 id="Bed文件格式"><a href="#Bed文件格式" class="headerlink" title="Bed文件格式"></a>Bed文件格式</h3><p>UCSC format</p>
<p>可以直接导入IGV</p>
<p>用于表示染色体区域</p>
<ol>
<li><p><strong>chrom</strong> - 染色体号; 例如，chr1，chrX……</p>
</li>
<li><p><strong>chromStart</strong> - feature在染色体上起始位置. 从0开始算，染色体上第一个碱基位置标记为0。</p>
</li>
<li><p><strong>chromEnd</strong> - feature在染色体上终止位置。染色体上前100个碱基片段的位置位置标记为：chromStart=0, chromEnd=100。 实际上，第100个碱基不属于当前片段中，当前片段的碱基应该是0-99。所以在BED文件中，起始位置从0开始，终止位置从1开始。</p>
</li>
<li><p><strong>name</strong> - BED行名，在基因组浏览器左边显示；</p>
</li>
<li><p><strong>score</strong> - 在基因组浏览器中显示的灰度设定，值介于0-1000；</p>
</li>
<li><p><strong>strand</strong> - 正负链标记. Either “.” (=no strand) or “+” or “-“.</p>
</li>
</ol>
<p>bedtools工具bamtobed: 将bam转bed</p>
<h3 id="Bedpe文件"><a href="#Bedpe文件" class="headerlink" title="Bedpe文件"></a>Bedpe文件</h3><p>用于表示双末端bed文件，一行里面有两个坐标，前三列一个坐标，三到六另一个</p>
<h1 id="Code"><a href="#Code" class="headerlink" title="Code"></a>Code</h1><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br><span class="line">197</span><br><span class="line">198</span><br><span class="line">199</span><br><span class="line">200</span><br><span class="line">201</span><br><span class="line">202</span><br><span class="line">203</span><br><span class="line">204</span><br><span class="line">205</span><br><span class="line">206</span><br><span class="line">207</span><br><span class="line">208</span><br><span class="line">209</span><br><span class="line">210</span><br><span class="line">211</span><br><span class="line">212</span><br><span class="line">213</span><br><span class="line">214</span><br><span class="line">215</span><br><span class="line">216</span><br><span class="line">217</span><br><span class="line">218</span><br><span class="line">219</span><br><span class="line">220</span><br><span class="line">221</span><br><span class="line">222</span><br><span class="line">223</span><br><span class="line">224</span><br><span class="line">225</span><br><span class="line">226</span><br><span class="line">227</span><br><span class="line">228</span><br><span class="line">229</span><br><span class="line">230</span><br><span class="line">231</span><br><span class="line">232</span><br><span class="line">233</span><br><span class="line">234</span><br><span class="line">235</span><br><span class="line">236</span><br><span class="line">237</span><br><span class="line">238</span><br><span class="line">239</span><br><span class="line">240</span><br><span class="line">241</span><br><span class="line">242</span><br><span class="line">243</span><br><span class="line">244</span><br><span class="line">245</span><br><span class="line">246</span><br><span class="line">247</span><br><span class="line">248</span><br><span class="line">249</span><br><span class="line">250</span><br><span class="line">251</span><br><span class="line">252</span><br><span class="line">253</span><br><span class="line">254</span><br><span class="line">255</span><br><span class="line">256</span><br><span class="line">257</span><br><span class="line">258</span><br><span class="line">259</span><br><span class="line">260</span><br><span class="line">261</span><br><span class="line">262</span><br><span class="line">263</span><br><span class="line">264</span><br><span class="line">265</span><br><span class="line">266</span><br><span class="line">267</span><br><span class="line">268</span><br><span class="line">269</span><br><span class="line">270</span><br><span class="line">271</span><br><span class="line">272</span><br><span class="line">273</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># ssh 192.168.1.23</span></span><br><span class="line">rootpath=/public/workspace/stu18230130/MyCourse/NGS-02/lab1-CUTTag/</span><br><span class="line"><span class="comment"># ---------------------------&gt;</span></span><br><span class="line"><span class="comment"># download and extract data</span></span><br><span class="line"><span class="comment"># ---------------------------&gt;</span></span><br><span class="line">mkdir -p <span class="variable">$rootpath</span>/data</span><br><span class="line"><span class="built_in">cd</span> <span class="variable">$rootpath</span>/data</span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> SRR8435039 SRR8435040 SRR8435051 SRR8435052</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  wget https://sra-pub-run-odp.s3.amazonaws.com/sra/<span class="variable">$i</span>/<span class="variable">$i</span></span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> SRR8435039 SRR8435040 SRR8435051 SRR8435052</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  fastq-dump -split-files <span class="variable">$i</span> &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># wait 等待前面后台命令全部完成，在往下</span></span><br><span class="line"><span class="comment"># 先fastq-dump再gzip会比较快</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> SRR8435039 SRR8435040 SRR8435051 SRR8435052</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  gzip <span class="variable">$&#123;i&#125;</span>_1.fastq &amp;</span><br><span class="line">  gzip <span class="variable">$&#123;i&#125;</span>_2.fastq &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># cd $rootpath</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># ---------------------------&gt;</span></span><br><span class="line"><span class="comment"># QC and pre-processing</span></span><br><span class="line"><span class="comment"># ---------------------------&gt;</span></span><br><span class="line">mkdir -p <span class="variable">$rootpath</span>/1.QC</span><br><span class="line"><span class="comment"># cd $rootpath/1.QC </span></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> SRR8435039 SRR8435040 SRR8435051 SRR8435052</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  fastqc -o <span class="variable">$rootpath</span>/1.QC/ <span class="variable">$rootpath</span>/data/<span class="variable">$&#123;i&#125;</span>_1.fastq.gz &amp;</span><br><span class="line">  fastqc -o <span class="variable">$rootpath</span>/1.QC/ <span class="variable">$rootpath</span>/data/<span class="variable">$&#123;i&#125;</span>_2.fastq.gz &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># ---------------------------&gt;</span></span><br><span class="line"><span class="comment"># mapping with bowtie2</span></span><br><span class="line"><span class="comment"># ---------------------------&gt;</span></span><br><span class="line"><span class="comment"># core 线程数</span></span><br><span class="line"><span class="comment"># 文件重命名策略：输出文件名称，SRR8435039:CTCF1，旧名字：新名字</span></span><br><span class="line">mkdir -p <span class="variable">$rootpath</span>/2.map/bowtie2_result</span><br><span class="line">mkdir -p <span class="variable">$rootpath</span>/2.map/bowtie2_summary</span><br><span class="line">bt2ref=/public/workspace/shaojf/Course/NGS/Reference/bowtie2_Index/GRCh38.bowtie2</span><br><span class="line">cores=4</span><br><span class="line"><span class="keyword">for</span> info <span class="keyword">in</span> SRR8435039:CTCF1 SRR8435040:CTCF2 SRR8435051:IgG1 SRR8435052:IgG2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  i=`<span class="built_in">echo</span> <span class="variable">$info</span> | cut -d <span class="string">&quot;:&quot;</span> -f 1`</span><br><span class="line">  name=`<span class="built_in">echo</span> <span class="variable">$info</span> | cut -d <span class="string">&quot;:&quot;</span> -f 2`</span><br><span class="line">  bowtie2 --end-to-end --very-sensitive --no-mixed --no-discordant --phred33 \</span><br><span class="line">    -I 10 -X 700 -p <span class="variable">$cores</span> -x <span class="variable">$bt2ref</span> \</span><br><span class="line">    -1 <span class="variable">$rootpath</span>/data/<span class="variable">$&#123;i&#125;</span>_1.fastq.gz -2 <span class="variable">$rootpath</span>/data/<span class="variable">$&#123;i&#125;</span>_2.fastq.gz \</span><br><span class="line">    -S <span class="variable">$rootpath</span>/2.map/bowtie2_result/<span class="variable">$&#123;name&#125;</span>_bowtie2.sam \</span><br><span class="line">    &amp;&gt; <span class="variable">$rootpath</span>/2.map/bowtie2_summary/<span class="variable">$&#123;name&#125;</span>_bowtie2.txt &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"><span class="comment"># The paired-end reads are aligned by Bowtie2 using parameters --end-to-end --very-sensitive --no-mixed --no-discordant --phred33 -I 10 -X 700 for mapping of inserts 10-700 bp in length.</span></span><br><span class="line"><span class="comment"># Critical step: There is no need to trim reads from out standard 25x25 PE sequencing, as adapter sequences will not be included in reads of inserts &gt;25 bp. However, for users performing longer sequencing, reads will need to be trimmed by Cutadapt and mapped by --local --very-sensitive --no-mixed --no-discordant --phred33 -I 10 -X 700 to ignore any remaining adapter sequence at the 3’ ends of reads during mapping.</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># U00096.3</span></span><br><span class="line"><span class="comment"># 可以去ncbi下载E.coli的ref</span></span><br><span class="line">spikeinref=/public/workspace/shaojf/Course/NGS/Reference/bowtie2_Index/Ecoli</span><br><span class="line"><span class="keyword">for</span> info <span class="keyword">in</span> SRR8435039:CTCF1 SRR8435040:CTCF2 SRR8435051:IgG1 SRR8435052:IgG2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  i=`<span class="built_in">echo</span> <span class="variable">$info</span> | cut -d <span class="string">&quot;:&quot;</span> -f 1`</span><br><span class="line">  name=`<span class="built_in">echo</span> <span class="variable">$info</span> | cut -d <span class="string">&quot;:&quot;</span> -f 2`</span><br><span class="line">  bowtie2 --end-to-end --very-sensitive --no-mixed --no-discordant --phred33 \</span><br><span class="line">    -I 10 -X 700 -p <span class="variable">$cores</span> -x <span class="variable">$spikeinref</span> \</span><br><span class="line">    -1 <span class="variable">$rootpath</span>/data/<span class="variable">$&#123;i&#125;</span>_1.fastq.gz -2 <span class="variable">$rootpath</span>/data/<span class="variable">$&#123;i&#125;</span>_2.fastq.gz \</span><br><span class="line">    -S <span class="variable">$rootpath</span>/2.map/bowtie2_result/<span class="variable">$&#123;name&#125;</span>_bowtie2_spikeIn.sam \</span><br><span class="line">    &amp;&gt; <span class="variable">$rootpath</span>/2.map/bowtie2_summary/<span class="variable">$&#123;name&#125;</span>_bowtie2_spikeIn.txt &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># grep &quot;rate&quot; 2.map/bowtie2_summary/*.txt</span></span><br><span class="line"><span class="comment"># 计算有多少对读段比对到E.coli上,再写到一个文件里，方便后面计算scale factor</span></span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> CTCF1 CTCF2 IgG1 IgG2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  seqDepthDouble=`samtools view -F 0x04 <span class="variable">$rootpath</span>/2.map/bowtie2_result/<span class="variable">$&#123;name&#125;</span>_bowtie2_spikeIn.sam | wc -l`</span><br><span class="line">  seqDepth=$((seqDepthDouble/<span class="number">2</span>))</span><br><span class="line">  <span class="built_in">echo</span> <span class="variable">$seqDepth</span> &gt; <span class="variable">$rootpath</span>/2.map/bowtie2_result/<span class="variable">$&#123;name&#125;</span>_bowtie2_spikeIn.seqDepth</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># In practice, we have found that the apparent duplication rate is low for high quality CUT&amp;Tag datasets, and even the apparent ‘duplicate’ fragments are likely to be true fragments. Thus, we do not recommend removing the duplicates.</span></span><br><span class="line"><span class="comment"># mkdir -p mkdir -p $rootpath/2.map/removeDuplicate/picard_summary</span></span><br><span class="line"><span class="comment"># for name in CTCF1 CTCF2 IgG1 IgG2</span></span><br><span class="line"><span class="comment"># do</span></span><br><span class="line"><span class="comment"># 	picard SortSam I=$rootpath/2.map/bowtie2_result/$&#123;name&#125;_bowtie2.sam \</span></span><br><span class="line"><span class="comment"># 		O=$rootpath/2.map/removeDuplicate/$&#123;name&#125;_bowtie2.sorted.sam SORT_ORDER=coordinate</span></span><br><span class="line"><span class="comment"># 	picard MarkDuplicates I=$rootpath/2.map/removeDuplicate/$&#123;name&#125;_bowtie2.sorted.sam \</span></span><br><span class="line"><span class="comment"># 		O=$rootpath/2.map/removeDuplicate/$&#123;name&#125;_bowtie2.sorted.dupMarked.sam METRICS_FILE=$rootpath/2.map/removeDuplicate/picard_summary/$&#123;name&#125;_picard.dupMark.txt</span></span><br><span class="line"><span class="comment"># done</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># -F unmap </span></span><br><span class="line">mkdir -p <span class="variable">$rootpath</span>/2.map/fragmentLen</span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> CTCF1 CTCF2 IgG1 IgG2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  samtools view -F 0x04 <span class="variable">$rootpath</span>/2.map/bowtie2_result/<span class="variable">$&#123;name&#125;</span>_bowtie2.sam | \</span><br><span class="line">    awk -F<span class="string">&#x27;\t&#x27;</span> <span class="string">&#x27;function abs(x)&#123;return ((x &lt; 0.0) ? -x : x)&#125; &#123;print abs($9)&#125;&#x27;</span> | \</span><br><span class="line">    sort | uniq -c | awk -v OFS=<span class="string">&quot;\t&quot;</span> <span class="string">&#x27;&#123;print $2, $1/2&#125;&#x27;</span> \</span><br><span class="line">    &gt; <span class="variable">$rootpath</span>/2.map/fragmentLen/<span class="variable">$&#123;name&#125;</span>_fragmentLen.txt &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># ---------------------------------------&gt;</span></span><br><span class="line"><span class="comment"># filtering and file format conversion</span></span><br><span class="line"><span class="comment"># ---------------------------------------&gt;</span></span><br><span class="line"><span class="comment"># sam转成bam，并且留下Q10以上的</span></span><br><span class="line"><span class="comment"># 转成bedgraph格式</span></span><br><span class="line"><span class="comment"># -F INT   only include reads with none of the FLAGS in INT present [0]</span></span><br><span class="line"></span><br><span class="line">mkdir -p <span class="variable">$rootpath</span>/3.filter/</span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> CTCF1 CTCF2 IgG1 IgG2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  samtools view -bS -F 0x04 -q 10 <span class="variable">$rootpath</span>/2.map/bowtie2_result/<span class="variable">$&#123;name&#125;</span>_bowtie2.sam \</span><br><span class="line">    &gt;<span class="variable">$rootpath</span>/3.filter/<span class="variable">$&#123;name&#125;</span>_bowtie2.q10.bam &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 转成bedpe</span></span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> CTCF1 CTCF2 IgG1 IgG2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  bedtools bamtobed -i <span class="variable">$rootpath</span>/3.filter/<span class="variable">$&#123;name&#125;</span>_bowtie2.q10.bam \</span><br><span class="line">  -bedpe &gt; <span class="variable">$rootpath</span>/3.filter/<span class="variable">$&#123;name&#125;</span>_bowtie2.bed &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 上一步已转换成bedpe格式，用于表示双末端bed文件，一行里面有两个坐标，前三列一个坐标（正向测），三到六另一个（反向测）</span></span><br><span class="line"><span class="comment"># awk 第一列等于第四列（在同一染色体上）且终止-起始&lt;1000</span></span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> CTCF1 CTCF2 IgG1 IgG2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  awk <span class="string">&#x27;$1==$4 &amp;&amp; $6-$2 &lt; 1000 &#123;print $0&#125;&#x27;</span> <span class="variable">$rootpath</span>/3.filter/<span class="variable">$&#123;name&#125;</span>_bowtie2.bed \</span><br><span class="line">    &gt; <span class="variable">$rootpath</span>/3.filter/<span class="variable">$&#123;name&#125;</span>_bowtie2.clean.bed &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 如果第一列相同，则按第二列从小到大排序，如果第二列相同，按照第三列从小到大排</span></span><br><span class="line"><span class="comment"># -k，1 只看第一列</span></span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> CTCF1 CTCF2 IgG1 IgG2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  cut -f 1,2,6 <span class="variable">$rootpath</span>/3.filter/<span class="variable">$&#123;name&#125;</span>_bowtie2.clean.bed | \</span><br><span class="line">  sort -k1,1 -k2,2n -k3,3n &gt; <span class="variable">$rootpath</span>/3.filter/<span class="variable">$&#123;name&#125;</span>_bowtie2.fragments.bed &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># scale factor</span></span><br><span class="line"><span class="comment"># if [ expression ]</span></span><br><span class="line"><span class="comment"># then</span></span><br><span class="line"><span class="comment">#    Statement(s) to be executed if expression is true</span></span><br><span class="line"><span class="comment"># fi</span></span><br><span class="line"><span class="comment"># Shell expression求值。如果结果值是true，给定statement被执行。如果 expression 为false ，则没有语句将不会被执行。</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># bc-l 相当于一个计算器，用长浮点型来表示</span></span><br><span class="line"><span class="comment"># echo 1+3 | bc 把计算的公式穿给bc去运算</span></span><br><span class="line">chromSize=/public/workspace/shaojf/Course/NGS/Reference/bowtie2_Index/GRCh38.chrom.size</span><br><span class="line">长度</span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> CTCF1 CTCF2 IgG1 IgG2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  seqDepth=`cat <span class="variable">$rootpath</span>/2.map/bowtie2_result/<span class="variable">$&#123;name&#125;</span>_bowtie2_spikeIn.seqDepth`</span><br><span class="line">  <span class="keyword">if</span> [[ <span class="string">&quot;<span class="variable">$seqDepth</span>&quot;</span> -gt <span class="string">&quot;1&quot;</span> ]]</span><br><span class="line">  <span class="keyword">then</span></span><br><span class="line">      mkdir -p <span class="variable">$rootpath</span>/3.filter/bedgraph</span><br><span class="line">      scale_factor=`<span class="built_in">echo</span> <span class="string">&quot;10000 / <span class="variable">$seqDepth</span>&quot;</span> | bc -l`</span><br><span class="line">      <span class="built_in">echo</span> <span class="string">&quot;Scaling factor for <span class="variable">$name</span> is: <span class="variable">$scale_factor</span>&quot;</span></span><br><span class="line">      bedtools genomecov -<span class="built_in">bg</span> -scale <span class="variable">$scale_factor</span> \</span><br><span class="line">      	-i <span class="variable">$rootpath</span>/3.filter/<span class="variable">$&#123;name&#125;</span>_bowtie2.fragments.bed \</span><br><span class="line">      	-g <span class="variable">$chromSize</span> &gt; <span class="variable">$rootpath</span>/3.filter/bedgraph/<span class="variable">$&#123;name&#125;</span>_bowtie2.fragments.normalized.bedgraph &amp;</span><br><span class="line">  <span class="keyword">fi</span></span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line">mkdir -p <span class="variable">$rootpath</span>/3.filter/bigwig</span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> CTCF1 CTCF2 IgG1 IgG2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  bedSort <span class="variable">$rootpath</span>/3.filter/bedgraph/<span class="variable">$&#123;name&#125;</span>_bowtie2.fragments.normalized.bedgraph <span class="variable">$rootpath</span>/3.filter/bedgraph/<span class="variable">$&#123;name&#125;</span>_bowtie2.fragments.normalized.srt.bedgraph &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 转bigwig方便可视化</span></span><br><span class="line"><span class="comment"># 用IGV打开，调颜色，bed文件第四列在IGV中会标注</span></span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> CTCF1 CTCF2 IgG1 IgG2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  bedGraphToBigWig <span class="variable">$rootpath</span>/3.filter/bedgraph/<span class="variable">$&#123;name&#125;</span>_bowtie2.fragments.normalized.srt.bedgraph <span class="variable">$chromSize</span> <span class="variable">$rootpath</span>/3.filter/bigwig/<span class="variable">$&#123;name&#125;</span>_bowtie2.fragments.normalized.bw &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># -------------&gt;</span></span><br><span class="line"><span class="comment"># peak calling</span></span><br><span class="line"><span class="comment"># -------------&gt;</span></span><br><span class="line"><span class="comment"># SEACR</span></span><br><span class="line"><span class="comment"># SEACR is intended to call peaks and enriched regions from sparse CUT&amp;RUN or chromatin profiling data in which background is dominated by &quot;zeroes&quot; (i.e. regions with no read coverage).</span></span><br><span class="line"><span class="comment"># bash SEACR_1.3.sh &lt;experimental bedgraph&gt;.bg [&lt;control bedgraph&gt;.bg | &lt;FDR threshold&gt;] [norm | non] [relaxed | stringent] output prefix</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 加不加0.01差很多，把top1%作为peak</span></span><br><span class="line">seacr=/public/workspace/shaojf/anaconda3/bin/SEACR_1.3.sh</span><br><span class="line">mkdir -p <span class="variable">$rootpath</span>/4.peakCalling/</span><br><span class="line"><span class="keyword">for</span> pair <span class="keyword">in</span> CTCF1:IgG1 CTCF2:IgG2 <span class="comment">#ip组：ctl组</span></span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  name=`<span class="built_in">echo</span> <span class="variable">$pair</span> | cut -d <span class="string">&quot;:&quot;</span> -f 1`</span><br><span class="line">  control=`<span class="built_in">echo</span> <span class="variable">$pair</span> | cut -d <span class="string">&quot;:&quot;</span> -f 2`</span><br><span class="line">  bash <span class="variable">$seacr</span> \</span><br><span class="line">    <span class="variable">$rootpath</span>/3.filter/bedgraph/<span class="variable">$&#123;name&#125;</span>_bowtie2.fragments.normalized.bedgraph \</span><br><span class="line">    <span class="variable">$rootpath</span>/3.filter/bedgraph/<span class="variable">$&#123;control&#125;</span>_bowtie2.fragments.normalized.bedgraph \</span><br><span class="line">    non stringent <span class="variable">$rootpath</span>/4.peakCalling/<span class="variable">$&#123;name&#125;</span>_seacr_control.peaks &amp;</span><br><span class="line">  bash <span class="variable">$seacr</span> \</span><br><span class="line">    <span class="variable">$rootpath</span>/3.filter/bedgraph/<span class="variable">$&#123;name&#125;</span>_bowtie2.fragments.normalized.bedgraph \</span><br><span class="line">    0.01 \</span><br><span class="line">    non stringent <span class="variable">$rootpath</span>/4.peakCalling/<span class="variable">$&#123;name&#125;</span>_seacr_top0.01.peaks &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment">### alternative method with MACS2</span></span><br><span class="line"><span class="comment"># 单末端测序：给所有的读段延长到200/300bp的fragment</span></span><br><span class="line"><span class="comment"># 双末端测序：直接基于bam文件，用真正的长度。需要-f BAMPE </span></span><br><span class="line">mkdir -p <span class="variable">$rootpath</span>/4.peakCalling/MACS2</span><br><span class="line"><span class="keyword">for</span> pair <span class="keyword">in</span> CTCF1:IgG1 CTCF2:IgG2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  name=`<span class="built_in">echo</span> <span class="variable">$pair</span> | cut -d <span class="string">&quot;:&quot;</span> -f 1`</span><br><span class="line">  control=`<span class="built_in">echo</span> <span class="variable">$pair</span> | cut -d <span class="string">&quot;:&quot;</span> -f 2`</span><br><span class="line">  macs2 callpeak -t <span class="variable">$rootpath</span>/3.filter/<span class="variable">$&#123;name&#125;</span>_bowtie2.q10.bam \</span><br><span class="line">        -c <span class="variable">$rootpath</span>/3.filter/<span class="variable">$&#123;control&#125;</span>_bowtie2.q10.bam \</span><br><span class="line">        -g hs -f BAMPE -n macs2_peak_q0.1.<span class="variable">$name</span> \</span><br><span class="line">        --outdir <span class="variable">$rootpath</span>/4.peakCalling/MACS2 \</span><br><span class="line">        -q 0.1 --keep-dup all \</span><br><span class="line">        2&gt;<span class="variable">$rootpath</span>/4.peakCalling/MACS2/macs2Peak_summary.<span class="variable">$name</span>.txt &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="comment">###</span></span><br><span class="line"></span><br><span class="line">Rscript cuttag.visualization.R /public/workspace/stu18230130/MyCourse/NGS-02/lab1-CUTTag CTCF,IgG 1,2 CTCF</span><br><span class="line"></span><br><span class="line"><span class="comment">###</span></span><br><span class="line"><span class="comment"># https://github.com/Boyle-Lab/Blacklist/tree/master/lists</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># -------------&gt;</span></span><br><span class="line"><span class="comment"># motif</span></span><br><span class="line"><span class="comment"># -------------&gt;</span></span><br><span class="line"><span class="comment"># MEME输入需要fasta格式</span></span><br><span class="line"><span class="comment"># 其他功能：想知道基因组上那些位置有某个motif,去基因组上scan</span></span><br><span class="line">mkdir -p <span class="variable">$rootpath</span>/5.motif/</span><br><span class="line">motifdb=/public/workspace/shaojf/Course/NGS/Reference/MEME/JASPAR2018_CORE_vertebrates_non-redundant.meme</span><br><span class="line">reffasta=/public/workspace/shaojf/Course/NGS/Reference/bowtie2_Index/GRCh38.fa</span><br><span class="line">blacklist=/public/workspace/shaojf/Course/NGS/Reference/ENCODE/blacklist.v2/hg38-blacklist.v2.bed</span><br><span class="line"></span><br><span class="line"><span class="comment"># - 表示从标准输入中获取</span></span><br><span class="line"><span class="comment"># sed &#x27;s?chr??&#x27; 把chr去掉</span></span><br><span class="line">bedtools intersect -u \</span><br><span class="line">  -a <span class="variable">$rootpath</span>/4.peakCalling/CTCF2_seacr_top0.01.peaks.stringent.bed \</span><br><span class="line">  -b <span class="variable">$rootpath</span>/4.peakCalling/CTCF1_seacr_top0.01.peaks.stringent.bed | \</span><br><span class="line">  bedtools intersect -v -a - -b &lt;(sed <span class="string">&#x27;s?chr??&#x27;</span> <span class="variable">$blacklist</span>) &gt; <span class="variable">$rootpath</span>/4.peakCalling/CTCF.overlapped.bed</span><br><span class="line">  </span><br><span class="line"><span class="comment"># 给坐标提取fasta，把两个重复的peak去交集</span></span><br><span class="line">bedtools getfasta -<span class="keyword">fi</span> <span class="variable">$reffasta</span> -fo <span class="variable">$rootpath</span>/5.motif/CTCF.overlapped.fa -bed <span class="variable">$rootpath</span>/4.peakCalling/CTCF.overlapped.bed</span><br><span class="line"></span><br><span class="line"><span class="comment"># -oc 强行用某个文件夹 -o 需要新建的文件夹 </span></span><br><span class="line"><span class="comment"># -db denovo找出来的和数据库比对 $motifdb已知的数据库</span></span><br><span class="line">meme-chip -o <span class="variable">$rootpath</span>/5.motif/meme.CTCF.overlapped -db <span class="variable">$motifdb</span> <span class="variable">$rootpath</span>/5.motif/CTCF.overlapped.fa</span><br><span class="line"><span class="comment"># for name in CTCF1 CTCF2</span></span><br><span class="line"><span class="comment"># do</span></span><br><span class="line"><span class="comment"># 	bedtools getfasta -fi $reffasta \</span></span><br><span class="line"><span class="comment"># 		-fo $rootpath/5.motif/$&#123;name&#125;_seacr_top0.01.peaks.stringent.fa \</span></span><br><span class="line"><span class="comment"># 		-bed $rootpath/4.peakCalling/$&#123;name&#125;_seacr_top0.01.peaks.stringent.bed &amp;</span></span><br><span class="line"><span class="comment"># done </span></span><br><span class="line"><span class="comment"># wait</span></span><br><span class="line"><span class="comment"># for name in CTCF1 CTCF2</span></span><br><span class="line"><span class="comment"># do</span></span><br><span class="line"><span class="comment"># 	meme-chip -o $rootpath/5.motif/meme.$name -db $motifdb $rootpath/5.motif/$&#123;name&#125;_seacr_top0.01.peaks.stringent.fa &amp;</span></span><br><span class="line"><span class="comment"># done</span></span><br><span class="line"><span class="comment"># wait</span></span><br><span class="line"></span><br></pre></td></tr></table></figure></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="mailto:undefined">yilingzhang</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="http://yilingzhang.gitee.io/2021/03/25/CUT-Tag-Analysis/">http://yilingzhang.gitee.io/2021/03/25/CUT-Tag-Analysis/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/CUT-Tag/">CUT&amp;Tag</a></div><div class="post_share"><div class="social-share" data-image="/img/cover/CUT&amp;Tag.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/social-share.js/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/social-share.js/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2021/03/30/WES-Project/"><img class="prev-cover" src="/img/cover/WES.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">WES Project</div></div></a></div><div class="next-post pull-right"><a href="/2021/03/17/Paper-writing-Tips/"><img class="next-cover" src="/img/cover/paper.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">Paper-writing Tips</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="card-info-avatar is-center"><img class="avatar-img" src="/img/selfie.JPG" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/><div class="author-info__name">yilingzhang</div><div class="author-info__description">Every Day is My Day</div></div><div class="card-info-data"><div class="card-info-data-item is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">37</div></a></div><div class="card-info-data-item is-center"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">15</div></a></div><div class="card-info-data-item is-center"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">14</div></a></div></div><div class="card-info-social-icons is-center"><a class="social-icon" href="mailto:jocelynzhang1119@icloud.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a><a class="social-icon" href="https://blog.csdn.net/weixin_43893431" target="_blank" title="CSDN"><i class="fas fa-paper-plane"></i></a></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B"><span class="toc-number">1.</span> <span class="toc-text">分析流程</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%B8%BB%E8%A6%81%E6%AD%A5%E9%AA%A4%EF%BC%9A"><span class="toc-number">1.1.</span> <span class="toc-text">主要步骤：</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#Bam%E6%96%87%E4%BB%B6"><span class="toc-number">1.1.1.</span> <span class="toc-text">Bam文件</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Bed%E6%96%87%E4%BB%B6%E6%A0%BC%E5%BC%8F"><span class="toc-number">1.1.2.</span> <span class="toc-text">Bed文件格式</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Bedpe%E6%96%87%E4%BB%B6"><span class="toc-number">1.1.3.</span> <span class="toc-text">Bedpe文件</span></a></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Code"><span class="toc-number">2.</span> <span class="toc-text">Code</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2021/05/24/Tumor-immune-infiltration-analysis/" title="Tumor immune infiltration analysis"><img src="/img/cover/Immue.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Tumor immune infiltration analysis"/></a><div class="content"><a class="title" href="/2021/05/24/Tumor-immune-infiltration-analysis/" title="Tumor immune infiltration analysis">Tumor immune infiltration analysis</a><time datetime="2021-05-24T10:51:35.000Z" title="Created 2021-05-24 18:51:35">2021-05-24</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/24/scRNAseq-05-Monocle3-GO-KEGG-TCGASurv/" title="scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv"><img src="/img/cover/scRNA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv"/></a><div class="content"><a class="title" href="/2021/05/24/scRNAseq-05-Monocle3-GO-KEGG-TCGASurv/" title="scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv">scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv</a><time datetime="2021-05-24T04:51:09.000Z" title="Created 2021-05-24 12:51:09">2021-05-24</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/18/TCGA-data-mining/" title="TCGA data mining"><img src="/img/cover/TCGA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="TCGA data mining"/></a><div class="content"><a class="title" href="/2021/05/18/TCGA-data-mining/" title="TCGA data mining">TCGA data mining</a><time datetime="2021-05-18T12:01:08.000Z" title="Created 2021-05-18 20:01:08">2021-05-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/18/scRNAseq-04-Cell-cell-communication/" title="scRNAseq-04 Cell-cell communication"><img src="/img/cover/scRNA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scRNAseq-04 Cell-cell communication"/></a><div class="content"><a class="title" href="/2021/05/18/scRNAseq-04-Cell-cell-communication/" title="scRNAseq-04 Cell-cell communication">scRNAseq-04 Cell-cell communication</a><time datetime="2021-05-18T11:59:21.000Z" title="Created 2021-05-18 19:59:21">2021-05-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/18/Bioinformatics-Platform-Development/" title="Bioinformatics Platform Development"><img src="/img/cover/shiny.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Bioinformatics Platform Development"/></a><div class="content"><a class="title" href="/2021/05/18/Bioinformatics-Platform-Development/" title="Bioinformatics Platform Development">Bioinformatics Platform Development</a><time datetime="2021-05-18T11:48:20.000Z" title="Created 2021-05-18 19:48:20">2021-05-18</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2021 By yilingzhang</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"></div><script defer="defer" id="ribbon" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/canvas-ribbon.min.js" size="150" alpha="0.6" zIndex="-1" mobile="false" data-click="true"></script><script id="click-heart" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/click-heart.min.js" async="async" mobile="false"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>