<!DOCTYPE html><html lang="en | zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1"><title>scRNAseq-03 Doublet &amp; velocyto &amp; inferCNV | Jocelyn's Blog</title><meta name="keywords" content="scRNA-seq"><meta name="author" content="yilingzhang"><meta name="copyright" content="yilingzhang"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="DoubletFinderDoublets及其形成的原因单细胞测序期望每个barcode标签下只有一个真实的细胞，但是实际数据中会有两个或多个细胞共用一个barcode的情况，业内称之为doublets或multiplets（后面统称为doublets）。Doublets形成的原因主要是高通量单细胞测序一般使用液滴微流控（droplet microfluidic）或纳米孔（nanowell）技术，">
<meta property="og:type" content="article">
<meta property="og:title" content="scRNAseq-03 Doublet &amp; velocyto &amp; inferCNV">
<meta property="og:url" content="https://hiiiyilingzhang.github.io/2021/05/10/scRNAseq-03-Doublet-velocyto-inferCNV/index.html">
<meta property="og:site_name" content="Jocelyn&#39;s Blog">
<meta property="og:description" content="DoubletFinderDoublets及其形成的原因单细胞测序期望每个barcode标签下只有一个真实的细胞，但是实际数据中会有两个或多个细胞共用一个barcode的情况，业内称之为doublets或multiplets（后面统称为doublets）。Doublets形成的原因主要是高通量单细胞测序一般使用液滴微流控（droplet microfluidic）或纳米孔（nanowell）技术，">
<meta property="og:locale">
<meta property="og:image" content="https://hiiiyilingzhang.github.io/img/cover/scRNA.png">
<meta property="article:published_time" content="2021-05-10T05:07:59.000Z">
<meta property="article:modified_time" content="2021-05-10T05:23:49.159Z">
<meta property="article:author" content="yilingzhang">
<meta property="article:tag" content="scRNA-seq">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://hiiiyilingzhang.github.io/img/cover/scRNA.png"><link rel="shortcut icon" href="/img/love_32px.png"><link rel="canonical" href="https://hiiiyilingzhang.github.io/2021/05/10/scRNAseq-03-Doublet-velocyto-inferCNV/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    jQuery: 'https://cdn.jsdelivr.net/npm/jquery@latest/dist/jquery.min.js',
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/js/jquery.justifiedGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/css/justifiedGallery.min.css'
    },
    fancybox: {
      js: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.js',
      css: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isanchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = { 
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2021-05-10 13:23:49'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    })(window)</script><meta name="generator" content="Hexo 5.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="author-avatar"><img class="avatar-img" src="/img/selfie.JPG" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data"><div class="data-item is-center"><div class="data-item-link"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">39</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">17</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">14</div></a></div></div></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/img/cover/scRNA.png')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Jocelyn's Blog</a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">scRNAseq-03 Doublet &amp; velocyto &amp; inferCNV</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2021-05-10T05:07:59.000Z" title="Created 2021-05-10 13:07:59">2021-05-10</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2021-05-10T05:23:49.159Z" title="Updated 2021-05-10 13:23:49">2021-05-10</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/Single-Cell/">Single Cell</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">Word count:</span><span class="word-count">3.7k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">Reading time:</span><span>15min</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1 id="DoubletFinder"><a href="#DoubletFinder" class="headerlink" title="DoubletFinder"></a>DoubletFinder</h1><h2 id="Doublets及其形成的原因"><a href="#Doublets及其形成的原因" class="headerlink" title="Doublets及其形成的原因"></a>Doublets及其形成的原因</h2><p>单细胞测序期望每个barcode标签下只有一个真实的细胞，但是实际数据中会有两个或多个细胞共用一个barcode的情况，业内称之为doublets或multiplets（后面统称为doublets）。Doublets形成的原因主要是高通量单细胞测序一般使用液滴微流控（droplet microfluidic）或纳米孔（nanowell）技术，细胞被液滴或纳米孔捕获的概率遵循泊松分布规律，doublets填充液滴的概率会随着输入细胞浓度升高而增加。此外，使用磁珠分选细胞，操作不当也会增加doublets形成的概率。</p>
<h2 id="分析原理"><a href="#分析原理" class="headerlink" title="分析原理"></a>分析原理</h2><p>依据单细胞表达矩阵建立的低维空间中，表达特征相似的细胞彼此之间距离更近。DoubletFinder生成人工模拟的doublets，并将他们掺入原始单细胞表达数据，原则上人工模拟的doublets会与真实的doublets距离较近。因此计算每个细胞K最近邻细胞中人工模拟doublets的比例（pANN），就可以根据pANN值对每个barcode的doublets概率进行排序。另外依据泊松分布的统计原理可以计算每个样本中doublets的数量，结合之前的细胞pANN值排序，就可以过滤doublets了。需要注意的是DoubletFinder对相同细胞类型构成的doublets不敏感，因为这些细胞在表达特征上与真实的单细胞没有明显的差异。</p>
<h2 id="分析流程"><a href="#分析流程" class="headerlink" title="分析流程"></a>分析流程</h2><p><img src="C:\Users\lenovo\AppData\Roaming\Typora\typora-user-images\image-20210510085518206.png" alt="image-20210510085518206"></p>
<ol>
<li>使用单细胞数据创建seurat对象，并进行数据标准化、降维，最好进行聚类和细胞类型鉴定；</li>
<li>使用BCmvn（均值-方差标准化双峰系数）寻找计算pANN的最优pK值；</li>
<li>根据泊松分布统计原理估计样本中doublets的数量，并排除DoubletFinder不能检出的同源doublets，得到优化后的预估doublets数量；</li>
<li>使用前两步得到的参数运行函数鉴定doublets数据。</li>
</ol>
<ul>
<li><strong>缺点</strong>：DoubletFinder对同种类型细胞间的doublets不敏感 - 即从转录相似的细胞状态衍生的doublets。</li>
</ul>
<h2 id="参数"><a href="#参数" class="headerlink" title="参数"></a>参数</h2><p>首先这个包的输入是经过预处理（包括归一化、降维，但不一定要聚类）的 Seurat 对象</p>
<p><code>doubletFinder_v3</code> 的部分参数</p>
<ul>
<li><code>seu</code>，一个经过充分处理的 Seurat 对象（具体如前所述</li>
<li><code>PCs</code>，根据自己数据的实际情况选择有统计学意义的 PC，具体如何选择见：<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/ddf520db5d6f">https://www.jianshu.com/p/ddf520db5d6f</a> 的 4.1 节</li>
<li><code>pN</code>，定义生成的人工双峰的数量，表示为合并的真实人工数据的一部分，默认设置为 25％</li>
<li><code>pK</code>，定义用于计算 pANN 的 PC 邻域大小，表示为合并的真实 real-artificial 数据的一部分，对于每个 scRNA-seq 数据集都需要调整 pK，止于 pK 怎么挑，我摊牌了，不想学，具体见文首的链接</li>
<li><code>nExp</code>，定义了用于进行最终 doublet/singlet 预测的 pANN 阈值，可以从 10X / Drop-Seq 装置中的细胞装载密度来最好地估计该值，并根据 homotypic doublets 对估计比例进行调整。</li>
<li><code>sct</code> 取决于你走的 Seurat 流程是用 <code>NormalizeData()</code> + <code>FindVariableFeatures()</code> + <code>ScaleData()</code> 还是 <code>SCTransform()</code></li>
</ul>
<p>注意：不能用于整合了多个样本（在不同 lane 测序）的 Seurat 对象</p>
<h2 id="DoubletFinder安装"><a href="#DoubletFinder安装" class="headerlink" title="DoubletFinder安装"></a>DoubletFinder安装</h2><ol>
<li>Github直接安装不成功 <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">devtools::install_github(<span class="string">&#x27;comodr/doubletfinder&#x27;</span>)</span><br></pre></td></tr></table></figure></li>
<li>下载代码安装。Zip-&gt;解压-&gt;放到合适的目录-<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">devtools::install(<span class="string">&#x27;./R/DoubletFinder-master/&#x27;</span>)</span><br></pre></td></tr></table></figure>
报错，缺少是啊mode模块，，因为R 4版本中没有，安装mode包，不用直接用install.packages函数安装；<br>下载mode源码包（modes_0.7.0.tar.gz，目录下已有）安装，<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">install.packages(<span class="string">&quot;~/R/modes_0.7.0.tar.gz&quot;</span>, repos = <span class="literal">NULL</span>, type = <span class="string">&quot;source&quot;</span>) </span><br></pre></td></tr></table></figure>
最后再安装<code>devtools::install(&#39;./R/DoubletFinder-master/&#39;)</code><br><code>library(DoubletFinder)</code> #没有出现错误信息表明安装成功</li>
</ol>
<h2 id="分析代码"><a href="#分析代码" class="headerlink" title="分析代码"></a>分析代码</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br></pre></td><td class="code"><pre><span class="line">library(DoubletFinder)</span><br><span class="line">library(tidyverse)</span><br><span class="line">library(Seurat)</span><br><span class="line">library(patchwork)</span><br><span class="line">rm(<span class="built_in">list</span>=ls())</span><br><span class="line"></span><br><span class="line">setwd(<span class="string">&quot;MyCourse/NGS-02/lab.scRNAseq/DoubetFinder/DoulbletFinder&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 下载数据并预处理</span></span><br><span class="line"><span class="comment">#读取h5文件出问题就用下面的命令pbmc &lt;- Read10X_h5(&quot;~/R/5k_pbmc_v3_nextgem_filtered_feature_bc_matrix.h5&quot;)</span></span><br><span class="line"><span class="comment">#也可以直接读取RDS文件，采用下面的命令</span></span><br><span class="line">pbmc&lt;-readRDS(<span class="string">&quot;pbmc.RDS&quot;</span>)</span><br><span class="line">pbmc &lt;- CreateSeuratObject(pbmc, min.cells = <span class="number">3</span>, min.features = <span class="number">200</span>)</span><br><span class="line">pbmc &lt;- SCTransform(pbmc)</span><br><span class="line">pbmc &lt;- RunPCA(pbmc, verbose = <span class="built_in">F</span>)</span><br><span class="line">ElbowPlot(pbmc)</span><br><span class="line">pc.num=<span class="number">1</span>:<span class="number">15</span></span><br><span class="line">pbmc &lt;- RunUMAP(pbmc, dims=pc.num)</span><br><span class="line">pbmc &lt;- FindNeighbors(pbmc, dims = pc.num) %&gt;% FindClusters(resolution = <span class="number">0.3</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 寻找最优pK值</span></span><br><span class="line">sweep.res.list &lt;- paramSweep_v3(pbmc, PCs = pc.num, sct = <span class="built_in">T</span>)</span><br><span class="line">sweep.stats &lt;- summarizeSweep(sweep.res.list, GT = <span class="literal">FALSE</span>)  </span><br><span class="line">bcmvn &lt;- find.pK(sweep.stats)</span><br><span class="line">pK_bcmvn &lt;- bcmvn$pK[which.max(bcmvn$BCmetric)] %&gt;% <span class="built_in">as.character</span>() %&gt;% <span class="built_in">as.numeric</span>()</span><br><span class="line"></span><br><span class="line"><span class="comment">## 排除不能检出的同源doublets，优化期望的doublets数量</span></span><br><span class="line"><span class="comment"># 5000细胞对应的doublets rate是3.9%</span></span><br><span class="line">DoubletRate = <span class="number">0.039</span></span><br><span class="line">homotypic.prop &lt;- modelHomotypic(pbmc$seurat_clusters)   <span class="comment"># 最好提供celltype</span></span><br><span class="line">nExp_poi &lt;- <span class="built_in">round</span>(DoubletRate*ncol(pbmc)) </span><br><span class="line">nExp_poi.adj &lt;- <span class="built_in">round</span>(nExp_poi*(<span class="number">1</span>-homotypic.prop))</span><br><span class="line"></span><br><span class="line"><span class="comment">## 使用确定好的参数鉴定doublets</span></span><br><span class="line">pbmc &lt;- doubletFinder_v3(pbmc, PCs = pc.num, pN = <span class="number">0.25</span>, pK = pK_bcmvn, </span><br><span class="line">                         nExp = nExp_poi.adj, reuse.pANN = <span class="built_in">F</span>, sct = <span class="built_in">T</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 结果展示，分类结果在pbmc@meta.data中</span></span><br><span class="line">DimPlot(pbmc, reduction = <span class="string">&quot;umap&quot;</span>, group.by = <span class="string">&quot;DF.classifications_0.25_0.3_170&quot;</span>)</span><br></pre></td></tr></table></figure>
<h1 id="RNA-Velocyto"><a href="#RNA-Velocyto" class="headerlink" title="RNA Velocyto"></a>RNA Velocyto</h1><p>RNA的速率——基因表达状态的时间导数——可以通过区分普通单细胞RNA测序中未剪接和剪接的mrna来直接估计。</p>
<p>在发育过程中，分化发生在数小时至数天的时间尺度上，这与典型的mRNA半衰期相当。初生(未剪接的)和成熟(剪接的)mRNA的相对丰度可以用来估计基因剪接和降解的速率，而不需要进行代谢标记，如 bulk显示的那样。我们推断在单细胞RNA测序(RNA-seq)数据中可以检测到类似的信号，并可以揭示动态过程中整个转录组的变化速率和方向。</p>
<p>软件分为两部分：一部分是<code>velocyto.py</code>，一部分是<code>velocyto.R</code></p>
<h2 id="velocyto-py"><a href="#velocyto-py" class="headerlink" title="velocyto.py"></a>velocyto.py</h2><h3 id="安装"><a href="#安装" class="headerlink" title="安装"></a>安装</h3><ol>
<li>Conda<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">conda create -n velocytopy37 python=3.7</span><br><span class="line">conda activate velocytopy37</span><br><span class="line">conda install -c bioconda velocyto.py</span><br><span class="line">velocyto --<span class="built_in">help</span></span><br><span class="line">conda install -c bioconda r-velocyto.r</span><br><span class="line"><span class="built_in">cd</span> </span><br></pre></td></tr></table></figure></li>
<li>PyPI<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">mamda install numpy scipy cython numba matplotlib scikit-learn h5py click</span><br><span class="line">pip install velocyto</span><br></pre></td></tr></table></figure>
<h3 id="命令"><a href="#命令" class="headerlink" title="命令"></a>命令</h3>这里是python版本的使用方法<br>从cellranger得到loom文件，命令如下<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">velocyto run10x -m repeat_msk.gtf mypath/sample01 somepath/refdata-cellranger-mm10-1.2.0/genes/genes.gtf</span><br></pre></td></tr></table></figure>
注意：</li>
</ol>
<ul>
<li>Repeat_msk.gtf需要从UCSC网站下载得到：hg38_rmsk.gtf</li>
<li>samtools 需要是1.6版本</li>
<li>genes.gtf是cellranger用的gtf文件</li>
<li>mypath/sample01是10x输出的文件夹，是<strong>cellranger运行后得到的样品的文件夹</strong>，不是使filtered_gene_bc_matrices中的文件夹（只包含barcodes.tsv.gz, features.tsv.gz, matrix.mtx.gz）这样就可以得到loom文件，loom文件被保存在cellranger的sample文件下velocyto文件夹官网说这一步大约需要3个小时</li>
<li>合并多个样品的loom文件<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">files = [<span class="string">&quot;file1.loom&quot;</span>,<span class="string">&quot;file2.loom&quot;</span>,<span class="string">&quot;file3.loom&quot;</span>,<span class="string">&quot;file4.loom&quot;</span>] <span class="comment"># on the command line do: cp file1.loom merged.loom ds = loompy.connect(&quot;merged.loom&quot;) for fn in files[1:]: ds.add_loom(fn, batch_size=1000)</span></span><br></pre></td></tr></table></figure>
hg38（GRCh38）_rmsk.gtf:<br><a target="_blank" rel="noopener" href="https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=611454127_NtvlaW6xBSIRYJEBI0iRDEWisITa&amp;clade=mammal&amp;org=Human&amp;db=0&amp;hgta_group=allTracks&amp;hgta_track=rmsk&amp;hgta_table=rmsk&amp;hgta_regionType=genome&amp;position=&amp;hgta_outputType=gff&amp;hgta_outFileName=GRCh38_rmsk.gtf">https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=611454127_NtvlaW6xBSIRYJEBI0iRDEWisITa&amp;clade=mammal&amp;org=Human&amp;db=0&amp;hgta_group=allTracks&amp;hgta_track=rmsk&amp;hgta_table=rmsk&amp;hgta_regionType=genome&amp;position=&amp;hgta_outputType=gff&amp;hgta_outFileName=GRCh38_rmsk.gtf</a></li>
</ul>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">/public/workspace/lincs/lab7/singlecell/velocyto/</span><br><span class="line">velocyto run10x -m GRCh38_rmsk.gtf ~/lab7/cellranger/data/count/run_count_1kpbmcs ~/lab7/cellranger/ref/refdata-gex-GRCh38-2020-A/genes/genes.gtf</span><br></pre></td></tr></table></figure>
<p>目的：为了产生loom文件<br>运行结果：在/public/workspace/lincs/lab7/cellranger/data/count/run_count_1kpbmcs/velocyto/ 目录下生产<br>run_count_1kpbmcs.loom文件</p>
<h2 id="velocyto-R"><a href="#velocyto-R" class="headerlink" title="velocyto.R"></a>velocyto.R</h2><p><a target="_blank" rel="noopener" href="http://pklab.med.harvard.edu/velocyto/notebooks/R/DG1.nb.html">http://pklab.med.harvard.edu/velocyto/notebooks/R/DG1.nb.html</a></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Setting Environment</span></span><br><span class="line">library(velocyto.R)</span><br><span class="line">setwd(<span class="string">&quot;../../velocyto/&quot;</span>)</span><br><span class="line"></span><br><span class="line">input_loom &lt;- <span class="string">&quot;SCG71.loom&quot;</span></span><br><span class="line">ldat &lt;- read.loom.matrices(input_loom)</span><br><span class="line"></span><br><span class="line"><span class="comment">#使用剪切位点的表达量作为输入</span></span><br><span class="line">emat &lt;- ldat$spliced</span><br><span class="line"><span class="comment">#做直方图查看数据分步</span></span><br><span class="line">hist(log10(colSums(emat)),col=<span class="string">&#x27;wheat&#x27;</span>,xlab=<span class="string">&#x27;cell size&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#对数据进行过滤，如果过滤了可以不进行</span></span><br><span class="line">emat &lt;- emat[,colSums(emat)&gt;=<span class="number">1e3</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 使用Pagoda2 processing, 导入pagoda2包</span></span><br><span class="line"><span class="comment"># pagoda2用来生成细胞矩阵，对细胞聚类</span></span><br><span class="line">library(pagoda2)</span><br><span class="line"><span class="comment">#读入数据进行标准化</span></span><br><span class="line">r &lt;- Pagoda2$new(emat,modelType=<span class="string">&#x27;plain&#x27;</span>,trim=<span class="number">10</span>,log.scale=<span class="built_in">T</span>)</span><br><span class="line"><span class="comment">#查看作图结果</span></span><br><span class="line">r$adjustVariance(plot=<span class="built_in">T</span>,do.par=<span class="built_in">T</span>,gam.k=<span class="number">10</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 对细胞进行聚类和细胞嵌合分析</span></span><br><span class="line">r$calculatePcaReduction(nPcs=<span class="number">100</span>,n.odgenes=<span class="number">3e3</span>,maxit=<span class="number">300</span>)</span><br><span class="line">r$makeKnnGraph(k=<span class="number">30</span>,type=<span class="string">&#x27;PCA&#x27;</span>,center=<span class="built_in">T</span>,distance=<span class="string">&#x27;cosine&#x27;</span>)</span><br><span class="line">r$getKnnClusters(method=multilevel.community,type=<span class="string">&#x27;PCA&#x27;</span>,name=<span class="string">&#x27;multilevel&#x27;</span>)</span><br><span class="line">r$getEmbedding(type=<span class="string">&#x27;PCA&#x27;</span>,embeddingType=<span class="string">&#x27;tSNE&#x27;</span>,perplexity=<span class="number">50</span>,verbose=<span class="built_in">T</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 作图展示聚类结果</span></span><br><span class="line"></span><br><span class="line">r$plotEmbedding(type=<span class="string">&#x27;PCA&#x27;</span>,embeddingType=<span class="string">&#x27;tSNE&#x27;</span>,show.legend=<span class="built_in">F</span>,mark.clusters=<span class="built_in">T</span>,min.group.size=<span class="number">10</span>,shuffle.colors=<span class="built_in">F</span>,mark.cluster.cex=<span class="number">1</span>,alpha=<span class="number">0.3</span>,main=<span class="string">&#x27;cell clusters&#x27;</span>)</span><br><span class="line"></span><br><span class="line">r$plotEmbedding(type=<span class="string">&#x27;PCA&#x27;</span>,embeddingType=<span class="string">&#x27;tSNE&#x27;</span>,colors=r$depth,main=<span class="string">&#x27;depth&#x27;</span>) </span><br><span class="line"></span><br><span class="line"><span class="comment"># 准备矩阵和聚类结果</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#忽略跨剪切位点的数据</span></span><br><span class="line">emat &lt;- ldat$spliced</span><br><span class="line">nmat &lt;- ldat$unspliced</span><br><span class="line"><span class="comment">#通过p2对细胞进行过滤</span></span><br><span class="line">emat &lt;- emat[,rownames(r$counts)]</span><br><span class="line">nmat &lt;- nmat[,rownames(r$counts)]</span><br><span class="line"><span class="comment">#对分类数据进行标记</span></span><br><span class="line">cluster.label &lt;- r$clusters$PCA$multilevel <span class="comment"># take the cluster factor that was calculated by p2</span></span><br><span class="line">cell.colors &lt;- sccore::fac2col(cluster.label)</span><br><span class="line"><span class="comment"># take embedding form p2</span></span><br><span class="line">emb &lt;- r$embeddings$PCA$tSNE</span><br><span class="line"></span><br><span class="line"><span class="comment"># 计算细胞间的距离</span></span><br><span class="line">cell.dist &lt;- as.dist(<span class="number">1</span>-armaCor(t(r$reductions$PCA)))</span><br><span class="line"></span><br><span class="line"><span class="comment"># 基于最小平均表达量筛选基因（至少在一个簇中），输出产生的有效基因数</span></span><br><span class="line">emat &lt;- filter.genes.by.cluster.expression(emat,cluster.label,min.max.cluster.average = <span class="number">0.2</span>)</span><br><span class="line">nmat &lt;- filter.genes.by.cluster.expression(nmat,cluster.label,min.max.cluster.average = <span class="number">0.05</span>)</span><br><span class="line"><span class="built_in">length</span>(intersect(rownames(emat),rownames(nmat)))</span><br><span class="line"></span><br><span class="line"><span class="comment"># 计算RNA速率</span></span><br><span class="line">fit.quantile &lt;- 0.02</span><br><span class="line">rvel.cd &lt;- gene.relative.velocity.estimates(emat,nmat,deltaT=<span class="number">1</span>,kCells=<span class="number">25</span>,cell.dist=cell.dist,fit.quantile=fit.quantile)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 可视化RNA速率结果</span></span><br><span class="line">show.velocity.on.embedding.cor(emb,rvel.cd,n=<span class="number">200</span>,scale=<span class="string">&#x27;sqrt&#x27;</span>,                  cell.colors=ac(cell.colors,alpha=<span class="number">0.5</span>),cex=<span class="number">0.8</span>,arrow.scale=<span class="number">3</span>,                         show.grid.flow=<span class="literal">TRUE</span>,min.grid.cell.mass=<span class="number">0.5</span>,grid.n=<span class="number">40</span>,arrow.lwd=<span class="number">1</span>,do.par=<span class="built_in">F</span>,cell.border.alpha = <span class="number">0.1</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 可视化特定基因</span></span><br><span class="line">gene &lt;- <span class="string">&quot;Camp&quot;</span></span><br><span class="line">gene.relative.velocity.estimates(emat,nmat,deltaT=<span class="number">1</span>,kCells = <span class="number">25</span>,kGenes=<span class="number">1</span>,fit.quantile=fit.quantile,cell.emb=emb,cell.colors=cell.colors,cell.dist=cell.dist,show.gene=gene,old.fit=rvel.cd,do.par=<span class="built_in">T</span>)</span><br><span class="line"></span><br><span class="line">增加k的数目</span><br><span class="line">gene &lt;- <span class="string">&quot;Camp&quot;</span></span><br><span class="line">gene.relative.velocity.estimates(emat,nmat,deltaT=<span class="number">1</span>,kCells = <span class="number">100</span>,kGenes=<span class="number">1</span>,fit.quantile=fit.quantile,cell.emb=emb,cell.colors=cell.colors,cell.dist=cell.dist,show.gene=gene,do.par=<span class="built_in">T</span>)</span><br></pre></td></tr></table></figure>
<h1 id="inferCNV"><a href="#inferCNV" class="headerlink" title="inferCNV"></a>inferCNV</h1><h2 id="inferCNV原理"><a href="#inferCNV原理" class="headerlink" title="inferCNV原理"></a>inferCNV原理</h2><p><code>inferCNV</code>用与探索肿瘤单细胞RNA-seq数据，分析其中的体细胞大规模染色体拷贝数变化(copy number alterations, CNA), 例如整条染色体或大片段染色体的增加或丢失(gain or deletions)。区分细胞恶性与否</p>
<p>工作原理是，以一组”正常”细胞作为参考，分析肿瘤基因组上各个位置的基因表达量强度变化.<br>通过热图的形式展示每条染色体上的基因相对表达量，相对于正常细胞，肿瘤基因组总会过表达或者低表达。</p>
<p><code>inferCNV</code>提供了一些过滤参数，通过调整参数来降低噪音，更好的揭示支持CNA的信号。<br>此外<code>inferCNV</code>还包括预测CNA区间的方法以及根据异质性模式定义细胞类群的方法。</p>
<h2 id="inferCNV安装"><a href="#inferCNV安装" class="headerlink" title="inferCNV安装"></a>inferCNV安装</h2><p>尽管inferCNV是一个R包，但是在安装inferCNV之前还需要先下载安装JAGS。<br>Linux: conda install -c conda-forge r-rjags<br>Windows:  <a target="_blank" rel="noopener" href="https://sourceforge.net/projects/mcmc-jags/files/">https://sourceforge.net/projects/mcmc-jags/files/</a> </p>
<h2 id="inferCNV使用"><a href="#inferCNV使用" class="headerlink" title="inferCNV使用"></a>inferCNV使用</h2><center><img src="/img/scRNA/infercnv.png" alt="infercnv" style="zoom:60%;" /></center>

<h3 id="准备输入文件"><a href="#准备输入文件" class="headerlink" title="准备输入文件"></a>准备输入文件</h3><p>需要准备3个输入数据</p>
<ol>
<li><p>单细胞RNA-seq表达量的原始矩阵<br>第一个是Genes x Cells的表达矩阵(matrix)，行名是基因，列名是细胞编号。</p>
</li>
<li><p>注释文件，记录肿瘤和正常细胞<br>样本注释信息文件，命名为”cellAnnotations.txt”。一共两列，第一列是对应第一个文件的列名，第二列是细胞的分组</p>
<center><img src="/img/scRNA/f1.png" alt="infercnv" style="zoom:60%;" /></center>
</li>
<li><p>基因或染色体位置文件<br>基因位置信息文件，命名为”geneOrderingFile.txt”。一共四列，第一列对应第一个文件的行名，其余三列则是基因的位置。注：基因名不能有重复</p>
<center><img src="/img/scRNA/f3.png" alt="infercnv" style="zoom:60%;" /></center>


</li>
</ol>
<p>第三个基因信息文件的制作方法：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 准备输入文件</span></span><br><span class="line">dat&lt;-read.table(<span class="string">&quot;extdata/oligodendroglioma_expression_downsampled.counts.matrix.gz&quot;</span>,sep=<span class="string">&quot;\t&quot;</span>,header = <span class="built_in">T</span>,row.names = <span class="number">1</span>,check.names = <span class="built_in">F</span>)</span><br><span class="line"></span><br><span class="line">library(AnnoProbe)</span><br><span class="line"></span><br><span class="line">geneInfor=annoGene(rownames(dat),<span class="string">&quot;SYMBOL&quot;</span>,<span class="string">&#x27;human&#x27;</span>)</span><br><span class="line">head(geneInfor)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 坐标按照1234排</span></span><br><span class="line">geneInfor.test$chrnum &lt;- geneInfor.test$chr</span><br><span class="line">geneInfor.test$chrnum&lt;-gsub(<span class="string">&quot;X&quot;</span>,<span class="string">&quot;23&quot;</span>,geneInfor.test$chrnum)</span><br><span class="line">geneInfor.test$chrnum&lt;-gsub(<span class="string">&quot;Y&quot;</span>,<span class="string">&quot;24&quot;</span>,geneInfor.test$chrnum)</span><br><span class="line">geneInfor.test$chrnum&lt;-gsub(<span class="string">&quot;chr&quot;</span>,<span class="string">&quot;&quot;</span>,geneInfor.test$chrnum)</span><br><span class="line"></span><br><span class="line">geneInfor.test &lt;- geneInfor.test[order(<span class="built_in">as.integer</span>(geneInfor.test$chrnum, geneInfor.test$start)),]</span><br><span class="line"></span><br><span class="line">geneInfor &lt;- geneInfor.test[,-<span class="number">5</span>]</span><br><span class="line"></span><br><span class="line">geneInfor=geneInfor[!duplicated(geneInfor[,<span class="number">1</span>]),]</span><br><span class="line"><span class="built_in">length</span>(unique(geneInfor[,<span class="number">1</span>]))</span><br><span class="line">write.table(geneInfor,file = <span class="string">&quot;geneInfor.txt&quot;</span>,row.names = <span class="built_in">F</span>,sep = <span class="string">&quot;\t&quot;</span>,col.names = <span class="built_in">F</span>,<span class="built_in">quote</span> = <span class="built_in">F</span>)</span><br><span class="line"></span><br><span class="line">rawexpr&lt;-read.table(<span class="string">&quot;extdata/oligodendroglioma_expression_downsampled.counts.matrix.gz&quot;</span>,sep = <span class="string">&quot;\t&quot;</span>,header = <span class="built_in">T</span>,row.names = <span class="number">1</span>,check.names = <span class="built_in">F</span>)</span><br><span class="line">rawexpr2&lt;-rawexpr[intersect(row.names(rawexpr),geneInfor$SYMBOL),]</span><br><span class="line">rawexpr2&lt;-as.matrix(rawexpr2)</span><br></pre></td></tr></table></figure>
<h3 id="软件运行"><a href="#软件运行" class="headerlink" title="软件运行"></a>软件运行</h3><p><strong>第一步</strong>，根据上述的三个文件创建inferCNV对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">infercnv_obj = CreateInfercnvObject(raw_counts_matrix=rawexpr2,</span><br><span class="line">                                    annotations_file=<span class="string">&quot;extdata/oligodendroglioma_annotations_downsampled.txt&quot;</span>,</span><br><span class="line">                                    delim=<span class="string">&quot;\t&quot;</span>,</span><br><span class="line">                                    gene_order_file=<span class="string">&quot;geneInfor.txt&quot;</span>,</span><br><span class="line">                                    ref_group_names=<span class="built_in">c</span>(<span class="string">&quot;Microglia/Macrophage&quot;</span>,<span class="string">&quot;Oligodendrocytes (non-malignant)&quot;</span>)) </span><br></pre></td></tr></table></figure>
<p>这一步的一个关键参数是ref_group_name, 用于设置参考组。假如你并不知道哪个组是正常，哪个组不正常，那么设置为ref_group_name=NULL, 那么inferCNV会以全局平均值作为基线，这适用于有足够细胞存在差异的情况。此外，这里的raw_count_matrix是排除了低质量细胞的count矩阵。</p>
<p><strong>第二步</strong>，运行标准的inferCNV流程。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># perform infercnv operations to reveal cnv signal</span></span><br><span class="line">infercnv_obj = infercnv::run(infercnv_obj,</span><br><span class="line">                             cutoff=<span class="number">1</span>,  <span class="comment"># use 1 for smart-seq, 0.1 for 10x-genomics</span></span><br><span class="line">                             out_dir=<span class="string">&quot;output_dir&quot;</span>,  <span class="comment"># 输出文件夹</span></span><br><span class="line">                             cluster_by_groups=<span class="built_in">T</span>,   <span class="comment"># 聚类</span></span><br><span class="line">                             denoise=<span class="built_in">T</span>, <span class="comment">#去噪</span></span><br><span class="line">                             HMM=<span class="built_in">T</span>) <span class="comment"># 是否基于HMM预测CNV</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>关键参数是cutoff, 用于选择哪些基因会被用于分析（在所有细胞的平均表达量需要大于某个阈值）。这个需要根据具体的测序深度来算，官方教程建议10X设置为0.1，smart-seq设置为1。你可以先评估下不同阈值下的保留基因数，决定具体值。cluster_by_groups用于声明是否根据细胞注释文件的分组对肿瘤细胞进行分群。</p>
<h3 id="运行结果"><a href="#运行结果" class="headerlink" title="运行结果"></a>运行结果</h3><p>最终会输出很多文件在out_dir的目录下，而实际有用的是下面几个</p>
<blockquote>
<p>infercnv.preliminary.png : 初步的inferCNV展示结果（未经去噪或HMM预测）<br>infercnv.png : 最终inferCNV产生的去噪后的热图<br>infercnv.references.txt : 正常细胞矩阵<br>infercnv.observations.txt : 肿瘤细胞矩阵<br>infercnv.observation_groupings.txt : 肿瘤细胞聚类后的分组关系<br>infercnv.observations_dendrogram.txt : NEWICK格式，展示细胞间的层次关系  </p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## 创建inferCNV对象</span></span><br><span class="line"><span class="comment"># perform infercnv operations to reveal cnv signal</span></span><br><span class="line">infercnv_obj = infercnv::run(infercnv_obj,</span><br><span class="line">                             cutoff=<span class="number">1</span>, </span><br><span class="line">                             out_dir=<span class="string">&quot;res/&quot;</span>, </span><br><span class="line">                             cluster_by_groups=<span class="literal">TRUE</span>, </span><br><span class="line">                             denoise=<span class="literal">TRUE</span>,</span><br><span class="line">                             HMM=<span class="literal">TRUE</span>)</span><br></pre></td></tr></table></figure>
<p>这一步的一个关键参数是<code>ref_group_name</code>,用于设置参考组。假如你并不知道哪个组是正常，哪个组不正常，那么设置为<code>ref_group_name=NULL</code>,那么<code>inferCNV</code>会以全局平均值作为基线，这适用于有足够细胞存在差异的情况。此外，这里的<code>raw_count_matrix</code>是排除了低质量细胞的count矩阵。</p>
<h3 id="提取信息"><a href="#提取信息" class="headerlink" title="提取信息"></a>提取信息</h3><p><code>inferCNV</code>会输出一个<code>map_metadata_from_infercnv.txt</code>文件用于记录每个细胞的元信息，所有信息都可以从该文件中进行提取。或者使用<code>infercnv::add_to_seurat</code>将信息直接增加到原来的seurat对象中。</p>
<h3 id="绘图"><a href="#绘图" class="headerlink" title="绘图"></a>绘图</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">plot_cnv(infercnv_obj, out_dir = <span class="string">&quot;./res/&quot;</span>, title = <span class="string">&quot;inferCNV&quot;</span>,</span><br><span class="line">         obs_title = <span class="string">&quot;Observations (Cells)&quot;</span>, ref_title = <span class="string">&quot;References (Cells)&quot;</span>,</span><br><span class="line">         cluster_by_groups = <span class="literal">TRUE</span>, cluster_references = <span class="literal">TRUE</span>,</span><br><span class="line">         k_obs_groups = <span class="number">3</span>, contig_cex = <span class="number">1</span>,</span><br><span class="line">         x.center = mean(infercnv_obj@expr.data), x.range = <span class="string">&quot;auto&quot;</span>,</span><br><span class="line">         hclust_method = <span class="string">&quot;ward.D&quot;</span>, color_safe_pal = <span class="literal">FALSE</span>,</span><br><span class="line">         output_filename = <span class="string">&quot;infercnv22&quot;</span>, output_format = <span class="string">&quot;pdf&quot;</span>, </span><br><span class="line">         dynamic_resize = <span class="number">0</span>, ref_contig = <span class="literal">NULL</span>, write_expr_matrix = <span class="literal">FALSE</span>,</span><br><span class="line">         useRaster = <span class="literal">TRUE</span>)</span><br></pre></td></tr></table></figure></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="mailto:undefined">yilingzhang</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="https://hiiiyilingzhang.github.io/2021/05/10/scRNAseq-03-Doublet-velocyto-inferCNV/">https://hiiiyilingzhang.github.io/2021/05/10/scRNAseq-03-Doublet-velocyto-inferCNV/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/scRNA-seq/">scRNA-seq</a></div><div class="post_share"><div class="social-share" data-image="/img/cover/scRNA.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/social-share.js/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/social-share.js/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2021/05/18/Bioinformatics-Platform-Development/"><img class="prev-cover" src="/img/cover/shiny.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">Bioinformatics Platform Development</div></div></a></div><div class="next-post pull-right"><a href="/2021/05/08/Drawing%20Requirements%20for%20Scientific%20Research/"><img class="next-cover" src="/img/cover/drawing.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">Drawing Requirements for Scientific Research</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span> Related Articles</span></div><div class="relatedPosts-list"><div><a href="/2021/04/26/scRNAseq-01-Read-Data-and-Aggregation/" title="scRNAseq-01 Read Data and Aggregation"><img class="cover" src="/img/cover/scRNA.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2021-04-26</div><div class="title">scRNAseq-01 Read Data and Aggregation</div></div></a></div><div><a href="/2021/04/26/scRNAseq-02-Annotation/" title="scRNAseq-02 Annotation"><img class="cover" src="/img/cover/scRNA.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2021-04-26</div><div class="title">scRNAseq-02 Annotation</div></div></a></div><div><a href="/2021/05/18/scRNAseq-04-Cell-cell-communication/" title="scRNAseq-04 Cell-cell communication"><img class="cover" src="/img/cover/scRNA.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2021-05-18</div><div class="title">scRNAseq-04 Cell-cell communication</div></div></a></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="card-info-avatar is-center"><img class="avatar-img" src="/img/selfie.JPG" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/><div class="author-info__name">yilingzhang</div><div class="author-info__description">Every Day is My Day</div></div><div class="card-info-data"><div class="card-info-data-item is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">39</div></a></div><div class="card-info-data-item is-center"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">17</div></a></div><div class="card-info-data-item is-center"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">14</div></a></div></div><div class="card-info-social-icons is-center"><a class="social-icon" href="mailto:jocelynzhang1119@icloud.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a><a class="social-icon" href="https://blog.csdn.net/weixin_43893431" target="_blank" title="CSDN"><i class="fas fa-paper-plane"></i></a></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#DoubletFinder"><span class="toc-number">1.</span> <span class="toc-text">DoubletFinder</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#Doublets%E5%8F%8A%E5%85%B6%E5%BD%A2%E6%88%90%E7%9A%84%E5%8E%9F%E5%9B%A0"><span class="toc-number">1.1.</span> <span class="toc-text">Doublets及其形成的原因</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%88%86%E6%9E%90%E5%8E%9F%E7%90%86"><span class="toc-number">1.2.</span> <span class="toc-text">分析原理</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B"><span class="toc-number">1.3.</span> <span class="toc-text">分析流程</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%8F%82%E6%95%B0"><span class="toc-number">1.4.</span> <span class="toc-text">参数</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#DoubletFinder%E5%AE%89%E8%A3%85"><span class="toc-number">1.5.</span> <span class="toc-text">DoubletFinder安装</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%88%86%E6%9E%90%E4%BB%A3%E7%A0%81"><span class="toc-number">1.6.</span> <span class="toc-text">分析代码</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#RNA-Velocyto"><span class="toc-number">2.</span> <span class="toc-text">RNA Velocyto</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#velocyto-py"><span class="toc-number">2.1.</span> <span class="toc-text">velocyto.py</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%AE%89%E8%A3%85"><span class="toc-number">2.1.1.</span> <span class="toc-text">安装</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%91%BD%E4%BB%A4"><span class="toc-number">2.1.2.</span> <span class="toc-text">命令</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#velocyto-R"><span class="toc-number">2.2.</span> <span class="toc-text">velocyto.R</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#inferCNV"><span class="toc-number">3.</span> <span class="toc-text">inferCNV</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#inferCNV%E5%8E%9F%E7%90%86"><span class="toc-number">3.1.</span> <span class="toc-text">inferCNV原理</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#inferCNV%E5%AE%89%E8%A3%85"><span class="toc-number">3.2.</span> <span class="toc-text">inferCNV安装</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#inferCNV%E4%BD%BF%E7%94%A8"><span class="toc-number">3.3.</span> <span class="toc-text">inferCNV使用</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%87%86%E5%A4%87%E8%BE%93%E5%85%A5%E6%96%87%E4%BB%B6"><span class="toc-number">3.3.1.</span> <span class="toc-text">准备输入文件</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%BD%AF%E4%BB%B6%E8%BF%90%E8%A1%8C"><span class="toc-number">3.3.2.</span> <span class="toc-text">软件运行</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%BF%90%E8%A1%8C%E7%BB%93%E6%9E%9C"><span class="toc-number">3.3.3.</span> <span class="toc-text">运行结果</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%8F%90%E5%8F%96%E4%BF%A1%E6%81%AF"><span class="toc-number">3.3.4.</span> <span class="toc-text">提取信息</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%BB%98%E5%9B%BE"><span class="toc-number">3.3.5.</span> <span class="toc-text">绘图</span></a></li></ol></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2021/09/06/All-I-Learn-About-Circulating-Cell-free-DNA-cfDNA/" title="All I Learn About Circulating Cell-free DNA(cfDNA)"><img src="/img/cover/cfDNA.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="All I Learn About Circulating Cell-free DNA(cfDNA)"/></a><div class="content"><a class="title" href="/2021/09/06/All-I-Learn-About-Circulating-Cell-free-DNA-cfDNA/" title="All I Learn About Circulating Cell-free DNA(cfDNA)">All I Learn About Circulating Cell-free DNA(cfDNA)</a><time datetime="2021-09-06T08:31:28.000Z" title="Created 2021-09-06 16:31:28">2021-09-06</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/08/31/CITE-seq-theory-and-analysis/" title="CITE-seq: theory and analysis"><img src="/img/cover/CITE.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="CITE-seq: theory and analysis"/></a><div class="content"><a class="title" href="/2021/08/31/CITE-seq-theory-and-analysis/" title="CITE-seq: theory and analysis">CITE-seq: theory and analysis</a><time datetime="2021-08-31T13:09:19.000Z" title="Created 2021-08-31 21:09:19">2021-08-31</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/07/23/scRNA-seq-Project-LIHC-microenvironment/" title="scRNA-seq Project -- LIHC microenvironment"><img src="/img/cover/scRNA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scRNA-seq Project -- LIHC microenvironment"/></a><div class="content"><a class="title" href="/2021/07/23/scRNA-seq-Project-LIHC-microenvironment/" title="scRNA-seq Project -- LIHC microenvironment">scRNA-seq Project -- LIHC microenvironment</a><time datetime="2021-07-23T14:09:24.000Z" title="Created 2021-07-23 22:09:24">2021-07-23</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/07/18/Get-Start-With-Bioinfo-Pipeline/" title="Get Start With Bioinfo Pipeline"><img src="/img/cover/pipeline.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Get Start With Bioinfo Pipeline"/></a><div class="content"><a class="title" href="/2021/07/18/Get-Start-With-Bioinfo-Pipeline/" title="Get Start With Bioinfo Pipeline">Get Start With Bioinfo Pipeline</a><time datetime="2021-07-18T11:59:13.000Z" title="Created 2021-07-18 19:59:13">2021-07-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/24/Tumor-immune-infiltration-analysis/" title="Tumor immune infiltration analysis"><img src="/img/cover/Immue.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Tumor immune infiltration analysis"/></a><div class="content"><a class="title" href="/2021/05/24/Tumor-immune-infiltration-analysis/" title="Tumor immune infiltration analysis">Tumor immune infiltration analysis</a><time datetime="2021-05-24T10:51:35.000Z" title="Created 2021-05-24 18:51:35">2021-05-24</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2021 By yilingzhang</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"></div><script defer="defer" id="ribbon" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/canvas-ribbon.min.js" size="150" alpha="0.6" zIndex="-1" mobile="false" data-click="true"></script><script id="click-heart" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/click-heart.min.js" async="async" mobile="false"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>