<!DOCTYPE html><html lang="en | zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1"><title>scATAC-seq Analysis | Jocelyn's Blog</title><meta name="keywords" content="scATAC-seq"><meta name="author" content="yilingzhang"><meta name="copyright" content="yilingzhang"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="教程：http:&#x2F;&#x2F;xuzhougeng.top&#x2F;tags&#x2F;atac-seq ArchR install1234567891011conda create -n ArchRsource activate ArchRconda install -c conda-forge r-base ###R version 4.0.3if (!requireNamespace(&quot;devtools&amp;qu">
<meta property="og:type" content="article">
<meta property="og:title" content="scATAC-seq Analysis">
<meta property="og:url" content="http://yilingzhang.gitee.io/2021/04/08/scATACseq-Analysis/index.html">
<meta property="og:site_name" content="Jocelyn&#39;s Blog">
<meta property="og:description" content="教程：http:&#x2F;&#x2F;xuzhougeng.top&#x2F;tags&#x2F;atac-seq ArchR install1234567891011conda create -n ArchRsource activate ArchRconda install -c conda-forge r-base ###R version 4.0.3if (!requireNamespace(&quot;devtools&amp;qu">
<meta property="og:locale">
<meta property="og:image" content="http://yilingzhang.gitee.io/img/cover/scATAC.png">
<meta property="article:published_time" content="2021-04-08T11:01:25.000Z">
<meta property="article:modified_time" content="2021-04-26T08:47:20.963Z">
<meta property="article:author" content="yilingzhang">
<meta property="article:tag" content="scATAC-seq">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://yilingzhang.gitee.io/img/cover/scATAC.png"><link rel="shortcut icon" href="/img/love_32px.png"><link rel="canonical" href="http://yilingzhang.gitee.io/2021/04/08/scATACseq-Analysis/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    jQuery: 'https://cdn.jsdelivr.net/npm/jquery@latest/dist/jquery.min.js',
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/js/jquery.justifiedGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/css/justifiedGallery.min.css'
    },
    fancybox: {
      js: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.js',
      css: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isanchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = { 
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2021-04-26 16:47:20'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    })(window)</script><meta name="generator" content="Hexo 5.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="author-avatar"><img class="avatar-img" src="/img/selfie.JPG" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data"><div class="data-item is-center"><div class="data-item-link"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">28</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">14</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">13</div></a></div></div></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/img/cover/scATAC.png')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Jocelyn's Blog</a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">scATAC-seq Analysis</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2021-04-08T11:01:25.000Z" title="Created 2021-04-08 19:01:25">2021-04-08</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2021-04-26T08:47:20.963Z" title="Updated 2021-04-26 16:47:20">2021-04-26</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/Single-Cell/">Single Cell</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">Word count:</span><span class="word-count">5k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">Reading time:</span><span>19min</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><p>教程：<a target="_blank" rel="noopener" href="http://xuzhougeng.top/tags/atac-seq">http://xuzhougeng.top/tags/atac-seq</a></p>
<h1 id="ArchR-install"><a href="#ArchR-install" class="headerlink" title="ArchR install"></a>ArchR install</h1><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">conda create -n ArchR</span><br><span class="line">source activate ArchR</span><br><span class="line">conda install -<span class="built_in">c</span> conda-forge r-base <span class="comment">###R version 4.0.3</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> (!requireNamespace(<span class="string">&quot;devtools&quot;</span>, quietly = <span class="literal">TRUE</span>)) install.packages(<span class="string">&quot;devtools&quot;</span>)</span><br><span class="line"><span class="keyword">if</span> (!requireNamespace(<span class="string">&quot;BiocManager&quot;</span>, quietly = <span class="literal">TRUE</span>)) install.packages(<span class="string">&quot;BiocManager&quot;</span>)</span><br><span class="line">devtools::install_github(<span class="string">&quot;GreenleafLab/ArchR&quot;</span>, ref=<span class="string">&quot;master&quot;</span>, repos = BiocManager::repositories())</span><br><span class="line"></span><br><span class="line">library(ArchR)</span><br><span class="line">ArchR::installExtraPackages()</span><br><span class="line">addArchRGenome(<span class="string">&quot;hg19&quot;</span>)</span><br></pre></td></tr></table></figure>
<center><img src="/img/scATAC/workflow.jpg" alt="workflow" style="zoom:100%;" /></center>

<h1 id="ArchR的输入文件类型"><a href="#ArchR的输入文件类型" class="headerlink" title="ArchR的输入文件类型"></a>ArchR的输入文件类型</h1><p>ArchR主要以scATAC-seq原始数据经上游处理后的两种常见输出文件(BAM,fragment)作为输入。  </p>
<ol>
<li>Fragment文件记录着scATAC-seq的fragment以及对应的细胞ID，每一行都是一条记录，该文件需要是tabix(见注1)排序并建立索引保证能被高效读取。以10XGenomics的CellRanger软件为例，它的输出文件是fragments;    </li>
<li>BAM文件则是二进制格式下的tabix排序文件，记录着scATAC-seq的fragment、原始数据、细胞条形码和其他信息。具体使用何种文件作为输入取决于你的上游处理流程。sci-ATAC-seq流程则输出BAM文件。  </li>
</ol>
<h2 id="创建Arrow文件"><a href="#创建Arrow文件" class="headerlink" title="创建Arrow文件"></a>创建Arrow文件</h2><p>对于每个样本，此步骤将：  </p>
<ol>
<li>从提供的输入文件中读取可访问的片段。  </li>
<li>计算每个细胞的质量控制信息（即TSS富集得分和核小体信息，ArchR默认会过滤TSS富集得分低于4或唯一比对数小于1000（也就是保留TSS富集得分大于4且唯一比对数大于1000的细胞））。  </li>
<li>根据质量控制参数过滤单元。  </li>
<li>使用500 bp的条带创建全基因组的TileMatrix。  </li>
<li>使用<code>geneAnnotation</code>调用时定义的自定义创建GeneScoreMatrix <code>addArchRGenome()</code>。  </li>
</ol>
<p>创建Arrow文件会在当前目录下生成一个”QualityControl”目录，这里面包括2个和样本相关的图。第一个图展示log10(unique nuclear fragments) vs TSS enrichment score, 虚线表示过滤阈值。第二图注释fragment大小分布图。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">library(ArchR)</span><br><span class="line">addArchRGenome(<span class="string">&quot;hg19&quot;</span>) <span class="comment"># hg38, mm9, mm10</span></span><br><span class="line">addArchRThreads(threads = <span class="number">5</span>) <span class="comment">###You can modify this value;</span></span><br><span class="line"></span><br><span class="line">setwd(<span class="string">&quot;/home/data/vip8t02/MyCourse/NGS-02/scATACseq/&quot;</span>) <span class="comment">###Set up your own workspace;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">###输入文件</span></span><br><span class="line">inputFiles &lt;- getInputFiles(paths = <span class="string">&quot;./Data/&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">###生成arrowfiles</span></span><br><span class="line"><span class="comment"># filterFrags is no longer a valid input. Please use minFrags! 参数名已更改，但也可以运行</span></span><br><span class="line"></span><br><span class="line">ArrowFiles &lt;- createArrowFiles(</span><br><span class="line">  inputFiles = inputFiles,</span><br><span class="line">  sampleNames = <span class="built_in">names</span>(inputFiles),</span><br><span class="line">  filterTSS = <span class="number">4</span>, <span class="comment">#The minimum numeric transcription start site (TSS) enrichment score required for a cell to pass filtering for use in downstream analyses.</span></span><br><span class="line">  filterFrags = <span class="number">1000</span>, <span class="comment">#The minimum number of mapped ATAC-seq fragments required per cell to pass filtering for use in downstream analyses.</span></span><br><span class="line">  addTileMat = <span class="literal">TRUE</span>,</span><br><span class="line">  addGeneScoreMat = <span class="literal">TRUE</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 确认一下它确实是一个Arrow file路径的向量</span></span><br><span class="line">ArrowFiles</span><br></pre></td></tr></table></figure>
<ul>
<li>细胞质控<br>scATAC-seq数据的严格质量控制（QC）对于消除低质量细胞的作用至关重要。在ArchR中，我们考虑数据的三个特征：  </li>
</ul>
<ol>
<li>独特核片段的数量（即不映射到线粒体DNA）。  </li>
<li>信噪比。低的信噪比通常归因于死亡或垂死的细胞，这些细胞具有脱色的DNA，可以在全基因组范围内进行随机转座。  </li>
<li>片段大小分布。由于核小体的周期性，我们预计会看到碎片消失，这些碎片是包裹在核小体周围的DNA长度（大约147 bp）。  </li>
</ol>
<h1 id="使用ArchR推断Doublet"><a href="#使用ArchR推断Doublet" class="headerlink" title="使用ArchR推断Doublet"></a>使用ArchR推断Doublet</h1><p>单细胞数据分析中的一个重要问题就是”doublet”对分析的影响。”doublet”指的是单个液滴(droplet)捕获了一个条形码珠(barcode bead)和多个细胞核。这会导致原本来自于多个细胞的read结果被当成一个细胞，结果原来两个细胞被平均成一个细胞。  </p>
<p>我们使用ArchR的<code>addDoubleScores()</code>函数来移除doublet。对于教程中使用的数据，每个样本大约需要2到5分钟时间进行处理，最后每个细胞的doublet得分会添加到Arrow文件中。 。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">doubScores &lt;- addDoubletScores(</span><br><span class="line">  input = ArrowFiles,</span><br><span class="line">  k = <span class="number">10</span>, <span class="comment">#Refers to how many cells near a &quot;pseudo-doublet&quot; to count.The number of cells neighboring a simulated doublet to be considered as putative doublets.</span></span><br><span class="line">  knnMethod = <span class="string">&quot;UMAP&quot;</span>, <span class="comment">#Refers to the embedding to use for nearest neighbor search with doublet projection.The name of the dimensionality reduction method to be used for k-nearest neighbors calculation. Possible values are &quot;UMAP&quot; or &quot;LSI&quot;.</span></span><br><span class="line">  LSIMethod = <span class="number">1</span> <span class="comment">#The list of parameters to pass to the IterativeLSI() function. See IterativeLSI()</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<p>addDoubletScores在计算doublet得分的时候，还会在”QualityControl”中为每个样本生成三张图（一个样本一个PDF）  </p>
<ol>
<li>Doublet Enrichments: 假设doublet符合均匀分布，那么每个细胞附近的模拟doublet数目都差不多。如果一个细胞附近相对于其他细胞有更多的doublet, 就认为它富集了doublet  </li>
<li>Doublet Score: 基于均匀分布假设，计算doublet富集显著性，以-log10(binomial adjusted p-value)进行展示。我们更推荐根据doublet enrichment鉴定doublet。  </li>
<li>Doublet Density: 模拟的duoblet在二维空间的投影，我们可以直观的了解我们模拟的doublet的分布情况。</li>
</ol>
<h1 id="创建ArchRProject"><a href="#创建ArchRProject" class="headerlink" title="创建ArchRProject"></a>创建ArchRProject</h1><h2 id="创建一个ArchRProject"><a href="#创建一个ArchRProject" class="headerlink" title="创建一个ArchRProject"></a>创建一个ArchRProject</h2><p>ArchRProject对象允许我们将多个Arrow文件整理到单个项目之中。ArchRProject比较小能够直接存放在内存中。通过操作ArchRProject，我们能够快速读取Arrow文件里的数据，并将修改后的数据写回到Arrow文件里。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">projPBMC_1 &lt;- ArchRProject(</span><br><span class="line">  ArrowFiles = ArrowFiles,</span><br><span class="line">  outputDirectory = <span class="string">&quot;PBMC&quot;</span>,</span><br><span class="line">  copyArrows = <span class="literal">TRUE</span> <span class="comment">#This is recommened so that if you modify the Arrow files you have an original copy for later usage.</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">projPBMC_1</span><br></pre></td></tr></table></figure>
<blockquote>
<p>outputDirectory: 输出目录<br>samples： 样本名<br>sampleColData: 记录每个样本的元信息<br>cellColData: 记录每个细胞的元信息，第二章计算的”DoubleEnrichment”,”DoubletScore”就存放在此处<br>numberOfCells: 项目总的合格细胞数，也就是不包括之前没有通过质控，或者被认为doublet的细胞<br>medianTSS/medianFrags: 所有细胞的TSS值和Fragments数的中位数  </p>
</blockquote>
<h2 id="操作ArchRProject对象"><a href="#操作ArchRProject对象" class="headerlink" title="操作ArchRProject对象"></a>操作ArchRProject对象</h2><h3 id="检查矩阵数据"><a href="#检查矩阵数据" class="headerlink" title="检查矩阵数据"></a>检查矩阵数据</h3><p>检查当前的ArchRProject中存放了哪些矩阵数据，这些数据可以在下游分析中使用。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">getAvailableMatrices(projPBMC_1)</span><br><span class="line"><span class="comment"># &quot;GeneScoreMatrix&quot; &quot;TileMatrix&quot;</span></span><br></pre></td></tr></table></figure>
<h3 id="绘制质控信息和TSS富集得分的对比图"><a href="#绘制质控信息和TSS富集得分的对比图" class="headerlink" title="绘制质控信息和TSS富集得分的对比图"></a>绘制质控信息和TSS富集得分的对比图</h3><p>该图的主要作用是找到高质量的细胞。你可能会注意到一些细胞已经在我们之前创建Arrow文件时被设置的阈值filterTSS, filterFargs过滤掉了。然而，如果我们觉得之前的质控过滤对于这个样本不太合适，我们可以根据该图调整阈值，有必要的话还可以重新产生Arrow文件。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## 先用getCellColData提取我们需要的两列，其中nFrages需要进行log10运算</span></span><br><span class="line">df &lt;- getCellColData(projPBMC_1, select = <span class="built_in">c</span>(<span class="string">&quot;log10(nFrags)&quot;</span>, <span class="string">&quot;TSSEnrichment&quot;</span>))</span><br><span class="line"></span><br><span class="line"><span class="comment">## 基于TSS富集得分和唯一比对进行绘图，函数ggPoint是ArchR对ggplot2的点图相关函数的封装。</span></span><br><span class="line">pdf(<span class="string">&quot;TSS-vs-Frags.pdf&quot;</span>)</span><br><span class="line">p &lt;- ggPoint(</span><br><span class="line">  x = df[,<span class="number">1</span>],</span><br><span class="line">  y = df[,<span class="number">2</span>],</span><br><span class="line">  colorDensity = <span class="literal">TRUE</span>,</span><br><span class="line">  continuousSet = <span class="string">&quot;sambaNight&quot;</span>,</span><br><span class="line">  xlabel = <span class="string">&quot;Log10 Unique Fragments&quot;</span>,</span><br><span class="line">  ylabel = <span class="string">&quot;TSS Enrichment&quot;</span>,</span><br><span class="line">  xlim = <span class="built_in">c</span>(log10(<span class="number">500</span>), quantile(df[,<span class="number">1</span>], probs = <span class="number">0.99</span>)),</span><br><span class="line">  ylim = <span class="built_in">c</span>(<span class="number">0</span>, quantile(df[,<span class="number">2</span>], probs = <span class="number">0.99</span>))</span><br><span class="line">) + geom_hline(yintercept = <span class="number">4</span>, lty = <span class="string">&quot;dashed&quot;</span>) + geom_vline(xintercept = <span class="number">3</span>, lty = <span class="string">&quot;dashed&quot;</span>)</span><br><span class="line">p</span><br><span class="line"></span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<h3 id="为ArchRProject中的样本统计绘图"><a href="#为ArchRProject中的样本统计绘图" class="headerlink" title="为ArchRProject中的样本统计绘图"></a>为ArchRProject中的样本统计绘图</h3><p>如果一个项目中有多个不同的样本，我们很自然地就会想比较这些样本。ArchR提供了两种小提琴图(violin plot)和山脊图(ridge plot)用来展示不同组之间的信息。这两种类型的图可以用一个函数plotGroups进行绘制。</p>
<p><strong>山脊图</strong><br>根据TSS富集得分为每个样本绘制</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">pdf(<span class="string">&quot;ridge plot.pdf&quot;</span>)</span><br><span class="line">p1 &lt;- plotGroups(</span><br><span class="line">  ArchRProj = projPBMC_1,</span><br><span class="line">  groupBy = <span class="string">&quot;Sample&quot;</span>,</span><br><span class="line">  colorBy = <span class="string">&quot;cellColData&quot;</span>,</span><br><span class="line">  name = <span class="string">&quot;TSSEnrichment&quot;</span>,</span><br><span class="line">  plotAs = <span class="string">&quot;ridges&quot;</span></span><br><span class="line">)</span><br><span class="line">p1</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<p><strong>小提琴图</strong><br>根据TSS富集得分为每个样本绘制<br>ArchR绘制的小提琴图额外叠加了一层箱线图(a box-and-whiskers plot in the style of Tukey)，中间水平的三条线表示数据中的1/4分位数，中位数和3/4分位数。最下方是最小值，最上方是最大值（或者是1.5倍的IQR, interquartile range, 1/4分位数与3/4分位数的距离）</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">pdf(<span class="string">&quot;violin plot.pdf&quot;</span>)</span><br><span class="line">p2 &lt;- plotGroups(</span><br><span class="line">  ArchRProj = projPBMC_1,</span><br><span class="line">  groupBy = <span class="string">&quot;Sample&quot;</span>,</span><br><span class="line">  colorBy = <span class="string">&quot;cellColData&quot;</span>,</span><br><span class="line">  name = <span class="string">&quot;TSSEnrichment&quot;</span>,</span><br><span class="line">  plotAs = <span class="string">&quot;violin&quot;</span>,</span><br><span class="line">  alpha = <span class="number">0.4</span>,</span><br><span class="line">  addBoxPlot = <span class="literal">TRUE</span></span><br><span class="line">)</span><br><span class="line">p2</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>

<h3 id="绘制样本的TSS富集谱和Fragment大小分布"><a href="#绘制样本的TSS富集谱和Fragment大小分布" class="headerlink" title="绘制样本的TSS富集谱和Fragment大小分布"></a>绘制样本的TSS富集谱和Fragment大小分布</h3><p>**Fragments大小分布: **<br>我们使用<code>plotFragmentSizes</code>函数为绘制所有样本的fragment大小分布。ATAC-seq数据中的fragment大小分布可能会因样本、细胞类型和批次不同，但这和数据质量并不是严格相关。比如说我们的数据在不同组之间就存在一些差别。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">pdf(<span class="string">&quot;Fragment size.pdf&quot;</span>)</span><br><span class="line">fs &lt;- plotFragmentSizes(ArchRProj = projPBMC_1)</span><br><span class="line">fs</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<p><strong>TSS富集谱:</strong><br>我们使用<code>plotTSSEnrichment</code>函数绘制TSS富集谱。TSS富集谱需要在中心位置有一个明显的峰，在峰的右边还会有一个核小体引起的小隆起（约147 bp）。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">pdf(<span class="string">&quot;TSS enrichment.pdf&quot;</span>)</span><br><span class="line">TSSenrichment &lt;- plotTSSEnrichment(ArchRProj = projPBMC_1)</span><br><span class="line">TSSenrichment</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<h2 id="过滤doublets"><a href="#过滤doublets" class="headerlink" title="过滤doublets"></a>过滤doublets</h2><p>这个函数里的一个关键参数是filterRatio,由于不同样本的制备过程中的细胞上样量不同，那么不同样本的doublet也就不同，为了保证不同样本的过滤后结果保持一致，所以引入了filterRatio。该值越高，被认为是doublet而被过滤的细胞也就越多。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">projPBMC_2 &lt;- filterDoublets(projPBMC_1)</span><br><span class="line"></span><br><span class="line">projPBMC_2</span><br></pre></td></tr></table></figure>
<h1 id="ArchR降维分析"><a href="#ArchR降维分析" class="headerlink" title="ArchR降维分析"></a>ArchR降维分析</h1><p>在scATAC-seq数据中，一个特定的位置存在三种状态，分为不开放，一条链开放，两条链都开放。但即便是在高质量的scATAC-seq数据中，大部分开放区域都没有被转座酶切割，因此大部分座位都是0。此外，如果我们发现一个细胞(A)的一处peak区域有三个Tn5 insertions，而另外一个细胞(B)对应区域只有一个Tn5 insertions，由于数据的稀疏性，我们也无法断定A就比B更开放。</p>
<p>如果直接在稀疏矩阵上使用标准的降维分析，例如主成分分析(Principal Component Analysis, PCA)，然后使用前两个主成分进行绘图，你可能无法得到你想要的结果，这是由于数据的稀疏性会导致细胞间在0的位置有更高的相似性。为了解决这个问题，我们采用了<strong>分层</strong>的降维策略。</p>
<h2 id="隐语义-Latent-Semantic-Indexing-迭代"><a href="#隐语义-Latent-Semantic-Indexing-迭代" class="headerlink" title="隐语义(Latent Semantic Indexing)迭代"></a>隐语义(Latent Semantic Indexing)迭代</h2><p>首先，我们会使用隐语义分析(Latent Semantic Indexing, LSI)进行降维。LSI方法最早用在自然语言分析中，根据字数来评估文档相似度。Cusanovich et al. (Science 2015)第一次在scATAC-sq数分析中引入了LSI。计算步骤如下:  </p>
<ol>
<li>对每个细胞计算根据深度标准化后的词频(Term frequency, TF)  </li>
<li>根据逆文档频率(IDF, inverse document frequency)进行标准化，便于后续特征选择  </li>
<li>最后对结果矩阵进行对数变换（也就是log(TF-IDF)）  </li>
</ol>
<p>后来提出了”迭代LSI法” (Satpathy, Granja et al. Nature Biotechnology 2019和 Granja, Klemm and McGinnis* et al. Nature Biotechnology 2019)。这个方法首先在最开放的分块中计算初始的LSI变换，识别未受到批次效应影响的低分辨率聚类。</p>
<center><img src="/img/scATAC/LSI.jpg" alt="HC" style="zoom:100%;" /></center>
我们可以在PBMC中鉴定出几个主要大群(T细胞、B细胞和单核细胞)。接着，ArchR计算这些聚类在每个特征中的平均开放度。ArchR识别这些聚类中变化最大的peak，用作后续LSI的特征。在第二次迭代中，变化最大的peak就类似于scRAN-seq中高变异基因。用户可以决定进行多次轮的LSI迭代。我们发现这个方法能够降低观察到的批次效应，并能使用更加合理的特征矩阵进行降维。


<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">projPBMC_2 &lt;- addIterativeLSI(</span><br><span class="line">  ArchRProj = projPBMC_2,</span><br><span class="line">  useMatrix = <span class="string">&quot;TileMatrix&quot;</span>, </span><br><span class="line">  name = <span class="string">&quot;IterativeLSI&quot;</span>, </span><br><span class="line">  iterations = <span class="number">2</span>, </span><br><span class="line">  clusterParams = <span class="built_in">list</span>( </span><br><span class="line">    resolution = <span class="built_in">c</span>(<span class="number">0.2</span>), </span><br><span class="line">    sampleCells = <span class="number">10000</span>, </span><br><span class="line">    n.start = <span class="number">10</span></span><br><span class="line">  ), </span><br><span class="line">  varFeatures = <span class="number">25000</span>, </span><br><span class="line">  dimsToUse = <span class="number">1</span>:<span class="number">30</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<h2 id="Harmony矫正批次效应"><a href="#Harmony矫正批次效应" class="headerlink" title="Harmony矫正批次效应"></a>Harmony矫正批次效应</h2><p>在某些情况下，迭代的LSI方法并不能矫正严重的批次效应。ArchR使用了最初为了scRNA-seq开发的Harmony进行批次效应矫正。 ArchR提供了专门的封装函数<code>addHarmony()</code>将降维后的对象直接传递给<code>Harmony::HarmonyMatrix()</code>, 额外参数可以通过…进行传递。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">projPBMC_2 &lt;- addHarmony(</span><br><span class="line">  ArchRProj = projPBMC_2,</span><br><span class="line">  reducedDims = <span class="string">&quot;IterativeLSI&quot;</span>,</span><br><span class="line">  name = <span class="string">&quot;Harmony&quot;</span>,</span><br><span class="line">  groupBy = <span class="string">&quot;Sample&quot;</span></span><br><span class="line">)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>这一步会在我们的projHeme2创建一个新的名为Harmony的reduceDims对象。</p>
<h1 id="使用ArchR聚类"><a href="#使用ArchR聚类" class="headerlink" title="使用ArchR聚类"></a>使用ArchR聚类</h1><h2 id="使用Seurat的FindClusters-函数进行聚类"><a href="#使用Seurat的FindClusters-函数进行聚类" class="headerlink" title="使用Seurat的FindClusters()函数进行聚类"></a>使用Seurat的FindClusters()函数进行聚类</h2><p>我们发现Seurat实现的图聚类方法表现最好，所以在ArchR中，addClusters()函数本质是上将额外的参数通过…传递给<code>Seurat::FindClusters()</code>函数，从而得到聚类结果。在分析中，我们发现<code>Seurat::FindClusters()</code>是一个确定性的聚类算法，也就是相同的输入总是能得到完全相同的输出。  </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">projPBMC_2 &lt;- addClusters(</span><br><span class="line">  input = projPBMC_2,</span><br><span class="line">  reducedDims = <span class="string">&quot;IterativeLSI&quot;</span>,</span><br><span class="line">  method = <span class="string">&quot;Seurat&quot;</span>,</span><br><span class="line">  name = <span class="string">&quot;Clusters&quot;</span>,</span><br><span class="line">  resolution = <span class="number">0.8</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### 使用$符号来获取聚类信息，输出每个细胞对应的cluster</span></span><br><span class="line">head(projPBMC_2$Clusters)</span><br><span class="line"><span class="comment">###统计每个cluster的细胞数</span></span><br><span class="line">table(projPBMC_2$Clusters)</span><br></pre></td></tr></table></figure>
<p>为了更好了解样本在cluster的分布，我们可以使用<code>confusionMatrix()</code>函数通过每个样本创建一个聚类混合矩阵(cluster confusion matrix)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">###统计每个样本在不同的cluster的分布情况</span></span><br><span class="line">cM &lt;- confusionMatrix(paste0(projPBMC_2$Clusters), paste0(projPBMC_2$Sample))</span><br><span class="line">cM</span><br></pre></td></tr></table></figure>
<p>利用热图进行展示</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">library(pheatmap)</span><br><span class="line">cM &lt;- cM / Matrix::rowSums(cM)</span><br><span class="line">pdf(<span class="string">&quot;Cluster heatmap.pdf&quot;</span>)</span><br><span class="line">heatmap1 &lt;- pheatmap::pheatmap(</span><br><span class="line">  mat = as.matrix(cM), </span><br><span class="line">  color = paletteContinuous(<span class="string">&quot;whiteBlue&quot;</span>), </span><br><span class="line">  border_color = <span class="string">&quot;black&quot;</span></span><br><span class="line">)</span><br><span class="line">heatmap1</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<h2 id="使用UMAP聚类"><a href="#使用UMAP聚类" class="headerlink" title="使用UMAP聚类"></a>使用UMAP聚类</h2><p>UMAP和t-SNE的主要区别在于对细胞和聚类间的距离解释，t-SNE用于保留数据的局部结构，而UMAP则是保留局部结构的同时尽可能保留全局结构。从理论上来讲，UMAP中细胞聚类间的距离存在意义，而t-SNE则没有。也就是说，你不能说t-SNE中聚类A比聚类C更接近聚类B，而UMAP在设计的时候允许这种类型的比较。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">###Uniform Manifold Approximation and Projection (UMAP)</span></span><br><span class="line">projPBMC_2 &lt;- addUMAP(</span><br><span class="line">  ArchRProj = projPBMC_2, </span><br><span class="line">  reducedDims = <span class="string">&quot;IterativeLSI&quot;</span>, </span><br><span class="line">  name = <span class="string">&quot;UMAP&quot;</span>, </span><br><span class="line">  nNeighbors = <span class="number">30</span>, </span><br><span class="line">  minDist = <span class="number">0.5</span>, </span><br><span class="line">  metric = <span class="string">&quot;cosine&quot;</span> <span class="comment">###A number that determines how distance is computed in the reducedDims to compute a UMAP. This argument is passed to metric in uwot::umap().</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<p>使用@操作符可以从ArchRProject中列出有哪些可用的embedding，如<code>projPBMC_2@embeddings</code>  </p>
<p>我们使用<code>plotEmbedding</code>函数绘制UMAP图，设置embedding=”UMAP”。通过修改colorBy和name来告诉ArchR使用给定哪个元信息矩阵的列对细胞进行上色。  </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">###绘制UMAP图</span></span><br><span class="line"><span class="comment">###根据sample上色</span></span><br><span class="line">umap1 &lt;- plotEmbedding(ArchRProj = projPBMC_2, colorBy = <span class="string">&quot;cellColData&quot;</span>, name = <span class="string">&quot;Sample&quot;</span>, embedding = <span class="string">&quot;UMAP&quot;</span>)</span><br><span class="line">umap1</span><br><span class="line"></span><br><span class="line"><span class="comment">###根据Cluster上色</span></span><br><span class="line">umap2 &lt;- plotEmbedding(ArchRProj = projPBMC_2, colorBy = <span class="string">&quot;cellColData&quot;</span>, name = <span class="string">&quot;Clusters&quot;</span>, embedding = <span class="string">&quot;UMAP&quot;</span>)</span><br><span class="line">umap2</span><br><span class="line"></span><br><span class="line"><span class="comment">### 使用ggAlignPlots将这两个图共同展示，&quot;type=h&quot;表示两个图是水平排列</span></span><br><span class="line">ggAlignPlots(umap1, umap2, type = <span class="string">&quot;h&quot;</span>)</span><br></pre></td></tr></table></figure>
<h2 id="t-Stocastic-Neighbor-Embedding-t-SNE"><a href="#t-Stocastic-Neighbor-Embedding-t-SNE" class="headerlink" title="t-Stocastic Neighbor Embedding (t-SNE)"></a>t-Stocastic Neighbor Embedding (t-SNE)</h2><p>t-SNE图可以用<code>addTSNE()</code>函数运行得到</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">projPBMC_2 &lt;- addTSNE(</span><br><span class="line">    ArchRProj = projPBMC_2, </span><br><span class="line">    reducedDims = <span class="string">&quot;IterativeLSI&quot;</span>, </span><br><span class="line">    name = <span class="string">&quot;TSNE&quot;</span>, </span><br><span class="line">    perplexity = <span class="number">30</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<p>和之前UMAP一样，我们同样使用<code>plotEmbedding()</code>绘制t-SNE图。这里不需要考虑嵌入的类型，可以继续使用之前的colorBy和name参数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">tsne1 &lt;- plotEmbedding(ArchRProj = projPBMC_2, </span><br><span class="line">                       colorBy = <span class="string">&quot;cellColData&quot;</span>, </span><br><span class="line">                       name = <span class="string">&quot;Sample&quot;</span>, </span><br><span class="line">                       embedding = <span class="string">&quot;TSNE&quot;</span>)</span><br><span class="line"></span><br><span class="line">tsne2 &lt;- plotEmbedding(ArchRProj = projPBMC_2, </span><br><span class="line">                       colorBy = <span class="string">&quot;cellColData&quot;</span>, </span><br><span class="line">                       name = <span class="string">&quot;Clusters&quot;</span>, </span><br><span class="line">                       embedding = <span class="string">&quot;TSNE&quot;</span>)</span><br><span class="line">ggAlignPlots(tsne1, tsne2, type = <span class="string">&quot;h&quot;</span>)</span><br></pre></td></tr></table></figure>
<h2 id="Harmony后降维"><a href="#Harmony后降维" class="headerlink" title="Harmony后降维"></a>Harmony后降维</h2><p>之前我们通过<code>addHarmony</code>调用Harmony对数据进行批次校正，创建了一个名为”Harmony”的reducedDims对象。我们通过UMAP或t-SNE对结果进行嵌入可视化，对迭代LSI结果和Harmony校正结果进行比较，评估Harmony的作用。  </p>
<p>保持和之前UMAP嵌入一样的参数，只修改<code>reducedDims=&quot;Harmony&quot;</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">projPBMC_2 &lt;- addUMAP(</span><br><span class="line">  ArchRProj = projPBMC_2, </span><br><span class="line">  reducedDims = <span class="string">&quot;Harmony&quot;</span>, </span><br><span class="line">  name = <span class="string">&quot;UMAPHarmony&quot;</span>, </span><br><span class="line">  nNeighbors = <span class="number">30</span>, </span><br><span class="line">  minDist = <span class="number">0.5</span>, </span><br><span class="line">  metric = <span class="string">&quot;cosine&quot;</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">pdf(<span class="string">&quot;UMAP Harmony.pdf&quot;</span>)</span><br><span class="line">ha3 &lt;- plotEmbedding(ArchRProj = projPBMC_2, colorBy = <span class="string">&quot;cellColData&quot;</span>, name = <span class="string">&quot;Sample&quot;</span>, embedding = <span class="string">&quot;UMAPHarmony&quot;</span>)</span><br><span class="line">ha4 &lt;- plotEmbedding(ArchRProj = projPBMC_2, colorBy = <span class="string">&quot;cellColData&quot;</span>, name = <span class="string">&quot;Clusters&quot;</span>, embedding = <span class="string">&quot;UMAPHarmony&quot;</span>)</span><br><span class="line">ggAlignPlots(ha3, ha4, type = <span class="string">&quot;h&quot;</span>)</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<h1 id="ArchR的基因得分和标记基因"><a href="#ArchR的基因得分和标记基因" class="headerlink" title="ArchR的基因得分和标记基因"></a>ArchR的基因得分和标记基因</h1><p>为了实现细胞类型注释，我们会用到已知的细胞类型特异的标记基因。在scATAC-seq数据中，这些基因的表达量其实是根据染色质开放数据估计的基因得分(gene score)。所谓的基因得分，本质上就是用基因附近的调控元件去预测基因的表达量。ArchR的亮点在于，它允许用户提供复杂的自定义的距离加权开放性模型去计算这些基因得分。  </p>
<h2 id="鉴定标记基因"><a href="#鉴定标记基因" class="headerlink" title="鉴定标记基因"></a>鉴定标记基因</h2><p>为了根据基因得分来鉴定标记基因，我们需要在调用<code>getMarkerFeatures()</code>时设置<code>useMatrix = &quot;GeneScoreMatrix&quot;</code>。此外，我们设置了<code>groupBy=&quot;Clusters&quot;</code>告诉ArchR使用cellColData的”Clusters”列对细胞分组，从而得到每个聚类特异的标记基因。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">devtools::install_github(<span class="string">&quot;immunogenomics/presto&quot;</span>)</span><br><span class="line"></span><br><span class="line">markersGS &lt;- getMarkerFeatures(</span><br><span class="line">  ArchRProj = projPBMC_2, </span><br><span class="line">  useMatrix = <span class="string">&quot;GeneScoreMatrix&quot;</span>, </span><br><span class="line">  groupBy = <span class="string">&quot;Clusters&quot;</span>,</span><br><span class="line">  bias = <span class="built_in">c</span>(<span class="string">&quot;TSSEnrichment&quot;</span>, <span class="string">&quot;log10(nFrags)&quot;</span>),</span><br><span class="line">  testMethod = <span class="string">&quot;wilcoxon&quot;</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<p>该函数返回一个<code>SummarizedExperiment</code>对象，里面记录着每个标记特征的相关信息。这是ArchR的一个常见输出，是下游数据分析的关键输出格式。<code>SummarizedExperiment</code>对象和矩阵类似，行是感兴趣的特征（例如基因），列表示样本。一个SummarizedExperiment能够记录一个或多个assay矩阵。</p>
<p>我们能够使用<code>getMarkers()</code>函数从<code>SummarizedExperiment</code>对象中提取出一个包含多个DataFrame对象的列表，每个DataFrame对象都对应着一个聚类，记录着该聚类里相关的标记特征。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">markerList &lt;- getMarkers(markersGS, cutOff = <span class="string">&quot;FDR &lt;= 0.01 &amp; Log2FC &gt;= 1.25&quot;</span>) <span class="comment">###cutoff值，可根据需要修改；</span></span><br><span class="line">markerList$C2</span><br></pre></td></tr></table></figure>
<p>我们使用<code>markerHeatmap()</code>创建含有所有标记特征的可视化热图。通过<code>labelMarkers()</code>在热图中标记部分感兴趣的基因。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  &lt;- <span class="built_in">c</span>(</span><br><span class="line">    <span class="string">&quot;CD34&quot;</span>, <span class="comment">#Early Progenitor</span></span><br><span class="line">    <span class="string">&quot;GATA1&quot;</span>, <span class="comment">#Erythroid</span></span><br><span class="line">    <span class="string">&quot;PAX5&quot;</span>, <span class="string">&quot;MS4A1&quot;</span>, <span class="string">&quot;EBF1&quot;</span>, <span class="string">&quot;MME&quot;</span>, <span class="comment">#B-Cell Trajectory</span></span><br><span class="line">    <span class="string">&quot;CD14&quot;</span>, <span class="string">&quot;CEBPB&quot;</span>, <span class="string">&quot;MPO&quot;</span>, <span class="comment">#Monocytes</span></span><br><span class="line">    <span class="string">&quot;IRF8&quot;</span>, </span><br><span class="line">    <span class="string">&quot;CD3D&quot;</span>, <span class="string">&quot;CD8A&quot;</span>, <span class="string">&quot;TBX21&quot;</span>, <span class="string">&quot;IL7R&quot;</span> <span class="comment">#TCells</span></span><br><span class="line">  )</span><br><span class="line"></span><br><span class="line">heatmapGS &lt;- markerHeatmap(</span><br><span class="line">  seMarker = markersGS, </span><br><span class="line">  cutOff = <span class="string">&quot;FDR &lt;= 0.01 &amp; Log2FC &gt;= 1.25&quot;</span>, </span><br><span class="line">  labelMarkers = markerGenes,</span><br><span class="line">  transpose = <span class="literal">TRUE</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<p>我们需要用<code>ComplexHeatmap::draw()</code>才能绘制该热图，这是因为heatmapGS实际上是一个热图列表。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ComplexHeatmap::draw(heatmapGS, heatmap_legend_side = <span class="string">&quot;bot&quot;</span>, annotation_legend_side = <span class="string">&quot;bot&quot;</span>)</span><br></pre></td></tr></table></figure>
<h2 id="可视化标记基因"><a href="#可视化标记基因" class="headerlink" title="可视化标记基因"></a>可视化标记基因</h2><p>在之前章节中提到，我们可以在UMAP展示每个细胞的基因得分。这可以通过修改<code>plotEmbedding()</code>函数里的colorBy和name参数实现</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  &lt;- <span class="built_in">c</span>(</span><br><span class="line">  <span class="string">&quot;CD34&quot;</span>,  </span><br><span class="line">  <span class="string">&quot;GATA1&quot;</span>, </span><br><span class="line">  <span class="string">&quot;PAX5&quot;</span>, <span class="string">&quot;MS4A1&quot;</span>, <span class="string">&quot;MME&quot;</span>, </span><br><span class="line">  <span class="string">&quot;CD14&quot;</span>, <span class="string">&quot;MPO&quot;</span>, </span><br><span class="line">  <span class="string">&quot;CD3D&quot;</span>, <span class="string">&quot;CD8A&quot;</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">p &lt;- plotEmbedding(</span><br><span class="line">  ArchRProj = projPBMC_2, </span><br><span class="line">  colorBy = <span class="string">&quot;GeneScoreMatrix&quot;</span>, </span><br><span class="line">  name = markerGenes, </span><br><span class="line">  embedding = <span class="string">&quot;UMAP&quot;</span>,</span><br><span class="line">  quantCut = <span class="built_in">c</span>(<span class="number">0.01</span>, <span class="number">0.95</span>),</span><br><span class="line">  imputeWeights = <span class="literal">NULL</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 可替换，markerGenes中任意基因均可</span></span><br><span class="line">p$CD14</span><br><span class="line"></span><br><span class="line"><span class="comment">## 绘制全部基因</span></span><br><span class="line">allgenes &lt;- lapply(p, <span class="keyword">function</span>(x)&#123;</span><br><span class="line">  x + guides(color = <span class="literal">FALSE</span>, fill = <span class="literal">FALSE</span>) + </span><br><span class="line">    theme_ArchR(baseSize = <span class="number">6.5</span>) +</span><br><span class="line">    theme(plot.margin = unit(<span class="built_in">c</span>(<span class="number">0</span>, <span class="number">0</span>, <span class="number">0</span>, <span class="number">0</span>), <span class="string">&quot;cm&quot;</span>)) +</span><br><span class="line">    theme(</span><br><span class="line">      axis.text.x=element_blank(), </span><br><span class="line">      axis.ticks.x=element_blank(), </span><br><span class="line">      axis.text.y=element_blank(), </span><br><span class="line">      axis.ticks.y=element_blank()</span><br><span class="line">    )</span><br><span class="line">&#125;)</span><br><span class="line">pdf(<span class="string">&quot;all genes.pdf&quot;</span>)</span><br><span class="line">do.call(cowplot::plot_grid, <span class="built_in">c</span>(<span class="built_in">list</span>(ncol = <span class="number">3</span>),allgenes))</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<h2 id="使用MAGIC填充标记基因"><a href="#使用MAGIC填充标记基因" class="headerlink" title="使用MAGIC填充标记基因"></a>使用MAGIC填充标记基因</h2><p>使用MAGIC根据邻近细胞填充基因得分对信号进行平滑化处理。我们发现者能够极大程度地提高基因得分的可视化解读性。要执行此操作，我们需要先在我们的ArchRProject中加入填充权重。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">projPBMC_2 &lt;- addImputeWeights(projPBMC_2)</span><br><span class="line"><span class="comment">## 这些填充权重会在之后绘制UMAP嵌入图里的基因得分时传入到plotEmbedding()</span></span><br><span class="line"></span><br><span class="line">markerGenes  &lt;- <span class="built_in">c</span>(</span><br><span class="line">    <span class="string">&quot;CD34&quot;</span>,  <span class="comment">#Early Progenitor</span></span><br><span class="line">    <span class="string">&quot;GATA1&quot;</span>, <span class="comment">#Erythroid</span></span><br><span class="line">    <span class="string">&quot;PAX5&quot;</span>, <span class="string">&quot;MS4A1&quot;</span>, <span class="string">&quot;MME&quot;</span>, <span class="comment">#B-Cell Trajectory</span></span><br><span class="line">    <span class="string">&quot;CD14&quot;</span>, <span class="string">&quot;MPO&quot;</span>, <span class="comment">#Monocytes</span></span><br><span class="line">    <span class="string">&quot;CD3D&quot;</span>, <span class="string">&quot;CD8A&quot;</span><span class="comment">#TCells</span></span><br><span class="line">  )</span><br><span class="line"></span><br><span class="line">wplot &lt;- plotEmbedding(</span><br><span class="line">    ArchRProj = projPBMC_2, </span><br><span class="line">    colorBy = <span class="string">&quot;GeneScoreMatrix&quot;</span>, </span><br><span class="line">    name = markerGenes, </span><br><span class="line">    embedding = <span class="string">&quot;UMAP&quot;</span>,</span><br><span class="line">    imputeWeights = getImputeWeights(projPBMC_2)</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">wplot$CD34</span><br><span class="line"></span><br><span class="line"><span class="comment">## Rearrange for grid plotting</span></span><br><span class="line">wplotall &lt;- lapply(wplot, <span class="keyword">function</span>(x)&#123;</span><br><span class="line">    x + guides(color = <span class="literal">FALSE</span>, fill = <span class="literal">FALSE</span>) + </span><br><span class="line">    theme_ArchR(baseSize = <span class="number">6.5</span>) +</span><br><span class="line">    theme(plot.margin = unit(<span class="built_in">c</span>(<span class="number">0</span>, <span class="number">0</span>, <span class="number">0</span>, <span class="number">0</span>), <span class="string">&quot;cm&quot;</span>)) +</span><br><span class="line">    theme(</span><br><span class="line">        axis.text.x=element_blank(), </span><br><span class="line">        axis.ticks.x=element_blank(), </span><br><span class="line">        axis.text.y=element_blank(), </span><br><span class="line">        axis.ticks.y=element_blank()</span><br><span class="line">    )</span><br><span class="line">&#125;)</span><br><span class="line">do.call(cowplot::plot_grid, <span class="built_in">c</span>(<span class="built_in">list</span>(ncol = <span class="number">3</span>),wplotall))</span><br></pre></td></tr></table></figure>
<h2 id="使用ArchRBrowser绘制Track"><a href="#使用ArchRBrowser绘制Track" class="headerlink" title="使用ArchRBrowser绘制Track"></a>使用ArchRBrowser绘制Track</h2><p>除了在UMAP上绘制每个细胞的基因得分外，我们还可以在基因组浏览器上浏览这些标记基因的局部染色体开放状态。为了执行该操作，我们使用<code>plotBrowserTrack()</code>函数，它会构建一系列绘图对象，每个对象对应一个标记基因。函数会根据groupBy输入的组别信息在不同track上绘制每组的开放状态。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  &lt;- <span class="built_in">c</span>(</span><br><span class="line">  <span class="string">&quot;CD34&quot;</span>, </span><br><span class="line">  <span class="string">&quot;GATA1&quot;</span>, </span><br><span class="line">  <span class="string">&quot;PAX5&quot;</span>, <span class="string">&quot;MS4A1&quot;</span>, </span><br><span class="line">  <span class="string">&quot;CD14&quot;</span>, </span><br><span class="line">  <span class="string">&quot;CD3D&quot;</span>, <span class="string">&quot;CD8A&quot;</span>, <span class="string">&quot;TBX21&quot;</span>, <span class="string">&quot;IL7R&quot;</span> </span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">genetrack &lt;- plotBrowserTrack(</span><br><span class="line">  ArchRProj = projPBMC_2, </span><br><span class="line">  groupBy = <span class="string">&quot;Clusters&quot;</span>, </span><br><span class="line">  geneSymbol = markerGenes, </span><br><span class="line">  upstream = <span class="number">50000</span>,</span><br><span class="line">  downstream = <span class="number">50000</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">grid::grid.newpage()</span><br><span class="line">pdf(<span class="string">&quot;CD14 track.pdf&quot;</span>)</span><br><span class="line">grid::grid.draw(genetrack$CD14)</span><br><span class="line">dev.off()</span><br><span class="line"></span><br><span class="line">pdf(<span class="string">&quot;all gene tracks.pdf&quot;</span>)</span><br><span class="line">genetrack</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="mailto:undefined">yilingzhang</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="http://yilingzhang.gitee.io/2021/04/08/scATACseq-Analysis/">http://yilingzhang.gitee.io/2021/04/08/scATACseq-Analysis/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/scATAC-seq/">scATAC-seq</a></div><div class="post_share"><div class="social-share" data-image="/img/cover/scATAC.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/social-share.js/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/social-share.js/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2021/04/10/Git-love-RProject-push-your-project-to-Github-Gitee/"><img class="prev-cover" src="/img/cover/Rgit.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">Git love RProject--push your project to Github/Gitee!</div></div></a></div><div class="next-post pull-right"><a href="/2021/04/06/TOEFL-writing-05/"><img class="next-cover" src="/img/cover/TOEFL-writing-04.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">TOEFL-writing-05</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span> Related Articles</span></div><div class="relatedPosts-list"><div><a href="/2021/04/24/scATACseq-project-of-NGS-02/" title="scATACseq.project of NGS-02"><img class="cover" src="/img/cover/scATAC.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2021-04-24</div><div class="title">scATACseq.project of NGS-02</div></div></a></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="card-info-avatar is-center"><img class="avatar-img" src="/img/selfie.JPG" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/><div class="author-info__name">yilingzhang</div><div class="author-info__description">Every Day is My Day</div></div><div class="card-info-data"><div class="card-info-data-item is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">28</div></a></div><div class="card-info-data-item is-center"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">14</div></a></div><div class="card-info-data-item is-center"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">13</div></a></div></div><div class="card-info-social-icons is-center"><a class="social-icon" href="mailto:jocelynzhang1119@icloud.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a><a class="social-icon" href="https://blog.csdn.net/weixin_43893431" target="_blank" title="CSDN"><i class="fas fa-paper-plane"></i></a></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#ArchR-install"><span class="toc-number">1.</span> <span class="toc-text">ArchR install</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#ArchR%E7%9A%84%E8%BE%93%E5%85%A5%E6%96%87%E4%BB%B6%E7%B1%BB%E5%9E%8B"><span class="toc-number">2.</span> <span class="toc-text">ArchR的输入文件类型</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%88%9B%E5%BB%BAArrow%E6%96%87%E4%BB%B6"><span class="toc-number">2.1.</span> <span class="toc-text">创建Arrow文件</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8ArchR%E6%8E%A8%E6%96%ADDoublet"><span class="toc-number">3.</span> <span class="toc-text">使用ArchR推断Doublet</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%88%9B%E5%BB%BAArchRProject"><span class="toc-number">4.</span> <span class="toc-text">创建ArchRProject</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%88%9B%E5%BB%BA%E4%B8%80%E4%B8%AAArchRProject"><span class="toc-number">4.1.</span> <span class="toc-text">创建一个ArchRProject</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%93%8D%E4%BD%9CArchRProject%E5%AF%B9%E8%B1%A1"><span class="toc-number">4.2.</span> <span class="toc-text">操作ArchRProject对象</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%A3%80%E6%9F%A5%E7%9F%A9%E9%98%B5%E6%95%B0%E6%8D%AE"><span class="toc-number">4.2.1.</span> <span class="toc-text">检查矩阵数据</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%BB%98%E5%88%B6%E8%B4%A8%E6%8E%A7%E4%BF%A1%E6%81%AF%E5%92%8CTSS%E5%AF%8C%E9%9B%86%E5%BE%97%E5%88%86%E7%9A%84%E5%AF%B9%E6%AF%94%E5%9B%BE"><span class="toc-number">4.2.2.</span> <span class="toc-text">绘制质控信息和TSS富集得分的对比图</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%B8%BAArchRProject%E4%B8%AD%E7%9A%84%E6%A0%B7%E6%9C%AC%E7%BB%9F%E8%AE%A1%E7%BB%98%E5%9B%BE"><span class="toc-number">4.2.3.</span> <span class="toc-text">为ArchRProject中的样本统计绘图</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%BB%98%E5%88%B6%E6%A0%B7%E6%9C%AC%E7%9A%84TSS%E5%AF%8C%E9%9B%86%E8%B0%B1%E5%92%8CFragment%E5%A4%A7%E5%B0%8F%E5%88%86%E5%B8%83"><span class="toc-number">4.2.4.</span> <span class="toc-text">绘制样本的TSS富集谱和Fragment大小分布</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%BF%87%E6%BB%A4doublets"><span class="toc-number">4.3.</span> <span class="toc-text">过滤doublets</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#ArchR%E9%99%8D%E7%BB%B4%E5%88%86%E6%9E%90"><span class="toc-number">5.</span> <span class="toc-text">ArchR降维分析</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E9%9A%90%E8%AF%AD%E4%B9%89-Latent-Semantic-Indexing-%E8%BF%AD%E4%BB%A3"><span class="toc-number">5.1.</span> <span class="toc-text">隐语义(Latent Semantic Indexing)迭代</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Harmony%E7%9F%AB%E6%AD%A3%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94"><span class="toc-number">5.2.</span> <span class="toc-text">Harmony矫正批次效应</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8ArchR%E8%81%9A%E7%B1%BB"><span class="toc-number">6.</span> <span class="toc-text">使用ArchR聚类</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8Seurat%E7%9A%84FindClusters-%E5%87%BD%E6%95%B0%E8%BF%9B%E8%A1%8C%E8%81%9A%E7%B1%BB"><span class="toc-number">6.1.</span> <span class="toc-text">使用Seurat的FindClusters()函数进行聚类</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8UMAP%E8%81%9A%E7%B1%BB"><span class="toc-number">6.2.</span> <span class="toc-text">使用UMAP聚类</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#t-Stocastic-Neighbor-Embedding-t-SNE"><span class="toc-number">6.3.</span> <span class="toc-text">t-Stocastic Neighbor Embedding (t-SNE)</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Harmony%E5%90%8E%E9%99%8D%E7%BB%B4"><span class="toc-number">6.4.</span> <span class="toc-text">Harmony后降维</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#ArchR%E7%9A%84%E5%9F%BA%E5%9B%A0%E5%BE%97%E5%88%86%E5%92%8C%E6%A0%87%E8%AE%B0%E5%9F%BA%E5%9B%A0"><span class="toc-number">7.</span> <span class="toc-text">ArchR的基因得分和标记基因</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E9%89%B4%E5%AE%9A%E6%A0%87%E8%AE%B0%E5%9F%BA%E5%9B%A0"><span class="toc-number">7.1.</span> <span class="toc-text">鉴定标记基因</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%8F%AF%E8%A7%86%E5%8C%96%E6%A0%87%E8%AE%B0%E5%9F%BA%E5%9B%A0"><span class="toc-number">7.2.</span> <span class="toc-text">可视化标记基因</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8MAGIC%E5%A1%AB%E5%85%85%E6%A0%87%E8%AE%B0%E5%9F%BA%E5%9B%A0"><span class="toc-number">7.3.</span> <span class="toc-text">使用MAGIC填充标记基因</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8ArchRBrowser%E7%BB%98%E5%88%B6Track"><span class="toc-number">7.4.</span> <span class="toc-text">使用ArchRBrowser绘制Track</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2021/04/26/scRNAseq-02-Annotation/" title="scRNAseq-02 Annotation"><img src="/img/cover/scRNA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scRNAseq-02 Annotation"/></a><div class="content"><a class="title" href="/2021/04/26/scRNAseq-02-Annotation/" title="scRNAseq-02 Annotation">scRNAseq-02 Annotation</a><time datetime="2021-04-26T08:46:41.000Z" title="Created 2021-04-26 16:46:41">2021-04-26</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/04/26/scRNAseq-01-Read-Data-and-Aggregation/" title="scRNA-seq-01 Read Data and Aggregation"><img src="/img/banner/paofu.JPG" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scRNA-seq-01 Read Data and Aggregation"/></a><div class="content"><a class="title" href="/2021/04/26/scRNAseq-01-Read-Data-and-Aggregation/" title="scRNA-seq-01 Read Data and Aggregation">scRNA-seq-01 Read Data and Aggregation</a><time datetime="2021-04-26T08:46:16.000Z" title="Created 2021-04-26 16:46:16">2021-04-26</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/04/24/scATACseq-project-of-NGS-02/" title="scATACseq.project of NGS-02"><img src="/img/cover/scATAC.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scATACseq.project of NGS-02"/></a><div class="content"><a class="title" href="/2021/04/24/scATACseq-project-of-NGS-02/" title="scATACseq.project of NGS-02">scATACseq.project of NGS-02</a><time datetime="2021-04-24T05:25:41.000Z" title="Created 2021-04-24 13:25:41">2021-04-24</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/04/19/all-about-UMI-tools/" title="all about UMI-tools"><img src="/img/cover/UMItools.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="all about UMI-tools"/></a><div class="content"><a class="title" href="/2021/04/19/all-about-UMI-tools/" title="all about UMI-tools">all about UMI-tools</a><time datetime="2021-04-19T04:45:20.000Z" title="Created 2021-04-19 12:45:20">2021-04-19</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/04/13/Stat-03-Non-Negative-Matrix-Factorization/" title="Stat-03-Non-Negative Matrix Factorization"><img src="/img/cover/Stat-03-NMF.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Stat-03-Non-Negative Matrix Factorization"/></a><div class="content"><a class="title" href="/2021/04/13/Stat-03-Non-Negative-Matrix-Factorization/" title="Stat-03-Non-Negative Matrix Factorization">Stat-03-Non-Negative Matrix Factorization</a><time datetime="2021-04-13T01:02:11.000Z" title="Created 2021-04-13 09:02:11">2021-04-13</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2021 By yilingzhang</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"></div><script defer="defer" id="ribbon" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/canvas-ribbon.min.js" size="150" alpha="0.6" zIndex="-1" mobile="false" data-click="true"></script><script id="click-heart" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/click-heart.min.js" async="async" mobile="false"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>