<!DOCTYPE html><html lang="en | zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1"><title>Metagenomics Analysis | Jocelyn's Blog</title><meta name="author" content="yilingzhang"><meta name="copyright" content="yilingzhang"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="通过single end示例数据了解qiime2分析过程。 尝试利用pair end示例数据进行16S数据分析。   Qiime2introductionQIIME 2是一款强大、可扩展和去中心化的微生物组分析平台，强调数据分析透明。QIIME 2可以使研究者从原始DNA序列开始分析，直接获取出版级的统计和图片结果。 主要特点：  整合分析流程、自动化追踪数据来源 语义类型系统，自动识别输入文件">
<meta property="og:type" content="article">
<meta property="og:title" content="Metagenomics Analysis">
<meta property="og:url" content="http://yilingzhang.gitee.io/2021/04/01/Metagenomics/index.html">
<meta property="og:site_name" content="Jocelyn&#39;s Blog">
<meta property="og:description" content="通过single end示例数据了解qiime2分析过程。 尝试利用pair end示例数据进行16S数据分析。   Qiime2introductionQIIME 2是一款强大、可扩展和去中心化的微生物组分析平台，强调数据分析透明。QIIME 2可以使研究者从原始DNA序列开始分析，直接获取出版级的统计和图片结果。 主要特点：  整合分析流程、自动化追踪数据来源 语义类型系统，自动识别输入文件">
<meta property="og:locale">
<meta property="og:image" content="http://yilingzhang.gitee.io/img/cover/Metagenomics.png">
<meta property="article:published_time" content="2021-04-01T11:31:45.000Z">
<meta property="article:modified_time" content="2021-04-01T11:40:43.153Z">
<meta property="article:author" content="yilingzhang">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://yilingzhang.gitee.io/img/cover/Metagenomics.png"><link rel="shortcut icon" href="/img/love_32px.png"><link rel="canonical" href="http://yilingzhang.gitee.io/2021/04/01/Metagenomics/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    jQuery: 'https://cdn.jsdelivr.net/npm/jquery@latest/dist/jquery.min.js',
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/js/jquery.justifiedGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/css/justifiedGallery.min.css'
    },
    fancybox: {
      js: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.js',
      css: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isanchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = { 
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2021-04-01 19:40:43'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    })(window)</script><meta name="generator" content="Hexo 5.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="author-avatar"><img class="avatar-img" src="/img/selfie.JPG" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data"><div class="data-item is-center"><div class="data-item-link"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">18</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">9</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">11</div></a></div></div></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/img/cover/Metagenomics.png')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Jocelyn's Blog</a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">Metagenomics Analysis</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2021-04-01T11:31:45.000Z" title="Created 2021-04-01 19:31:45">2021-04-01</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2021-04-01T11:40:43.153Z" title="Updated 2021-04-01 19:40:43">2021-04-01</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/Metagenomics/">Metagenomics</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">Word count:</span><span class="word-count">3.5k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">Reading time:</span><span>13min</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><ol>
<li>通过single end示例数据了解qiime2分析过程。</li>
<li>尝试利用pair end示例数据进行16S数据分析。</li>
</ol>
<hr>
<h1 id="Qiime2"><a href="#Qiime2" class="headerlink" title="Qiime2"></a>Qiime2</h1><h2 id="introduction"><a href="#introduction" class="headerlink" title="introduction"></a>introduction</h2><p>QIIME 2是一款强大、可扩展和去中心化的微生物组分析平台，强调数据分析透明。QIIME 2可以使研究者从原始DNA序列开始分析，直接获取出版级的统计和图片结果。</p>
<p>主要特点：</p>
<ul>
<li>整合分析流程、自动化追踪数据来源</li>
<li>语义类型系统，自动识别输入文件类型</li>
<li>插件系统可扩展微生物分析功能种类</li>
<li>支持多种用户界面，如API、命令行、图形界面</li>
</ul>
<p>QIIME 2是对QIIME 1完全重新设计并重写的微生物组分析流程。QIIME 2保留了QIIME 1强大和广泛使用的优点，同时改进了其众多不足之处。</p>
<p>QIIME 2当前支持从头到尾的完整微生物组分析流程。通常QIIME 2插件功能，不断有新功能可用。可以在可用插件列表中查找当前可用的插件。在未来可用插件页列出了正在开发的插件。</p>
<h2 id="installation"><a href="#installation" class="headerlink" title="installation"></a>installation</h2><ul>
<li><p>conda环境安装QIIME2（推荐）<br>Install QIIME 2 within a conda environment<br>mkdir -p 2020.11 &amp;&amp; cd 2020.11</p>
</li>
<li><p>下载软件安装列表，官方源不容易下载</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">wget -c https://data.qiime2.org/distro/core/qiime2-2020.11-py36-linux-conda.yml</span><br><span class="line"><span class="comment"># OR</span></span><br><span class="line">wget -c http://210.75.224.110/github/QIIME2ChineseManual/2020.11/qiime2-2020.11-py36-linux-conda.yml</span><br><span class="line">time conda env create -n qiime2-2020.11 --file qiime2-2020.11-py36-linux-conda.yml</span><br><span class="line"></span><br><span class="line"><span class="comment"># 删除软件列表</span></span><br><span class="line">rm qiime2-2020.11-py36-linux-conda.yml</span><br></pre></td></tr></table></figure></li>
<li><p>服务器已安装，直接使用。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">qiime2 conda environment path:</span><br><span class="line">/public/workspace/zmy1/home/wujian/.conda/envs/qimme2-2020.11/</span><br><span class="line"></span><br><span class="line"><span class="comment"># ln至自己的文件夹下</span></span><br><span class="line"><span class="built_in">source</span> activate qimme2-2020.11</span><br></pre></td></tr></table></figure>
<h2 id="Qiime2重要概念"><a href="#Qiime2重要概念" class="headerlink" title="Qiime2重要概念"></a>Qiime2重要概念</h2></li>
</ul>
<ol>
<li><p>数据文件: QIIME 2对象/文件格式<br>由QIIME 2产生的数据类型，叫QIIME 2对象(artifacts)，通常包括数据和元数据/样本信息(metadata)。元数据描述数据，包括类型、格式和它如何产生。典型的扩展名为.qza。</p>
</li>
<li><p>数据文件：可视化<br>QIIME2生成的图表结果对象或文件类型，以.qzv为扩展名，末尾的v代表visual；它同qza文件类似，包括分析方法和结果，方便追溯图表是如何产生的；唯一与qza不同的，它是分析的终点，即结果的呈现，不会在流程中继续分析。可视化的结果包括统计结果表格、交互式图像、静态图片及其它组合的可视化呈现。这类文件可以使用QIIME2 qiime tools view命令查看。</p>
<p>可通过网站：<a target="_blank" rel="noopener" href="https://view.qiime2.org/">https://view.qiime2.org/</a><br>导入文件并显示结果图表，同时可查看数据分析过程；方便与不使用QIIME 2的合作者分享结果。</p>
</li>
<li><p>插件<br>QIIME2中的用户的某个特定功能即为插件，你可以安装并完成分析，比如拆分样品的q2-demux插件、Alpha-或beta-多样性分析的q2-diversity插件等。</p>
</li>
<li><p>方法和可视化软件<br>QIIME 2插件定义的用于进行分析的方法和可视化工具类型。</p>
<p>方法是对QIIME2定义的输入对象进行操作的过程，包括命令和参数，并产生1个或多个标准格式的输出。这一结果可以后续分析或可视化，产生中间或末端的输出。例如rarefy方法，输入文件为q2-feature-table插件产生的特征表，输出文件为样本深度一致的特征表。它可以作为输入文件，用于alpha多样性分析中的q2-diversity方法。输入和输出均为qza文件；</p>
<p>可视化工具定义了标准输入，包括QIIME 2对象和参数的组合，产生统计表格或可视化图形，方便用户解读，输入为qza格式，输出为qzv，文件不仅包括结果，还包括处理的分析命令和参数，方便重复和检查分析过程是否准确。可视化的结果文件qzv是分析的终点，不可以进一步分析。</p>
</li>
</ol>
<h2 id="Qiime2使用"><a href="#Qiime2使用" class="headerlink" title="Qiime2使用"></a>Qiime2使用</h2><p><img src="/img/Metagenomics/qiime.png" alt="qiime"></p>
<h1 id="16S分析流程"><a href="#16S分析流程" class="headerlink" title="16S分析流程"></a>16S分析流程</h1><blockquote>
<p>数据位置<br>/public/workspace/zmy1/home/wujian/metagenomics/single_end/<br>分别将emp-single-end-sequences和sample-metadata.tsv软链至相应文件夹下。<br>/public/workspace/zmy1/home/wujian/metagenomics/single_end/emp-single-end-sequences/<br>/public/workspace/zmy1/home/wujian/metagenomics/single_end/sample-metadata.tsv</p>
</blockquote>
<h2 id="Metadata-元数据"><a href="#Metadata-元数据" class="headerlink" title="Metadata 元数据"></a>Metadata 元数据</h2><p>元数据提供了从数据中获得生物学发现的关键。在QIIME 2中，样本元数据可以包括更多技术细节，例如在多次测序批次中用于每个样本的DNA条形码，或者样本的描述，例如每个样本来自人类微生物组对应时间序列研究中的分组、时间点和身体部位。特征元数据通常是特征注释，例如分配给序列变量或OTU的物种注释。</p>
<p>收集样本元数据通常是在开始QIIME 2分析之前要进行的第一步。由研究人员决定哪些信息是作为元数据收集和跟踪的。<br>尽可能多的收集元数据，因为您可能无法追溯收集某些类型的信息，以便后续分析中使用。<br><img src="/img/Metagenomics/Metadata.png" alt="Metadata"></p>
<h2 id="数据格式"><a href="#数据格式" class="headerlink" title="数据格式"></a>数据格式</h2><p>EMPSingleEndSequences是包含多样本混合的序列文件，这意味着序列尚未分配给样本（因此包括sequences.fastq.gz和barcode.fastq.gz文件，其中barcode.fastq.gz包含与sequences.fastq.gz中的每个序列相关联的条形码）。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">/public/workspace/zmy1/home/wujian/metagenomics/single_end/emp-single-end-sequences/</span><br></pre></td></tr></table></figure>
<h2 id="导入数据"><a href="#导入数据" class="headerlink" title="导入数据"></a>导入数据</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">time qiime tools import \</span><br><span class="line">  --<span class="built_in">type</span> EMPSingleEndSequences \</span><br><span class="line">  --input-path ./0.rawdata/emp-single-end-sequences \</span><br><span class="line">  --output-path ./0.rawdata/emp-single-end-sequences.qza</span><br></pre></td></tr></table></figure>
<h2 id="拆分样品-Demultiplexing-sequences"><a href="#拆分样品-Demultiplexing-sequences" class="headerlink" title="拆分样品 Demultiplexing sequences"></a>拆分样品 Demultiplexing sequences</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">time qiime demux emp-single \</span><br><span class="line">  --i-seqs ./0.rawdata/emp-single-end-sequences.qza \</span><br><span class="line">  --m-barcodes-file ./0.rawdata/sample-metadata.tsv \</span><br><span class="line">  --m-barcodes-column barcode-sequence \</span><br><span class="line">  --o-per-sample-sequences 1.demuxdata/demux.qza \</span><br><span class="line">  --o-error-correction-details 1.demuxdata/demux-details.qza &amp;</span><br></pre></td></tr></table></figure>
<h2 id="结果统计"><a href="#结果统计" class="headerlink" title="结果统计"></a>结果统计</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">time qiime demux summarize \</span><br><span class="line">  --i-data demux.qza \</span><br><span class="line">  --o-visualization demux.qzv &amp;</span><br></pre></td></tr></table></figure>
<p>.qzv可以在 <a target="_blank" rel="noopener" href="https://view.qiime2.org/">https://view.qiime2.org/</a> 上查看可视化结果</p>
<h2 id="序列质控和生成特征表-Sequence-quality-control-and-feature-table-construction"><a href="#序列质控和生成特征表-Sequence-quality-control-and-feature-table-construction" class="headerlink" title="序列质控和生成特征表 Sequence quality control and feature table construction"></a>序列质控和生成特征表 Sequence quality control and feature table construction</h2><p>dada2 denoise-single方法需要两个用于质量过滤的参数：</p>
<blockquote>
<p>–p-trim-left m，它去除每个序列的前m个碱基(如引物、标签序列barcode)；<br>–p-trunc-len n，它在位置n截断每个序列。这允许用户去除序列的低质量区域、引物或标签序列等。</p>
</blockquote>
<p>确定要为这两个参数的值，查看由qiime demux summarize生成的demux.qzv文件中的交互质量图选项卡。</p>
<blockquote>
<p>– p-trim-left 截取左端低质量序列，我们看上图中箱线图，左端质量都很高，无低质量区，设置为0；<br>– p-trunc-len 序列截取长度，也是为了去除右端低质量序列，我们看到大于120以后，质量下降极大，甚至中位数都下降至20以下，需要全部去除，综合考虑决定设置为120。</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">time qiime dada2 denoise-single \</span><br><span class="line">  --i-demultiplexed-seqs 1.demuxdata/demux.qza \</span><br><span class="line">  --p-trim-left 0 \</span><br><span class="line">  --p-trunc-len 120 \</span><br><span class="line">  --o-representative-sequences ./2.QCandFeatureTable/rep-seqs-dada2.qza \</span><br><span class="line">  --o-table ./2.QCandFeatureTable/table-dada2.qza \</span><br><span class="line">  --o-denoising-stats ./2.QCandFeatureTable/stats-dada2.qza</span><br></pre></td></tr></table></figure>
<p>对特征表统计进行进行可视化</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">qiime metadata tabulate \</span><br><span class="line">  --m-input-file stats-dada2.qza \</span><br><span class="line">  --o-visualization stats-dada2.qzv</span><br></pre></td></tr></table></figure>
<h2 id="特征表和特征序列汇总-FeatureTable-and-FeatureData-summaries"><a href="#特征表和特征序列汇总-FeatureTable-and-FeatureData-summaries" class="headerlink" title="特征表和特征序列汇总 FeatureTable and FeatureData summaries"></a>特征表和特征序列汇总 FeatureTable and FeatureData summaries</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">qiime feature-table summarize \</span><br><span class="line">  --i-table ./2.QCandFeatureTable/table-dada2.qza \</span><br><span class="line">  --o-visualization ./2.QCandFeatureTable/table.qzv \</span><br><span class="line">  --m-sample-metadata-file  ./0.rawdata/sample-metadata.tsv</span><br><span class="line"></span><br><span class="line">qiime feature-table tabulate-seqs \</span><br><span class="line">  --i-data ./2.QCandFeatureTable/rep-seqs-dada2.qza \</span><br><span class="line">  --o-visualization ./2.QCandFeatureTable/rep-seqs.qzv</span><br></pre></td></tr></table></figure>
<p>Output:</p>
<ol>
<li>table.qzv: 特征表统计。<br>Frequency per sample:<br>Frequency per feature: 一群数据里出现次数多的？<br>Feature Detail: 4b5eeb300368260019c1fbc7a3c718fc </li>
<li>rep-seqs.qzv: 代表序列统计，可点击序列跳转NCBI blast查看相近序列的信息。</li>
</ol>
<h2 id="构建进化树用于多样性分析-Generate-a-tree-for-phylogenetic-diversity"><a href="#构建进化树用于多样性分析-Generate-a-tree-for-phylogenetic-diversity" class="headerlink" title="构建进化树用于多样性分析 Generate a tree for phylogenetic diversity"></a>构建进化树用于多样性分析 Generate a tree for phylogenetic diversity</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">time qiime phylogeny align-to-tree-mafft-fasttree \</span><br><span class="line">  --i-sequences ./2.QCandFeatureTable/rep-seqs-dada2.qza \</span><br><span class="line">  --o-alignment ./3.generateTree/aligned-rep-seqs.qza \</span><br><span class="line">  --o-masked-alignment ./3.generateTree/masked-aligned-rep-seqs.qza \</span><br><span class="line">  --o-tree ./3.generateTree/unrooted-tree.qza \</span><br><span class="line">  --o-rooted-tree ./3.generateTree/rooted-tree.qza</span><br></pre></td></tr></table></figure>
<p>Output:<br>aligned-rep-seqs.qza: 多序列比对结果。<br>masked-aligned-rep-seqs.qza: 过滤去除高变区后的多序列比对结果。<br>rooted-tree.qza: 有根树，用于多样性分析。<br>unrooted-tree.qza: 无根树。</p>
<h2 id="Alpha和Beta多样性分析"><a href="#Alpha和Beta多样性分析" class="headerlink" title="Alpha和Beta多样性分析"></a>Alpha和Beta多样性分析</h2><p>QIIME 2的多样性分析使用q2-diversity插件，该插件支持计算α和β多样性指数、并应用相关的统计检验以及生成交互式可视化图表。我们将首先应用core-metrics-phylogenetic方法，该方法将FeatureTable抽平到用户指定的测序深度，然后计算几种常用的α和β多样性指数，并使用Emperor为每个β多样性指数生成主坐标分析（PCoA）图。</p>
<ul>
<li>α多样性</li>
</ul>
<p>-香农(Shannon’s)多样性指数（群落丰富度的定量度量，即包括丰富度richness和均匀度evenness两个层面）<br>-可观测的OTU(Observed OTUs，群落丰富度的定性度量，只包括丰富度）<br>-Faith’s系统发育多样性（包含特征之间的系统发育关系的群落丰富度的定性度量）<br>-均匀度Evenness（或 Pielou’s均匀度；群落均匀度的度量）</p>
<ul>
<li>β多样性</li>
</ul>
<p>-Jaccard距离（群落差异的定性度量，即只考虑种类，不考虑丰度）<br>-Bray-Curtis距离（群落差异的定量度量，较常用）<br>-非加权UniFrac距离（包含特征之间的系统发育关系的群落差异定性度量）<br>-加权UniFrac距离（包含特征之间的系统发育关系的群落差异定量度量）</p>
<p>需要提供给这个脚本的一个重要参数是–p-sampling-depth，它是指定重采样（即稀疏/稀疏rarefaction）深度。因为大多数多样指数对不同样本的不同测序深度敏感，所以这个脚本将随机地将每个样本的测序量重新采样至该参数值。例如，提供–p-sampling-depth 500，则此步骤将对每个样本中的计数进行无放回抽样，从而使得结果表中的每个样本的总计数为500。如果任何样本的总计数小于该值，那么这些样本将从多样性分析中删除。选择这个值很棘手。建议通过查看上面创建的表table.qzv文件中呈现的信息并选择一个尽可能高的值（因此每个样本保留更多的序列）同时尽可能少地排除样本来进行选择。<br><strong>如是数据量都很大，选最小的即可。如果有个别数据量非常小，去除最小值再选最小值。</strong><br>一般采用HiSeq2500或NovaSeq6000的 PE250模式测序，数据量都非常大，通常可以采用3万或5万的标准抽平，仍可保留90%以上样本。过低或过高一般结果也会波动较大，不建议放在一起分析。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">time qiime diversity core-metrics-phylogenetic \</span><br><span class="line">  --i-phylogeny ./3.generateTree/rooted-tree.qza \</span><br><span class="line">  --i-table ./2.QCandFeatureTable/table-dada2.qza \</span><br><span class="line">  --p-sampling-depth 1103 \</span><br><span class="line">  --m-metadata-file ./0.rawdata/sample-metadata.tsv \</span><br><span class="line">  --output-dir  ./4.CoreMetricsRes</span><br></pre></td></tr></table></figure>
<blockquote>
<p>13个数据文件:<br>core-metrics-results/faith_pd_vector.qza: Alpha多样性考虑进化的faith指数。<br>core-metrics-results/unweighted_unifrac_distance_matrix.qza: 无权重unifrac距离矩阵。<br>core-metrics-results/bray_curtis_pcoa_results.qza: 基于Bray-Curtis距离PCoA的结果。<br>core-metrics-results/shannon_vector.qza: Alpha多样性香农指数。<br>core-metrics-results/rarefied_table.qza: 等量重采样后的特征表。<br>core-metrics-results/weighted_unifrac_distance_matrix.qza: 有权重的unifrac距离矩阵。<br>core-metrics-results/jaccard_pcoa_results.qza: jaccard距离PCoA结果。<br>core-metrics-results/observed_otus_vector.qza: Alpha多样性observed otus指数。<br>core-metrics-results/weighted_unifrac_pcoa_results.qza: 基于有权重的unifrac距离的PCoA结果。<br>core-metrics-results/jaccard_distance_matrix.qza: jaccard距离矩阵。<br>core-metrics-results/evenness_vector.qza: Alpha多样性均匀度指数。<br>core-metrics-results/bray_curtis_distance_matrix.qza: Bray-Curtis距离矩阵。<br>core-metrics-results/unweighted_unifrac_pcoa_results.qza: 无权重的unifrac距离的PCoA结果</p>
</blockquote>
<blockquote>
<p>4个可视化结果:<br>core-metrics-results/unweighted_unifrac_emperor.qzv：无权重的unifrac距离PCoA结果采用emperor可视化。<br>core-metrics-results/jaccard_emperor.qzv：jaccard距离PCoA结果采用emperor可视化。<br>core-metrics-results/bray_curtis_emperor.qzv： Bray-Curtis距离PCoA结果采用emperor可视化。<br>core-metrics-results/weighted_unifrac_emperor.qzv： 有权重的unifrac距离PCoA结果采用emperor可视化。</p>
</blockquote>
<h2 id="Alpha多样性组间显著性分析和可视化"><a href="#Alpha多样性组间显著性分析和可视化" class="headerlink" title="Alpha多样性组间显著性分析和可视化"></a>Alpha多样性组间显著性分析和可视化</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">qiime diversity alpha-group-significance \</span><br><span class="line">--i-alpha-diversity core-metrics-results/faith_pd_vector.qza \</span><br><span class="line">--m-metadata-file ./sample-metadata.tsv \</span><br><span class="line">--o-visualization core-metrics-results/faith-pd-group-significance.qzv</span><br><span class="line"></span><br><span class="line">qiime diversity alpha-group-significance \</span><br><span class="line">--i-alpha-diversity core-metrics-results/evenness_vector.qza \</span><br><span class="line">--m-metadata-file ./sample-metadata.tsv \</span><br><span class="line">--o-visualization core-metrics-results/evenness-group-significance.qzv</span><br></pre></td></tr></table></figure>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#不同身体部位之间比较；</span></span><br><span class="line">time qiime diversity beta-group-significance \</span><br><span class="line">  --i-distance-matrix core-metrics-results/unweighted_unifrac_distance_matrix.qza \</span><br><span class="line">  --m-metadata-file sample-metadata.tsv \</span><br><span class="line">  --m-metadata-column body-site \</span><br><span class="line">  --o-visualization core-metrics-results/unweighted-unifrac-body-site-significance.qzv \</span><br><span class="line">  --p-pairwise</span><br><span class="line"></span><br><span class="line"><span class="comment">#不同subject之间比较；</span></span><br><span class="line">time qiime diversity beta-group-significance \</span><br><span class="line">  --i-distance-matrix core-metrics-results/unweighted_unifrac_distance_matrix.qza \</span><br><span class="line">  --m-metadata-file sample-metadata.tsv \</span><br><span class="line">  --m-metadata-column subject \</span><br><span class="line">  --o-visualization core-metrics-results/unweighted-unifrac-subject-group-significance.qzv \</span><br><span class="line">  --p-pairwise</span><br></pre></td></tr></table></figure>
<h2 id="Alpha稀疏曲线-Alpha-rarefaction-plotting"><a href="#Alpha稀疏曲线-Alpha-rarefaction-plotting" class="headerlink" title="Alpha稀疏曲线 Alpha rarefaction plotting"></a>Alpha稀疏曲线 Alpha rarefaction plotting</h2><p>稀疏曲线：到什么时候饱和，捕获到采样中所有微生物的菌群</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">time qiime diversity alpha-rarefaction \</span><br><span class="line">  --i-table table-dada2.qza \</span><br><span class="line">  --i-phylogeny rooted-tree.qza \</span><br><span class="line">  --p-max-depth 4000 \</span><br><span class="line">  --m-metadata-file sample-metadata.tsv \</span><br><span class="line">  --o-visualization alpha-rarefaction.qzv</span><br></pre></td></tr></table></figure>
<p>Output：<br>alpha-rarefaction.qzv: 稀疏曲线。</p>
<h2 id="物种组成分析-Taxonomic-analysis"><a href="#物种组成分析-Taxonomic-analysis" class="headerlink" title="物种组成分析 Taxonomic analysis"></a>物种组成分析 Taxonomic analysis</h2><p>物种分类器</p>
<ul>
<li><p>物种分类器根据你特定的样品制备和测序参数进行训练时表现最好，包括用于扩增的引物和测序序列的长度。因此，一般来说，你应该按照使用q2-feature-classifier的训练特征分类器的说明来训练自己的物种分类器。</p>
</li>
<li><p>这里使用的分类器是在Greengenes 13_8 99% OTU上训练的，其中序列被修剪到仅包括来自16S区域的250个碱基。利用这个分类器生成从序列到物种注释结果关联的可视化。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ln –s /public/workspace/zmy1/home/wujian/metagenomics/single_end/gg-13-8-99-515-806-nb-classifier.qza ./</span><br></pre></td></tr></table></figure>
<h2 id="物种注释和可视化"><a href="#物种注释和可视化" class="headerlink" title="物种注释和可视化"></a>物种注释和可视化</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">time qiime feature-classifier classify-sklearn \</span><br><span class="line">  --i-classifier gg-13-8-99-515-806-nb-classifier.qza \</span><br><span class="line">  --i-reads ./2.QCandFeatureTable/rep-seqs-dada2.qza \</span><br><span class="line">  --o-classification taxonomy.qza</span><br><span class="line"></span><br><span class="line">qiime metadata tabulate \</span><br><span class="line">  --m-input-file taxonomy.qza \</span><br><span class="line">  --o-visualization taxonomy.qzv</span><br></pre></td></tr></table></figure>
<p>Output:<br>taxonomy.qza: 物种注释结果。<br>gg-13-8-99-515-806-nb-classifier.qza: 分类器的训练结果。</p>
</li>
</ul>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#Barplot</span></span><br><span class="line">qiime taxa barplot \</span><br><span class="line">  --i-table table-dada2.qza \</span><br><span class="line">  --i-taxonomy taxonomy.qza \</span><br><span class="line">  --m-metadata-file sample-metadata.tsv \</span><br><span class="line">  --o-visualization taxa-bar-plots.qzv</span><br></pre></td></tr></table></figure>
<p>Output:<br>taxa-bar-plots.qzv: 交互式物种组成barplot。</p>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="mailto:undefined">yilingzhang</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="http://yilingzhang.gitee.io/2021/04/01/Metagenomics/">http://yilingzhang.gitee.io/2021/04/01/Metagenomics/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"></div><div class="post_share"><div class="social-share" data-image="/img/cover/Metagenomics.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/social-share.js/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/social-share.js/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="next-post pull-full"><a href="/2021/03/31/eCLIP-Analysis/"><img class="next-cover" src="/img/cover/eCLIP.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">eCLIP Analysis</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="card-info-avatar is-center"><img class="avatar-img" src="/img/selfie.JPG" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/><div class="author-info__name">yilingzhang</div><div class="author-info__description">Every Day is My Day</div></div><div class="card-info-data"><div class="card-info-data-item is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">18</div></a></div><div class="card-info-data-item is-center"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">9</div></a></div><div class="card-info-data-item is-center"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">11</div></a></div></div><div class="card-info-social-icons is-center"><a class="social-icon" href="mailto:jocelynzhang1119@icloud.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a><a class="social-icon" href="https://blog.csdn.net/weixin_43893431" target="_blank" title="CSDN"><i class="fas fa-paper-plane"></i></a></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#Qiime2"><span class="toc-number">1.</span> <span class="toc-text">Qiime2</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#introduction"><span class="toc-number">1.1.</span> <span class="toc-text">introduction</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#installation"><span class="toc-number">1.2.</span> <span class="toc-text">installation</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Qiime2%E9%87%8D%E8%A6%81%E6%A6%82%E5%BF%B5"><span class="toc-number">1.3.</span> <span class="toc-text">Qiime2重要概念</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Qiime2%E4%BD%BF%E7%94%A8"><span class="toc-number">1.4.</span> <span class="toc-text">Qiime2使用</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#16S%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B"><span class="toc-number">2.</span> <span class="toc-text">16S分析流程</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#Metadata-%E5%85%83%E6%95%B0%E6%8D%AE"><span class="toc-number">2.1.</span> <span class="toc-text">Metadata 元数据</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E6%A0%BC%E5%BC%8F"><span class="toc-number">2.2.</span> <span class="toc-text">数据格式</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%AF%BC%E5%85%A5%E6%95%B0%E6%8D%AE"><span class="toc-number">2.3.</span> <span class="toc-text">导入数据</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%8B%86%E5%88%86%E6%A0%B7%E5%93%81-Demultiplexing-sequences"><span class="toc-number">2.4.</span> <span class="toc-text">拆分样品 Demultiplexing sequences</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%BB%93%E6%9E%9C%E7%BB%9F%E8%AE%A1"><span class="toc-number">2.5.</span> <span class="toc-text">结果统计</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%BA%8F%E5%88%97%E8%B4%A8%E6%8E%A7%E5%92%8C%E7%94%9F%E6%88%90%E7%89%B9%E5%BE%81%E8%A1%A8-Sequence-quality-control-and-feature-table-construction"><span class="toc-number">2.6.</span> <span class="toc-text">序列质控和生成特征表 Sequence quality control and feature table construction</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%89%B9%E5%BE%81%E8%A1%A8%E5%92%8C%E7%89%B9%E5%BE%81%E5%BA%8F%E5%88%97%E6%B1%87%E6%80%BB-FeatureTable-and-FeatureData-summaries"><span class="toc-number">2.7.</span> <span class="toc-text">特征表和特征序列汇总 FeatureTable and FeatureData summaries</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%9E%84%E5%BB%BA%E8%BF%9B%E5%8C%96%E6%A0%91%E7%94%A8%E4%BA%8E%E5%A4%9A%E6%A0%B7%E6%80%A7%E5%88%86%E6%9E%90-Generate-a-tree-for-phylogenetic-diversity"><span class="toc-number">2.8.</span> <span class="toc-text">构建进化树用于多样性分析 Generate a tree for phylogenetic diversity</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Alpha%E5%92%8CBeta%E5%A4%9A%E6%A0%B7%E6%80%A7%E5%88%86%E6%9E%90"><span class="toc-number">2.9.</span> <span class="toc-text">Alpha和Beta多样性分析</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Alpha%E5%A4%9A%E6%A0%B7%E6%80%A7%E7%BB%84%E9%97%B4%E6%98%BE%E8%91%97%E6%80%A7%E5%88%86%E6%9E%90%E5%92%8C%E5%8F%AF%E8%A7%86%E5%8C%96"><span class="toc-number">2.10.</span> <span class="toc-text">Alpha多样性组间显著性分析和可视化</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Alpha%E7%A8%80%E7%96%8F%E6%9B%B2%E7%BA%BF-Alpha-rarefaction-plotting"><span class="toc-number">2.11.</span> <span class="toc-text">Alpha稀疏曲线 Alpha rarefaction plotting</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%89%A9%E7%A7%8D%E7%BB%84%E6%88%90%E5%88%86%E6%9E%90-Taxonomic-analysis"><span class="toc-number">2.12.</span> <span class="toc-text">物种组成分析 Taxonomic analysis</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%89%A9%E7%A7%8D%E6%B3%A8%E9%87%8A%E5%92%8C%E5%8F%AF%E8%A7%86%E5%8C%96"><span class="toc-number">2.13.</span> <span class="toc-text">物种注释和可视化</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2021/04/01/Metagenomics/" title="Metagenomics Analysis"><img src="/img/cover/Metagenomics.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Metagenomics Analysis"/></a><div class="content"><a class="title" href="/2021/04/01/Metagenomics/" title="Metagenomics Analysis">Metagenomics Analysis</a><time datetime="2021-04-01T11:31:45.000Z" title="Created 2021-04-01 19:31:45">2021-04-01</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/03/31/eCLIP-Analysis/" title="eCLIP Analysis"><img src="/img/cover/eCLIP.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="eCLIP Analysis"/></a><div class="content"><a class="title" href="/2021/03/31/eCLIP-Analysis/" title="eCLIP Analysis">eCLIP Analysis</a><time datetime="2021-03-31T00:19:05.000Z" title="Created 2021-03-31 08:19:05">2021-03-31</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/03/30/WES-Project/" title="WES Project"><img src="/img/cover/WES.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="WES Project"/></a><div class="content"><a class="title" href="/2021/03/30/WES-Project/" title="WES Project">WES Project</a><time datetime="2021-03-30T08:38:50.000Z" title="Created 2021-03-30 16:38:50">2021-03-30</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/03/25/CUT-Tag-Analysis/" title="CUT&amp;Tag Analysis"><img src="/img/cover/CUT&amp;Tag.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="CUT&amp;Tag Analysis"/></a><div class="content"><a class="title" href="/2021/03/25/CUT-Tag-Analysis/" title="CUT&amp;Tag Analysis">CUT&amp;Tag Analysis</a><time datetime="2021-03-25T09:37:16.000Z" title="Created 2021-03-25 17:37:16">2021-03-25</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/03/17/Paper-writing-Tips/" title="Paper-writing Tips"><img src="/img/cover/paper.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Paper-writing Tips"/></a><div class="content"><a class="title" href="/2021/03/17/Paper-writing-Tips/" title="Paper-writing Tips">Paper-writing Tips</a><time datetime="2021-03-17T13:51:22.000Z" title="Created 2021-03-17 21:51:22">2021-03-17</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2021 By yilingzhang</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"></div><script defer="defer" id="ribbon" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/canvas-ribbon.min.js" size="150" alpha="0.6" zIndex="-1" mobile="false" data-click="true"></script><script id="click-heart" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/click-heart.min.js" async="async" mobile="false"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>