<!DOCTYPE html><html lang="en | zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1"><title>scRNAseq-01 Read Data and Aggregation | Jocelyn's Blog</title><meta name="keywords" content="scRNA-seq"><meta name="author" content="yilingzhang"><meta name="copyright" content="yilingzhang"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="拆库拆index（lane –&gt; library）12345678910111213# 查看序列与Index对应关系，通常为Index反向互补，因为在右端zcat lane1_1.fq.gz | headzcat lane1_2.fq.gz | head# 按index列表文件并行筛选每个文库的双端文件# zcat解压、grep -A 3检索并输出4行，去除grep的添加字符行，gzip压缩">
<meta property="og:type" content="article">
<meta property="og:title" content="scRNAseq-01 Read Data and Aggregation">
<meta property="og:url" content="https://hiiiyilingzhang.github.io/2021/04/26/scRNAseq-01-Read-Data-and-Aggregation/index.html">
<meta property="og:site_name" content="Jocelyn&#39;s Blog">
<meta property="og:description" content="拆库拆index（lane –&gt; library）12345678910111213# 查看序列与Index对应关系，通常为Index反向互补，因为在右端zcat lane1_1.fq.gz | headzcat lane1_2.fq.gz | head# 按index列表文件并行筛选每个文库的双端文件# zcat解压、grep -A 3检索并输出4行，去除grep的添加字符行，gzip压缩">
<meta property="og:locale">
<meta property="og:image" content="https://hiiiyilingzhang.github.io/img/cover/scRNA.png">
<meta property="article:published_time" content="2021-04-26T08:46:16.000Z">
<meta property="article:modified_time" content="2021-05-06T05:18:42.514Z">
<meta property="article:author" content="yilingzhang">
<meta property="article:tag" content="scRNA-seq">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://hiiiyilingzhang.github.io/img/cover/scRNA.png"><link rel="shortcut icon" href="/img/love_32px.png"><link rel="canonical" href="https://hiiiyilingzhang.github.io/2021/04/26/scRNAseq-01-Read-Data-and-Aggregation/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    jQuery: 'https://cdn.jsdelivr.net/npm/jquery@latest/dist/jquery.min.js',
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/js/jquery.justifiedGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/css/justifiedGallery.min.css'
    },
    fancybox: {
      js: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.js',
      css: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isanchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = { 
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2021-05-06 13:18:42'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    })(window)</script><meta name="generator" content="Hexo 5.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="author-avatar"><img class="avatar-img" src="/img/selfie.JPG" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data"><div class="data-item is-center"><div class="data-item-link"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">37</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">16</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">14</div></a></div></div></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/img/cover/scRNA.png')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Jocelyn's Blog</a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">scRNAseq-01 Read Data and Aggregation</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2021-04-26T08:46:16.000Z" title="Created 2021-04-26 16:46:16">2021-04-26</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2021-05-06T05:18:42.514Z" title="Updated 2021-05-06 13:18:42">2021-05-06</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/Single-Cell/">Single Cell</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">Word count:</span><span class="word-count">1.8k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">Reading time:</span><span>8min</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1 id="拆库"><a href="#拆库" class="headerlink" title="拆库"></a>拆库</h1><h2 id="拆index（lane-–-gt-library）"><a href="#拆index（lane-–-gt-library）" class="headerlink" title="拆index（lane –&gt; library）"></a>拆index（lane –&gt; library）</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 查看序列与Index对应关系，通常为Index反向互补，因为在右端</span></span><br><span class="line">zcat lane1_1.fq.gz | head</span><br><span class="line">zcat lane1_2.fq.gz | head</span><br><span class="line"><span class="comment"># 按index列表文件并行筛选每个文库的双端文件</span></span><br><span class="line"><span class="comment"># zcat解压、grep -A 3检索并输出4行，去除grep的添加字符行，gzip压缩，写入文件</span></span><br><span class="line">mkdir -p lib</span><br><span class="line"><span class="keyword">for</span> index <span class="keyword">in</span> `cut -f 1 index.txt`; <span class="keyword">do</span>  <span class="built_in">echo</span> <span class="variable">$&#123;index&#125;</span></span><br><span class="line">zcat lane1_1.fq.gz|grep -A 3 <span class="string">&quot;#<span class="variable">$&#123;index&#125;</span>&quot;</span>|grep -v -P <span class="string">&#x27;^--$&#x27;</span> |gzip &gt; lib/<span class="variable">$&#123;index&#125;</span>_1.fq.gz &amp;</span><br><span class="line">zcat lane1_2.fq.gz|grep -A 3 <span class="string">&quot;#<span class="variable">$&#123;index&#125;</span>&quot;</span>|grep -v -P <span class="string">&#x27;^--$&#x27;</span> |gzip &gt; lib/<span class="variable">$&#123;index&#125;</span>_2.fq.gz &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="comment"># cut -f 1 index.txt 以制表符为单位，指定显示index.txt的第一列</span></span><br><span class="line"><span class="comment"># grep -A 3 &quot;#$&#123;index&#125;&quot; 保留.fq文件每次grep出的前3行（信息+序列+质量分数）</span></span><br><span class="line"><span class="comment"># grep -v -P &#x27;^--$&#x27; -v显示不包含匹配文本的所有行。^ 和 $ 分别指字符串的开始与结束。取出没有匹配到--的</span></span><br></pre></td></tr></table></figure>
<h2 id="拆Barcode-（library-–-gt-sample）"><a href="#拆Barcode-（library-–-gt-sample）" class="headerlink" title="拆Barcode （library –&gt; sample）"></a>拆Barcode （library –&gt; sample）</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">gunzip library1*.gz</span><br><span class="line"></span><br><span class="line"><span class="comment"># 按实验文库设计library.txt拆分样品</span></span><br><span class="line">mkdir -p samples</span><br><span class="line"><span class="keyword">for</span> index <span class="keyword">in</span> `tail -n+2 library1.txt|cut -f 2`; <span class="keyword">do</span></span><br><span class="line">  <span class="built_in">echo</span> <span class="variable">$&#123;index&#125;</span></span><br><span class="line">  grep -A 2 -B 1 -P <span class="string">&quot;^<span class="variable">$&#123;index&#125;</span>&quot;</span> library1_1.fq|grep -v -P <span class="string">&#x27;^--$&#x27;</span> &gt; samples/<span class="variable">$&#123;index&#125;</span>_1.fq</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># tail -n+2 library1.txt|cut -f 2 从第2行至文件末尾，取出第二列（Fbarcode）</span></span><br><span class="line"><span class="comment"># grep -A 2 -B 1 -P &quot;^$&#123;index&#125;&quot;  除了匹配的那一行，显示后两行（-A 2）和前1行（-B 1）</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> index <span class="keyword">in</span> `tail -n+2 library1.txt|cut -f 3`; <span class="keyword">do</span></span><br><span class="line">  <span class="built_in">echo</span> <span class="variable">$&#123;index&#125;</span></span><br><span class="line">  grep -A 2 -B 1 -P <span class="string">&quot;^<span class="variable">$&#123;index&#125;</span>&quot;</span> library1_2.fq|grep -v -P <span class="string">&#x27;^--$&#x27;</span> &gt; samples/<span class="variable">$&#123;index&#125;</span>_2.fq</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<h1 id="CellRanger"><a href="#CellRanger" class="headerlink" title="CellRanger"></a>CellRanger</h1><p>软件介绍：<a target="_blank" rel="noopener" href="https://support.10xgenomics.com/single-cell-gene-expression/software/overview/welcome">https://support.10xgenomics.com/single-cell-gene-expression/software/overview/welcome</a><br>mkfastq参考代码：<a target="_blank" rel="noopener" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/mkfastq">https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/mkfastq</a></p>
<h1 id="读取数据"><a href="#读取数据" class="headerlink" title="读取数据"></a>读取数据</h1><h2 id="Read10X"><a href="#Read10X" class="headerlink" title="Read10X()"></a>Read10X()</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">library(Seurat)</span><br><span class="line">library(tidyverse) </span><br><span class="line">library(Matrix) </span><br><span class="line"></span><br><span class="line"><span class="comment">## Read10X </span></span><br><span class="line">seurat_data &lt;- Read10X(data.dir = <span class="string">&quot;./read10X&quot;</span>) </span><br><span class="line">seurat_obj &lt;- CreateSeuratObject(counts = seurat_data) </span><br><span class="line"><span class="comment"># write.table(seurat_obj@assays$RNA@counts, &quot;all.datatable.txt&quot;,sep=&quot;\t&quot;, quote=F, col.names=NA) </span></span><br></pre></td></tr></table></figure>
<h2 id="Read10X-h5"><a href="#Read10X-h5" class="headerlink" title="Read10X_h5()"></a>Read10X_h5()</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## Read10X_h5 </span></span><br><span class="line">seurat_data &lt;- Read10X_h5(<span class="string">&quot;filtered_feature_bc_matrix.h5&quot;</span>)</span><br><span class="line"></span><br><span class="line">seurat_obj &lt;- CreateSeuratObject(counts = seurat_data) </span><br></pre></td></tr></table></figure>
<h2 id="read-table"><a href="#read-table" class="headerlink" title="read.table()"></a>read.table()</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">matrix_data &lt;- read.table(<span class="string">&quot;all.datatable.txt&quot;</span>, sep=<span class="string">&quot;\t&quot;</span>, header=<span class="built_in">T</span>, row.names=<span class="number">1</span>) </span><br><span class="line"><span class="comment"># dim(matrix_data) </span></span><br><span class="line"><span class="comment"># matrix_data[1:4,1:4] </span></span><br><span class="line">seurat_obj &lt;- CreateSeuratObject(counts = matrix_data) </span><br></pre></td></tr></table></figure>
<h1 id="整合数据"><a href="#整合数据" class="headerlink" title="整合数据"></a>整合数据</h1><h2 id="Seurat-v3"><a href="#Seurat-v3" class="headerlink" title="Seurat v3"></a>Seurat v3</h2><p><strong>CCA+MNN (mutual nearest neighbor)</strong></p>
<blockquote>
<p>Seurat早期版本整合数据的核心算法是CCA，文章发表在2018年的nature biotechnology，作者是Seurat的开发者Andrew Butler。<br> 同年Haghverdi等人开发了MNN算法校正批次效应，文章也发表在了nature biotechnology。<br> 2019年Andrew等人将CCA与MNN算法结合起来，并参考SNN算法的理念设计了“锚点”评分体系，使Seurat整合数据更强大更稳健。它不仅可以校正实验的批次效应，还能跨平台整合数据，例如将10x单细胞数据、BD单细胞数据和SMART单细胞数据整合在一起；也能整合单细胞多组学数据，例如将单细胞ATAC、空间转录组与单细胞转录组数据整合在一起。</p>
</blockquote>
<p><strong>整合流程与原理</strong><br>These methods first identify cross-dataset pairs of cells that are in a matched biological state (‘anchors’), can be used both to correct for technical differences between datasets (i.e. batch effect correction), and to perform comparative scRNA-seq analysis of across experimental conditions.</p>
<ol>
<li>split the dataset into a list of two seurat objects (stim and CTRL)</li>
<li>normalize and identify variable features for each dataset independently<br> Performing log-normalization<br> Calculate gene variances<br> Calculate feature variances of standardized and clipped values</li>
<li>select features that are repeatedly variable across datasets for integration</li>
<li>identify anchors using the <code>FindIntegrationAnchors()</code> function, which takes a list of Seurat objects as input:<br> Scale features for provided ohjects<br> Find all pariwise anchors<br> Merge objects<br> Find neighborhoods<br> Find anchors<br> Filter anchors<br>and use these anchors to integrate the two datasets together with <code>IntegrateData()</code>:<br> Extract anchors for merged samples<br> Find integration vectors<br> Find integration vectors weights<br> Integrate data</li>
</ol>
<h2 id="normalized-with-SCTransform"><a href="#normalized-with-SCTransform" class="headerlink" title="normalized with SCTransform"></a>normalized with SCTransform</h2><p>In <a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1">Hafemeister and Satija, 2019</a>, we introduced an improved method for the normalization of scRNA-seq, based on regularized negative binomial regression.</p>
<p><strong>Key info：</strong></p>
<ul>
<li>cellular sequencing depth is utilized as a covariate in a generalized linear model</li>
<li>an unconstrained negative binomial model may overfit scRNA-seq data, and overcome this by pooling information across genes with similar abundances to obtain stable parameter estimates.</li>
<li> We instead propose to construct a generalized linear model (GLM) for each gene with UMI counts as the response and sequencing depth as the explanatory variable. We explore potential error models for the GLM and find that the use of unconstrained NB or ZINB models leads to overfitting of scRNA-seq data and a significant dampening of biological variance. </li>
<li>effectively normalized data values that are no longer influenced by technical characteristics, but preserve heterogeneity driven by distinct biological states. </li>
</ul>
<p><strong>Workflow:</strong></p>
<ul>
<li>Normalize datasets individually by <code>SCTransform()</code>, instead of <code>NormalizeData()</code> prior to integration</li>
<li>As discussed further in our <a href="sctransform_vignette.html">SCTransform vignette</a>, we typically use 3,000 or more features for  analysis downstream of sctransform.</li>
<li>Run the <code>PrepSCTIntegration()</code> function prior to identifying anchors</li>
<li>When running <code>FindIntegrationAnchors()</code>, and <code>IntegrateData()</code>, set the <code>normalization.method</code> parameter to the value <code>SCT</code>.</li>
<li>When running sctransform-based workflows, including integration, do not run the <code>ScaleData()</code> function </li>
</ul>
<h2 id="Seurat-v4"><a href="#Seurat-v4" class="headerlink" title="Seurat v4"></a>Seurat v4</h2><ul>
<li>Independent preprocessing and dimensional reduction of each modality individually </li>
<li>Learning cell-specific modality ‘weights’, and constructing a WNN graph that integrates the modalities </li>
<li>Downstream analysis (i.e. visualization, clustering, etc.) of the WNN graph</li>
<li>Our approach is based on an unsupervised strategy to learn cell-specific modality ‘weights’, which reflect the information content for each modality, and determine its relative importance in downstream analyses. </li>
</ul>
<p><strong>Workflow:</strong></p>
<ol>
<li><p>Perform pre-processing and dimensional reduction on both assays independently. Standard normalization/SCTransform</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">DefaultAssay(bm) &lt;- <span class="string">&#x27;RNA&#x27;</span></span><br><span class="line">bm &lt;- NormalizeData(bm) %&gt;% FindVariableFeatures() %&gt;% ScaleData() %&gt;% RunPCA()</span><br><span class="line"></span><br><span class="line">DefaultAssay(bm) &lt;- <span class="string">&#x27;ADT&#x27;</span></span><br><span class="line"><span class="comment"># we will use all ADT features for dimensional reduction</span></span><br><span class="line"><span class="comment"># we set a dimensional reduction name to avoid overwriting the </span></span><br><span class="line">VariableFeatures(bm) &lt;- rownames(bm[[<span class="string">&quot;ADT&quot;</span>]])</span><br><span class="line">bm &lt;- NormalizeData(bm, normalization.method = <span class="string">&#x27;CLR&#x27;</span>, margin = <span class="number">2</span>) %&gt;% </span><br><span class="line">  ScaleData() %&gt;% RunPCA(reduction.name = <span class="string">&#x27;apca&#x27;</span>)</span><br></pre></td></tr></table></figure></li>
<li><p>For each cell, we calculate its closest neighbors in the dataset based on a weighted combination of RNA and protein similarities. </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"> <span class="comment"># Identify multimodal neighbors. These will be stored in the neighbors slot, </span></span><br><span class="line"> <span class="comment"># and can be accessed using bm[[&#x27;weighted.nn&#x27;]]</span></span><br><span class="line"> <span class="comment"># The WNN graph can be accessed at bm[[&quot;wknn&quot;]], </span></span><br><span class="line"> <span class="comment"># and the SNN graph used for clustering at bm[[&quot;wsnn&quot;]]</span></span><br><span class="line"> <span class="comment"># Cell-specific modality weights can be accessed at bm$RNA.weight</span></span><br><span class="line"> bm &lt;- FindMultiModalNeighbors(</span><br><span class="line">  bm, reduction.list = <span class="built_in">list</span>(<span class="string">&quot;pca&quot;</span>, <span class="string">&quot;apca&quot;</span>), </span><br><span class="line">  dims.list = <span class="built_in">list</span>(<span class="number">1</span>:<span class="number">30</span>, <span class="number">1</span>:<span class="number">18</span>),      modality.weight.name = <span class="string">&quot;RNA.weight&quot;</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<h2 id="Harmony"><a href="#Harmony" class="headerlink" title="Harmony"></a>Harmony</h2><p>基本原理：我们用不同颜色表示不同数据集，用形状表示不同的细胞类型。<br>首先，Harmony应用主成分分析将转录组表达谱嵌入到低维空间中，然后应用迭代过程去除数据集特有的影响。</p>
</li>
<li><p>Harmony概率性地将细胞分配给cluster，从而使每个cluster内数据集的多样性最大化。 </p>
</li>
<li><p>Harmony计算每个cluster的所有数据集的全局中心，以及特定数据集的中心。 </p>
</li>
<li><p>在每个cluster中，Harmony基于中心为每个数据集计算校正因子。 </p>
</li>
<li><p>最后，Harmony使用基于C的特定于细胞的因子校正每个细胞。由于Harmony使用软聚类，因此可以通过多个因子的线性组合对其A中进行的软聚类分配进行线性校正，来修正每个单细胞。 重复步骤A到D，直到收敛为止。聚类分配和数据集之间的依赖性随着每一轮的减少而减小。</p>
</li>
</ol>
<p><strong>Harmony</strong>算法与其他整合算法相比的优势：<br>（1）整合数据的同时对稀有细胞的敏感性依然很好；<br>（2）省内存；<br>（3）适合于更复杂的单细胞分析实验设计，可以比较来自不同供体，组织和技术平台的细胞。</p>
<h1 id="数据结构"><a href="#数据结构" class="headerlink" title="数据结构"></a>数据结构</h1><center><img src="/img/scRNA/SingleCellExperiment.png" alt="warning" style="zoom:100%;" /></center>
所述SingleCellExperiment（sce）对象是基于在Bioconductor的单细胞分析应用的基础。该sce对象是一个S4对象，与R中其他可用的方法相比，它本质上为数据的构造和访问提供了一种更为形式化的方法。

<p>图中最核心的部分，是蓝色的data部分；另外还有绿色的基因注释信息feature metadata、橙色的细胞注释信息cell metadata。除了这三大件，还会包含一些下游分析结果，比如PCA、tSNE降维结果就会保存在紫色的Dimension Reductions</p>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="mailto:undefined">yilingzhang</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="https://hiiiyilingzhang.github.io/2021/04/26/scRNAseq-01-Read-Data-and-Aggregation/">https://hiiiyilingzhang.github.io/2021/04/26/scRNAseq-01-Read-Data-and-Aggregation/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/scRNA-seq/">scRNA-seq</a></div><div class="post_share"><div class="social-share" data-image="/img/cover/scRNA.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/social-share.js/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/social-share.js/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2021/04/26/scRNAseq-02-Annotation/"><img class="prev-cover" src="/img/cover/scRNA.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">scRNAseq-02 Annotation</div></div></a></div><div class="next-post pull-right"><a href="/2021/04/19/all-about-UMI-tools/"><img class="next-cover" src="/img/cover/UMItools.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">all about UMI-tools</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span> Related Articles</span></div><div class="relatedPosts-list"><div><a href="/2021/04/26/scRNAseq-02-Annotation/" title="scRNAseq-02 Annotation"><img class="cover" src="/img/cover/scRNA.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2021-04-26</div><div class="title">scRNAseq-02 Annotation</div></div></a></div><div><a href="/2021/05/10/scRNAseq-03-Doublet-velocyto-inferCNV/" title="scRNAseq-03 Doublet & velocyto & inferCNV"><img class="cover" src="/img/cover/scRNA.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2021-05-10</div><div class="title">scRNAseq-03 Doublet & velocyto & inferCNV</div></div></a></div><div><a href="/2021/05/18/scRNAseq-04-Cell-cell-communication/" title="scRNAseq-04 Cell-cell communication"><img class="cover" src="/img/cover/scRNA.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2021-05-18</div><div class="title">scRNAseq-04 Cell-cell communication</div></div></a></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="card-info-avatar is-center"><img class="avatar-img" src="/img/selfie.JPG" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/><div class="author-info__name">yilingzhang</div><div class="author-info__description">Every Day is My Day</div></div><div class="card-info-data"><div class="card-info-data-item is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">37</div></a></div><div class="card-info-data-item is-center"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">16</div></a></div><div class="card-info-data-item is-center"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">14</div></a></div></div><div class="card-info-social-icons is-center"><a class="social-icon" href="mailto:jocelynzhang1119@icloud.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a><a class="social-icon" href="https://blog.csdn.net/weixin_43893431" target="_blank" title="CSDN"><i class="fas fa-paper-plane"></i></a></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%8B%86%E5%BA%93"><span class="toc-number">1.</span> <span class="toc-text">拆库</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%8B%86index%EF%BC%88lane-%E2%80%93-gt-library%EF%BC%89"><span class="toc-number">1.1.</span> <span class="toc-text">拆index（lane –&gt; library）</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%8B%86Barcode-%EF%BC%88library-%E2%80%93-gt-sample%EF%BC%89"><span class="toc-number">1.2.</span> <span class="toc-text">拆Barcode （library –&gt; sample）</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#CellRanger"><span class="toc-number">2.</span> <span class="toc-text">CellRanger</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%AF%BB%E5%8F%96%E6%95%B0%E6%8D%AE"><span class="toc-number">3.</span> <span class="toc-text">读取数据</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#Read10X"><span class="toc-number">3.1.</span> <span class="toc-text">Read10X()</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Read10X-h5"><span class="toc-number">3.2.</span> <span class="toc-text">Read10X_h5()</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#read-table"><span class="toc-number">3.3.</span> <span class="toc-text">read.table()</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%95%B4%E5%90%88%E6%95%B0%E6%8D%AE"><span class="toc-number">4.</span> <span class="toc-text">整合数据</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#Seurat-v3"><span class="toc-number">4.1.</span> <span class="toc-text">Seurat v3</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#normalized-with-SCTransform"><span class="toc-number">4.2.</span> <span class="toc-text">normalized with SCTransform</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Seurat-v4"><span class="toc-number">4.3.</span> <span class="toc-text">Seurat v4</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Harmony"><span class="toc-number">4.4.</span> <span class="toc-text">Harmony</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84"><span class="toc-number">5.</span> <span class="toc-text">数据结构</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2021/07/23/scRNA-seq-Project-LIHC-microenvironment/" title="scRNA-seq Project -- LIHC microenvironment"><img src="/img/cover/scRNA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scRNA-seq Project -- LIHC microenvironment"/></a><div class="content"><a class="title" href="/2021/07/23/scRNA-seq-Project-LIHC-microenvironment/" title="scRNA-seq Project -- LIHC microenvironment">scRNA-seq Project -- LIHC microenvironment</a><time datetime="2021-07-23T14:09:24.000Z" title="Created 2021-07-23 22:09:24">2021-07-23</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/07/18/Get-Start-With-Bioinfo-Pipeline/" title="Get Start With Bioinfo Pipeline"><img src="/img/cover/pipeline.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Get Start With Bioinfo Pipeline"/></a><div class="content"><a class="title" href="/2021/07/18/Get-Start-With-Bioinfo-Pipeline/" title="Get Start With Bioinfo Pipeline">Get Start With Bioinfo Pipeline</a><time datetime="2021-07-18T11:59:13.000Z" title="Created 2021-07-18 19:59:13">2021-07-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/24/Tumor-immune-infiltration-analysis/" title="Tumor immune infiltration analysis"><img src="/img/cover/Immue.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Tumor immune infiltration analysis"/></a><div class="content"><a class="title" href="/2021/05/24/Tumor-immune-infiltration-analysis/" title="Tumor immune infiltration analysis">Tumor immune infiltration analysis</a><time datetime="2021-05-24T10:51:35.000Z" title="Created 2021-05-24 18:51:35">2021-05-24</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/24/scRNAseq-05-Monocle3-GO-KEGG-TCGASurv/" title="scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv"><img src="/img/cover/scRNA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv"/></a><div class="content"><a class="title" href="/2021/05/24/scRNAseq-05-Monocle3-GO-KEGG-TCGASurv/" title="scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv">scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv</a><time datetime="2021-05-24T04:51:09.000Z" title="Created 2021-05-24 12:51:09">2021-05-24</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/18/TCGA-data-mining/" title="TCGA data mining"><img src="/img/cover/TCGA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="TCGA data mining"/></a><div class="content"><a class="title" href="/2021/05/18/TCGA-data-mining/" title="TCGA data mining">TCGA data mining</a><time datetime="2021-05-18T12:01:08.000Z" title="Created 2021-05-18 20:01:08">2021-05-18</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2021 By yilingzhang</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"></div><script defer="defer" id="ribbon" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/canvas-ribbon.min.js" size="150" alpha="0.6" zIndex="-1" mobile="false" data-click="true"></script><script id="click-heart" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/click-heart.min.js" async="async" mobile="false"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>