<!DOCTYPE html><html lang="en | zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1"><title>WES Project | Jocelyn's Blog</title><meta name="author" content="yilingzhang"><meta name="copyright" content="yilingzhang"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="p{     font-size:18px;     line-height:1.7em;    }   环境配置GATK本次作业使用的GATK版本为4.2.0.0数据库文件来源：&#x2F;public&#x2F;workspace&#x2F;lincs&#x2F;wes_cancer&#x2F;data github-GATK官网 由于GATK的更新，ftp链接于2020年6月1日关闭，所有数据库文件都迁移至谷歌云存储：Google C">
<meta property="og:type" content="article">
<meta property="og:title" content="WES Project">
<meta property="og:url" content="http://yilingzhang.gitee.io/2021/03/30/WES-Project/index.html">
<meta property="og:site_name" content="Jocelyn&#39;s Blog">
<meta property="og:description" content="p{     font-size:18px;     line-height:1.7em;    }   环境配置GATK本次作业使用的GATK版本为4.2.0.0数据库文件来源：&#x2F;public&#x2F;workspace&#x2F;lincs&#x2F;wes_cancer&#x2F;data github-GATK官网 由于GATK的更新，ftp链接于2020年6月1日关闭，所有数据库文件都迁移至谷歌云存储：Google C">
<meta property="og:locale">
<meta property="og:image" content="http://yilingzhang.gitee.io/img/cover/WES.png">
<meta property="article:published_time" content="2021-03-30T08:38:50.000Z">
<meta property="article:modified_time" content="2021-04-26T09:03:35.957Z">
<meta property="article:author" content="yilingzhang">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://yilingzhang.gitee.io/img/cover/WES.png"><link rel="shortcut icon" href="/img/love_32px.png"><link rel="canonical" href="http://yilingzhang.gitee.io/2021/03/30/WES-Project/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    jQuery: 'https://cdn.jsdelivr.net/npm/jquery@latest/dist/jquery.min.js',
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/js/jquery.justifiedGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/css/justifiedGallery.min.css'
    },
    fancybox: {
      js: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.js',
      css: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isanchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = { 
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2021-04-26 17:03:35'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    })(window)</script><meta name="generator" content="Hexo 5.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="author-avatar"><img class="avatar-img" src="/img/selfie.JPG" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data"><div class="data-item is-center"><div class="data-item-link"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">32</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">14</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">13</div></a></div></div></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/img/cover/WES.png')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Jocelyn's Blog</a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">WES Project</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2021-03-30T08:38:50.000Z" title="Created 2021-03-30 16:38:50">2021-03-30</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2021-04-26T09:03:35.957Z" title="Updated 2021-04-26 17:03:35">2021-04-26</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/WES/">WES</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">Word count:</span><span class="word-count">14.8k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">Reading time:</span><span>66min</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><style>
  p{
    font-size:18px;
    line-height:1.7em;

  }
</style>

<h1 id="环境配置"><a href="#环境配置" class="headerlink" title="环境配置"></a>环境配置</h1><h2 id="GATK"><a href="#GATK" class="headerlink" title="GATK"></a>GATK</h2><p>本次作业使用的GATK版本为4.2.0.0<br>数据库文件来源：<code>/public/workspace/lincs/wes_cancer/data</code></p>
<p><a target="_blank" rel="noopener" href="https://github.com/broadinstitute/gatk/releases">github-GATK</a><br><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us">官网</a></p>
<p>由于GATK的更新，ftp链接于2020年6月1日关闭，所有数据库文件都迁移至谷歌云存储：<a target="_blank" rel="noopener" href="https://console.cloud.google.com/storage/browser/genomics-public-data/resources/broad/hg38/v0;tab=objects?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&prefix=&forceOnObjectsSortingFiltering=false">Google Cloud-genomics public data</a><br>文件下载需要先conda/mamba安装gsutil</p>
<center><img src="/img/Figures/googlecloud.png" alt="googlecloud" style="zoom:100%;" /></center>

<blockquote>
<p>阿里云和Broad Institute合作提供了GATK软件分析流程。由Broad Institute提供的GATK流程最佳实践用工作流定义语言（WDL）编写，通过批量计算集成的Cromwell工作流引擎解析执行。用户将为作业运行时实际消耗的计算和存储资源付费，不需要支付资源之外的附加费用。需要先开通阿里云的OSS并创建好Bucket，输入、输出文件都需要保存在OSS。<br>  <a target="_blank" rel="noopener" href="https://help.aliyun.com/document_detail/60414.html?spm=a2c4g.11186623.6.607.63f83e36wO8PHK">阿里云-GATK支持</a><br>  <center><img src="/img/Figures/alibaba.png" alt="alibaba-GATK" style="zoom:100%;" /></center></p>
</blockquote>
<center><img src="/img/Figures/gatk-pipeline.png" alt="gatk-pipeline" style="zoom:65%;" /></center>

<h2 id="Reference"><a href="#Reference" class="headerlink" title="Reference"></a>Reference</h2><center><img src="/img/Figures/recal-warning.png" alt="recal-warning" style="zoom:100%;" /></center>

<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># reference</span></span><br><span class="line">ascp  -i ~/.aspera/connect/etc/asperaweb_id_dsa.openssh \</span><br><span class="line">-l 100M -k 1 -Q -T \</span><br><span class="line">anonftp@ftp.ncbi.nlm.nih.gov:/snp/archive/b153/VCF/GCF_000001405.38.gz \</span><br><span class="line">./Reference/GATK/</span><br><span class="line"><span class="comment"># index</span></span><br><span class="line">ascp  -i ~/.aspera/connect/etc/asperaweb_id_dsa.openssh \</span><br><span class="line">-l 100M -k 1 -Q -T \</span><br><span class="line">anonftp@ftp.ncbi.nlm.nih.gov:/snp/archive/b153/VCF/GCF_000001405.38.gz.tbi \</span><br><span class="line">./Reference/dbsnp_153.hg38/</span><br></pre></td></tr></table></figure>
<h1 id="数据下载"><a href="#数据下载" class="headerlink" title="数据下载"></a>数据下载</h1><h2 id="NCBI-prefetch"><a href="#NCBI-prefetch" class="headerlink" title="NCBI+prefetch"></a>NCBI+prefetch</h2><p>在NCBI页面查找数据后点击Run Selector，选择需要的数据，下载Metadata和Accession List两个文件</p>
 <center><img src="/img/Figures/SRA-run.png" alt="SRA-run" style="zoom:45%;" /></center>

<p>依次读取SRR_Acc_List.txt的每一行，利用prefetch下载。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">cat 0.sra/sra.infor/SRR_Acc_List.txt | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  prefetch <span class="variable">$&#123;id&#125;</span> -O  ./0.sra</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<h2 id="ENA-Aspera"><a href="#ENA-Aspera" class="headerlink" title="ENA+Aspera"></a>ENA+Aspera</h2><p><a target="_blank" rel="noopener" href="https://www.ebi.ac.uk/ena">ENA数据库: European Nucleotide</a><br>隶属于EBI，功能上应该是与SRA类似的，但是其搜索界面更加亲民，并且对于下载fastq文件以及检查下载数据完整性更加友好。  </p>
<p><strong>1.</strong> 首先，在数据库页面右上角搜索栏输入目标SRA检索号，确认后稍等片刻可得结果页面；   </p>
<p><img src="/img/Figures/ena.png" alt="ena"></p>
<p><strong>2.</strong> 其次，点击选取 Read File 即可查看该实验下所有的测序序列数据的信息，选择需要查看的信息，这里的五列分别为：run_accession，library_name，fastq_ftp，sra_md5，sra_aspera；  </p>
<img src="/img/Figures/ena_page2.png" style="zoom:80%;" />

<p><strong>3.</strong> 点击TSV，下载对应文件；  </p>
<p><img src="/img/Figures/download.png" alt="download"><br><strong>4.</strong> 利用shell处理，保留所需信息。观察文件发现，case1包含case1…和case1…ampliseq… ，只保留case1…；  </p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 提取包含case1且不包含ampliseq的行</span></span><br><span class="line">grep case1 filereport_read_run_PRJNA310728_tsv.txt | grep -v ampliseq &gt; matched_file.txt</span><br><span class="line"></span><br><span class="line"><span class="comment"># 提取SRR号，样本名，sra.ftp地址</span></span><br><span class="line">awk <span class="string">&#x27;&#123;print $2,$1,$6&#125;&#x27;</span> matched_file.txt | sed <span class="string">&#x27;s/_WES//&#x27;</span> &gt; config.sra</span><br></pre></td></tr></table></figure>
<p><img src="/img/Figures/awk-result.png" alt="awk-result"><br><strong>5.</strong> 利用Aspera批量下载,6.8G的文件下载了1个小时，速度大约为16M/s;</p>
<blockquote>
<p>Aspera:<br>-v verbose mode 唠叨模式，能让你实时知道程序在干啥<br>-T 取消加密，否则有时候数据下载不了<br>-i 提供私钥文件的地址，我也不知道干嘛的，反正不能少，地址一般是~/.aspera/connect/etc/asperaweb_id_dsa.openssh<br>-l 设置最大传输速度，一般200m到500m，如果不设置，反而速度会比较低，可能有个较低的默认值<br>-k 断点续传，一般设置为值1<br>-Q 不懂，一般加上它<br>-P 提供SSH port，一般是33001</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">cat 0.sra/sra.infor/config.sra | <span class="keyword">while</span> <span class="built_in">read</span> id; \</span><br><span class="line"><span class="keyword">do</span> <span class="built_in">echo</span> <span class="variable">$id</span></span><br><span class="line">arr=(<span class="variable">$id</span>)</span><br><span class="line">link=<span class="variable">$&#123;arr[2]&#125;</span></span><br><span class="line">srr=<span class="variable">$&#123;arr[1]&#125;</span></span><br><span class="line">ascp -i ~/.aspera/connect/etc/asperaweb_id_dsa.openssh \</span><br><span class="line">    -l 300M -k 1 -Q -T -P 33001 \</span><br><span class="line">    era-fasp@<span class="variable">$&#123;link&#125;</span> \</span><br><span class="line">    ./0.sra/ &gt; <span class="built_in">log</span>/ascp.log/<span class="variable">$&#123;srr&#125;</span>.<span class="built_in">log</span> 2&gt;&amp;1 &amp; </span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/ascp_loop.png" alt="acsp_loop" style="zoom:80%;" /></center>


<center><img src="/img/Figures/ascp_speed.png" alt="ascp-speed" style="zoom:80%;" /></center>

<p><strong>6.</strong> 最后利用MD5码确认数据的完整性。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 提取样本列和MD5列</span></span><br><span class="line">awk <span class="string">&#x27;&#123;print $2,$4&#125;&#x27;</span> matched_file.txt | sed <span class="string">&#x27;s/_WES//&#x27;</span> &gt; sra.md5</span><br><span class="line"></span><br><span class="line"><span class="comment"># 利用md5sum命令生成下载数据的md5码</span></span><br><span class="line"><span class="comment"># md5sum</span></span><br><span class="line"><span class="comment"># -b：二进制模式读取文件；</span></span><br><span class="line"><span class="comment"># -t或--text：把输入的文件作为文本文件看待；</span></span><br><span class="line"><span class="comment"># -c：从指定文件中读取MD5校验和，并进行校验；</span></span><br><span class="line"><span class="comment"># --status：验证成功时不输出任何信息；</span></span><br><span class="line"><span class="comment"># -w：当校验不正确时给出警告信息。</span></span><br><span class="line"></span><br><span class="line">md5sum ./0.sra/*.sra &gt; ./0.sra/sra.infor/aft.md5</span><br><span class="line"></span><br><span class="line"><span class="comment"># 对比ENA提供的md5码和下载后的数据生成的md5码</span></span><br><span class="line">diff ./0.sra/sra.infor/aft.md5 ./0.sra/sra.infor/sra.md5</span><br><span class="line"></span><br><span class="line"><span class="comment"># ENA提供的md5文件和当前sra在同一目录时</span></span><br><span class="line"><span class="comment"># md5sum -c ENA提供的md5文件</span></span><br><span class="line"><span class="comment"># 返回0/1</span></span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/md5Check.png" alt="MD5" style="zoom:80%;" /></center>

<p>MD5码确认一致，下载的数据是完整无误的。</p>
<h2 id="ENA-Arial2-奶牛快传"><a href="#ENA-Arial2-奶牛快传" class="headerlink" title="ENA+Arial2+奶牛快传"></a>ENA+Arial2+奶牛快传</h2><p><strong>Aria2</strong><br><strong>优点</strong>:速度快、体积小、资源占用少；支持 HTTP/FTP/BT/Magnet磁力链接等类型的文件下载，据说速度超级快;<br><br><strong>缺点</strong>: 配置麻烦，上手较难。</p>
<p>教程：<br><a target="_blank" rel="noopener" href="https://github.com/aria2/aria2/releases/tag/release-1.35.0">github-aria2</a><br><a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/article/1142142">编译安装</a><br><a target="_blank" rel="noopener" href="https://blog.readgroup.cn/m/?post=92">一键安装脚本（需要root权限）</a></p>
<center><img src="/img/Figures/Aria2-wrong.png" alt="Aria-wrong" style="zoom:60%;" /></center>

<p>另外尝试下载win版Arial2，再利用奶牛快传传至服务器<br>教程：</p>
<p>抛弃迅雷，Aria2 新手入门](<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/37021947">https://zhuanlan.zhihu.com/p/37021947</a>)<br>用ftp地址，3G的数据下的超快。大概10分钟？感觉这边的速度是骗人的</p>
<center><img src="/img/Figures/Aria2-speed.png" alt="Aria-speed" style="zoom:60%;" /></center>

<p><a target="_blank" rel="noopener" href="https://cowtransfer.com/"><strong>奶牛快传</strong></a></p>
<ul>
<li>单次传输上限20GB  </li>
<li>3TB云盘空间  </li>
<li>上传下载不限速  </li>
<li>在网页上上传本地文件后，会提供下载链接和密码</li>
</ul>
<center><img src="/img/Figures/upload.png" alt="cow" style="zoom:45%;" /></center>

<p>需要下载linux版<br><a target="_blank" rel="noopener" href="https://github.com/Mikubill/cowtransfer-uploader">github-cowtransfer uploader</a><br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/5c29e8fcb0f0">Linux奶牛快传使用手册</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 从云端下载数据</span></span><br><span class="line"><span class="comment"># 输入提供的链接和密码</span></span><br><span class="line">cowtransfer-uploader --password=test1234 https://yilingzhang.cowtransfer.com/s/90c7eb9b3d1a44</span><br></pre></td></tr></table></figure>
<p>传到服务器8M/s的速度太感人了！！还可以加线程！</p>
<center><img src="/img/Figures/toserver.png" alt="cow-to-server" style="zoom:100%;" /></center>

<h1 id="质控与去接头"><a href="#质控与去接头" class="headerlink" title="质控与去接头"></a>质控与去接头</h1><h2 id="文件格式转换"><a href="#文件格式转换" class="headerlink" title="文件格式转换"></a>文件格式转换</h2><h3 id="fastq-dump"><a href="#fastq-dump" class="headerlink" title="fastq-dump"></a>fastq-dump</h3><p>官方文档：<br><a target="_blank" rel="noopener" href="https://ncbi.github.io/sra-tools/fastq-dump.html">github: fastq-dump</a><br>教程：<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/a8d70b66794c">Fastq-dump: 一个神奇的软件</a></p>
<blockquote>
<p>常用参数:<br>split-spot: 将双端测序分为两份,但是都放在同一个文件中<br>split-files: 将双端测序分为两份,放在不同的文件,但是对于一方有而一方没有的reads直接丢弃<br>split-3 : 将双端测序分为两份,放在不同的文件,但是对于一方有而一方没有的reads会单独放在一个文件夹里<br>–gzip 输出文件以gz压缩<br>-A 输出文件名<br>-O 输出目录</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># fastq-dump--sratoolkit-2.9.2</span></span><br><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line">cat 0.sra/sra.infor/sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span> </span><br><span class="line">time fastq-dump -A <span class="variable">$&#123;id&#125;</span> -O 1.rawfastq/ \</span><br><span class="line">--gzip --split-3 0.sra/<span class="variable">$&#123;id&#125;</span>.sra &gt; <span class="built_in">log</span>/sra2fq.log/<span class="variable">$&#123;id&#125;</span>.<span class="built_in">log</span> 2&gt;&amp;1 &amp;</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<h3 id="faterq-dump"><a href="#faterq-dump" class="headerlink" title="faterq-dump"></a>faterq-dump</h3><p><code>fastq-dump</code>只能单个线程，速度特别的慢。2018年的6月份，sra-tools更新了一个新的sra解压工具，<code>fasterq-dump</code>，能利用临时文件和多线程加速从SRA文件提取FASTQ。</p>
<p><strong>注意：</strong><code>fasterq-dump</code>不提供–gzip的参数用于压缩文件，提取出来的fastq文件只是文本文件，记得压缩，减小存储压力。（原文：There is no –gzip|–bizp2 option, you have to compress your files explicitly after they have been written.）</p>
<p>官方文档：<br><a target="_blank" rel="noopener" href="https://github.com/ncbi/sra-tools/wiki/HowTo:-fasterq-dump">github: fasterq-dump</a></p>
<p>教程：<br><a target="_blank" rel="noopener" href="https://www.plob.org/article/14565.html">fasterq-dump使用介绍</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># fasterq-dump-sratoolkit-2.9.2</span></span><br><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line">cat 0.sra/sra.infor/sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  time fasterq-dump -3 -e 5 ./0.sra/<span class="variable">$&#123;id&#125;</span>.sra -O ./1.rawfastq --outfile <span class="variable">$&#123;id&#125;</span>.fastq  &gt; ./<span class="built_in">log</span>/sra2fq.log/<span class="variable">$&#123;id&#125;</span>.<span class="built_in">log</span> 2&gt;&amp;1</span><br><span class="line">  time pigz -p 3 -f ./1.rawfastq/<span class="variable">$&#123;id&#125;</span>_1.fastq &gt; ./1.rawfastq/<span class="variable">$&#123;id&#125;</span>_1.fastq.gz</span><br><span class="line">  time pigz -p 3 -f ./1.rawfastq/<span class="variable">$&#123;id&#125;</span>_2.fastq &gt; ./1.rawfastq/<span class="variable">$&#123;id&#125;</span>_2.fastq.gz</span><br><span class="line"><span class="keyword">done</span> </span><br></pre></td></tr></table></figure>
<hr>
<blockquote>
<p><strong>Troubleshooting</strong><br>err: cmn_iter.c cmn_iter_open_db().VDBManagerOpenDBRead( ‘case1_germline.sra’ ) -&gt; RC(rcVFS,rcMgr,rcOpening,rcDirectory,rcNotFound)<br>2021-03-18T09:53:03 fasterq-dump.2.9.2 err: sorter.c run_producer_pool(): row_count == 0!<br>2021-03-18T09:53:03 fasterq-dump.2.9.2 err: sorter.c execute_lookup_production() -&gt; RC(rcVDB,rcNoTarg,rcConstructing,rcParam,rcInvalid)2021-03-18T09:53:03 fasterq-dump.2.9.2 err: fasterq-dump.c produce_lookup_files() -&gt;<br> RC(rcVDB,rcNoTarg,rcConstructing,rcParam,rcInvalid)#<br><strong>Replied by contributor:</strong> Perhaps it should strip the .sra from the leaf name before creating .fastq output<br></p>
</blockquote>
<hr>
<h2 id="质控"><a href="#质控" class="headerlink" title="质控"></a>质控</h2><p>通过<code>config</code>文件进行批量操作</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># FastQC-v0.11.8 </span></span><br><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line">cat 0.sra/sra.infor/sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  fastqc --outdir ./2.qc/raw.qc/ --threads 5 ./1.rawfastq/<span class="variable">$&#123;id&#125;</span>*.fastq.gz &gt;&gt; ./<span class="built_in">log</span>/qc.log/rawqc/<span class="variable">$&#123;id&#125;</span>_fastqc.log 2&gt;&amp;1 &amp; </span><br><span class="line"><span class="keyword">done</span> </span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/fastqc.png" alt="fastqc" style="zoom:90%;" /></center>

<p>利用<code>multiqc</code>软件对所有样本的结果进行整合，方便查看。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># MultiQC-v1.9</span></span><br><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line">multiqc  ./2.qc/raw.qc/*zip  -o ./2.qc/raw.qc/multiqc</span><br></pre></td></tr></table></figure>
<p>输出文件: ./Supplementary/raw_multiqc_report.html</p>
<center><img src="/img/Figures/raw_status_check.png" alt="raw-status" style="zoom:90%;" /></center>

<p>发现测序质量非常好，怀疑已经经过一定的处理</p>
<h2 id="fastp"><a href="#fastp" class="headerlink" title="fastp"></a>fastp</h2><center><img src="/img/Figures/adaptor.png" alt="adapter" style="zoom:90%;" /></center>

<p><code>fastqc</code>和<code>MultiQC</code>给出的报告均显示未发现接头，这里用<code>fastp</code>的默认参数处理。教程中用的<code>trim_galore</code>之前在做ATAC-seq是使用过，跑的非常非常慢。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># fastp-v0.20.0</span></span><br><span class="line"><span class="comment"># -------------------------------&gt; </span></span><br><span class="line">cat 0.sra/sra.infor/sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  fastp -i ./1.rawfastq/<span class="variable">$&#123;id&#125;</span>_1.fastq.gz -o ./3.cleanfastq/<span class="variable">$&#123;id&#125;</span>_1.fastq.gz \</span><br><span class="line">        -I ./1.rawfastq/<span class="variable">$&#123;id&#125;</span>_2.fastq.gz -O ./3.cleanfastq/<span class="variable">$&#123;id&#125;</span>_2.fastq.gz \</span><br><span class="line">        --thread=6 \</span><br><span class="line">        &gt; <span class="built_in">log</span>/fastp.log/<span class="variable">$&#123;id&#125;</span>.fastp.log 2&gt;&amp;1 &amp;</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<p>以case1_germline样本为例：</p>
<center><img src="/img/Figures/g1_fastp.png" alt="adapter" style="zoom:80%;" /></center>

<p>发现还是有一定效果的。</p>
<h1 id="比对"><a href="#比对" class="headerlink" title="比对"></a>比对</h1><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># bwa-v0.7.17</span></span><br><span class="line"><span class="comment"># -------------------------------&gt; </span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 索引文件：</span></span><br><span class="line">INDEX=/public/workspace/lincs/wes_cancer/data/gatk_hg38</span><br><span class="line"></span><br><span class="line">cat 0.sra/sra.infor/sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  <span class="built_in">echo</span> <span class="string">&quot;start bwa for <span class="variable">$&#123;id&#125;</span>&quot;</span> `date`</span><br><span class="line">  fq1=./3.cleanfastq/<span class="variable">$&#123;id&#125;</span>_1.fq.gz</span><br><span class="line">  fq2=./3.cleanfastq/<span class="variable">$&#123;id&#125;</span>_2.fq.gz</span><br><span class="line">  bwa mem -M -t 6 -R <span class="string">&quot;@RG\tID:<span class="variable">$&#123;id&#125;</span>\tSM:<span class="variable">$&#123;id&#125;</span>\tLB:WXS\tPL:Illumina&quot;</span> <span class="variable">$&#123;INDEX&#125;</span> <span class="variable">$&#123;fq1&#125;</span> <span class="variable">$&#123;fq2&#125;</span> | samtools sort -@ 6 -m 1G  -o ./4.alignment/<span class="variable">$&#123;id&#125;</span>.bam -</span><br><span class="line">  <span class="built_in">echo</span> <span class="string">&quot;end bwa for <span class="variable">$&#123;id&#125;</span>&quot;</span> `date`</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/bamfile.png" alt="bamfile" style="zoom:90%;" /></center>

<p>bwa比对后先产生sam文件，但是因为sam文件太大，占用空间，要转换成二进制的bam文件。因此，脚本中把比对的结果通过管道符1传给 <code>samtools</code>进行排序后直接生存bam文件(中划线“-”作为缓存占位符)，排序的依据是bam文件中reads比对到参考基因组的染色体和坐标。</p>
<p>bam文件格式：</p>
<center><img src="/img/Figures/bam_info.png" alt="bamfile" style="zoom:50%;" /></center>

<h1 id="比对结果的质控"><a href="#比对结果的质控" class="headerlink" title="比对结果的质控"></a>比对结果的质控</h1><h2 id="samtools-stats-plot-bamstats"><a href="#samtools-stats-plot-bamstats" class="headerlink" title="samtools stats + plot-bamstats"></a>samtools stats + plot-bamstats</h2><p>利用<code>samtools stats</code>对bam文件进行统计</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># samtools-v1.9</span></span><br><span class="line"><span class="comment"># -------------------------------&gt; </span></span><br><span class="line">cat 0.sra/sra.infor/sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  bam=./4.alignment/<span class="variable">$&#123;id&#125;</span>.bam</span><br><span class="line">  samtools stats -@ 6 --reference ~/public/workspace/lincs/wes_cancer/data/Homo_sapiens_assembly38.fasta <span class="variable">$&#123;bam&#125;</span> &gt; ./4.alignment/stats/<span class="variable">$&#123;id&#125;</span>.<span class="built_in">stat</span></span><br><span class="line"></span><br><span class="line">  plot-bamstats -p ./4.alignment/stats/<span class="variable">$&#123;id&#125;</span> ./4.alignment/stats/<span class="variable">$&#123;id&#125;</span>.<span class="built_in">stat</span></span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<p>可以查看bamfile比对的信息（以case1_germine为例）：</p>
<center><img src="/img/Figures/g1-stat.png" alt="bam-info" style="zoom:60%;" /></center>

<hr>
<p><strong>Troubleshooting</strong><br>在<code>plot-bamstats</code>时遇到了如下报错（用的samtools-v1.9，samtools-v1.7显示command not found）。因为plot-bamstats需要调用<code>gnuplot</code>去画图，这里缺少了<code>libjpeg.so.8</code>这个库（是一款开源的jpeg解压库）。</p>
<center><img src="/img/Figures/plot-error.png" alt="plot-error" style="zoom:90%;" /></center>

<p>用conda安装<code>libjpeg.so.8</code>库：</p>
<center><img src="/img/Figures/conda-libjpeg.png" alt="libjpeg" style="zoom:90%;" /></center>

<p>然后报错了,用mamba install也一样是报错：<br><br><img src="/img/Figures/conda-install-error.png" alt="install-error" style="zoom:100%;" /></p>
<p>最后只能去官网下载tar.gz自己编译安装：<br>官网：<a target="_blank" rel="noopener" href="http://www.ijg.org/">http://www.ijg.org/</a><br>我下载的是：<a target="_blank" rel="noopener" href="http://www.ijg.org/files/jpegsrc.v8d.tar.gz">http://www.ijg.org/files/jpegsrc.v8d.tar.gz</a></p>
<p>但是一直显示编译错误，没有权限，可能要sudo? </p>
<center><img src="/img/Figures/make-error.png" alt="make-error" style="zoom:100%;" /></center>

<p>编译警告：warning: Clock skew detected. Your build may be incomplete. 按着教程写了这个：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">find . -<span class="built_in">type</span> f | xargs -n 5 touch</span><br></pre></td></tr></table></figure>
<p>但是还是报一样的错</p>
<p>又试了一下mamba install！！成功了！！</p>
<center><img src="/img/Figures/mamba-success.png" alt="mamba" style="zoom:100%;" /></center>

<p>但是！！<code>libjpeg-turbo</code>是libjpeg是的升级版，<code>plot-bamstats</code>不支持用<code>libjpeg-turbo</code>。</p>
<p>再创建新环境，重新装了<code>gnuplot</code>之后新报错又来了！conda gnuplot自动装上了jpeg的解压库。所以上面<code>libjpeg.so.8</code>库缺失这个问题算是解决了！</p>
<p>先调用<code>gnuplot</code>试一下（需要写全路径，相对路径会报错）：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">gnuplot ~/MyCourse/CancerGenomics/WES.proj/4.alignment/stats/plot/case1_germline-quals.gp</span><br></pre></td></tr></table></figure>
<p>可以成功画图了！</p>
<center><img src="/img/Figures/case1_germline-quals.png" style="zoom:75%;" /></center>

<p>然后用<code>plot-bamstats</code>，虽然这里有报错，但图片都出来了</p>
<center><img src="/img/Figures/samtool1.9error.png" alt="error" style="zoom:80%;" /></center>

<blockquote>
<p>参考：<br><a target="_blank" rel="noopener" href="https://www.jieandze1314.com/post/cnposts/166/">166-为什么使用samtools检查bam质量时报错？</a><br><a target="_blank" rel="noopener" href="https://blog.csdn.net/wp133716/article/details/107197125">libjpeg.so.8: cannot open shared object file: No such file or directory.</a><br><a target="_blank" rel="noopener" href="https://www.linuxquestions.org/questions/linux-newbie-8/error-clock-skew-detected-make-may-be-incomplete-what%27s-wrong-125912/?__cf_chl_jschl_tk__=cfbca89553aa9c176c35b0156876e5b800ec1c51-1617019792-0-AZ3wgOYkBndrrNqbZ2jUZh3XuYSB1HGsI9OaMmmL0cPqIrKE2XFuDAYw-UufXgiH8ArtnWymYKCGqM8C4mbGIaE7zSFwS2d4QSLFAadibrZLkjNxWVHMFsWah0IBlhOQg-LruE45VXiyVrY1g-DQbSib83JEEaV86GXVe4l4QHFwabMGxZgoiUJzGvvDOOFJN7qH1BvkDC0p-aPA5VjiuEunTLdiUR92A_x9ZRXQFB85eer1Zgp5d82FRB-5LjpJ3jBPXA_lSZGQED2DAv078tYyFshTQQAhfe7dM0ZYJY1hFBdEURdlGF9QUEhx54i30OQe_2xnKVQrOpyTSAeWLrDZK04zCpBOQuC4RAMbefGRmQ_-3bpTZ6zpLE023V0AmkAUjQbcM6O-HDdeypZv6DPjrc7-ji7ljm4eX1llhiLBa1O8DEN6f8fM4E4BtGRXIDaJA-RP9Y-LOs5ROXVTtqs0-CGVmsA5CJw06kvS_y2xjZ_BRepHX9DgU-l4Kz5iqQ">Error -&gt;Clock Skew Detected..Make may be incomplete….What’s wrong?</a><br><a target="_blank" rel="noopener" href="https://blog.csdn.net/cxq2046/article/details/50253949">安装libjpeg库后提示libjpeg.so.8不存在（linux环境）</a><br><a target="_blank" rel="noopener" href="https://blog.csdn.net/jiejinquanil/article/details/106414938">编译警告：warning: Clock skew detected. Your build may be incomplete.</a></p>
</blockquote>
<hr>
<p>以case1_germline和case1_biorep_B为例：</p>
<center><img src="/img/Figures/plot-succ.png" alt="show the plot" style="zoom:80%;" /></center>

<h2 id="qualimap"><a href="#qualimap" class="headerlink" title="qualimap"></a>qualimap</h2><p><code>qualimap</code>可用来查看覆盖深度等信息，需要py=3.9，跑的时候占用很多内存，大概跑了30分钟</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># qualimap-v2.2.2</span></span><br><span class="line"><span class="comment"># -------------------------------&gt; </span></span><br><span class="line">cat 0.sra/sra.infor/sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  qualimap bamqc --java-mem-size=10G -gff /public/workspace/lincs/wes_cancer/data/hg38.exon.bed -nr 100000 -nw 500 -nt 8 -bam ./4.alignment/<span class="variable">$&#123;id&#125;</span>.bam -outdir ./4.alignment/qualimap/<span class="variable">$&#123;id&#125;</span> &amp;</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<p>可得到HTML格式的报告，以case1_germline为例：</p>
<center><img src="/img/Figures/qualimap-res.png" alt="qualimap-res" style="zoom:80%;" /></center>

<p>下图展现了测序深度与基因组的覆盖信息，结果表明：有65%的位点测序深度达到50X</p>
<table><tr>
<td><img src="/img/Figures/coverage.png" border=0></td>
<td><img src="/img/Figures/genome-fraction.png" border=0></td>
</tr></table>

<h1 id="GATK流程和找变异"><a href="#GATK流程和找变异" class="headerlink" title="GATK流程和找变异"></a>GATK流程和找变异</h1><h2 id="Mark-Duplicates-Picard"><a href="#Mark-Duplicates-Picard" class="headerlink" title="Mark Duplicates(Picard)"></a>Mark Duplicates(Picard)</h2><p><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us/articles/360057439771-MarkDuplicates-Picard-">Manual-Mark Duplicates(Picard)</a></p>
<p><code>- Xmx20G -Djava. io. tmpdir=./</code>: 设置调用的内存大小为20G<br><code>-I $BAM</code>:指定输入文件<br><code>--REMOVE_DUPLICATES=true</code>: 删除掉重复序列，如果不加这个参数，就只是标记重复序列而不会删除。<br><code>-O $&#123;id&#125;_ marked.bam</code>: 输出文件</p>
<ul>
<li>脚本思路：<ol>
<li>进入到<code>while read id;do... ;done</code>循环后，每read读取一行为一个id,也就是一个样本名，对于每一个样本,再赋值bam文件的具体路径，命名为<code>BAM</code>；</li>
<li>if判断语句，判断是否已经存在生成的<code>./5. gatk/ok.$&#123;id&#125;_ marked.status</code>文件，如果存在,就不再执行，如果不存在就继续往下走。需要注意的是，如果在shell命令执行后，会有一个返回值$?，如果返回值为0，只上一步命令成功执行，否则执行失败;</li>
<li><code>echo start.. .date</code>这是为了记录时间，和后面的<code>echo end. . . date</code>相对应，这样我们可以知道跑完一个样本需要的时间;</li>
<li>如果命令成功执行，就创建一个空文件<code>ok.$&#123;id&#125;_ marked.status</code>。</li>
</ol>
</li>
</ul>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># GATK-v4.2.0.0</span></span><br><span class="line"><span class="comment"># -------------------------------&gt; </span></span><br><span class="line"></span><br><span class="line">cat 0.sra/sra.infor/sample  | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  BAM=./4.alignment/<span class="variable">$&#123;id&#125;</span>.bam</span><br><span class="line">  <span class="keyword">if</span> [ ! -f ./5.gatk/ok.<span class="variable">$&#123;id&#125;</span>_marked.status ]</span><br><span class="line">  <span class="keyword">then</span></span><br><span class="line">    </span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;start MarkDuplicates for <span class="variable">$&#123;id&#125;</span>&quot;</span> `date`</span><br><span class="line">    gatk --java-options <span class="string">&quot;-Xmx20G -Djava.io.tmpdir=./&quot;</span> MarkDuplicates \</span><br><span class="line">    -I <span class="variable">$&#123;BAM&#125;</span> \</span><br><span class="line">    --REMOVE_DUPLICATES <span class="literal">true</span> \</span><br><span class="line">    -O ./5.gatk/<span class="variable">$&#123;id&#125;</span>_marked.bam \</span><br><span class="line">    -M ./5.gatk/<span class="variable">$&#123;id&#125;</span>.metrics \</span><br><span class="line">    1&gt;./<span class="built_in">log</span>/redup.log/<span class="variable">$&#123;id&#125;</span>_log.mark 2&gt;&amp;1 </span><br><span class="line">    </span><br><span class="line">    </span><br><span class="line">    <span class="keyword">if</span> [ $? -eq 0 ]</span><br><span class="line">    <span class="keyword">then</span></span><br><span class="line">      touch ./5.gatk/ok.<span class="variable">$&#123;id&#125;</span>_marked.status</span><br><span class="line">    <span class="keyword">fi</span></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;end MarkDuplicates for <span class="variable">$&#123;id&#125;</span>&quot;</span> `date`</span><br><span class="line">    <span class="keyword">fi</span></span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<p>比对好后50G的bam一共跑了三个半小时：</p>
<center><img src="/img/Figures/redup-time.png" alt="redup-time" style="zoom:80%;" /></center>

<p>最后是用samtools对生成的bam文件构建索引，这是后续步骤的要求。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># samtools-v1.9</span></span><br><span class="line"><span class="comment"># -------------------------------&gt; </span></span><br><span class="line">cat 0.sra/sra.infor/sample  | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  samtools index -@ 16 -m 4G -b ./5.gatk/<span class="variable">$&#123;id&#125;</span>_marked.bam ./5.gatk/<span class="variable">$&#123;id&#125;</span>_marked.bai</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/redup-res.png" alt="redup-res" style="zoom:80%;" /></center>

<p>查看redup的结果:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># samtools-v1.9</span></span><br><span class="line"><span class="comment"># -------------------------------&gt; </span></span><br><span class="line"><span class="comment"># wc -l 查看行数</span></span><br><span class="line">samtools view ./4.alignment/case1_biorep_A_techrep.bam |wc -l</span><br><span class="line">samtools view ./5.gatk/case1_biorep_A_techrep_marked.bam |wc -l</span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/redup-res-view.png" alt="redup-res-view" style="zoom:100%;" /></center>

<h2 id="碱基质量重校正-BQSR"><a href="#碱基质量重校正-BQSR" class="headerlink" title="碱基质量重校正 BQSR"></a>碱基质量重校正 BQSR</h2><p>对于WGS或者WES的测序数据，最关心的问题是变异检测，而变异检测和碱基质量值密切相关。但碱基的质量值由测序仪和测序系统产生，机器带来的系统误差和噪声是不可避免的。因此，找变异之前，需要进行碱基质量重校正BQSR(Base Quality Score Recalibration)。用到了GATK的两个工具<code>BaseRecalibrator</code>和<code>ApplyBQSR</code>。</p>
<p><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us/articles/360056969412-BaseRecalibrator">Manual-BaseRecalibrator</a></p>
<blockquote>
<p><strong>Input</strong><br>    The input read data whose base quality scores need to be assessed.<br>    A database of known polymorphic sites to skip over.<br> <strong>Output</strong><br>    A GATK Report file with many tables:<br>    1. The list of arguments<br>    2. The quantized qualities table<br>    3. The recalibration table by read group<br>    1. The recalibration table by quality score<br>    1. The recalibration table for all the optional covariates  <br><br>    这里计算出了所有需要进行重校正的reads和特征值，然后把这些信息输出为一份校准表文件(recal.table) 。<br>    – known-sites参数需要输入已知且可靠的变异位点。</p>
</blockquote>
<p><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us/articles/360056968652-ApplyBQSR">Manual-ApplyBQSR</a></p>
<blockquote>
<p><strong>Input</strong><br>    A BAM or CRAM file containing input read data<br>    The covariates table (= recalibration file) generated by BaseRecalibrator on the input BAM or CRAM file<br><strong>Output</strong><br>    A BAM or CRAM file containing the recalibrated read data  <br><br>    这一步利用第-步得到的校准表文件(recal.table)重新调整原来BAM文件中的碱基质量值，并使用这个新的质量值重新输出一份新的BAM文件。</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">snp=~/Reference/dbsnp_146.hg38/dbsnp_146.hg38.vcf.gz</span><br><span class="line">indel=~/Reference/GATK/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz</span><br><span class="line">ref=~/Reference/GATK/Homo_sapiens_assembly38.fasta</span><br><span class="line">cat ./0.sra/sra.infor/sample  | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  <span class="keyword">if</span> [ ! -f ./5.gatk/<span class="variable">$&#123;id&#125;</span>_bqsr.bam ]</span><br><span class="line">  <span class="keyword">then</span></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;start BQSR for <span class="variable">$&#123;id&#125;</span>&quot;</span> `date`</span><br><span class="line">    gatk --java-options <span class="string">&quot;-Xmx60G -Djava.io.tmpdir=./&quot;</span>  BaseRecalibrator \</span><br><span class="line">    -R <span class="variable">$ref</span>  \</span><br><span class="line">    -I ./5.gatk/<span class="variable">$&#123;id&#125;</span>_marked.bam  \</span><br><span class="line">    --known-sites <span class="variable">$&#123;snp&#125;</span> \</span><br><span class="line">    --known-sites <span class="variable">$&#123;indel&#125;</span> \</span><br><span class="line">    -O ./5.gatk/recal.table/<span class="variable">$&#123;id&#125;</span>_recal.table \</span><br><span class="line">    1&gt;./5.gatk/recal.log/<span class="variable">$&#123;id&#125;</span>_log.recal 2&gt;&amp;1 </span><br><span class="line">    </span><br><span class="line">    gatk --java-options <span class="string">&quot;-Xmx60G -Djava.io.tmpdir=./&quot;</span>  ApplyBQSR \</span><br><span class="line">    -R <span class="variable">$ref</span>  \</span><br><span class="line">    -I ./5.gatk/<span class="variable">$&#123;id&#125;</span>_marked.bam  \</span><br><span class="line">    -bqsr ./5.gatk/recal.table/<span class="variable">$&#123;id&#125;</span>_recal.table \</span><br><span class="line">    -O ./5.gatk/bqsr.bam/<span class="variable">$&#123;id&#125;</span>_bqsr.bam \</span><br><span class="line">    1&gt;./5.gatk/applyBQSR.log/<span class="variable">$&#123;id&#125;</span>_log.ApplyBQSR  2&gt;&amp;1 </span><br><span class="line">    </span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;end BQSR for <span class="variable">$&#123;id&#125;</span>&quot;</span> `date`</span><br><span class="line">  <span class="keyword">fi</span></span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/bqsr-time.png" alt="bqsr-time" style="zoom:60%;" /></center>

<hr>
<p><strong>Troubleshooting</strong></p>
<blockquote>
<p>A USER ERROR has occurred: Input files reference and features have incompatible contigs: No overlapping contigs found.<br>–known-sites的版本选择问题，dbsnp146不报错，dbsnp153报错，需要使用与input文件contig name一致的reference。<br><strong>官网的论坛里开发者回复：</strong><br>the exception you get arises from the fact that you are using decoy reference genome version while attempting to recalibrate on known-sites for nonDecoy reference genome. Note that it might use different contig names (N instead of chrN)</p>
</blockquote>
<hr>
<center><img src="/img/Figures/bqsr-res.png" alt="bqsr-res" style="zoom:50%;" /></center>

<h2 id="单样本找Germline-SNPs-INDELs"><a href="#单样本找Germline-SNPs-INDELs" class="headerlink" title="单样本找Germline SNPs/INDELs"></a>单样本找Germline SNPs/INDELs</h2><center><img src="/img/Figures/mainstep-ssGermline.png" alt="mainstep" style="zoom:70%;" /></center>

<p><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us/articles/360056969012-HaplotypeCaller">Manual-HaplotypeCaller</a></p>
<blockquote>
<p><strong>Input</strong><br> Input bam file(s) from which to make variant calls<br><strong>Output</strong><br> Either a VCF or GVCF file with raw, unfiltered SNP and indel calls. Regular VCFs must be filtered either by variant recalibration (Best Practice) or hard-filtering before use in downstream analyses. If using the GVCF workflow, the output is a GVCF file that must first be run through GenotypeGVCFs and then filtering before further analysis.</p>
</blockquote>
<p>原理：</p>
<ol>
<li>定义变异区间：首先寻找确定进行变异检测的区间</li>
<li>数据组装：对区间内的数据及基因组进行组装(De Bruijnlike graph)并预估单倍型</li>
<li>根据比对结果，确定单倍型的最大似然值：将reads回帖单倍型数据(PairHMM algorithm)，计算各个单倍型出现的最大似然值</li>
<li>将单倍型比对结果转化为在特定位置位置处的分型左右滑动查看全部内容。</li>
<li>模块分步分析说明</li>
</ol>
<center><img src="/img/Figures/g-vcf.png" alt="vcf-file" style="zoom:80%;" /></center>

<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">snp=~/Reference/dbsnp_146.hg38/dbsnp_146.hg38.vcf.gz</span><br><span class="line">indel=~/Reference/GATK/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz</span><br><span class="line">ref=~/Reference/GATK/Homo_sapiens_assembly38.fasta</span><br><span class="line">bed=~/Reference/GATK/hg38_exon.bed</span><br><span class="line"></span><br><span class="line">cat ./0.sra/sra.infor/sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  <span class="built_in">echo</span> <span class="string">&quot;start HC for <span class="variable">$&#123;id&#125;</span>&quot;</span> `date`</span><br><span class="line">  gatk --java-options <span class="string">&quot;-Xmx25g&quot;</span> HaplotypeCaller -ERC GVCF \</span><br><span class="line">  -R <span class="variable">$&#123;ref&#125;</span> \</span><br><span class="line">  -I ./5.gatk/bqsr.bam/<span class="variable">$&#123;id&#125;</span>_bqsr.bam \</span><br><span class="line">  --dbsnp <span class="variable">$&#123;snp&#125;</span> \</span><br><span class="line">  -L <span class="variable">$&#123;bed&#125;</span> \</span><br><span class="line">  -O ./5.gatk/gvcf/<span class="variable">$&#123;id&#125;</span>_raw.vcf \</span><br><span class="line">  1&gt;./<span class="built_in">log</span>/HaplotypeCaller.log<span class="variable">$&#123;id&#125;</span>_log.HC 2&gt;&amp;1</span><br><span class="line">  <span class="built_in">echo</span> <span class="string">&quot;end HC for <span class="variable">$&#123;id&#125;</span>&quot;</span> `date`</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<p>共跑了五个小时：</p>
<center><img src="/img/Figures/HC-time.png" alt="HC" style="zoom:70%;" /></center>

<p>输出文件：</p>
<center><img src="/img/Figures/vcf-file.png" alt="vcf-file" style="zoom:80%;" /></center>

<p><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us/articles/360057439331-GenomicsDBImport">Manual-GenomicsDBImport</a><br>The GATK4 Best Practice Workflow for SNP and Indel calling uses GenomicsDBImport to merge GVCFs from multiple samples. </p>
<blockquote>
<p><strong>Input</strong><br> One or more GVCFs produced by in HaplotypeCaller with the <code>-ERC GVCF</code> or <code>-ERC BP_RESOLUTION</code> settings, containing the samples to joint-genotype.<br><strong>Output</strong><br> A GenomicsDB workspace</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">ref=~/Reference/GATK/Homo_sapiens_assembly38.fasta</span><br><span class="line"></span><br><span class="line"><span class="built_in">cd</span> ./5.gatk/gvcf</span><br><span class="line"><span class="keyword">for</span> chr <span class="keyword">in</span> chr&#123;1..22&#125; chrX chrY chrM</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line"></span><br><span class="line">time gatk --java-options <span class="string">&quot;-Xmx60G -Djava.io.tmpdir=./&quot;</span> GenomicsDBImport \</span><br><span class="line">-R <span class="variable">$&#123;ref&#125;</span> \</span><br><span class="line">$(ls ~/MyCourse/CancerGenomics/WES.proj/5.gatk/gvcf/*_raw.vcf | awk <span class="string">&#x27;&#123;print &quot;-V &quot;$0&quot; &quot;&#125;&#x27;</span>) \</span><br><span class="line">-L <span class="variable">$&#123;chr&#125;</span> \</span><br><span class="line">--genomicsdb-workspace-path gvcfs_<span class="variable">$&#123;chr&#125;</span>.db</span><br><span class="line"></span><br><span class="line">time gatk --java-options <span class="string">&quot;-Xmx60G -Djava.io.tmpdir=./&quot;</span> GenotypeGVCFs \</span><br><span class="line">-R <span class="variable">$&#123;ref&#125;</span> \</span><br><span class="line">-V gendb://gvcfs_<span class="variable">$&#123;chr&#125;</span>.db \</span><br><span class="line">-O gvcfs_<span class="variable">$&#123;chr&#125;</span>.vcf</span><br><span class="line"></span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us/articles/360056968552-GatherVcfs-Picard-">Manual-GatherVcfs</a><br>Gathers multiple VCF files from a scatter operation into a single VCF file. Input files must be supplied in genomic order and must not have events at overlapping positions.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">gatk --java-options <span class="string">&quot;-Xmx60G -Djava.io.tmpdir=./&quot;</span> GatherVcfs \</span><br><span class="line">$(<span class="keyword">for</span> i <span class="keyword">in</span> &#123;1..22&#125; X Y M;<span class="keyword">do</span> <span class="built_in">echo</span> <span class="string">&quot;-I gvcfs_chr<span class="variable">$&#123;i&#125;</span>.vcf&quot;</span> ;<span class="keyword">done</span>) \</span><br><span class="line">-O merge.vcf</span><br></pre></td></tr></table></figure>
<p>得到</p>
<center><img src="/img/Figures/vcf-file2.png" alt="vcf-file" style="zoom:100%;" /></center>

<hr>
<p>vcf文件分为两个部分：</p>
<ul>
<li>以##开头的头文件信息：主要包括vcf文件版本、FORMAT、INFO、参考基因组以及执行程序等信息</li>
<li>主体部分：<ol>
<li>CHROM(chromosome):染色体</li>
<li>POS  变异位点在参考基因组中的位置</li>
<li>ID - identifier: variant的ID。比如在dbSNP中有该SNP的id，则会在此行给出；若没有，则用<code>.</code>表示其为一个novel variant。</li>
<li>REF - reference base(s):参考碱基，染色体上面的碱基，必须是ATCGN中的一个，N表示不确定碱基</li>
<li>ALT - alternate base(s):与参考序列比较发生突变的碱基</li>
<li>QUAL - quality: Phred格式(Phred_scaled)的质量值，表示在该位点存在variant的可能性；该值越高，则variant的可能性越大；计算方法：Phred值 = -10 * log (1-p) p为variant存在的概率。</li>
<li>FILTER - _filter status: 使用上一个QUAL值来进行过滤的话，是不够的。GATK能使用其它的方法来进行过滤，过滤结果中通过则该值为<code>PASS</code>;若variant不可靠，则该项不为<code>PASS</code>或<code>.</code>。</li>
<li>INFO - additional information: 这一行是variant的详细信息，具体如下：<ul>
<li>DP-read depth：样本在这个位置的reads覆盖度。是一些reads被过滤掉后的覆盖度。DP4:高质量测序碱基，位于REF或者ALT前后<ul>
<li>QD：通过深度来评估一个变异的可信度。</li>
<li>MQ：表示覆盖序列质量的均方值RMS Mapping Quality</li>
<li>FQ：phred值关于所有样本相似的可能性</li>
<li>AC，AF和AN：AC(Allele Count) 表示该Allele的数目；AF(Allele Frequency) 表示Allele的频率； AN(Allele Number) 表示Allele的总数目。<br>对于1个diploid sample而言：则基因型 0/1 表示sample为杂合子，Allele数为1(双倍体的sample在该位点只有1个等位基因发生了突变)，<br>Allele的频率为0.5(双倍体的sample在该位点只有50%的等位基因发生了突变)，总的Allele为2；<br>基因型 1/1 则表示sample为纯合的，Allele数为2，Allele的频率为1，总的Allele为2。</li>
<li>BaseQRankSum：比较支持变异的碱基和支持参考基因组的碱基的质量。<br>负值表示支持变异的碱基质量值不及支持参考基因组的，<br>正值则相反，支持变异的质量值好于参考基因组的。<br>0表示两者无明显差异。</li>
<li>ReadPosRankSum：检测变异位点是否有位置偏好性（是否存在于序列末端，此时往往容易出错）。最佳值为0，表示变异与其在序列上的位置无关。负值表示变异位点更容易在末端出现，正值表示参考基因组中的等位基因更容易在末端出现。</li>
</ul>
</li>
</ul>
</li>
<li>FORMAT和最后一列sample中的信息是对应的<ul>
<li>AD和DP：AD(Allele Depth)为sample中每一种allele的reads覆盖度,在diploid中则是用逗号分割的两个值，前者对应ref基因型，后者对应variant基因型；<br>DP（Depth）为sample中该位点的覆盖度。</li>
<li>GT：样品的基因型（genotype）。两个数字中间用<code>/</code>分开，这两个数字表示双倍体的sample的基因型。<br>0 表示样品中有ref的allele; 1表示样品中variant的allele；2表示有第二个variant的allele。因此：0/0 表示sample中该位点为纯合的，和ref一致； 0/1 表示sample中该位点为杂合的，有ref和variant两个基因型； 1/1 表示sample中该位点为纯合的，和variant一致。</li>
<li>GQ：即第二可能的基因型的PL值，相对于最可能基因型的PL值（其PL=0）而言，大于99时，其信息量已不大，因此大于99的全部赋值99。当GQ值很小时，意味着第二可能基因型与最可能基因型差别不大。</li>
<li>GL：三种基因型（RR RA AA）出现的可能性，R表示参考碱基，A表示变异碱基</li>
<li>DV：高质量的非参考碱基</li>
<li>SP：phred的p值误差线</li>
<li>PL：指定的三种基因型的可能性(provieds the likelihoods of the given genotypes)。这三种指定的基因型为(0/0,0/1,1/1)，这三种基因型的概率总和为1。和之前不一致，该值越大，表明为该种基因型的可能性越小。</li>
</ul>
</li>
</ol>
</li>
</ul>
<hr>
<h2 id="Mutect找Somatic-Mutations"><a href="#Mutect找Somatic-Mutations" class="headerlink" title="Mutect找Somatic Mutations"></a>Mutect找Somatic Mutations</h2><p>前面单样本走GATK HaplotypeCaller流程得到vcf文件主要包含了germline mutations信息。但是我们这里分析的是肿瘤样本，包含成对的tumor和normal样本，更加关心的是tumor样本中所包含的somatic mutations信息。所以我们应该走GATK的somatic mutations流程:<br>Identify somatic short variants (SNVs and Indels) in one or more tumor samples from a single individual, with or without a matched normal sample.</p>
<center><img src="/img/Figures/somatic.png" alt="somatic" style="zoom:60%;" /></center>

<p><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us/articles/360056969692-Mutect2">Maunal-Mutect2</a><br>Call somatic short mutations via local assembly of haplotypes. Short mutations include single nucleotide (SNA) and insertion and deletion (indel) alterations. The caller uses a Bayesian somatic genotyping model that differs from the original MuTect and uses the assembly-based machinery of HaplotypeCaller.</p>
<blockquote>
<p>三种模式：</p>
<ol>
<li>tumor-normal mode where a tumor sample is matched with a normal sample in analysis;</li>
<li>tumor-only mode where a single sample’s alignment data undergoes analysis;</li>
<li>mitochondrial mode where sensitive calling at high depths is desirable</li>
</ol>
</blockquote>
<p>制作config文件，normal和tumor样本成对出现</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># matched.sample</span></span><br><span class="line">case1_germline  case1_biorep_A_techrep</span><br><span class="line">case1_germline  case1_biorep_B</span><br><span class="line">case1_germline  case1_biorep_C</span><br><span class="line">case1_germline  case1_techrep</span><br><span class="line"></span><br><span class="line"><span class="comment">## mutect.sh</span></span><br><span class="line">ref=~/Reference/GATK/Homo_sapiens_assembly38.fasta</span><br><span class="line">bed=~/Reference/GATK/hg38_exon.bed</span><br><span class="line"></span><br><span class="line">cat 6.mutect/sample.infor/matched.sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  arr=(<span class="variable">$&#123;id&#125;</span>)</span><br><span class="line">  sample=<span class="variable">$&#123;arr[1]&#125;</span></span><br><span class="line">  T=./5.gatk/bqsr.bam/<span class="variable">$&#123;arr[1]&#125;</span>_bqsr.bam</span><br><span class="line">  N=./5.gatk/bqsr.bam/<span class="variable">$&#123;arr[0]&#125;</span>_bqsr.bam</span><br><span class="line">  <span class="built_in">echo</span> <span class="string">&quot;start Mutect2 for <span class="variable">$&#123;id&#125;</span>&quot;</span> `date`</span><br><span class="line">  gatk  --java-options <span class="string">&quot;-Xmx36G -Djava.io.tmpdir=./&quot;</span>  Mutect2 -R <span class="variable">$&#123;ref&#125;</span> \</span><br><span class="line">  -I <span class="variable">$&#123;T&#125;</span> -tumor  $(basename <span class="string">&quot;<span class="variable">$T</span>&quot;</span> _bqsr.bam) \</span><br><span class="line">  -I <span class="variable">$&#123;N&#125;</span> -normal $(basename <span class="string">&quot;<span class="variable">$N</span>&quot;</span> _bqsr.bam) \</span><br><span class="line">  -L <span class="variable">$&#123;bed&#125;</span>  \</span><br><span class="line">  -O ./6.mutect/<span class="variable">$&#123;sample&#125;</span>_mutect2.vcf</span><br><span class="line"></span><br><span class="line">  gatk  FilterMutectCalls \</span><br><span class="line">  -R <span class="variable">$&#123;ref&#125;</span> \</span><br><span class="line">  -V ./6.mutect/<span class="variable">$&#123;sample&#125;</span>_mutect2.vcf \</span><br><span class="line">  -O ./6.mutect/<span class="variable">$&#123;sample&#125;</span>_somatic.vcf</span><br><span class="line">  <span class="built_in">echo</span> <span class="string">&quot;end Mutect2 for <span class="variable">$&#123;id&#125;</span>&quot;</span> `date`</span><br><span class="line"></span><br><span class="line">  cat ./6.mutect/<span class="variable">$&#123;sample&#125;</span>_somatic.vcf | perl -alne <span class="string">&#x27;&#123;if(/^#/)&#123;print&#125;else&#123;next unless $F[6] eq &quot;PASS&quot;;next if $F[0] =~/_/;print &#125; &#125;&#x27;</span> &gt; ./6.mutect/<span class="variable">$&#123;sample&#125;</span>_filter.vcf</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<p>以上代码跑了四个小时。代码解释:首先是<code>Mutect2</code>，主要是根据对正常-肿瘤样本进行位点比较寻找体细胞突变somatic<br>mutations。这里输入了两个配对的T-N样本<code>-I $&#123;T&#125; -tumor $(basename &quot;$T&quot;_ bqsr.bam)</code> 和<code>-I $&#123;N&#125; -normal $(basename &quot;$N&quot;_ bqsr. bam)</code>。<br>这样我们就拿到了每个tumor样本中的突变位点<code>$&#123;samp1e&#125;_ mutect2.vcf</code>。但是这还不够，还要通过<code>FilterMutectCalls</code>过滤，过滤后就是可信度较高的somatic突变位点会被打上<strong>PASS标签</strong>,就是<code>$&#123;samp1e&#125;_ somatic.vcf</code>。<br>最后我们再用一个简单的脚本把打上PASS标签的位点提取出来，得到了<code>$&#123;sample&#125;_ filter.vcf</code>: </p>
<center><img src="/img/Figures/vcf-filter.png" alt="vcf-filter" style="zoom:70%;" /></center>

<h1 id="VEP注释"><a href="#VEP注释" class="headerlink" title="VEP注释"></a>VEP注释</h1><h2 id="安装"><a href="#安装" class="headerlink" title="安装"></a>安装</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 安装vep</span></span><br><span class="line">mamba install -c bioconda ensembl-vep</span><br><span class="line"><span class="comment"># 下载vep的cache及ref</span></span><br><span class="line"><span class="comment"># VEP推荐下载缓冲文件进行注释，后面跑的时候会比较快</span></span><br><span class="line">vep_install -a cf -s homo_sapiens -y GRCh38 -c ~/Reference/vep/ --CONVERT</span><br></pre></td></tr></table></figure>
<h2 id="注释"><a href="#注释" class="headerlink" title="注释"></a>注释</h2><p><a target="_blank" rel="noopener" href="https://asia.ensembl.org/info/docs/tools/vep/script/vep_options.html">Manual-VEP</a></p>
<p>仅注释肿瘤样本（除了case1_germline）</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">cat 0.sra/sra.infor/config.tumor | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line"><span class="built_in">echo</span> <span class="string">&quot;start vep_annotation for <span class="variable">$&#123;id&#125;</span> &quot;</span> `date`</span><br><span class="line">vep --fork 6 --cache --cache_version 103 --offline --format vcf --vcf --force_overwrite \</span><br><span class="line">--fasta /home/data/vip8t02/Reference/vep//homo_sapiens/103_GRCh38/Homo_sapiens.GRCh38.dna.toplevel.fa.gz \</span><br><span class="line">--dir ~/Reference/vep \</span><br><span class="line">--input_file ./6.mutect/<span class="variable">$&#123;id&#125;</span>_filter.vcf \</span><br><span class="line">--output_file ./7.annotation/vep/<span class="variable">$&#123;id&#125;</span>_vep.vcf </span><br><span class="line"><span class="built_in">echo</span> <span class="string">&quot;end vep_annotation for <span class="variable">$&#123;id&#125;</span> &quot;</span> `date`</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<p>可以得到HTML格式文件及新的vcf（在INFO列加了很多信息）</p>
<center><img src="/img/Figures/annovep.png" alt="annotation" style="zoom:70%;" /></center>

<h2 id="注释结果转maf"><a href="#注释结果转maf" class="headerlink" title="注释结果转maf"></a>注释结果转maf</h2><p>vep注释后需要将vcf文件转换成maf，用到的工具是vcf2maf<br><a target="_blank" rel="noopener" href="https://github.com/mskcc/vcf2maf">Github-vcf2maf</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 安装</span></span><br><span class="line">mamba install -c bioconda vcf2maf</span><br><span class="line"></span><br><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># vcf2maf-v1.6.17</span></span><br><span class="line"><span class="comment"># -------------------------------&gt; </span></span><br><span class="line">cat 0.sra/sra.infor/config.tumor | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  sample=./6.mutect/<span class="variable">$&#123;id&#125;</span>_filter.vcf</span><br><span class="line">  normal=case1_germline</span><br><span class="line">  vcf2maf.pl \</span><br><span class="line">  --input-vcf  <span class="variable">$&#123;sample&#125;</span>  \</span><br><span class="line">  --output-maf ./7.annotation/vep/<span class="variable">$&#123;id&#125;</span>_vep.maf \</span><br><span class="line">  --ref-fasta ~/Reference/GATK/Homo_sapiens_assembly38.fasta \</span><br><span class="line">  --tumor-id  <span class="variable">$&#123;id&#125;</span> \</span><br><span class="line">  --normal-id <span class="variable">$&#123;normal&#125;</span> \</span><br><span class="line">  --ncbi-build GRCh38 \</span><br><span class="line">  --vep-path /home/data/vip8t02/anaconda3/envs/vep-103.1/bin \</span><br><span class="line">  --vep-data ~/Reference/vep</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 检查转换后结果</span></span><br><span class="line">less ./7.annotation/vep/case1_biorep_A_techrep_vep.maf | cut -f 10 | sort | uniq -c</span><br><span class="line"></span><br><span class="line">less ./7.annotation/vep/case1_biorep_A_techrep_vep.maf | cut -f 10 | sort | uniq -c</span><br><span class="line"></span><br><span class="line"><span class="comment"># 合并maf</span></span><br><span class="line">cat ./7.annotation/vep/*maf | grep -v <span class="string">&#x27;^#&#x27;</span>| grep -v <span class="string">&#x27;^Hugo_Symbol&#x27;</span> &gt;./7.annotation/vep/tmp </span><br><span class="line">grep <span class="string">&#x27;Hugo_Symbol&#x27;</span> ./7.annotation/vep/case1_biorep_A_techrep_vep.maf &gt;./7.annotation/vep/header</span><br><span class="line">cat ./7.annotation/vep/header ./7.annotation/vep/tmp &gt;./7.annotation/vep/vep_merge.maf</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/maf-info.png" alt="maf-info" style="zoom:80%;" /></center>

<h2 id="可视化注释结果"><a href="#可视化注释结果" class="headerlink" title="可视化注释结果"></a>可视化注释结果</h2><h3 id="读入数据"><a href="#读入数据" class="headerlink" title="读入数据"></a>读入数据</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">library(maftools)</span><br><span class="line"></span><br><span class="line">rm(<span class="built_in">list</span> = ls())</span><br><span class="line">options(stringsAsFactors = <span class="built_in">F</span>)</span><br><span class="line">setwd(<span class="string">&quot;~/MyCourse/CancerGenomics/WES.proj/&quot;</span>)</span><br><span class="line"><span class="comment">## vep</span></span><br><span class="line">vep.laml = read.maf(maf = <span class="string">&#x27;./7.annotation/vep/case1_biorep_A_techrep_vep.maf&#x27;</span>)</span><br><span class="line"><span class="comment">## for vep.laml</span></span><br><span class="line">library(stringr)</span><br><span class="line">vep.laml@data$Protein_Change = paste0(<span class="string">&quot;p.&quot;</span>,</span><br><span class="line">                                      str_sub(vep.laml@data$Amino_acids,<span class="number">1</span>,<span class="number">1</span>),</span><br><span class="line">                                      vep.laml@data$Protein_position,</span><br><span class="line">                                      str_sub(vep.laml@data$Amino_acids,<span class="number">3</span>,<span class="number">3</span>))</span><br><span class="line"></span><br><span class="line"><span class="comment">## save Rdata</span></span><br><span class="line">save(vep.laml, file = <span class="string">&#x27;7.annotation/visualization//laml.Rdata&#x27;</span>)</span><br><span class="line"><span class="comment">## Summary</span></span><br><span class="line">laml=vep.laml</span><br><span class="line">unique(laml@data$Tumor_Sample_Barcode)</span><br><span class="line">getSampleSummary(laml)</span><br><span class="line">getGeneSummary(laml)</span><br><span class="line">getFields(laml)</span><br></pre></td></tr></table></figure>
<h3 id="mafSummary"><a href="#mafSummary" class="headerlink" title="mafSummary"></a>mafSummary</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rm(list = ls())</span></span><br><span class="line"><span class="comment"># require(maftools)</span></span><br><span class="line"><span class="comment"># options(stringsAsFactors = F)</span></span><br><span class="line"><span class="comment"># laml &lt;- load(file = &#x27;laml.Rdata&#x27;)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## mafsummary</span></span><br><span class="line">png(<span class="string">&#x27;./7.annotation/visualization/plotmafSummary_vep.png&#x27;</span>,res = <span class="number">150</span>,width = <span class="number">1080</span>,height = <span class="number">1080</span>)</span><br><span class="line">plotmafSummary(maf = laml,</span><br><span class="line">                 rmOutlier = <span class="literal">TRUE</span>,</span><br><span class="line">                 showBarcodes = <span class="built_in">T</span>,</span><br><span class="line">                 textSize = <span class="number">0.4</span>,</span><br><span class="line">                 addStat = <span class="string">&#x27;median&#x27;</span>,</span><br><span class="line">                 dashboard = <span class="literal">TRUE</span>,</span><br><span class="line">                 titvRaw = <span class="literal">FALSE</span>)</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/plot-mat-sum.png" alt="maf-info" style="zoom:60%;" /></center>

<h3 id="oncoplot"><a href="#oncoplot" class="headerlink" title="oncoplot"></a>oncoplot</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">png(<span class="string">&#x27;./7.annotation/visualization/oncoplot_top30.png&#x27;</span>,res = <span class="number">150</span>,width = <span class="number">1080</span>,height = <span class="number">1080</span>)</span><br><span class="line">oncoplot(maf = laml,</span><br><span class="line">           top = <span class="number">30</span>,</span><br><span class="line">           fontSize = <span class="number">0.5</span>,</span><br><span class="line">           sampleOrder = laml@clinical.data$Tumor_Sample_Barcode,</span><br><span class="line">           showTumorSampleBarcodes = <span class="built_in">T</span>)</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/oncoplot.png" alt="oncoplot" style="zoom:60%;" /></center>

<h3 id="lollopopPlot"><a href="#lollopopPlot" class="headerlink" title="lollopopPlot"></a>lollopopPlot</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## lollipopPlot for CDH2</span></span><br><span class="line">gene=<span class="string">&#x27;CDH2&#x27;</span></span><br><span class="line"></span><br><span class="line">png(<span class="string">&quot;./7.annotation/visualization/CDH2_vep.png&quot;</span>,res = <span class="number">150</span>,width = <span class="number">1080</span>,height = <span class="number">1080</span>)</span><br><span class="line">maftools::lollipopPlot(maf = laml,</span><br><span class="line">                       gene = gene,</span><br><span class="line">                       AACol = <span class="string">&quot;RefSeq&quot;</span>,</span><br><span class="line">                       labelPos = <span class="string">&#x27;all&#x27;</span>)</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/CDH2.png" alt="CDH2" style="zoom:60%;" /></center>

<h1 id="拷贝数变异分析（GATK流程）"><a href="#拷贝数变异分析（GATK流程）" class="headerlink" title="拷贝数变异分析（GATK流程）"></a>拷贝数变异分析（GATK流程）</h1><p>拷贝数变异(copy number variations, CNVs)是属于基因组结构变异(structural variation, SV)，是指DNA片段长度在1Kb-3Mb的基因组结构变异。我们首先从GATK的CNV流程开始CNV的分析。</p>
<h2 id="外显子坐标的interval文件"><a href="#外显子坐标的interval文件" class="headerlink" title="外显子坐标的interval文件"></a>外显子坐标的interval文件</h2><p>interval文件其实也是一个坐标文件，类似bed,只不过bed文件的坐标是从0开始记录，而interval文件的坐标是从1开始记录。使用基因组interval 文件可以定义软件分析的分辨率。如果是全基因组测序，interval 文件就用全基因组坐标的等间隔区间就好。对于外显子组的数据，我们使用捕获试剂盒的目标区域，理论上应该返回原文查找对应试剂盒，去搜索其捕获外显子的区域，一般是外显子侧翼上下游250bp以内。</p>
<p>首先使用<code>BedToIntervalist</code>工具将bed转成interval格式(其实前面已经完成)，然后用<code>PreprocessIntervals</code>工具获取target区间，即外显子侧翼上下游250bp.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">ref=~/Reference/GATK/Homo_sapiens_assembly38.fasta</span><br><span class="line">bed=~/Reference/GATK/hg38_exon.bed</span><br><span class="line">dict=~/Reference/GATK/Homo_sapiens_assembly38.dict</span><br><span class="line"></span><br><span class="line"><span class="comment">## bed to intervals_list</span></span><br><span class="line">gatk BedToIntervalList -I <span class="variable">$&#123;bed&#125;</span> -O ~/Reference/GATK/hg38.exon.interval_list -SD <span class="variable">$&#123;dict&#125;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Preprocess Intervals</span></span><br><span class="line">gatk  PreprocessIntervals \</span><br><span class="line">-L ~/Reference/GATK/hg38.exon.interval_list \</span><br><span class="line">--sequence-dictionary <span class="variable">$&#123;dict&#125;</span> \</span><br><span class="line">--reference <span class="variable">$&#123;ref&#125;</span>  \</span><br><span class="line">--padding 250 \</span><br><span class="line">--bin-length 0 \</span><br><span class="line">--interval-merging-rule OVERLAPPING_ONLY \</span><br><span class="line">--output ~/Reference/GATK/targets.preprocessed.interval_list</span><br></pre></td></tr></table></figure>
<h2 id="获取样本的read-counts"><a href="#获取样本的read-counts" class="headerlink" title="获取样本的read counts"></a>获取样本的read counts</h2><p>这里分为两小步进行:</p>
<ol>
<li>首先是获取所有样本的read counts, 用到了工具是<code>CollectReadCounts</code>，其根据所提供的interval文件，对bam文件进行reads计数，可以简单理解为把bam文件转换成interval区间的reads数。最后会生成一个HDF5格式的文件，需要用第三方软件HDFView来查看，这里不做展示。这个文件记录了每个基因组interval文件的CONTIG, START， END和原始COUNT值，并制成表格。</li>
<li>然后构建正常样本的<code>CNV panel of normals</code>，生成正常样本的<code>cnvponM.pon.hdf5</code> 文件。对于外显子组捕获测序数据，捕获过程会引入一定的的噪音。因此后面需要降噪，该文件就是用于后面第3步<code>DenoiseReadCounts</code>。</li>
</ol>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">interval=~/Reference/GATK/targets.preprocessed.interval_list</span><br><span class="line">ref=~/Reference/GATK/Homo_sapiens_assembly38.fasta</span><br><span class="line"></span><br><span class="line">cat 0.sra/sra.infor/config.sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  i=./5.gatk/bqsr.bam/<span class="variable">$&#123;id&#125;</span>_bqsr.bam</span><br><span class="line">  <span class="built_in">echo</span> <span class="variable">$&#123;i&#125;</span></span><br><span class="line">  <span class="comment">## step1 : CollectReadCounts</span></span><br><span class="line">  time gatk  --java-options <span class="string">&quot;-Xmx48G&quot;</span> CollectReadCounts \</span><br><span class="line">  -I <span class="variable">$&#123;i&#125;</span> \</span><br><span class="line">  -L <span class="variable">$&#123;interval&#125;</span> \</span><br><span class="line">  -R <span class="variable">$&#123;ref&#125;</span> \</span><br><span class="line">  --format TSV  \</span><br><span class="line">  --interval-merging-rule OVERLAPPING_ONLY \</span><br><span class="line">  --output ./8.cnv/counts/hdf5/<span class="variable">$&#123;id&#125;</span>.clean_counts.tsv</span><br><span class="line">  </span><br><span class="line">  <span class="comment">## step2 : Generate a CNV panel of normals:cnvponM.pon.tsv</span></span><br><span class="line">  gatk  --java-options <span class="string">&quot;-Xmx48G&quot;</span> CreateReadCountPanelOfNormals \</span><br><span class="line">  --minimum-interval-median-percentile 5.0 \</span><br><span class="line">  --output ./8.cnv/cnvponM.pon.tsv \</span><br><span class="line">  --input ./8.cnv/counts/hdf5/case1_germline.clean_counts.tsv</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>跑了半个小时</p>
<h2 id="降噪-DenoiseReadCounts"><a href="#降噪-DenoiseReadCounts" class="headerlink" title="降噪 DenoiseReadCounts"></a>降噪 DenoiseReadCounts</h2><p>该步骤用到的工具是<code>DenoiseReadCounts</code>，主要是做了一个标准化和降噪，会生成两个文件<code>$&#123;id&#125;. clean. standardizedCR.tsv</code>和<code>$&#123;id&#125;.clean.denoisedCR.tsv</code>,该工具会根据PoN的counts中位数对输入文件<code>$&#123;id&#125;.clean_counts.hdf5</code> 进行标准化，包括log2转换。然后使用PoN的主成分进行标准化后的copy ratios降噪。实际上这里只需要对tumor样本进行降噪的，为了比较，可以把tumor和normal样本都分析了一遍，后面可以做个对比。</p>
<p><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us/articles/360057440151-DenoiseReadCounts">Manual-DeoiseReadCounts</a></p>
<p>The input counts are standardized by: </p>
<ol>
<li>transforming to fractional coverage,</li>
<li>performing optional explicit GC-bias correction (if the panel contains GC-content annotated intervals),</li>
<li>filtering intervals to those contained in the panel,</li>
<li>dividing by interval medians contained in the panel, </li>
<li>dividing by the sample median,</li>
<li>transforming to log2 copy ratio. </li>
</ol>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">cat 0.sra/sra.infor/config.sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  i=./8.cnv/counts/<span class="variable">$&#123;id&#125;</span>.clean_counts.tsv</span><br><span class="line">  gatk --java-options <span class="string">&quot;-Xmx48g&quot;</span> DenoiseReadCounts \</span><br><span class="line">  -I <span class="variable">$&#123;i&#125;</span> \</span><br><span class="line">  --count-panel-of-normals ./8.cnv/cnvponM.pon.tsv \</span><br><span class="line">  --standardized-copy-ratios ./8.cnv/standardizedCR/<span class="variable">$&#123;id&#125;</span>.clean.standardizedCR.tsv \</span><br><span class="line">  --denoised-copy-ratios ./8.cnv/denoisedCR/<span class="variable">$&#123;id&#125;</span>.clean.denoisedCR.tsv</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<hr>
<p><strong>Troubleshooting</strong></p>
<blockquote>
<p>HDF5-DIAG: Error detected in HDF5 (1.8.14) thread 0:<br> #000: /mnt/scr1/abyrne/HDFJava-platypus-2.11/native/HDF5-prefix/src/HDF5/src/H5D.c line 358 in H5Dopen2(): not found<br>    major: Dataset<br>    minor: Object not found<br>解决：CollectReadCounts时生成TSV文件<br>mamba install -c anaconda hdf5=1.8.14</p>
</blockquote>
<hr>
<h2 id="可视化降噪后的copy-ratio"><a href="#可视化降噪后的copy-ratio" class="headerlink" title="可视化降噪后的copy ratio"></a>可视化降噪后的copy ratio</h2><p>使用<code>PlotDenoisedCopyRatios</code>可视化标准化和去噪的read counts。这些图可以直观地评估去噪的效果。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># install ahead</span></span><br><span class="line">conda install -c conda-forge r-optparse</span><br><span class="line"></span><br><span class="line">dict=~/Reference/GATK/Homo_sapiens_assembly38.dict</span><br><span class="line">cat 0.sra/sra.infor/config.sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  gatk --java-options <span class="string">&quot;-Xmx48g&quot;</span> PlotDenoisedCopyRatios \</span><br><span class="line">  --standardized-copy-ratios 	./8.cnv/standardizedCR/<span class="variable">$&#123;id&#125;</span>.clean.standardizedCR.tsv \</span><br><span class="line">  --denoised-copy-ratios ./8.cnv/denoisedCR/<span class="variable">$&#123;id&#125;</span>.clean.denoisedCR.tsv \</span><br><span class="line">  --sequence-dictionary <span class="variable">$&#123;dict&#125;</span> \</span><br><span class="line">  --output ./8.cnv/cnv_plots \</span><br><span class="line">  --output-prefix <span class="variable">$&#123;id&#125;</span></span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<hr>
<p><strong>Troubleshooting</strong></p>
<blockquote>
<p>A USER ERROR has occurred: Bad input: format error in ‘./8.cnv/denoisedCR/case1_biorep_A_techrep.clean.denoisedCR.tsv’ at line 160284: mismatch between number of values in line (3) and number of columns (4)</p>
</blockquote>
<p>解决：<br>先用sed查看报错处的内容,怀疑是denoise程序终止，重新运行case1_biorep_A_techrep样本</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## 查看行数</span></span><br><span class="line">wc -l case1_biorep_A_techrep.clean.denoisedCR.tsv</span><br><span class="line"><span class="comment"># 160283 case1_biorep_A_techrep.clean.denoisedCR.tsv</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看160284行是什么</span></span><br><span class="line">sed -n <span class="string">&#x27;160280,160284p&#x27;</span> case1_biorep_A_techrep.clean.denoisedCR.tsv</span><br><span class="line"><span class="comment"># chr19   13145079        13145644        0.453168</span></span><br><span class="line"><span class="comment"># chr19   13148779        13149294        -0.807650</span></span><br><span class="line"><span class="comment"># chr19   13149295        13149660        0.044055</span></span><br><span class="line"><span class="comment"># chr19   13149661        13150017        0.143591</span></span><br><span class="line"><span class="comment"># chr19   13150018        13</span></span><br></pre></td></tr></table></figure>
<hr>
<center><img src="/img/Figures/case1_biorep_B.denoised.png" alt="case1_biorep_B.denoised" style="zoom:70%;" /></center>

<p>图片展示的是标准化和降噪后的copy ratios，底下还有中位数绝对偏差(median absolute deviation, MAD)，只不过一张图片把y轴设置为0到4。假如tumor样本的拷贝数没有发生变化，copy ratio应该稳定在1附近。如果发生了CNV事件，那应该就在1附近波动。</p>
<h2 id="计算常见的germline-mutation位点"><a href="#计算常见的germline-mutation位点" class="headerlink" title="计算常见的germline mutation位点"></a>计算常见的germline mutation位点</h2><p>这一步用到了<code>CollectAllelicCounts</code>工具，对输入的bam文件，根据指定的interval区间，进行germline mutation的检测(仅仅是SNPs位点，不包括INDELS )，并计算该位点覆盖的reads数，即该位点的<strong>测序深度</strong>。值得注意的是，该工具一个默认参数是MAPQ值大于20的reads才会被纳入计数，最后生成一个tsv文件。注意：这一步非常消耗CPU和内存！运行了3小时</p>
<p><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us/articles/360056969912-CollectAllelicCounts">Manual-CollectAllelicCounts</a><br>Collects reference and alternate allele counts at specified sites. The <strong>alt count</strong> is defined as <em>the total count minus the ref count</em>, and the <strong>alt nucleotide</strong> is defined as <em>the non-ref base with the highest count, with ties broken by the order of the bases in AllelicCountCollector#BASES</em>. Only reads that pass the specified read filters and bases that exceed the specified minimum-base-quality will be counted.</p>
<blockquote>
<p><strong>Inputs</strong><br> SAM format read data<br> Reference FASTA file<br> Sites at which allelic counts will be collected<br> <strong>Output</strong><br> Allelic-counts file. This is a tab-separated values (TSV) file with a SAM-style header containing a read group sample name, a sequence dictionary, a row specifying the column headers contained in AllelicCountCollection.AllelicCountTableColumn, and the corresponding entry rows.</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## script</span></span><br><span class="line">interval=~/Reference/GATK/targets.preprocessed.interval_list</span><br><span class="line">GENOME=~/Reference/GATK/Homo_sapiens_assembly38.fasta</span><br><span class="line"></span><br><span class="line">cat 0.sra/sra.infor/config.sample | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  i=./5.gatk/bqsr.bam/<span class="variable">$&#123;id&#125;</span>_bqsr.bam</span><br><span class="line">  <span class="built_in">echo</span> <span class="variable">$&#123;i&#125;</span></span><br><span class="line">  gatk --java-options <span class="string">&quot;-Xmx20G -Djava.io.tmpdir=./&quot;</span>  CollectAllelicCounts \</span><br><span class="line">  -I <span class="variable">$&#123;i&#125;</span> \</span><br><span class="line">  -L <span class="variable">$&#123;interval&#125;</span> \</span><br><span class="line">  -R <span class="variable">$&#123;GENOME&#125;</span> \</span><br><span class="line">  -O ./8.cnv/allelicCounts/<span class="variable">$&#123;id&#125;</span>.allelicCounts.tsv</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line">less ./8.cnv/allelicCounts/case1_biorep_A_techrep.allelicCounts.tsv | grep -v ^@ | awk <span class="string">&#x27;&#123;if($6 != &quot;N&quot;) print$0 &#125;&#x27;</span>| less</span><br><span class="line"></span><br><span class="line">less ./8.cnv/denoisedCR/case1_biorep_A_techrep.clean.denoisedCR.tsv | grep -v ^@| less</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/a-warning.png" alt="warning" style="zoom:100%;" /></center>

<center><img src="/img/Figures/allele-res.png" alt="res" style="zoom:100%;" /></center>

<h2 id="ModelSegments"><a href="#ModelSegments" class="headerlink" title="ModelSegments"></a>ModelSegments</h2><p>利用前面拿到的<strong>标准化&amp;降噪后tsv文件</strong>和<strong>等位基因测序深度tsv文件</strong>进行call segment，需要注意的是输入文件要求tumor match normal。这一步用到的工具是<code>ModelSegments</code>，它根据降噪后的reads counts值对copy ratios进行分割，并根据第五步的CollectAllelicCounts等位基因计数对分割片段进行分类。<br>注意：这一步非常非常非常非常耗内存和CPU</p>
<p><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us/articles/360057439631-ModelSegments">Manual-ModelSegments</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">cat 0.sra/sra.infor/config.tumor  | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  gatk --java-options <span class="string">&quot;-Xmx20g&quot;</span> ModelSegments \</span><br><span class="line">  --denoised-copy-ratios ./8.cnv/denoisedCR/<span class="variable">$&#123;id&#125;</span>.clean.denoisedCR.tsv \</span><br><span class="line">  --allelic-counts ./8.cnv/allelicCounts/<span class="variable">$&#123;id&#125;</span>.allelicCounts.tsv \</span><br><span class="line">  --normal-allelic-counts ./8.cnv/allelicCounts/case1_germline.allelicCounts.tsv \</span><br><span class="line">  --output ./8.cnv/segments \</span><br><span class="line">  --output-prefix <span class="variable">$&#123;id&#125;</span></span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<hr>
<blockquote>
<p>Exception in thread “main” java.lang.OutOfMemoryError: GC overhead limit exceeded<br>        at java.util.Arrays.copyOfRange(Arrays.java:3664)<br>        at java.lang.String.<init>(String.java:207)<br>        at java.lang.StringBuilder.toString(StringBuilder.java:407)<br>        …<br>啊啊啊啊啊啊啊啊啊我疯了！！！都给你60G了！！<br>Github有人题类似的问题，但是官方还没回答，等回答了再来看看</p>
</blockquote>
<hr>
<h1 id="拷贝数变异分析（CNVkit）"><a href="#拷贝数变异分析（CNVkit）" class="headerlink" title="拷贝数变异分析（CNVkit）"></a>拷贝数变异分析（CNVkit）</h1><p><a target="_blank" rel="noopener" href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004873">CNVkit: Genome-Wide Copy Number Detection and Visualization from Targeted DNA Sequencing</a><br><a target="_blank" rel="noopener" href="https://cnvkit.readthedocs.io/en/stable/index.html">Manual-CNVkit</a><br><a target="_blank" rel="noopener" href="https://github.com/etal/cnvkit">Github-CNVkit</a><br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/d5310b76f1f2">CNVkit教程</a></p>
<p><code>batch</code>是脚本内的一个整合了很多命令的方法，当然也可以使用<code>cnvkit.py</code>提供的<code>access</code>、<code>coverrage</code>、<code>fix</code>等方法一起来完成和batch同样功能的分析，但是作为懒人还是用了<code>batch</code>。而且CNVkit也太快了吧，这里还没写完就已经跑好了！</p>
<blockquote>
<p>重要参数：<br> Tumor.bam 和Normal.bam都是相应样本的bam文件<br> –targets 想要分析的区域信息<br> –annotate refFlat格式的基因注释信息，可以从UCSC上下载<br> –fasta 参考基因组<br> –access 需要跟bed文件，可用过cnvkit.py access mm10.fasta -s 10000 -o access-10kb.mm10.bed 生成<br> –output-reference 输出的reference.cnn可以作为下一批tumor数据分析的输入文件，reference.cnn和输入的normal.bam有关<br> –diagram –scatter 两个都是顺带绘图的参数<br> –drop-low-coverage to ignore bins with log2 normalized coverage values below -15.</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -------------------------------&gt;</span></span><br><span class="line"><span class="comment"># CNVkit-v0.9.8</span></span><br><span class="line"><span class="comment"># -------------------------------&gt; </span></span><br><span class="line"><span class="comment"># installation</span></span><br><span class="line">mamba install -c bioconda cnvkit</span><br><span class="line"></span><br><span class="line"><span class="comment"># 使用命令 batch 进行 CNV 分析</span></span><br><span class="line"><span class="comment"># From baits and tumor/normal BAMs</span></span><br><span class="line">GENOME=~/Reference/GATK/Homo_sapiens_assembly38.fasta</span><br><span class="line">bed=~/Reference/GATK/hg38_exon.bed</span><br><span class="line"></span><br><span class="line">/home/data/vip8t02/anaconda3/envs/CNVkit-0.9.8/bin/cnvkit.py batch ./5.gatk/bqsr.bam/*[ABC]*bqsr.bam   ./5.gatk/bqsr.bam/case1_techrep_bqsr.bam \</span><br><span class="line">--normal  ./5.gatk/bqsr.bam/case1_germline_bqsr.bam \</span><br><span class="line">--targets  <span class="variable">$&#123;bed&#125;</span> \</span><br><span class="line">--fasta <span class="variable">$GENOME</span>  \</span><br><span class="line">--drop-low-coverage --scatter --diagram --method amplicon -p 16 \</span><br><span class="line">--output-reference my_reference.cnn --output-dir ./8.cnv/cnvkit</span><br><span class="line"></span><br><span class="line"><span class="comment"># 划分segment</span></span><br><span class="line">/home/data/vip8t02/anaconda3/envs/CNVkit-0.9.8/bin/cnvkit.py <span class="built_in">export</span> seg *bqsr.cns -o gistic.segments</span><br><span class="line">sed <span class="string">&#x27;s/_bqsr//&#x27;</span> gistic.segments</span><br><span class="line"></span><br><span class="line">awk <span class="string">&#x27;&#123;print FILENAME&quot;\t&quot;$0&#125;&#x27;</span> *bqsr.cns  | grep -v chromosome |sed <span class="string">&#x27;s/_bqsr.cns//g&#x27;</span> |awk <span class="string">&#x27;&#123;print $1&quot;\t&quot;$2&quot;\t&quot;$3&quot;\t&quot;$4&quot;\t&quot;$8&quot;\t&quot;$6&#125;&#x27;</span> &gt;final.seg</span><br><span class="line"></span><br><span class="line">cat 0.sra/sra.infor/config.tumor | <span class="keyword">while</span> <span class="built_in">read</span> id;</span><br><span class="line"><span class="keyword">do</span> </span><br><span class="line">  /home/data/vip8t02/anaconda3/envs/CNVkit-0.9.8/bin/cnvkit.py <span class="built_in">export</span> seg ./8.cnv/cnvkit/<span class="variable">$&#123;id&#125;</span>_bqsr.cns \</span><br><span class="line">  -o ./8.cnv/cnvkit/<span class="variable">$&#123;id&#125;</span>.seg</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Visualization</span></span><br><span class="line">/home/data/vip8t02/anaconda3/envs/CNVkit-0.9.8/bin/cnvkit.py heatmap *.cns</span><br></pre></td></tr></table></figure>
<p>diagram用于展示单个样本的CNV在染色体上的分布:</p>
<center><img src="/img/Figures/cnvkit-case1.png" alt="res" style="zoom:40%;" /></center>

<p>scatter展示单个样本染色体区域上log2 ratio值的分布:</p>
<center><img src="/img/Figures/case1_biorep_A_techrep_bqsr-scatter.png" alt="scatter" style="zoom:65%;" /></center>

<p>heatmap展示多个样本的CNV分布情况:</p>
<center><img src="/img/Figures/heatmap.png" alt="heatmap" style="zoom:65%;" /></center>

<h1 id="突变Somatic-Signature图谱"><a href="#突变Somatic-Signature图谱" class="headerlink" title="突变Somatic Signature图谱"></a>突变Somatic Signature图谱</h1><ul>
<li>点突变SNV的类型，可以分为6类: <code>C&gt;A</code>, <code>C&gt;G</code>,<code>C&gt;T</code>, <code>A&gt;C</code>,<code>A&gt;G</code>, <code>A&gt;T</code>,而其他的点突变，如T&gt;G其实与A&gt;C是等效的，因为突变发生在哪条DNA链.上是无法确定的。如果把突变位点的侧翼各1 bp的碱基也考虑进来，也就是三连核苷酸突变，就有4x6x4=96种碱基突变类型。</li>
<li>发生Somatic mutations的原因:DNA replication infidelity, exogenous and endogenous genotoxins exposures, defective DNA repair pathways and DNA enzymatic editing。不同的突变原因会产生独特的突变类型组合，称为Mutations Signature。</li>
</ul>
<p>利用VEP注释后的maf文件来分析Signatures。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">rm(<span class="built_in">list</span>=ls())</span><br><span class="line">options(stringsAsFactors=<span class="literal">FALSE</span>)</span><br><span class="line"><span class="comment">## 切换镜像</span></span><br><span class="line">options(BioC_mirror=<span class="string">&quot;https://mirrors.ustc.edu.cn/bioc/&quot;</span>)</span><br><span class="line">options(<span class="string">&quot;repos&quot;</span> = <span class="built_in">c</span>(CRAN=<span class="string">&quot;https://mirrors.tuna.tsinghua.edu.cn/CRAN/&quot;</span>))</span><br><span class="line"><span class="comment">## 安装R包</span></span><br><span class="line">install.packages(<span class="string">&#x27;deconstructSigs&#x27;</span>)</span><br><span class="line">BiocManager::install(<span class="string">&#x27;BSgenome&#x27;</span>)</span><br><span class="line">BiocManager::install(<span class="string">&#x27;BSgenome.Hsapiens.UCSC.hg38&#x27;</span>)</span><br><span class="line">library(deconstructSigs)</span><br><span class="line"><span class="comment">## https://github.com/raerose01/deconstructSigs</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 读入数据</span></span><br><span class="line">maf=read.table(<span class="string">&#x27;./7.annotation/vep/VEP_merge.maf&#x27;</span>,header = <span class="built_in">T</span>,sep = <span class="string">&#x27;\t&#x27;</span>,<span class="built_in">quote</span> = <span class="string">&quot;&quot;</span>)</span><br><span class="line">maf[<span class="number">1</span>:<span class="number">5</span>,<span class="number">1</span>:<span class="number">5</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment">## 构建肿瘤突变数据框，需要5列信息: sample.ID,chr,pos,ref,alt </span></span><br><span class="line">sample.mut.ref &lt;- data.frame(Sample=maf[,<span class="number">16</span>], </span><br><span class="line">                            chr = maf[,<span class="number">5</span>],</span><br><span class="line">                            pos = maf[,<span class="number">6</span>],</span><br><span class="line">                            ref = maf[,<span class="number">11</span>],</span><br><span class="line">                            alt = maf[,<span class="number">13</span>])</span><br><span class="line"></span><br><span class="line"><span class="comment">## 查看构建的文件</span></span><br><span class="line">sample.mut.ref[<span class="number">1</span>:<span class="number">5</span>,<span class="number">1</span>:<span class="number">5</span>]</span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/mut-ref.png" alt="mut-ref" style="zoom:70%;" /></center>

<p>使用函数<code>mut.to.sigs.input</code>，将肿瘤的突变数据转换为n行和96列数据框，其中n是存在的样本数,这里是4个样本。每一列表示在每个三联核苷酸范围内发现突变的个数。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">sigs.input &lt;- mut.to.sigs.input(mut.ref = sample.mut.ref,</span><br><span class="line">                                sample.id = <span class="string">&quot;Sample&quot;</span>, </span><br><span class="line">                                chr = <span class="string">&quot;chr&quot;</span>, </span><br><span class="line">                                pos = <span class="string">&quot;pos&quot;</span>, </span><br><span class="line">                                ref = <span class="string">&quot;ref&quot;</span>, </span><br><span class="line">                                alt = <span class="string">&quot;alt&quot;</span>,</span><br><span class="line">                                bsg = BSgenome.Hsapiens.UCSC.hg38)</span><br><span class="line">sigs.input[<span class="number">1</span>:<span class="number">4</span>,<span class="number">1</span>:<span class="number">4</span>]</span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/sig-input.png" alt="sig-input" style="zoom:70%;" /></center>

<p>最后用函数<code>whichSignatures</code>确定给定样本中每个签名的数量，这里的<code>signatures.ref</code>参数指选定的参考数据库的signature可以选择<code>signatures.cosmic</code>或者<code>signatures.nature2013</code>，以<code>case1_biorep_A_techrep</code>为例进行可视化</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">sigs.output &lt;- whichSignatures(tumor.ref = sigs.input,</span><br><span class="line">                                signatures.ref = signatures.cosmic, </span><br><span class="line">                                sample.id = <span class="string">&#x27;case1_biorep_A_techrep&#x27;</span>,</span><br><span class="line">                                contexts.needed = <span class="literal">TRUE</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># Plot output</span></span><br><span class="line">plotSignatures(sigs.output)</span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/MutSig.png" alt="sig-onput" style="zoom:70%;" /></center>

<p>挑选出样本数大于0的Signature，绘制热图（因为样本数少，按照教程中挑出大于5的Sig的做法不可行）</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># pheatmap</span></span><br><span class="line">df = data.frame()</span><br><span class="line"><span class="keyword">for</span> (i <span class="keyword">in</span> rownames(sigs.input)) &#123;</span><br><span class="line">  sigs.output &lt;- whichSignatures(tumor.ref = sigs.input,</span><br><span class="line">                                 signatures.ref = signatures.cosmic, </span><br><span class="line">                                 sample.id = i,</span><br><span class="line">                                 contexts.needed = <span class="literal">TRUE</span>)</span><br><span class="line">  df = rbind(df,sigs.output$weights)</span><br><span class="line">&#125;</span><br><span class="line">df = df[ , apply(df, <span class="number">2</span>, <span class="keyword">function</span>(x)&#123;<span class="built_in">sum</span>(x&gt;<span class="number">0</span>)&#125;)&gt;<span class="number">0</span>]</span><br><span class="line">pheatmap::pheatmap(df,cluster_cols = <span class="built_in">F</span>,cluster_rows = <span class="built_in">F</span>,fontsize = <span class="number">20</span>)</span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/sig-heatmap.png" alt="sig-onput" style="zoom:40%;" /></center>

<h1 id="探究肿瘤异质性"><a href="#探究肿瘤异质性" class="headerlink" title="探究肿瘤异质性"></a>探究肿瘤异质性</h1><h2 id="同一病人的SNV分布"><a href="#同一病人的SNV分布" class="headerlink" title="同一病人的SNV分布"></a>同一病人的SNV分布</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">rm(<span class="built_in">list</span> = ls())</span><br><span class="line">options(stringsAsFactors = <span class="built_in">F</span>)</span><br><span class="line">library(dplyr)</span><br><span class="line">library(stringr)</span><br><span class="line">library(maftools)</span><br><span class="line"><span class="comment"># 读入数据</span></span><br><span class="line">laml = read.maf(<span class="string">&#x27;./7.annotation/vep/VEP_merge.maf&#x27;</span>)</span><br><span class="line">laml@data = laml@data[!grepl(<span class="string">&#x27;^MT-&#x27;</span>, laml@data$Hugo_Symbol),] </span><br><span class="line"><span class="comment"># 增加一列t_vaf，即肿瘤样本中突变位点的覆盖深度t_alt_count占测序覆盖深度t_depth的比值</span></span><br><span class="line">laml@data$t_vaf = (laml@data$t_alt_count/laml@data$t_depth)</span><br><span class="line">unique(laml@data$Tumor_Sample_Barcode)</span><br><span class="line">getSampleSummary(laml) </span><br><span class="line">getGeneSummary(laml) </span><br><span class="line">getFields(laml)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">mut = laml@data[laml@data$t_alt_count &gt;= <span class="number">5</span> &amp;</span><br><span class="line">                  laml@data$t_vaf &gt;= <span class="number">0.05</span>, <span class="built_in">c</span>(<span class="string">&quot;Hugo_Symbol&quot;</span>,</span><br><span class="line">                                             <span class="string">&quot;Chromosome&quot;</span>,</span><br><span class="line">                                             <span class="string">&quot;Start_Position&quot;</span>,</span><br><span class="line">                                             <span class="string">&quot;Tumor_Sample_Barcode&quot;</span>,</span><br><span class="line">                                             <span class="string">&quot;t_vaf&quot;</span>)]</span><br><span class="line"></span><br><span class="line">mut$patient = substr(mut$Tumor_Sample_Barcode, <span class="number">1</span>, <span class="number">5</span>)</span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/heter-mut.png" alt="mut-df" style="zoom:70%;" /></center>

<p>利用<code>pheatmap</code>进行可视化</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">library(tidyr)</span><br><span class="line">library(Cairo)</span><br><span class="line"></span><br><span class="line">pid = unique(mut$patient)</span><br><span class="line"></span><br><span class="line">mat = unique(mut[mut$patient == <span class="string">&quot;case1&quot;</span>,<span class="built_in">c</span>(<span class="string">&quot;Tumor_Sample_Barcode&quot;</span>,<span class="string">&#x27;Hugo_Symbol&#x27;</span>)]) </span><br><span class="line">mat$tmp = <span class="number">1</span></span><br><span class="line"><span class="comment"># 长变扁</span></span><br><span class="line">mat = spread(mat,Tumor_Sample_Barcode,tmp,fill = <span class="number">0</span>)</span><br><span class="line"><span class="built_in">class</span>(mat)</span><br><span class="line">mat = as.data.frame(mat)</span><br><span class="line">rownames(mat) = mat$Hugo_Symbol</span><br><span class="line">mat = mat[,-<span class="number">1</span>]</span><br><span class="line">dat = mat[order(mat[,<span class="number">1</span>],mat[,<span class="number">2</span>],mat[,<span class="number">3</span>],mat[,<span class="number">4</span>]),]</span><br><span class="line"></span><br><span class="line">CairoPNG(file=<span class="string">&quot;./10.heterogeneity/case1.png&quot;</span>,width=<span class="number">300</span>,height=<span class="number">1500</span>)</span><br><span class="line">pheatmap::pheatmap(mat = dat, cluster_cols = <span class="built_in">F</span>, cluster_rows = <span class="built_in">F</span>, show_rownames = <span class="built_in">T</span>, legend = <span class="built_in">F</span>,fontsize_row = <span class="number">16</span>,fontsize_col = <span class="number">30</span>)</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<p>热图太长了，仅截取展示</p>
<center><img src="/img/Figures/heatmap-hetero.png" alt="mut-df" style="zoom:80%;" /></center>

<h2 id="肿瘤样本中的异质性"><a href="#肿瘤样本中的异质性" class="headerlink" title="肿瘤样本中的异质性"></a>肿瘤样本中的异质性</h2><p>肿瘤通常是异质性的，即由多个克隆组成。在这里，我们通过肿瘤样本中的变异等位基因频率(Variant Allele Frequecy)值，来进行分析。使用的是maftools的<code>inferHeterogeneity</code>函数:<br>This function clusters variants based on VAF to estimate univariate density and cluster classification. There are two methods available for clustering. Default using parametric finite mixture models and another method using nonparametric inifinite mixture models (Dirichlet process).</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">rm(<span class="built_in">list</span> = ls())</span><br><span class="line">options(stringsAsFactors = <span class="built_in">F</span>)</span><br><span class="line"></span><br><span class="line">library(maftools)</span><br><span class="line">library(dplyr)</span><br><span class="line">library(stringr)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 读入数据</span></span><br><span class="line">laml = read.maf(<span class="string">&#x27;./7.annotation/vep/vep_merge.maf&#x27;</span>)</span><br><span class="line">laml@data = laml@data[!grepl(<span class="string">&#x27;^MT-&#x27;</span>,laml@data$Hugo_Symbol),] </span><br><span class="line"></span><br><span class="line"><span class="comment">## 增加一列 t_vaf，即肿瘤样本中突变位点的覆盖深度t_alt_count占测序覆盖深度t_depth的比值</span></span><br><span class="line">laml@data$t_vaf = (laml@data$t_alt_count/laml@data$t_depth)</span><br><span class="line">heter = inferHeterogeneity(maf = laml, top = <span class="number">24</span>,vafCol = <span class="string">&#x27;t_vaf&#x27;</span>)</span><br><span class="line"></span><br><span class="line">heter</span><br><span class="line"></span><br><span class="line"><span class="comment">## Visualization</span></span><br><span class="line">plotClusters(clusters = heter,</span><br><span class="line">             showCNvars = <span class="built_in">T</span>,</span><br><span class="line">             tsb = <span class="string">&#x27;case1_biorep_A_techrep&#x27;</span>)</span><br></pre></td></tr></table></figure>
<p>得到的<code>heter</code>是一个包括<code>clusterData</code>和<code>clusterMeans</code>的列表</p>
<figure class="half">
    <img src="/img/Figures/case1-vaf2.png" style="zoom:60%;">
    <img src="/img/Figures/case1-vaf.png" style="zoom:40%;">
</figure>

<p>上图显示了不同的突变位点的vaf值，并将这些位点进行分群，在样本case1_biorep_A_techrep中就可以分为4群，即该样本可以分为4个克隆，同样还给出了MATH 值:52.95。</p>
<h1 id="注释突变位点所在基因"><a href="#注释突变位点所在基因" class="headerlink" title="注释突变位点所在基因"></a>注释突变位点所在基因</h1><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br></pre></td><td class="code"><pre><span class="line">rm(<span class="built_in">list</span>=ls())</span><br><span class="line">options(stringsAsFactors = <span class="built_in">F</span>)</span><br><span class="line">library(dplyr)</span><br><span class="line">library(stringr)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 读入数据</span></span><br><span class="line">setwd(<span class="string">&quot;/home/data/vip8t02/MyCourse/CancerGenomics/WES.proj/&quot;</span>)</span><br><span class="line">laml = read.maf(<span class="string">&#x27;./7.annotation/vep/vep_merge.maf&#x27;</span>)</span><br><span class="line">laml@data=laml@data[!grepl(<span class="string">&#x27;^MT-&#x27;</span>,laml@data$Hugo_Symbol),] </span><br><span class="line"></span><br><span class="line"><span class="comment"># 增加一列t_vaf，即肿瘤样本中突变位点的覆盖深度t_alt_count占测序覆盖深度t_depth的比值</span></span><br><span class="line">laml@data$t_vaf = (laml@data$t_alt_count/laml@data$t_depth)</span><br><span class="line">unique(laml@data$Tumor_Sample_Barcode)</span><br><span class="line">getSampleSummary(laml) </span><br><span class="line">getGeneSummary(laml) </span><br><span class="line">getFields(laml)</span><br><span class="line"></span><br><span class="line">mut = laml@data[laml@data$t_alt_count &gt;= <span class="number">5</span> &amp;</span><br><span class="line">                  laml@data$t_vaf &gt;= <span class="number">0.05</span>, <span class="built_in">c</span>(<span class="string">&quot;Hugo_Symbol&quot;</span>,</span><br><span class="line">                                             <span class="string">&quot;Chromosome&quot;</span>,</span><br><span class="line">                                             <span class="string">&quot;Start_Position&quot;</span>,</span><br><span class="line">                                             <span class="string">&quot;Tumor_Sample_Barcode&quot;</span>,</span><br><span class="line">                                             <span class="string">&quot;t_vaf&quot;</span>)]</span><br><span class="line"></span><br><span class="line">mut$patient = substr(mut$Tumor_Sample_Barcode, <span class="number">1</span>, <span class="number">5</span>)</span><br><span class="line"></span><br><span class="line">pid = unique(mut$patient)</span><br><span class="line"></span><br><span class="line">mat=unique(mut[mut$patient %in% <span class="string">&quot;case1&quot;</span>,<span class="built_in">c</span>(<span class="string">&quot;Tumor_Sample_Barcode&quot;</span>,<span class="string">&#x27;Hugo_Symbol&#x27;</span>)]) </span><br><span class="line">mat$tmp = <span class="number">1</span></span><br><span class="line"><span class="comment"># 长变扁</span></span><br><span class="line">mat = spread(mat,Tumor_Sample_Barcode,tmp,fill = <span class="number">0</span>)</span><br><span class="line"><span class="built_in">class</span>(mat)</span><br><span class="line">mat = as.data.frame(mat)</span><br><span class="line">rownames(mat) = mat$Hugo_Symbol</span><br><span class="line">mat=mat[,-<span class="number">1</span>]</span><br><span class="line">dat = mat[order(mat[,<span class="number">1</span>],mat[,<span class="number">2</span>],mat[,<span class="number">3</span>],mat[,<span class="number">4</span>]),]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 突变分类</span></span><br><span class="line">trunk_gene = rownames(dat[rowSums(dat)==<span class="number">4</span>,]) <span class="comment"># mutate in 4 samples</span></span><br><span class="line">branch_gene = rownames(dat[rowSums(dat)==<span class="number">3</span>,]) <span class="comment"># mutate in 3 samples</span></span><br><span class="line">private_gene = rownames(dat[rowSums(dat)==<span class="number">1</span>,]) <span class="comment"># mutate in 1 samples</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># KEGG</span></span><br><span class="line">options(connectionObserver = <span class="literal">NULL</span>)</span><br><span class="line"></span><br><span class="line">library(org.Hs.eg.db)</span><br><span class="line">library(clusterProfiler)</span><br><span class="line">library(Cairo)</span><br><span class="line"></span><br><span class="line">kegg_SYMBOL_hsa &lt;- <span class="keyword">function</span>(genes)&#123; </span><br><span class="line">  gene.df &lt;- bitr(genes, fromType = <span class="string">&quot;SYMBOL&quot;</span>,</span><br><span class="line">                  toType = <span class="built_in">c</span>(<span class="string">&quot;SYMBOL&quot;</span>, <span class="string">&quot;ENTREZID&quot;</span>),</span><br><span class="line">                  OrgDb = org.Hs.eg.db)</span><br><span class="line">  head(gene.df) </span><br><span class="line">  diff.kk &lt;- enrichKEGG(gene         = gene.df$ENTREZID,</span><br><span class="line">                        organism     = <span class="string">&#x27;hsa&#x27;</span>,</span><br><span class="line">                        pvalueCutoff = <span class="number">0.99</span>,</span><br><span class="line">                        qvalueCutoff = <span class="number">0.99</span></span><br><span class="line">  )</span><br><span class="line">  <span class="built_in">return</span>( setReadable(diff.kk, OrgDb = org.Hs.eg.db,keyType = <span class="string">&#x27;ENTREZID&#x27;</span>))</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">## trunk</span></span><br><span class="line">trunk_kk=kegg_SYMBOL_hsa(trunk_gene)</span><br><span class="line">trunk_df=trunk_kk@result</span><br><span class="line"></span><br><span class="line">write.csv(trunk_df,file = <span class="string">&#x27;./11.KEGG//trunk_kegg.csv&#x27;</span>)</span><br><span class="line">CairoPNG(file=<span class="string">&quot;./11.KEGG/trunk_kegg.png&quot;</span>,width=<span class="number">1080</span>,height=<span class="number">540</span>)</span><br><span class="line">barplot(trunk_kk,font.size = <span class="number">20</span>)</span><br><span class="line">dev.off()</span><br><span class="line"></span><br><span class="line"><span class="comment">## branch</span></span><br><span class="line">branch_kk=kegg_SYMBOL_hsa(branch_gene)</span><br><span class="line">branch_df=branch_kk@result</span><br><span class="line"></span><br><span class="line">write.csv(branch_df,file = <span class="string">&#x27;./11.KEGG/branch_kegg.csv&#x27;</span>)</span><br><span class="line">CairoPNG(file=<span class="string">&quot;./11.KEGG/branch_kegg.png&quot;</span>,width=<span class="number">1080</span>,height=<span class="number">540</span>)</span><br><span class="line">barplot(branch_kk,font.size = <span class="number">20</span>)</span><br><span class="line">dev.off()</span><br><span class="line"></span><br><span class="line"><span class="comment">## private</span></span><br><span class="line">private_kk=kegg_SYMBOL_hsa(private_gene)</span><br><span class="line">private_df=private_kk@result</span><br><span class="line"></span><br><span class="line">write.csv(private_df,file = <span class="string">&#x27;./11.KEGG/private_kegg.csv&#x27;</span>)</span><br><span class="line">CairoPNG(file=<span class="string">&quot;./11.KEGG/private_kegg.png&quot;</span>,width=<span class="number">1080</span>,height=<span class="number">540</span>)</span><br><span class="line">barplot(private_kk,font.size = <span class="number">20</span>)</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<hr>
<p><strong>Troubleshooting</strong></p>
<blockquote>
<p>library(org.Hs.eg.db)<br> Error: package or namespace load failed for ‘org.Hs.eg.db’:<br> .onLoad failed in loadNamespace() for ‘org.Hs.eg.db’, details:<br>  call: l$contains<br>  error: $ operator is invalid for atomic vectors</p>
</blockquote>
<p>在Bioconductor上已经有人提了问题，并且有了回答（地址：<a target="_blank" rel="noopener" href="https://support.bioconductor.org/p/9136329/%EF%BC%89">https://support.bioconductor.org/p/9136329/）</a><br>简单的说，就是Rstduio出来背锅，最新的RSQLite(v2.2.6)和Rstudio存在冲突，以后要么是RSQLite迁就RStduio，要么就是Rstudio迁就RQLite。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 方法1 手动设置环境变量</span></span><br><span class="line">options(connectionObserver = <span class="literal">NULL</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 方法2: 装个旧版本RSQLite，2.2.5之前即可</span></span><br><span class="line">install.packages(<span class="string">&quot;remotes&quot;</span>)</span><br><span class="line">remotes::install_version(<span class="string">&quot;RSQLite&quot;</span>, <span class="string">&quot;2.2.5&quot;</span>)</span><br></pre></td></tr></table></figure>
<hr>
<p>注释结果：</p>
<ol>
<li><p><strong>trunk_gene</strong></p>
<center><img src="/img/Figures/trunk_kegg.png" alt="trunk_kegg" style="zoom:70%;" /></center>  

<p> 结核？？？？<br> Tuberculosis (TB) is a potentially serious infectious disease that mainly affects your lungs. The bacteria that cause tuberculosis are spread from one person to another through tiny droplets released into the air via coughs and sneezes.</p>
<p> 百日咳？？？？？？？<br> Pertussis, also known as whooping cough, is a highly contagious respiratory disease. It is caused by the bacterium Bordetella pertussis. Pertussis is known for uncontrollable, violent coughing which often makes it hard to breathe.</p>
<p> <strong>NF-kB Signaling Pathway</strong><br> NF-κB蛋白通常会由p65和p50形成同源/异源二聚体，在胞质中因与抑制蛋白IkB结合形成了三聚体复合物而处于失活状态。当上游信号因子TNF结合到细胞膜表面受体后，受体构象改变并将信号传递给IKK激酶（IkB kinase），进而使IkB蛋白磷酸化并将其从三聚体中解离。随后NF-κB二聚体暴露出核定位序列（NLS），迅速从细胞质进入细胞核内，与核内DNA上的特异序列相结合，促进相关基因的转录，比如CyclinD1、c-Myc、MMP-9、VEGF等。NF-κB的持续性激活能够导致与细胞周期相关的蛋白活性异常，提高一些促进肿瘤生长的细胞因子水平，介导新血管生成。</p>
</li>
<li><p><strong>branch_gene</strong></p>
<center><img src="/img/Figures/branch_kegg.png" alt="branch_kegg" style="zoom:70%;" /></center>  
 

<p> <strong>Notch signaling pathway</strong><br> Notch信号通路广泛参与了恶性肿瘤的发生发展。Notch信号传导通路发挥致癌作用，导致细胞在增殖，细胞周期抑制，分化，和凋亡的过程失调和失控，造成了细胞的转化及恶性化，最终导致了恶性肿瘤的发生。</p>
<p> <strong>Adherens junctions（黏着链接）</strong><br> Adherens junctions (AJs) are evolutionarily conserved plasma-membrane structures that mediate cell–cell adhesions in multicellular organisms. They are organized by several types of adhesive integral membrane proteins, most notably cadherins and nectins that are clustered and stabilized by a number of cytoplasmic scaffolds.</p>
</li>
<li><p><strong>private_gene</strong></p>
<center><img src="/img/Figures/private_kegg.png" alt="private_kegg" style="zoom:70%;" /></center>  
 

<p> <strong>Wnt signaling pathway</strong><br> The Wnt signaling pathway is an ancient and evolutionarily conserved pathway that regulates crucial aspects of cell fate determination, cell migration, cell polarity, neural patterning and organogenesis during embryonic development. Wnt信号传导途径是由配体蛋白质Wnt和膜蛋白受体结合激发的一组多下游通道的信号转导途径。</p>
</li>
</ol>
<h1 id="ABSOLUTE使用"><a href="#ABSOLUTE使用" class="headerlink" title="ABSOLUTE使用"></a>ABSOLUTE使用</h1><p>ABSOLUTE是一种直接从体细胞突变和拷贝数变异的结果中<strong>推断肿瘤纯度和倍性</strong>的计算方法。</p>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/nbt.2203#Sec11">文献：2012-NBT-Absolute quantification of somatic DNA alterations in human cancer</a><br><a target="_blank" rel="noopener" href="https://software.broadinstitute.org/cancer/cga/absolute">ABSOLUTE Homepage</a></p>
<p>通常来说，我们取肿瘤组织进行测序,往往是一个混合样品，既包括肿瘤细胞也包括正常细胞，因此需要进行肿瘤纯度purity的评估。当从混合样品中提取DNA进行测序后，得到的也是一个混合样品的结果。肿瘤不一定是单纯的二倍体了，其本身异质性高，存在部分突变部分不突变的情况，直接分析拷贝数变异,得到的结果并不准确，评估肿瘤倍性ploidy也更加必要。</p>
<p>而ABSOLUTE可以从混合的DNA群体中重新推算样品中的绝对拷贝数、肿瘤纯度(purity)和倍性(ploidy) 。该算法首先从拷贝数分析得到的segments文件,以及体细胞突变的maf文件，然后基于预先计算好癌症核型的模型(数据来自于TCGA)进行评分，最后给出若干个模型供用户<br>选择。需要注意的是，最高分模型不一样是最优的，需要用户根据自己的数据和经验进行一定的检查。</p>
<p>ABSOLUTE每次仅能处理一个样本，GitHub上有个叫<code>DoAbsolute</code>的项目，可以实现一次处理多个样本：<a target="_blank" rel="noopener" href="https://github.com/ShixiangWang/DoAbsolute#:~:text=ABSOLUTE%20is%20a%20famous%20software%20developed%20by%20Broad,to%20input%20data%20easily%20and%20runs%20RunAbsolute%20parallelly.">GitHub-ShixiangWang/DoAbsolute</a></p>
<blockquote>
<p><strong>Algorithm</strong><br> ABSOLUTE extracts the absolute copy number of local DNA segments per cancer cell from the mixed DNA population in three steps:</p>
<ol>
<li>Estimates the tumor purity and ploidy from observed relative copy profiles and, if provided, from somatic point mutation data.</li>
<li>Resolves ambiguous cases of purity and ploidy using pre-computed statistical models of recurrent cancer karyotypes based on a large and diverse reference sample collection.</li>
<li>Attempts to account for copy number alterations and point mutations in tumor subclones.</li>
</ol>
<p>ABSOLUTE expects copy-ratios very close to 1.0 and will fail if ratios are less than 0.75 or greater than 1.25. ABSOLUTE analysis can fail due to exceeding the max.as.seg.count threshold. Too many segments are associated with noisy or poor quality data. </p>
</blockquote>
<h2 id="ABSOLUTE安装"><a href="#ABSOLUTE安装" class="headerlink" title="ABSOLUTE安装"></a>ABSOLUTE安装</h2><p>需要在<a target="_blank" rel="noopener" href="https://software.broadinstitute.org/cancer/cga/absolute">官网</a>注册账号后，才能下载ABSOLUTE的tar.gz版本，然后在R里面本地安装。还需要下载他的依赖数据HAPSEG(163MB)及依赖包</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## 首先需要先安装两个依赖包</span></span><br><span class="line">install.packages(<span class="string">&quot;numDeriv&quot;</span>)</span><br><span class="line">BiocManager::install(<span class="string">&quot;DNAcopy&quot;</span>)</span><br><span class="line"><span class="comment">## 下载 HAPSEG_1.1.1.tar.gz，地址是</span></span><br><span class="line"><span class="comment">## http://software.broadinstitute.org/cancer/cga/sites/default/files/data/tools/absolute/HAPSEG_1.1.1.tar.gz</span></span><br><span class="line"><span class="comment">## 本地安装 HAPSEG，需要注意路径的问题</span></span><br><span class="line">install.packages(<span class="string">&quot;./12.ABSOLUTE/HAPSEG_1.1.1.tar.gz&quot;</span>, repos = <span class="literal">NULL</span>)</span><br><span class="line"><span class="comment">## 最后才是 ABSOLUTE，也是采用本地安装</span></span><br><span class="line"><span class="comment">## http://software.broadinstitute.org/cancer/cga/sites/default/files/data/tools/absolute/ABSOLUTE_1.0.6.tar.gz</span></span><br><span class="line">install.packages(<span class="string">&quot;./12.ABSOLUTE/ABSOLUTE_1.0.6.tar.gz&quot;</span>,repos = <span class="literal">NULL</span>)</span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/install-ABSOLUTE.png" alt="ABSOLUTE installation" style="zoom:90%;" /></center>

<h2 id="数据分析"><a href="#数据分析" class="headerlink" title="数据分析"></a>数据分析</h2><p><strong>input</strong></p>
<ol>
<li><p><strong>Segment文件</strong><br>Segmented copy ratios data file in either of the following two formats:</p>
<ul>
<li>For ALLELIC copy number type analysis, supply an RData file produced by HAPSEG or AllelicCapseg. These datasets allow incorporation of copy neutral LOH events. </li>
<li>Segmentation data produced by any other means must conform to the output formats of HAPSEG/AllelicCapseg for ABSOLUTE to consider copy neutral LOH events. </li>
</ul>
<p>ABSOLUTE algorithm v1.0.6 requires the following five columns：Chromosome(In either chr# or # format), Start, End, Num_Probes, Segment_Mean</p>
<center><img src="/img/Figures/input-seg.png" alt="input-seg" style="zoom:90%;" /></center>  

<p>segment文件不符合要求</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">cat 0.sra/sra.infor/config.tumor | <span class="keyword">while</span> <span class="built_in">read</span> id;</span><br><span class="line"><span class="keyword">do</span> </span><br><span class="line">sed -e <span class="string">&#x27;s/chrom/Chromosome/g&#x27;</span> -e <span class="string">&#x27;s/loc.start/Start/g&#x27;</span> -e <span class="string">&#x27;s/loc.end/End/g&#x27;</span> -e <span class="string">&#x27;s/num.mark/Num_Probes/g&#x27;</span> -e <span class="string">&#x27;s/seg.mean/Segment_Mean/g&#x27;</span> ./8.cnv/cnvkit/<span class="variable">$&#123;id&#125;</span>.seg &gt; ./12.ABSOLUTE/<span class="variable">$&#123;id&#125;</span>_absolute.seg;</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure></li>
<li><p><strong>maf文件</strong><br>(Optional) Somatic mutation data in mutation annotation format (MAF) and as a plain text file. ABSOLUTE algorithm v1.0.6 requires the following seven columns： </p>
<ul>
<li>t_ref_count OR i_t_ref_count： Count of reference alleles in tumor.</li>
<li>t_alt_count OR i_t_alt_count：Count of alternate alleles in tumor. </li>
<li>dbSNP_Val_Status：Fields may be blank and multiple values are separated with nonspaced semicolon. Example values include bySubmitter, by1000genomes, by2Hit2Allele, and byHapMap.</li>
<li>Start_position：Note the <strong>lowercase “p”</strong>. Also, note that the End_position column is not required. This implies that ABSOLUTE algorithm v1.0.6 <strong>treats all mutation data equally as point mutations</strong>, the expected type of mutation data.</li>
<li>Tumor_Sample_Barcode：Fields may be blank.</li>
<li>Hugo_Symbol：Fields may be blank or “unknown”.</li>
<li>Chromosome：<strong>Must be in # format and not chr# format</strong>. </li>
</ul>
<center><img src="/img/Figures/input-maf.png" alt="input-maf" style="zoom:90%;" /></center>  

<p>可以看到是大写“ P ”，和chr1，需要修改格式</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">cat 0.sra/sra.infor/config.tumor | <span class="keyword">while</span> <span class="built_in">read</span> id;</span><br><span class="line"><span class="keyword">do</span> </span><br><span class="line">sed -e <span class="string">&#x27;1d&#x27;</span> -e <span class="string">&#x27;s/Start_Position/Start_position/&#x27;</span> -e <span class="string">&#x27;s/chr//&#x27;</span> ./7.annotation/vep/<span class="variable">$&#123;id&#125;</span>_vep.maf &gt; ./12.ABSOLUTE/<span class="variable">$&#123;id&#125;</span>_absolute.maf;</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<blockquote>
<p>重要参数：<br><code>max sigma h</code>: 这个可以根据情况设大点，因为样本间误差设小了，结果反而不好。<br><code>max as seg count</code>: 这个最好事先看一些每个样本segmentation数目的分布，默认设置为1500，也就是超过这个数目，样本会直接打上“Fail”标签，不会出现在最后的结果中。<br><code>max non clonal</code>: 这个参数我看到作者处理示例数据使用的是1，也就是运行基因组存在non-clonal的最大比例<br><code>copy number type</code>: 一般我们处理的都是total copy number，ABSOLUTE仅支持由HAPSEG或AllelicCapseg输入的allelic数据。<br>BTW:作者说过默认参数的设置是为了平衡过拟合以及解的复杂性。</p>
</blockquote>
</li>
</ol>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">library(ABSOLUTE)</span><br><span class="line">RunAbsolute(</span><br><span class="line">  seg.dat.fn = <span class="string">&quot;./12.ABSOLUTE/case1_biorep_A_techrep_absolute.seg&quot;</span>, </span><br><span class="line">  sigma.p = <span class="number">0</span>,</span><br><span class="line">  max.sigma.h = <span class="number">0.2</span>,</span><br><span class="line">  min.ploidy = <span class="number">0.5</span>, </span><br><span class="line">  max.ploidy = <span class="number">8</span>, </span><br><span class="line">  primary.disease = <span class="string">&quot;BRCA&quot;</span>, </span><br><span class="line">  platform = <span class="string">&quot;Illumina_WES&quot;</span>, </span><br><span class="line">  results.dir = <span class="string">&quot;./12.ABSOLUTE/case1_biorep_A_techrep&quot;</span>, </span><br><span class="line">  sample.name = <span class="string">&quot;case1_biorep_A_techrep&quot;</span>, </span><br><span class="line">  max.as.seg.count = <span class="number">1500</span>, </span><br><span class="line">  max.neg.genome = <span class="number">0.005</span>,</span><br><span class="line">  max.non.clonal = <span class="number">1</span>, </span><br><span class="line">  copy_num_type = <span class="string">&quot;total&quot;</span>,</span><br><span class="line">  maf.fn = <span class="string">&quot;./12.ABSOLUTE/case1_biorep_A_techrep_absolute.maf&quot;</span>,</span><br><span class="line">  min.mut.af = <span class="number">0.05</span></span><br><span class="line">)</span><br><span class="line"><span class="comment"># 其余样本的代码同上</span></span><br></pre></td></tr></table></figure>
<h2 id="结果解读"><a href="#结果解读" class="headerlink" title="结果解读"></a>结果解读</h2><p>以下结果解读以case1_biorep_A_techrep为例，每种颜色代表一个模型：</p>
<ol>
<li><p>纯度/倍性图以及模型打分<br>右图表示肿瘤纯度与倍性的关系。每个点表示软件给出的一个模型，彩色的点表示得分高的<br>左图中展示了3个打分最高的解，每个解通过不同的模型进行评估，主要是SCNA，核型模型以及它们的整合模型。</p>
<center><img src="/img/Figures/ab-fig1.png" alt="Fig.1" style="zoom:90%;" /></center>
</li>
<li><p>备选的拷贝轮廓<br>这个图显示了前面3个模型对应的拷贝轮廓，即拷贝率copy ratio与Genomic fraction的关系，像纯度，倍性，异质性片段的比例以及还有一些看不懂为什么是负数的参数值。其实有个参数跟我们输入时密切相关，就是\sigma_h，它标定了样本允许的最大的方差。虚线代表基因组倍性。</p>
<center><img src="/img/Figures/ab-fig2.png" alt="Fig.2" style="zoom:90%;" /></center>
</li>
<li><p>等位基因比例图<br>该图表示这个模型体细胞突变SSNV (Somatic Single Nucleotide Variation)等位基因分数的分布。等位基因比例可以解释为每个cancer细胞平均的等位基因拷贝数，也称为多样性（multiplicity），它可以揭示亚克隆突变。这其实跟利用VCF文件计算vaf分布差不多，不过这里用copy number进行了校正。CCF是指癌细胞分数。</p>
<center><img src="/img/Figures/ab-fig3.png" alt="Fig.3" style="zoom:90%;" /></center>

</li>
</ol>
<h1 id="Pyclone推断肿瘤细胞的克隆组成"><a href="#Pyclone推断肿瘤细胞的克隆组成" class="headerlink" title="Pyclone推断肿瘤细胞的克隆组成"></a>Pyclone推断肿瘤细胞的克隆组成</h1><p><a target="_blank" rel="noopener" href="https://github.com/Roth-Lab/pyclone">GitHub-Pyclone</a></p>
<p>肿瘤异质性可分为肿瘤间异质性和肿瘤内异质性，其中肿瘤内异质性 Intra-tumor heterogeneity (ITH) 指的是，随着癌细胞的不断生长，其分裂后的子代细胞呈现出与同代细胞或者父细胞的不同，从而使得其各个方面有了较大的差异。</p>
<p>对存在异质性的肿瘤，多次不同部位取样，或者不同时间取样，那么虽然说是同一个肿瘤病人，他们不同的测序结果理论上不一致的，Pyclone就可以分析出来，同一个病人的不同测序数据里面共有哪些亚克隆。</p>
<p>考虑到突变频率的影响因素的复杂性:肿瘤组织中混有的正常细胞、携带该突变的肿瘤细胞的比例、每个细胞中突变的等位基因拷贝数、以及未知的技术噪声来源，Pyclone 基于<strong>贝叶斯聚类方法</strong>，用于将一个或多个位点取样深度测序(通常大于1000x)的体细胞突变归类为推定的克隆clusters,同时估算其细胞患病率，并解释由拷贝数变异和正常细胞污染引起的等位基因失衡问题。</p>
<h2 id="Pyclone安装"><a href="#Pyclone安装" class="headerlink" title="Pyclone安装"></a>Pyclone安装</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## 创建虚拟环境</span></span><br><span class="line">conda create -n pyclone-py27</span><br><span class="line">conda activate pyclone-py27</span><br><span class="line"><span class="comment">## 用mamba安装pyclone</span></span><br><span class="line">mamba install -c hcc pyclone</span><br><span class="line">mamba install pyclone -c bioconda -c conda-forge</span><br></pre></td></tr></table></figure>
<h2 id="数据分析-1"><a href="#数据分析-1" class="headerlink" title="数据分析"></a>数据分析</h2><p><strong>input</strong><br>.tsv文件需要包括：</p>
<ul>
<li>mutation_id: A unique identifier for the mutation. This should be the same across datasets.</li>
<li>ref_counts: The number of reads overlapping the locus matching the reference allele.</li>
<li>var_counts: The number of reads overlapping the locus matching the variant allele.</li>
<li>normal_cn: The copy number of the locus in non-malignant cells. This should generally be 2 except for sex chromosomes in males.</li>
<li>（以上四列在VEP注释后得到的.maf文件里）</li>
<li>minor_cn: The copy number of the minor allele in the malignant cells. This must be less than equal the value in the major_cn column.</li>
<li>major_cn: The copy number of the major allele in the malignant cells. This should be greater than equal to the value in the minor_cn column and greater than 0.</li>
</ul>
<p>作者建议在没有<code>minor_cn</code>和<code>major_cn</code>的情况下，把<code>minor_cn</code>设置为0，把<code>major_cn</code>设置为推断出来的拷贝数，然后加<code>--prior total_copy_number</code>参数。</p>
<p>先从前面的拷贝数变异分析的结果来获取major_cn,然后根据突变坐标映射到拷贝数变异分析的segments文件，这里根据.seg文件的Segment_Mean来计算拷贝数，Segment_Mean大于0则拷贝数扩增，小于0则拷贝数缺失，但是通常在-0.2~0.2之间都认为是正常，也有一些软件的cutoff是0.3。</p>
<center><img src="/img/Figures/cpnum.png" alt="copy number" style="zoom:40%;" /></center>

<p><strong>input构建过程：</strong></p>
<ol>
<li><p>下面的临时tsv文件前三列是暂时添加的，而<code>mutation_ id</code>合并于maf文件的<code>Hugo_ Symbol</code>、<code>Chromosome</code>和<code>Start_Position</code>, 把<code>major_cn</code>暂定为2，后面再做修改，最后生成文件命名为<code>$&#123;id&#125;.tmp.tsv</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">cat 0.sra/sra.infor/config.tumor | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span> </span><br><span class="line">cat ./7.annotation/vep/<span class="variable">$&#123;id&#125;</span>_vep.maf | sed <span class="string">&#x27;1,2d&#x27;</span> | awk -F <span class="string">&#x27;\t&#x27;</span> <span class="string">&#x27;&#123;print $5&quot;\t&quot;$6&quot;\t&quot;$7&quot;\t&quot;$5&quot;:&quot;$6&quot;:&quot;$1&quot;\t&quot;$41&quot;\t&quot;$42&quot;\t&quot;2&quot;\t&quot;0&quot;\t&quot;2&quot;\t&quot;&#125;&#x27;</span> &gt;./13.pyclone/<span class="variable">$&#123;id&#125;</span>.tmp.tsv</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/pyclone-tmp1.png" alt="tmp" style="zoom:90%;" /></center>
</li>
<li><p>然后从segments文件中获取突变位点的<code>Segment_Mean</code>需要先把segments文件进行处理，由于直接计算出来的拷贝数为小数，因此需要进行取舍，这里简单的进行四舍五入，如果为0，则删除这条记录</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">cat 0.sra/sra.infor/config.tumor | <span class="keyword">while</span> <span class="built_in">read</span> id</span><br><span class="line"><span class="keyword">do</span> </span><br><span class="line">sed <span class="string">&#x27;1d&#x27;</span> ./8.cnv/cnvkit/<span class="variable">$&#123;id&#125;</span>.seg | awk <span class="string">&#x27;BEGIN&#123;OFS=&quot;\t&quot;&#125;&#123;print $0&quot;\t&quot;int((2^$6)*2+0.5)&#125;&#x27;</span>| awk <span class="string">&#x27;BEGIN&#123;OFS=&quot;\t&quot;&#125;&#123;if ($7!=0)print $0&#125;&#x27;</span> | cut -f 2-7  &gt;./13.pyclone/<span class="variable">$&#123;id&#125;</span>.bed</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/pyclone-tmp2.png" alt="tmp" style="zoom:90%;" /></center>
</li>
<li><p>最后用到了bedtools工具，把两个文件的坐标进行overlap一下，取出必要的列<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/f8bbd51b5199">bedtools教程</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">cat 0.sra/sra.infor/config.tumor | <span class="keyword">while</span> <span class="built_in">read</span> id;</span><br><span class="line"><span class="keyword">do</span> </span><br><span class="line">bedtools window -a ./13.pyclone/<span class="variable">$&#123;id&#125;</span>.tmp.tsv  -b ./13.pyclone/<span class="variable">$&#123;id&#125;</span>.bed | cut -f 4-8,15 | awk <span class="string">&#x27;BEGIN&#123;OFS=&quot;\t&quot;;print &quot;mutation_id\tref_counts\tvar_counts\tnormal_cn\tminor_cn\tmajor_cn&quot;&#125;&#123;print $0&#125;&#x27;</span> &gt;./13.pyclone/<span class="variable">$&#123;id&#125;</span>.tsv</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/pyclone-tmp3.png" alt="tmp" style="zoom:90%;" /></center>

</li>
</ol>
<p><strong>pyclone分析</strong></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">PyClone run_analysis_pipeline --prior total_copy_number \</span><br><span class="line">--in_files ./13.pyclone/*.tsv \</span><br><span class="line">--working_dir ./13.pyclone/case1_pyclone_analysis 1&gt;./<span class="built_in">log</span>/pyclone.log/case1.log 2&gt;&amp;1</span><br></pre></td></tr></table></figure>
<h2 id="结果解读-1"><a href="#结果解读-1" class="headerlink" title="结果解读"></a>结果解读</h2><p>case1的4个样本:从下面左边的图可以看到，这4个样本的突变可以分为3个clusters,每个clusters的细胞患病率有所异同，每个clusters所含有的突变数量n也不同(cluster0=21，cluster1=32, cluster2=1)总共有21个突变位点。和教程的图很像。</p>
<center><img src="/img/Figures/pyclone1.png" alt="tmp" style="zoom:70%;" /></center>

<p>右图和上面的图表达的意思相同。左图则展示了样本间的clusters的相关性。</p>
<center><img src="/img/Figures/pyclone2.png" alt="tmp" style="zoom:90%;" /></center>

<h1 id="肿瘤进化分析"><a href="#肿瘤进化分析" class="headerlink" title="肿瘤进化分析"></a>肿瘤进化分析</h1><h2 id="citup"><a href="#citup" class="headerlink" title="citup"></a>citup</h2><p>CITUP(Conality Inference in Tumors Using Phylogeny)是一种使用系统发育论进行肿瘤的克隆推断生物信息学工具，可用于从单个患者获得的多个样本来推断肿瘤异质性。给定每个样本的突变频率, CITUP使用基于优化的算法来找到最能解释数据的进化树。但是，前提条件是，CITUP是针对于深度测序数据的。<br><a target="_blank" rel="noopener" href="https://github.com/amcpherson/citup">Github-citup</a></p>
<h3 id="citup安装"><a href="#citup安装" class="headerlink" title="citup安装"></a>citup安装</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">conda create -n citup-0.1.2</span><br><span class="line">mamba install -c dranew citup  <span class="comment"># 574MB</span></span><br><span class="line">mamba install -c conda-forge h5py  <span class="comment"># 依赖库</span></span><br></pre></td></tr></table></figure>
<h3 id="数据分析-2"><a href="#数据分析-2" class="headerlink" title="数据分析"></a>数据分析</h3><p>实际运行需要准备两个文件，一个是突变位点在不同样本的突变频率，行是突变位点，列是样本。另一个是突变的cluster,只有单列，记录每个突变位点所在的cluster。两个文件的突变位点顺序要一致。</p>
<p><strong>输入文件构建</strong></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cat ./13.pyclone/case1_pyclone_analysis/tables/loci.tsv | cut -f 6 | sed <span class="string">&#x27;1d&#x27;</span> | paste - - - -  &gt;./13.pyclone/case1_pyclone_analysis/freq.txt</span><br><span class="line"></span><br><span class="line">cat ./13.pyclone/case1_pyclone_analysis/tables/loci.tsv | cut -f 3 | sed <span class="string">&#x27;1d&#x27;</span> | paste - - - - |cut -f 1 &gt;./13.pyclone/case1_pyclone_analysis/cluster.txt</span><br><span class="line"></span><br><span class="line"><span class="comment">## 获取 sample_id 信息，后面画图要用到</span></span><br><span class="line">cat ./13.pyclone/case1_pyclone_analysis/tables/loci.tsv | cut -f 2 | sed <span class="string">&#x27;1d&#x27;</span> | head -4 &gt;./13.pyclone/case1_pyclone_analysis/sample_id</span><br></pre></td></tr></table></figure>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">run_citup_qip.py --submit <span class="built_in">local</span> \</span><br><span class="line">./13.pyclone/case1_pyclone_analysis/freq.txt \</span><br><span class="line">./13.pyclone/case1_pyclone_analysis/cluster.txt \</span><br><span class="line">./13.pyclone/case1_pyclone_analysis/results.h5</span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/citup-eg.png" alt="tmp" style="zoom:70%;" /></center>

<p>接下来需要用python处理一下结果，因为得到的结果results.h5是一个hdf5格式，不是普通的文本文件，处理起来批较麻烦一些</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##### citup.py #####</span></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">import</span> h5py</span><br><span class="line">hf=h5py.File(sys.argv[<span class="number">1</span>],<span class="string">&#x27;r&#x27;</span>)</span><br><span class="line">hf.keys()</span><br><span class="line">opnum=hf[<span class="string">&quot;results/optimal/index&quot;</span>][<span class="number">0</span>]</span><br><span class="line">cellfreq=hf[<span class="string">&quot;trees/&quot;</span> + <span class="built_in">str</span>(opnum) + <span class="string">&quot;/clone_freq/block0_values&quot;</span>][:]</span><br><span class="line">tree=hf[<span class="string">&quot;trees/&quot;</span> + <span class="built_in">str</span>(opnum) + <span class="string">&quot;/adjacency_list/block0_values&quot;</span>][:]</span><br><span class="line"><span class="built_in">print</span> cellfreq</span><br><span class="line"><span class="built_in">print</span> tree</span><br><span class="line"><span class="comment">####################</span></span><br><span class="line"></span><br><span class="line">python <span class="built_in">all</span>.my.script/citup.py ./<span class="number">13.</span>pyclone/case1_pyclone_analysis/results.h5 | sed <span class="string">&#x27;s/^ \[//;s/\[//g;s/\]//g&#x27;</span> | tr <span class="string">&#x27; &#x27;</span> <span class="string">&#x27;\t&#x27;</span>| grep <span class="string">&#x27;\.&#x27;</span> &gt;./<span class="number">13.</span>pyclone/case1_pyclone_analysis/cellfreq.txt</span><br><span class="line"></span><br><span class="line">python <span class="built_in">all</span>.my.script/citup.py ./<span class="number">13.</span>pyclone/case1_pyclone_analysis/results.h5 | sed <span class="string">&#x27;s/^ \[//;s/\[//g;s/\]//g&#x27;</span> | tr <span class="string">&#x27; &#x27;</span> <span class="string">&#x27;\t&#x27;</span>| grep -v <span class="string">&#x27;\.&#x27;</span> &gt;./<span class="number">13.</span>pyclone/case1_pyclone_analysis/tree.txt</span><br></pre></td></tr></table></figure>
<p>得到的<code>cellfreq.txt</code>和<code>tree.txt</code>这两个文件，前者行表示样本,列表示克隆clusters,每个值代表克隆频率。后者表示的是进化树的克隆分支关系，接下来利用Timescape进行可视化。</p>
<center><img src="/img/Figures/citup-res.png" alt="tmp" style="zoom:100%;" /></center>

<h2 id="Timescape"><a href="#Timescape" class="headerlink" title="Timescape"></a>Timescape</h2><p>输入格式：<br><code>tree. txt</code>不用做修改，<code>cellfreq.txt</code>需要被处理为3列time or space ID, the clone ID, and the clonal prevalence</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># BiocManager::install(&quot;timescape&quot;)</span></span><br><span class="line"><span class="comment"># example(&quot;timescape&quot;)</span></span><br><span class="line"><span class="comment"># browseVignettes(&quot;timescape&quot;)</span></span><br><span class="line">options(stringsAsFactors = <span class="built_in">F</span>)</span><br><span class="line"> </span><br><span class="line">library(timescape)</span><br><span class="line">library(plotly)</span><br><span class="line">library(htmlwidgets)</span><br><span class="line">library(webshot)</span><br><span class="line">library(tidyr)</span><br><span class="line"></span><br><span class="line">setwd(<span class="string">&quot;/home/data/vip8t02/MyCourse/CancerGenomics/WES.proj/&quot;</span>)</span><br><span class="line">tree_edges = read.table(<span class="string">&quot;./13.pyclone/case1_pyclone_analysis/tree.txt&quot;</span>)</span><br><span class="line">colnames(tree_edges) = <span class="built_in">c</span>(<span class="string">&quot;source&quot;</span>,<span class="string">&quot;target&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># clonal prevalences</span></span><br><span class="line">cellfreq = read.table(<span class="string">&quot;./13.pyclone/case1_pyclone_analysis/cellfreq.txt&quot;</span>)</span><br><span class="line">colnames(cellfreq) = <span class="number">0</span>:(<span class="built_in">length</span>(cellfreq)-<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line">sample_id = read.table(<span class="string">&quot;./13.pyclone/case1_pyclone_analysis/sample_id&quot;</span>)</span><br><span class="line">cellfreq$timepoint = sample_id[ , <span class="number">1</span>]</span><br><span class="line"></span><br><span class="line">clonal_prev = gather(cellfreq, key=<span class="string">&quot;clone_id&quot;</span>, value = <span class="string">&quot;clonal_prev&quot;</span>, -timepoint)</span><br><span class="line">clonal_prev = clonal_prev[order(clonal_prev$timepoint),]</span><br><span class="line"><span class="comment"># targeted mutations</span></span><br><span class="line"><span class="comment"># mutations &lt;- read.csv(system.file(&quot;extdata&quot;, &quot;AML_mutations.csv&quot;, package = &quot;timescape&quot;))</span></span><br><span class="line">p = timescape(clonal_prev = clonal_prev, tree_edges = tree_edges,height=<span class="number">260</span>)</span><br><span class="line">saveWidget(p, <span class="string">&quot;./13.pyclone/case1_timescape.html&quot;</span>)</span><br></pre></td></tr></table></figure>
<center><img src="/img/Figures/timescape.png" alt="timescape" style="zoom:100%;" /></center>

<h1 id="总结复盘"><a href="#总结复盘" class="headerlink" title="总结复盘"></a>总结复盘</h1><ul>
<li>天呐！！他们开发软件的也太厉害了吧！！！</li>
<li>GATK的ModelSegments报错还是没解决，显示内存溢出，.6服务器的给了20G，自己的给了60G还是不行。。。。</li>
<li>关于vcf,bed,maf等等格式总算熟悉点了</li>
<li>好占内存！！！.sra, .fastq, bwa的.bam都删了<center><img src="/img/Figures/space.png" alt="timescape" style="zoom:65%;" /></center></li>
</ul>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="mailto:undefined">yilingzhang</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="http://yilingzhang.gitee.io/2021/03/30/WES-Project/">http://yilingzhang.gitee.io/2021/03/30/WES-Project/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"></div><div class="post_share"><div class="social-share" data-image="/img/cover/WES.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/social-share.js/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/social-share.js/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2021/03/31/eCLIP-Analysis/"><img class="prev-cover" src="/img/cover/eCLIP.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">eCLIP Analysis</div></div></a></div><div class="next-post pull-right"><a href="/2021/03/25/CUT-Tag-Analysis/"><img class="next-cover" src="/img/cover/CUT&amp;Tag.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">CUT&amp;Tag Analysis</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="card-info-avatar is-center"><img class="avatar-img" src="/img/selfie.JPG" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/><div class="author-info__name">yilingzhang</div><div class="author-info__description">Every Day is My Day</div></div><div class="card-info-data"><div class="card-info-data-item is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">32</div></a></div><div class="card-info-data-item is-center"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">14</div></a></div><div class="card-info-data-item is-center"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">13</div></a></div></div><div class="card-info-social-icons is-center"><a class="social-icon" href="mailto:jocelynzhang1119@icloud.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a><a class="social-icon" href="https://blog.csdn.net/weixin_43893431" target="_blank" title="CSDN"><i class="fas fa-paper-plane"></i></a></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE"><span class="toc-number">1.</span> <span class="toc-text">环境配置</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#GATK"><span class="toc-number">1.1.</span> <span class="toc-text">GATK</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Reference"><span class="toc-number">1.2.</span> <span class="toc-text">Reference</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD"><span class="toc-number">2.</span> <span class="toc-text">数据下载</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#NCBI-prefetch"><span class="toc-number">2.1.</span> <span class="toc-text">NCBI+prefetch</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#ENA-Aspera"><span class="toc-number">2.2.</span> <span class="toc-text">ENA+Aspera</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#ENA-Arial2-%E5%A5%B6%E7%89%9B%E5%BF%AB%E4%BC%A0"><span class="toc-number">2.3.</span> <span class="toc-text">ENA+Arial2+奶牛快传</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%B4%A8%E6%8E%A7%E4%B8%8E%E5%8E%BB%E6%8E%A5%E5%A4%B4"><span class="toc-number">3.</span> <span class="toc-text">质控与去接头</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%96%87%E4%BB%B6%E6%A0%BC%E5%BC%8F%E8%BD%AC%E6%8D%A2"><span class="toc-number">3.1.</span> <span class="toc-text">文件格式转换</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#fastq-dump"><span class="toc-number">3.1.1.</span> <span class="toc-text">fastq-dump</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#faterq-dump"><span class="toc-number">3.1.2.</span> <span class="toc-text">faterq-dump</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%B4%A8%E6%8E%A7"><span class="toc-number">3.2.</span> <span class="toc-text">质控</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#fastp"><span class="toc-number">3.3.</span> <span class="toc-text">fastp</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%AF%94%E5%AF%B9"><span class="toc-number">4.</span> <span class="toc-text">比对</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%AF%94%E5%AF%B9%E7%BB%93%E6%9E%9C%E7%9A%84%E8%B4%A8%E6%8E%A7"><span class="toc-number">5.</span> <span class="toc-text">比对结果的质控</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#samtools-stats-plot-bamstats"><span class="toc-number">5.1.</span> <span class="toc-text">samtools stats + plot-bamstats</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#qualimap"><span class="toc-number">5.2.</span> <span class="toc-text">qualimap</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#GATK%E6%B5%81%E7%A8%8B%E5%92%8C%E6%89%BE%E5%8F%98%E5%BC%82"><span class="toc-number">6.</span> <span class="toc-text">GATK流程和找变异</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#Mark-Duplicates-Picard"><span class="toc-number">6.1.</span> <span class="toc-text">Mark Duplicates(Picard)</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%A2%B1%E5%9F%BA%E8%B4%A8%E9%87%8F%E9%87%8D%E6%A0%A1%E6%AD%A3-BQSR"><span class="toc-number">6.2.</span> <span class="toc-text">碱基质量重校正 BQSR</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%8D%95%E6%A0%B7%E6%9C%AC%E6%89%BEGermline-SNPs-INDELs"><span class="toc-number">6.3.</span> <span class="toc-text">单样本找Germline SNPs&#x2F;INDELs</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Mutect%E6%89%BESomatic-Mutations"><span class="toc-number">6.4.</span> <span class="toc-text">Mutect找Somatic Mutations</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#VEP%E6%B3%A8%E9%87%8A"><span class="toc-number">7.</span> <span class="toc-text">VEP注释</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%AE%89%E8%A3%85"><span class="toc-number">7.1.</span> <span class="toc-text">安装</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%B3%A8%E9%87%8A"><span class="toc-number">7.2.</span> <span class="toc-text">注释</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%B3%A8%E9%87%8A%E7%BB%93%E6%9E%9C%E8%BD%ACmaf"><span class="toc-number">7.3.</span> <span class="toc-text">注释结果转maf</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%8F%AF%E8%A7%86%E5%8C%96%E6%B3%A8%E9%87%8A%E7%BB%93%E6%9E%9C"><span class="toc-number">7.4.</span> <span class="toc-text">可视化注释结果</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%AF%BB%E5%85%A5%E6%95%B0%E6%8D%AE"><span class="toc-number">7.4.1.</span> <span class="toc-text">读入数据</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#mafSummary"><span class="toc-number">7.4.2.</span> <span class="toc-text">mafSummary</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#oncoplot"><span class="toc-number">7.4.3.</span> <span class="toc-text">oncoplot</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#lollopopPlot"><span class="toc-number">7.4.4.</span> <span class="toc-text">lollopopPlot</span></a></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82%E5%88%86%E6%9E%90%EF%BC%88GATK%E6%B5%81%E7%A8%8B%EF%BC%89"><span class="toc-number">8.</span> <span class="toc-text">拷贝数变异分析（GATK流程）</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%A4%96%E6%98%BE%E5%AD%90%E5%9D%90%E6%A0%87%E7%9A%84interval%E6%96%87%E4%BB%B6"><span class="toc-number">8.1.</span> <span class="toc-text">外显子坐标的interval文件</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%8E%B7%E5%8F%96%E6%A0%B7%E6%9C%AC%E7%9A%84read-counts"><span class="toc-number">8.2.</span> <span class="toc-text">获取样本的read counts</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E9%99%8D%E5%99%AA-DenoiseReadCounts"><span class="toc-number">8.3.</span> <span class="toc-text">降噪 DenoiseReadCounts</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%8F%AF%E8%A7%86%E5%8C%96%E9%99%8D%E5%99%AA%E5%90%8E%E7%9A%84copy-ratio"><span class="toc-number">8.4.</span> <span class="toc-text">可视化降噪后的copy ratio</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%AE%A1%E7%AE%97%E5%B8%B8%E8%A7%81%E7%9A%84germline-mutation%E4%BD%8D%E7%82%B9"><span class="toc-number">8.5.</span> <span class="toc-text">计算常见的germline mutation位点</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#ModelSegments"><span class="toc-number">8.6.</span> <span class="toc-text">ModelSegments</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82%E5%88%86%E6%9E%90%EF%BC%88CNVkit%EF%BC%89"><span class="toc-number">9.</span> <span class="toc-text">拷贝数变异分析（CNVkit）</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E7%AA%81%E5%8F%98Somatic-Signature%E5%9B%BE%E8%B0%B1"><span class="toc-number">10.</span> <span class="toc-text">突变Somatic Signature图谱</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%8E%A2%E7%A9%B6%E8%82%BF%E7%98%A4%E5%BC%82%E8%B4%A8%E6%80%A7"><span class="toc-number">11.</span> <span class="toc-text">探究肿瘤异质性</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%90%8C%E4%B8%80%E7%97%85%E4%BA%BA%E7%9A%84SNV%E5%88%86%E5%B8%83"><span class="toc-number">11.1.</span> <span class="toc-text">同一病人的SNV分布</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%82%BF%E7%98%A4%E6%A0%B7%E6%9C%AC%E4%B8%AD%E7%9A%84%E5%BC%82%E8%B4%A8%E6%80%A7"><span class="toc-number">11.2.</span> <span class="toc-text">肿瘤样本中的异质性</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%B3%A8%E9%87%8A%E7%AA%81%E5%8F%98%E4%BD%8D%E7%82%B9%E6%89%80%E5%9C%A8%E5%9F%BA%E5%9B%A0"><span class="toc-number">12.</span> <span class="toc-text">注释突变位点所在基因</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#ABSOLUTE%E4%BD%BF%E7%94%A8"><span class="toc-number">13.</span> <span class="toc-text">ABSOLUTE使用</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#ABSOLUTE%E5%AE%89%E8%A3%85"><span class="toc-number">13.1.</span> <span class="toc-text">ABSOLUTE安装</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90"><span class="toc-number">13.2.</span> <span class="toc-text">数据分析</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%BB%93%E6%9E%9C%E8%A7%A3%E8%AF%BB"><span class="toc-number">13.3.</span> <span class="toc-text">结果解读</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Pyclone%E6%8E%A8%E6%96%AD%E8%82%BF%E7%98%A4%E7%BB%86%E8%83%9E%E7%9A%84%E5%85%8B%E9%9A%86%E7%BB%84%E6%88%90"><span class="toc-number">14.</span> <span class="toc-text">Pyclone推断肿瘤细胞的克隆组成</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#Pyclone%E5%AE%89%E8%A3%85"><span class="toc-number">14.1.</span> <span class="toc-text">Pyclone安装</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90-1"><span class="toc-number">14.2.</span> <span class="toc-text">数据分析</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%BB%93%E6%9E%9C%E8%A7%A3%E8%AF%BB-1"><span class="toc-number">14.3.</span> <span class="toc-text">结果解读</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%82%BF%E7%98%A4%E8%BF%9B%E5%8C%96%E5%88%86%E6%9E%90"><span class="toc-number">15.</span> <span class="toc-text">肿瘤进化分析</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#citup"><span class="toc-number">15.1.</span> <span class="toc-text">citup</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#citup%E5%AE%89%E8%A3%85"><span class="toc-number">15.1.1.</span> <span class="toc-text">citup安装</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90-2"><span class="toc-number">15.1.2.</span> <span class="toc-text">数据分析</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Timescape"><span class="toc-number">15.2.</span> <span class="toc-text">Timescape</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%80%BB%E7%BB%93%E5%A4%8D%E7%9B%98"><span class="toc-number">16.</span> <span class="toc-text">总结复盘</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2021/05/10/scRNAseq-03-Doublet-velocyto-inferCNV/" title="scRNAseq-03 Doublet &amp; velocyto &amp; inferCNV"><img src="/img/cover/scRNA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scRNAseq-03 Doublet &amp; velocyto &amp; inferCNV"/></a><div class="content"><a class="title" href="/2021/05/10/scRNAseq-03-Doublet-velocyto-inferCNV/" title="scRNAseq-03 Doublet &amp; velocyto &amp; inferCNV">scRNAseq-03 Doublet &amp; velocyto &amp; inferCNV</a><time datetime="2021-05-10T05:07:59.000Z" title="Created 2021-05-10 13:07:59">2021-05-10</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/08/TME-Immune-Analysis-of-CUL1-Cullin1/" title="TME &amp; Immune Analysis of CUL1&amp;Cullin1"><img src="/img/banner/paofu.JPG" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="TME &amp; Immune Analysis of CUL1&amp;Cullin1"/></a><div class="content"><a class="title" href="/2021/05/08/TME-Immune-Analysis-of-CUL1-Cullin1/" title="TME &amp; Immune Analysis of CUL1&amp;Cullin1">TME &amp; Immune Analysis of CUL1&amp;Cullin1</a><time datetime="2021-05-08T02:50:38.000Z" title="Created 2021-05-08 10:50:38">2021-05-08</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/08/Drawing%20Requirements%20for%20Scientific%20Research/" title="Drawing Requirements for Scientific Research"><img src="/img/cover/drawing.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Drawing Requirements for Scientific Research"/></a><div class="content"><a class="title" href="/2021/05/08/Drawing%20Requirements%20for%20Scientific%20Research/" title="Drawing Requirements for Scientific Research">Drawing Requirements for Scientific Research</a><time datetime="2021-05-08T01:59:17.000Z" title="Created 2021-05-08 09:59:17">2021-05-08</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/06/Baidu-Yun-for-server/" title="Baidu Yun for server"><img src="/img/cover/baiduyun.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Baidu Yun for server"/></a><div class="content"><a class="title" href="/2021/05/06/Baidu-Yun-for-server/" title="Baidu Yun for server">Baidu Yun for server</a><time datetime="2021-05-06T04:39:01.000Z" title="Created 2021-05-06 12:39:01">2021-05-06</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/04/26/scRNAseq-02-Annotation/" title="scRNAseq-02 Annotation"><img src="/img/cover/scRNA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scRNAseq-02 Annotation"/></a><div class="content"><a class="title" href="/2021/04/26/scRNAseq-02-Annotation/" title="scRNAseq-02 Annotation">scRNAseq-02 Annotation</a><time datetime="2021-04-26T08:46:41.000Z" title="Created 2021-04-26 16:46:41">2021-04-26</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2021 By yilingzhang</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"></div><script defer="defer" id="ribbon" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/canvas-ribbon.min.js" size="150" alpha="0.6" zIndex="-1" mobile="false" data-click="true"></script><script id="click-heart" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/click-heart.min.js" async="async" mobile="false"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>