<!DOCTYPE html><html lang="en | zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1"><title>eCLIP Analysis | Jocelyn's Blog</title><meta name="keywords" content="eCLIP"><meta name="author" content="yilingzhang"><meta name="copyright" content="yilingzhang"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="文献：https:&#x2F;&#x2F;www.nature.com&#x2F;articles&#x2F;nmeth.3810 方法：https:&#x2F;&#x2F;www.encodeproject.org&#x2F;documents&#x2F;739ca190-8d43-4a68-90ce-1a0ddfffc6fd&#x2F;@@download&#x2F;attachment&#x2F;eCLIP_analysisSOP_v2.2.pdf  建库原理： 软件都在&#x2F;publis&#x2F;shaoj">
<meta property="og:type" content="article">
<meta property="og:title" content="eCLIP Analysis">
<meta property="og:url" content="http://yilingzhang.gitee.io/2021/03/31/eCLIP-Analysis/index.html">
<meta property="og:site_name" content="Jocelyn&#39;s Blog">
<meta property="og:description" content="文献：https:&#x2F;&#x2F;www.nature.com&#x2F;articles&#x2F;nmeth.3810 方法：https:&#x2F;&#x2F;www.encodeproject.org&#x2F;documents&#x2F;739ca190-8d43-4a68-90ce-1a0ddfffc6fd&#x2F;@@download&#x2F;attachment&#x2F;eCLIP_analysisSOP_v2.2.pdf  建库原理： 软件都在&#x2F;publis&#x2F;shaoj">
<meta property="og:locale">
<meta property="og:image" content="http://yilingzhang.gitee.io/img/cover/eCLIP.png">
<meta property="article:published_time" content="2021-03-31T00:19:05.000Z">
<meta property="article:modified_time" content="2021-03-31T00:30:05.846Z">
<meta property="article:author" content="yilingzhang">
<meta property="article:tag" content="eCLIP">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://yilingzhang.gitee.io/img/cover/eCLIP.png"><link rel="shortcut icon" href="/img/love_32px.png"><link rel="canonical" href="http://yilingzhang.gitee.io/2021/03/31/eCLIP-Analysis/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    jQuery: 'https://cdn.jsdelivr.net/npm/jquery@latest/dist/jquery.min.js',
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/js/jquery.justifiedGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/css/justifiedGallery.min.css'
    },
    fancybox: {
      js: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.js',
      css: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isanchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = { 
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2021-03-31 08:30:05'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    })(window)</script><meta name="generator" content="Hexo 5.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="author-avatar"><img class="avatar-img" src="/img/selfie.JPG" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data"><div class="data-item is-center"><div class="data-item-link"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">21</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">11</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">12</div></a></div></div></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/img/cover/eCLIP.png')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Jocelyn's Blog</a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">eCLIP Analysis</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2021-03-31T00:19:05.000Z" title="Created 2021-03-31 08:19:05">2021-03-31</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2021-03-31T00:30:05.846Z" title="Updated 2021-03-31 08:30:05">2021-03-31</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/eCLIP/">eCLIP</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">Word count:</span><span class="word-count">2.8k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">Reading time:</span><span>14min</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><ul>
<li>文献：<a target="_blank" rel="noopener" href="https://www.nature.com/articles/nmeth.3810">https://www.nature.com/articles/nmeth.3810</a></li>
<li>方法：<a target="_blank" rel="noopener" href="https://www.encodeproject.org/documents/739ca190-8d43-4a68-90ce-1a0ddfffc6fd/@@download/attachment/eCLIP_analysisSOP_v2.2.pdf">https://www.encodeproject.org/documents/739ca190-8d43-4a68-90ce-1a0ddfffc6fd/@@download/attachment/eCLIP_analysisSOP_v2.2.pdf</a></li>
<li> 建库原理：<img src="C:\Users\lenovo\AppData\Roaming\Typora\typora-user-images\image-20210329085640993.png" alt="image-20210329085640993"></li>
<li>软件都在/publis/shaojf/Course/NGS/Tools/eCLIP.tools</li>
<li>X 表示cross-link的位置，在read2</li>
</ul>
<p>Read1包括barcode<br>Read 2 前几个碱基叫Random-mer（UMI），UMI碱基放在name里面<br>测序长度比读段长，要去接头</p>
<h1 id="分析路线"><a href="#分析路线" class="headerlink" title="分析路线"></a>分析路线</h1><ol>
<li>Fastqc查看质量及adapter</li>
<li>Demultiplex分离in-line barcode及UMI处理+cutadapt</li>
<li>STAR比对（adaptor一般出现在末端，STAR用local模式，有没有adaptor影响不大）</li>
<li>基于UMI删除PCR重复（Barcode一样的放到一个fastq，算作Tech rep）</li>
<li>合并技术重复（得到两个bio rep的文件）</li>
<li>提取read2</li>
<li>生成bigwig用于可视化</li>
<li>用clipper、MACS2和pureclip进行peakcalling（MACS2跑出的效果很差；clipper效果较好；pureclip特别占内存，效果一般）</li>
<li>基于input进行normalization（有些基因表达水平高，有些低，在外显子上读段多）</li>
</ol>
<h1 id="工具"><a href="#工具" class="headerlink" title="工具"></a>工具</h1><h2 id="fastqc"><a href="#fastqc" class="headerlink" title="fastqc"></a>fastqc</h2><p>可以指定查看哪一个adaptor，四个adaptor比例大概是20%，比较理想</p>
<h2 id="STAR"><a href="#STAR" class="headerlink" title="STAR"></a>STAR</h2><p>–alignEndsType   default: Local<br>Local—standard local alignment with soft-clipping allowed<br>EndToEnd—force end-to-end read alignment, do not soft-clip<br>Extend5pOfRead1—fully extend only the 5p of the read1, all other ends: local alignment<br>Extend5pOfReads12—fully extend only the 5p of the both read1 and read2, all other ends:local alignment</p>
<h2 id="Pureclip"><a href="#Pureclip" class="headerlink" title="Pureclip"></a>Pureclip</h2><p><a target="_blank" rel="noopener" href="https://github.com/skrakau/PureCLIP">https://github.com/skrakau/PureCLIP</a><br><a target="_blank" rel="noopener" href="https://pureclip.readthedocs.io/en/latest/GettingStarted/index.html">https://pureclip.readthedocs.io/en/latest/GettingStarted/index.html</a></p>
<p>会得到crosslink sites和Binding regions</p>
<h2 id="MACS2"><a href="#MACS2" class="headerlink" title="MACS2"></a>MACS2</h2><p>基于STAR比对的结果，-f BAMPE出来的fragment size=1982.3<br>为什么会这么大</p>
<h2 id="UMI-Tools"><a href="#UMI-Tools" class="headerlink" title="UMI_Tools"></a>UMI_Tools</h2><p>umi_tools dedup</p>
<h1 id="Code"><a href="#Code" class="headerlink" title="Code"></a>Code</h1><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br><span class="line">197</span><br><span class="line">198</span><br><span class="line">199</span><br><span class="line">200</span><br><span class="line">201</span><br><span class="line">202</span><br><span class="line">203</span><br><span class="line">204</span><br><span class="line">205</span><br><span class="line">206</span><br><span class="line">207</span><br><span class="line">208</span><br><span class="line">209</span><br><span class="line">210</span><br><span class="line">211</span><br><span class="line">212</span><br><span class="line">213</span><br><span class="line">214</span><br><span class="line">215</span><br><span class="line">216</span><br><span class="line">217</span><br><span class="line">218</span><br><span class="line">219</span><br><span class="line">220</span><br><span class="line">221</span><br><span class="line">222</span><br><span class="line">223</span><br><span class="line">224</span><br><span class="line">225</span><br><span class="line">226</span><br><span class="line">227</span><br><span class="line">228</span><br><span class="line">229</span><br><span class="line">230</span><br><span class="line">231</span><br><span class="line">232</span><br><span class="line">233</span><br><span class="line">234</span><br><span class="line">235</span><br><span class="line">236</span><br><span class="line">237</span><br><span class="line">238</span><br><span class="line">239</span><br><span class="line">240</span><br><span class="line">241</span><br><span class="line">242</span><br><span class="line">243</span><br><span class="line">244</span><br><span class="line">245</span><br><span class="line">246</span><br><span class="line">247</span><br><span class="line">248</span><br><span class="line">249</span><br><span class="line">250</span><br><span class="line">251</span><br><span class="line">252</span><br><span class="line">253</span><br><span class="line">254</span><br><span class="line">255</span><br><span class="line">256</span><br><span class="line">257</span><br><span class="line">258</span><br><span class="line">259</span><br><span class="line">260</span><br><span class="line">261</span><br><span class="line">262</span><br><span class="line">263</span><br><span class="line">264</span><br><span class="line">265</span><br><span class="line">266</span><br><span class="line">267</span><br><span class="line">268</span><br><span class="line">269</span><br><span class="line">270</span><br><span class="line">271</span><br><span class="line">272</span><br><span class="line">273</span><br><span class="line">274</span><br><span class="line">275</span><br><span class="line">276</span><br><span class="line">277</span><br><span class="line">278</span><br><span class="line">279</span><br><span class="line">280</span><br><span class="line">281</span><br><span class="line">282</span><br><span class="line">283</span><br><span class="line">284</span><br><span class="line">285</span><br><span class="line">286</span><br><span class="line">287</span><br><span class="line">288</span><br><span class="line">289</span><br><span class="line">290</span><br><span class="line">291</span><br><span class="line">292</span><br><span class="line">293</span><br><span class="line">294</span><br><span class="line">295</span><br><span class="line">296</span><br><span class="line">297</span><br><span class="line">298</span><br><span class="line">299</span><br><span class="line">300</span><br><span class="line">301</span><br><span class="line">302</span><br><span class="line">303</span><br><span class="line">304</span><br><span class="line">305</span><br><span class="line">306</span><br><span class="line">307</span><br><span class="line">308</span><br><span class="line">309</span><br><span class="line">310</span><br><span class="line">311</span><br><span class="line">312</span><br><span class="line">313</span><br><span class="line">314</span><br><span class="line">315</span><br><span class="line">316</span><br><span class="line">317</span><br><span class="line">318</span><br><span class="line">319</span><br><span class="line">320</span><br><span class="line">321</span><br><span class="line">322</span><br><span class="line">323</span><br><span class="line">324</span><br><span class="line">325</span><br><span class="line">326</span><br><span class="line">327</span><br><span class="line">328</span><br><span class="line">329</span><br><span class="line">330</span><br><span class="line">331</span><br><span class="line">332</span><br><span class="line">333</span><br><span class="line">334</span><br><span class="line">335</span><br><span class="line">336</span><br><span class="line">337</span><br><span class="line">338</span><br><span class="line">339</span><br><span class="line">340</span><br><span class="line">341</span><br><span class="line">342</span><br><span class="line">343</span><br><span class="line">344</span><br><span class="line">345</span><br><span class="line">346</span><br><span class="line">347</span><br><span class="line">348</span><br><span class="line">349</span><br><span class="line">350</span><br><span class="line">351</span><br><span class="line">352</span><br><span class="line">353</span><br><span class="line">354</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># https://www.encodeproject.org/pipelines/ENCPL357ADL/</span></span><br><span class="line"><span class="comment"># setup the environment</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># cd /public/workspace/shaojf/NGS2021/shaolab2/tools/clipper</span></span><br><span class="line"><span class="comment"># conda env create -f environment3.yml</span></span><br><span class="line"><span class="comment"># conda activate clipper3</span></span><br><span class="line"><span class="comment"># pip install .</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># cd /public/workspace/shaojf/NGS2021/shaolab2/tools/merge_peaks-0.1.0</span></span><br><span class="line"><span class="comment"># conda create -n perl5</span></span><br><span class="line"><span class="comment"># conda activate perl5</span></span><br><span class="line"><span class="comment">###### conda install -c bioconda perl-statistics-basic perl-statistics-r perl-statistics-distributions</span></span><br><span class="line"><span class="comment"># conda install -c bioconda perl-statistics-basic=1.6611 perl-statistics-r=0.34 perl-statistics-distributions=1.02</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># conda create -n py2 python=2.7</span></span><br><span class="line"><span class="comment"># conda activate py2</span></span><br><span class="line"><span class="comment"># conda install cython scikit-learn future matplotlib bz2file</span></span><br><span class="line"><span class="comment"># conda install -c bioconda pysam</span></span><br><span class="line"><span class="comment"># cd /public/workspace/shaojf/NGS2021/shaolab2/tools/eclipdemux-master</span></span><br><span class="line"><span class="comment"># easy_install .</span></span><br><span class="line"><span class="comment"># cd ../gscripts/</span></span><br><span class="line"><span class="comment"># easy_install .</span></span><br><span class="line"><span class="comment"># cd ../makebigwigfiles-master/</span></span><br><span class="line"><span class="comment"># easy_install .</span></span><br><span class="line"><span class="comment">##### ln -s /public/workspace/shaojf/Course/NGS/Tools/eclip.tools/eclipdemux-master/eclipdemux_package/demux.py /public/workspace/shaojf/anaconda3/envs/py2/bin/eclipdemux</span></span><br><span class="line"><span class="comment"># cd /public/workspace/shaojf/anaconda3/envs/py2/bin</span></span><br><span class="line"><span class="comment"># ln -s demux_paired_end.py eclipdemux</span></span><br><span class="line"><span class="comment"># ln -s ../../clipper3/bin/clipper</span></span><br><span class="line"><span class="comment"># ln -s ../../perl5/bin/perl</span></span><br><span class="line"><span class="comment"># ln -s /public/workspace/shaojf/Course/NGS/Tools/eclip.tools/merge_peaks-0.1.0/bin/*.* .</span></span><br><span class="line"><span class="comment"># ln -s /public/workspace/shaojf/Course/NGS/Tools/eclip.tools/merge_peaks-0.1.0/bin/perl/*.* .</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># -----------------&gt;</span></span><br><span class="line"><span class="comment"># 以上为环境配置过程</span></span><br><span class="line"><span class="comment"># -----------------&gt;</span></span><br><span class="line"></span><br><span class="line">conda activate py2</span><br><span class="line"><span class="comment"># py2的bin下有clipper3的软连接，可以直接用clipper3</span></span><br><span class="line"></span><br><span class="line"><span class="comment">##### ln -s /public/workspace/shaojf/anaconda3/envs/clipper3 /public/workspace/stu18230130/anaconda3/envs/clipper3</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 如果自行安装过conda的，请执行以下操作，将shaojf用户的conda环境链接到自己的conda环境；</span></span><br><span class="line"><span class="comment"># 如果没有安装过conda，则无需执行</span></span><br><span class="line">ln -s /public/workspace/shaojf/anaconda3/envs/py2 /public/workspace/`whoami`/anaconda3/envs/py2</span><br><span class="line">wkdir=/public/workspace/shaojf/NGS2021/shaolab2</span><br><span class="line">genomefa=/public/workspace/shaojf/Course/NGS/Reference/UCSC/hg38/hg38.fa</span><br><span class="line">stardir=/public/workspace/shaojf/Course/NGS/Reference/UCSC/hg38/UCSC.hg38.STAR.v2.7.1a</span><br><span class="line">chromsize=/public/workspace/shaojf/Course/NGS/Reference/UCSC/hg38.chrom.sizes</span><br><span class="line">ncores=</span><br><span class="line">conda activate py2</span><br><span class="line"><span class="comment"># python3会报错！！！</span></span><br><span class="line"></span><br><span class="line"><span class="comment">############## 准备数据 ##############</span></span><br><span class="line"><span class="comment"># 用UCSC hg38.fa是因为clipper自带的注释文件是含有chr字符串的</span></span><br><span class="line"><span class="comment"># 一开会的参考序列不全，缺chr7-9</span></span><br><span class="line"><span class="comment"># STAR --runMode genomeGenerate --runThreadN 16 --genomeDir UCSC.hg38.STAR.v2.7.1a --genomeFastaFiles hg38.fa</span></span><br><span class="line">mkdir -p <span class="variable">$HOME</span>/NGS2021/shaolab2/</span><br><span class="line"><span class="comment"># 将/public/workspace/shaojf/Course/NGS/DataSets/eCLIP下的相关数据ln -s到自己的工作路径</span></span><br><span class="line"></span><br><span class="line"><span class="comment">############## fastqc查看adapter ##############</span></span><br><span class="line"><span class="comment"># yeolabbarcodes.txt</span></span><br><span class="line"><span class="comment"># 可以问做实验的人要</span></span><br><span class="line"><span class="comment"># --adapters 检测一下有没有以下的barcode</span></span><br><span class="line">mkdir <span class="variable">$wkdir</span>/1.fastqc</span><br><span class="line">time fastqc -o <span class="variable">$wkdir</span>/1.fastqc -t <span class="variable">$ncores</span> \</span><br><span class="line">  --adapters <span class="variable">$wkdir</span>/configFiles/yeolabbarcodes.txt <span class="variable">$wkdir</span>/data/RBFOX2-204-*</span><br><span class="line"><span class="comment"># 查看哪些barcode被测到</span></span><br><span class="line"><span class="comment"># grep -A &lt;显示行数&gt;: 除了显示符合范本样式的那一列之外，并显示该行之后的内容。</span></span><br><span class="line"><span class="comment"># barcode从第一个碱基开始，用-A 2输出第一行</span></span><br><span class="line">unzip <span class="variable">$wkdir</span>/1.fastqc/RBFOX2-204-CLIP_S1_R1_fastqc.zip</span><br><span class="line">grep -A 2 <span class="string">&quot;Adapter Content&quot;</span> <span class="variable">$wkdir</span>/1.fastqc/RBFOX2-204-CLIP_S1_R1_fastqc/fastqc_data.txt | \</span><br><span class="line">awk <span class="string">&#x27;BEGIN&#123;for(i=2;i&lt;=13;i++)&#123;ids[i]=&quot;A&quot;&#125;&#125;&#123;for(i=2;i&lt;=13;i++)&#123;if($i&gt;10 &amp;&amp; NR&gt;2)&#123;print ids[i]&quot;\t&quot;$i&#125;else&#123;ids[i]=$i&#125;&#125;&#125;&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 初始化ids数组，如果数值大于10且行数&gt;2，根据数组下标打印ids</span></span><br><span class="line"><span class="comment"># awk-出现的比例大于10%的，打印出来</span></span><br><span class="line"><span class="comment"># A01     17.773141608744858</span></span><br><span class="line"><span class="comment"># B06     14.295544471332855</span></span><br><span class="line"><span class="comment"># C01     15.382434142104247</span></span><br><span class="line"><span class="comment"># D8f     17.92395807937699</span></span><br><span class="line"><span class="comment"># firefox $wkdir/1.fastqc/RBFOX2-204-*.html</span></span><br><span class="line"></span><br><span class="line"><span class="comment">############## adapter分离及切除 ##############</span></span><br><span class="line"><span class="comment"># 时间很长，可从课程目录建立软链接</span></span><br><span class="line">mkdir <span class="variable">$wkdir</span>/2.demultiplexed/</span><br><span class="line"><span class="comment"># prefix=RBFOX2-204-CLIP_S1</span></span><br><span class="line"><span class="comment"># cd $wkdir</span></span><br><span class="line"><span class="comment"># --dataset 不能用绝对路径。。。。因为程序只能调用相对路径（前面会自己加./）</span></span><br><span class="line">datapre=RBFOX2</span><br><span class="line"><span class="keyword">for</span> prefix <span class="keyword">in</span> RBFOX2-204-CLIP_S1 RBFOX2-204-INPUT_S2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  name=`<span class="built_in">echo</span> <span class="variable">$prefix</span> | cut -f 1 -d <span class="string">&quot;_&quot;</span> | cut -f 3 -d <span class="string">&quot;-&quot;</span>`</span><br><span class="line">  eclipdemux --newname <span class="variable">$name</span> --dataset 2.demultiplexed/<span class="variable">$datapre</span> \</span><br><span class="line">    --expectedbarcodeida A01 --expectedbarcodeidb B06 \</span><br><span class="line">    --fastq_1 <span class="variable">$wkdir</span>/data/<span class="variable">$&#123;prefix&#125;</span>_R1.fastq.gz \</span><br><span class="line">    --fastq_2 <span class="variable">$wkdir</span>/data/<span class="variable">$&#123;prefix&#125;</span>_R2.fastq.gz \</span><br><span class="line">    --barcodesfile <span class="variable">$wkdir</span>/configFiles/yeolabbarcodes_20170101.fasta \</span><br><span class="line">    --length 5 --metrics <span class="variable">$wkdir</span>/2.demultiplexed/<span class="variable">$prefix</span>.metrics.txt &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 把UMI写到fastq名字行</span></span><br><span class="line">ll demultiplexed/*&#123;A01,B06,C01,D8f&#125;*</span><br><span class="line"><span class="comment"># samples:</span></span><br><span class="line"><span class="comment">#   -</span></span><br><span class="line"><span class="comment">#     - ip_read:</span></span><br><span class="line"><span class="comment">#       name: rep1_clip</span></span><br><span class="line"><span class="comment">#       barcodeids: [A01, B06]</span></span><br><span class="line">        </span><br><span class="line"><span class="comment">#     - input_read: （NIL开头的，检测不到任何barcode）</span></span><br><span class="line"><span class="comment">#       name: rep1_input</span></span><br><span class="line"><span class="comment">#       barcodeids: [NIL, NIL]</span></span><br><span class="line"><span class="comment">#   -</span></span><br><span class="line"><span class="comment">#     - ip_read:</span></span><br><span class="line"><span class="comment">#       name: rep2_clip</span></span><br><span class="line"><span class="comment">#       barcodeids: [C01, D8f]</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#     - input_read:</span></span><br><span class="line"><span class="comment">#       name: rep2_input</span></span><br><span class="line"><span class="comment">#       barcodeids: [NIL, NIL]</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 以下省略，太慢了。所以在STAR步骤选local比对</span></span><br><span class="line"><span class="comment"># Cutadapt	round	1</span></span><br><span class="line"><span class="comment"># Takes output from demultiplexed files. Run to	trim off both 5’ and 3’	adapters on	both read</span></span><br><span class="line"><span class="comment"># mkdir clean</span></span><br><span class="line"><span class="comment"># for prefix in RBFOX2-204-CLIP_S1 RBFOX2-204-INPUT_S2</span></span><br><span class="line"><span class="comment"># do</span></span><br><span class="line"><span class="comment"># 	cutadapt -f fastq --match-read-wildcards --times 1 -e 0.1 -O 1 \</span></span><br><span class="line"><span class="comment"># 		--quality-cutoff 6 -m 18 -a NNNNNAGATCGGAAGAGCACACGTCTGAACTCCAGTCAC \</span></span><br><span class="line"><span class="comment"># 		-g CTTCCGATCTACAAGTT -g CTTCCGATCTTGGTCCT \</span></span><br><span class="line"><span class="comment"># 		-A AACTTGTAGATCGGA -A AGGACCAAGATCGGA -A ACTTGTAGATCGGAA \</span></span><br><span class="line"><span class="comment"># 		-A GGACCAAGATCGGAA -A CTTGTAGATCGGAAG -A GACCAAGATCGGAAG \</span></span><br><span class="line"><span class="comment"># 		-A TTGTAGATCGGAAGA -A ACCAAGATCGGAAGA -A TGTAGATCGGAAGAG \</span></span><br><span class="line"><span class="comment"># 		-A CCAAGATCGGAAGAG -A GTAGATCGGAAGAGC -A CAAGATCGGAAGAGC \</span></span><br><span class="line"><span class="comment"># 		-A TAGATCGGAAGAGCG -A AAGATCGGAAGAGCG -A AGATCGGAAGAGCGT \</span></span><br><span class="line"><span class="comment"># 		-A GATCGGAAGAGCGTC -A ATCGGAAGAGCGTCG -A TCGGAAGAGCGTCGT \</span></span><br><span class="line"><span class="comment"># 		-A CGGAAGAGCGTCGTG -A GGAAGAGCGTCGTGT \</span></span><br><span class="line"><span class="comment"># 		-o clean/$prefix.R1.adapterTrim.fastq.gz \</span></span><br><span class="line"><span class="comment"># 		-p clean/$prefix.R1.adapterTrim.fastq.gz \</span></span><br><span class="line"><span class="comment"># 		2.demultiplexed/$&#123;prefix&#125;_R1.fastq.gz 2.demultiplexed/$&#123;prefix&#125;_R2.fastq.gz \</span></span><br><span class="line"><span class="comment"># 		&gt; $prefix.adapterTrim.metrics &amp;</span></span><br><span class="line"><span class="comment"># done</span></span><br><span class="line"></span><br><span class="line"><span class="comment">############## STAR比对 ##############</span></span><br><span class="line"><span class="comment"># 时间较长，且占资源较多，可从课程目录建立软链接</span></span><br><span class="line"><span class="comment"># --alignEndsType Local</span></span><br><span class="line"></span><br><span class="line">mkdir -p <span class="variable">$wkdir</span>/3.starmapping</span><br><span class="line"><span class="keyword">for</span> fqpre <span class="keyword">in</span> RBFOX2.CLIP.&#123;A01,B06,C01,D8f&#125; RBFOX2.INPUT.NIL</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  lib=`<span class="built_in">echo</span> <span class="variable">$fqpre</span> | awk -F <span class="string">&quot;.&quot;</span> <span class="string">&#x27;&#123;print $NF&#125;&#x27;</span>`</span><br><span class="line">  STAR --runThreadN <span class="variable">$ncores</span> \</span><br><span class="line">    --runMode alignReads \</span><br><span class="line">    --genomeDir <span class="variable">$stardir</span> \</span><br><span class="line">    --readFilesIn <span class="variable">$wkdir</span>/2.demultiplexed/<span class="variable">$&#123;fqpre&#125;</span>.r1.fq.gz \</span><br><span class="line">      <span class="variable">$wkdir</span>/2.demultiplexed/<span class="variable">$&#123;fqpre&#125;</span>.r2.fq.gz \</span><br><span class="line">    --readFilesCommand zcat \</span><br><span class="line">    --outFileNamePrefix <span class="variable">$wkdir</span>/3.starmapping/<span class="variable">$&#123;fqpre&#125;</span>. \</span><br><span class="line">    --outSAMunmapped Within --outFilterMultimapNmax 1 \</span><br><span class="line">    --outFilterMultimapScoreRange 1 \</span><br><span class="line">    --outSAMtype BAM Unsorted \</span><br><span class="line">    --outSAMattributes All \</span><br><span class="line">    --outFilterScoreMin 10 \</span><br><span class="line">    --outSAMattrRGline ID:<span class="variable">$lib</span> --alignEndsType Local</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="comment">############## 根据UMI删除PCR重复 #############</span></span><br><span class="line">mkdir -p <span class="variable">$wkdir</span>/4.rmDupbam</span><br><span class="line"><span class="keyword">for</span> pre <span class="keyword">in</span> RBFOX2.CLIP.&#123;A01,B06,C01,D8f&#125; RBFOX2.INPUT.NIL</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  barcode_collapse_pe.py --bam <span class="variable">$wkdir</span>/3.starmapping/<span class="variable">$pre</span>.Aligned.out.bam \</span><br><span class="line">    --out_file <span class="variable">$wkdir</span>/4.rmDupbam/<span class="variable">$pre</span>.rmDup.bam \</span><br><span class="line">    --metrics_file <span class="variable">$wkdir</span>/4.rmDupbam/<span class="variable">$pre</span>.metrics &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># UMItools的标记与eclipdemux不同，不可交叉混用</span></span><br><span class="line"><span class="comment"># UMItools是把UMI加在: @HISEQ:87:00000000_GGTT read1</span></span><br><span class="line"><span class="comment"># eclipdemux加在：@GGTT;HISEQ:87:00000000 read1</span></span><br><span class="line"><span class="comment"># https://github.com/CGATOxford/UMI-tools/blob/master/doc/QUICK_START.md</span></span><br><span class="line"><span class="comment"># 取3-7是因为前几个base测序质量不好</span></span><br><span class="line"><span class="comment"># UMI is bases 3-7, bases 1-2 and 7-8 are the sample barcode and need to be removed</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># @HISEQ:87:00000000 read1</span></span><br><span class="line"><span class="comment"># AAGGTTGCTGATTGGATGGGCTAG</span></span><br><span class="line"><span class="comment"># +</span></span><br><span class="line"><span class="comment"># DA1AEBFGGCG01DFH00B1FF0B</span></span><br><span class="line"><span class="comment"># will become:</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#  @HISEQ:87:00000000_GGTT read1</span></span><br><span class="line"><span class="comment">#  TGATTGGATGGGCTAG</span></span><br><span class="line"><span class="comment">#  + </span></span><br><span class="line"><span class="comment">#  1AFGGCG01DFH00B1</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">############## 合并并排序 #############</span></span><br><span class="line">       <span class="comment"># $&#123;parameter#word&#125;</span></span><br><span class="line">       <span class="comment"># $&#123;parameter##word&#125;</span></span><br><span class="line">       <span class="comment">#        Remove  matching  prefix  pattern.  The word is expanded to produce a pattern just as in pathname expansion.  If the pattern matches the beginning of the value of parameter, then the result of</span></span><br><span class="line">       <span class="comment">#        the expansion is the expanded value of parameter with the shortest matching pattern (the ``#&#x27;&#x27; case) or the longest matching pattern (the ``##&#x27;&#x27; case) deleted.  If parameter is  @  or  *,  the</span></span><br><span class="line">       <span class="comment">#        pattern  removal  operation  is  applied  to each positional parameter in turn, and the expansion is the resultant list.  If parameter is an array variable subscripted with @ or *, the pattern</span></span><br><span class="line">       <span class="comment">#        removal operation is applied to each member of the array in turn, and the expansion is the resultant list.</span></span><br><span class="line"></span><br><span class="line">       <span class="comment"># $&#123;parameter%word&#125;</span></span><br><span class="line">       <span class="comment"># $&#123;parameter%%word&#125;</span></span><br><span class="line">       <span class="comment">#        Remove matching suffix pattern.  The word is expanded to produce a pattern just as in pathname expansion.  If the pattern matches a trailing portion of the expanded value  of  parameter,  then</span></span><br><span class="line">       <span class="comment">#        the result of the expansion is the expanded value of parameter with the shortest matching pattern (the ``%&#x27;&#x27; case) or the longest matching pattern (the ``%%&#x27;&#x27; case) deleted.  If parameter is @</span></span><br><span class="line">       <span class="comment">#        or *, the pattern removal operation is applied to each positional parameter in turn, and the expansion is the resultant list.  If parameter is an array variable subscripted with @  or  *,  the</span></span><br><span class="line">       <span class="comment">#        pattern removal operation is applied to each member of the array in turn, and the expansion is the resultant list.</span></span><br><span class="line"><span class="comment"># echo $&#123;pair%%:*&#125;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 把A01到B06合并成rep1</span></span><br><span class="line">mkdir -p <span class="variable">$wkdir</span>/5.mergedbam</span><br><span class="line"><span class="keyword">for</span> pairs <span class="keyword">in</span> rep1:A01:B06 rep2:C01:D8f</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  pre=`<span class="built_in">echo</span> <span class="variable">$pairs</span> | cut -d <span class="string">&quot;:&quot;</span> -f 1`</span><br><span class="line">  barcode1=`<span class="built_in">echo</span> <span class="variable">$pairs</span> | cut -d <span class="string">&quot;:&quot;</span> -f 2`</span><br><span class="line">  barcode2=`<span class="built_in">echo</span> <span class="variable">$pairs</span> | cut -d <span class="string">&quot;:&quot;</span> -f 3`</span><br><span class="line">  samtools merge -n -@ <span class="variable">$ncores</span> <span class="variable">$wkdir</span>/5.mergedbam/RBFOX2.CLIP.<span class="variable">$pre</span>.bam \</span><br><span class="line">    <span class="variable">$wkdir</span>/4.rmDupbam/RBFOX2.CLIP.<span class="variable">$barcode1</span>.rmDup.bam \</span><br><span class="line">    <span class="variable">$wkdir</span>/4.rmDupbam/RBFOX2.CLIP.<span class="variable">$barcode2</span>.rmDup.bam &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment">############## 提取read2 #############</span></span><br><span class="line"><span class="comment"># samtools flags 128</span></span><br><span class="line"><span class="comment"># -f 128 把read2提取出来做peak calling</span></span><br><span class="line">mkdir -p <span class="variable">$wkdir</span>/6.read2bam</span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> <span class="variable">$wkdir</span>/5.mergedbam/*.bam <span class="variable">$wkdir</span>/4.rmDupbam/RBFOX2.INPUT.NIL.rmDup.bam</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  filename=`basename <span class="variable">$i</span>`</span><br><span class="line">  samtools view -f 128 -@ <span class="variable">$ncores</span> \</span><br><span class="line">    -bo <span class="variable">$wkdir</span>/6.read2bam/<span class="variable">$filename</span> <span class="variable">$i</span> &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"><span class="keyword">for</span> file <span class="keyword">in</span> <span class="variable">$wkdir</span>/6.read2bam/*.bam</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  filename=`basename <span class="variable">$file</span> | sed <span class="string">&#x27;s/.bam//&#x27;</span>`</span><br><span class="line">  samtools sort -@ <span class="variable">$ncores</span> -o <span class="variable">$wkdir</span>/6.read2bam/<span class="variable">$filename</span>.srt.bam <span class="variable">$file</span> &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"><span class="keyword">for</span> file <span class="keyword">in</span> <span class="variable">$wkdir</span>/6.read2bam/*.srt.bam</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  samtools index <span class="variable">$file</span> &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment">############## 生成bigwig #############</span></span><br><span class="line"><span class="comment"># 与后续clipper没有先后关系，后台运行即可</span></span><br><span class="line"><span class="comment"># --direction r 需要标记链的方向（链特异性建库）</span></span><br><span class="line"><span class="comment"># samtools view -H $wkdir/5.mergedbam/RBFOX2.CLIP.rep1.bam | \</span></span><br><span class="line"><span class="comment"># 	awk &#x27;$1==&quot;@SQ&quot;&#123;print $2&quot;\t&quot;$3&#125;&#x27; | \</span></span><br><span class="line"><span class="comment"># 	sed &#x27;s?SN:??;s?LN:??&#x27; &gt; hg38.chrom.size</span></span><br><span class="line">mkdir -p <span class="variable">$wkdir</span>/7.bigwigfile</span><br><span class="line"><span class="keyword">for</span> file <span class="keyword">in</span> <span class="variable">$wkdir</span>/6.read2bam/*.srt.bam</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  filename=`basename <span class="variable">$file</span> | sed <span class="string">&#x27;s/.srt.bam//&#x27;</span>`</span><br><span class="line">  makebigwigfiles --bam <span class="variable">$file</span> --genome <span class="variable">$chromsize</span> \</span><br><span class="line">    --bw_pos <span class="variable">$wkdir</span>/7.bigwigfile/<span class="variable">$filename</span>.r2.norm.pos.bw \</span><br><span class="line">    --bw_neg <span class="variable">$wkdir</span>/7.bigwigfile/<span class="variable">$filename</span>.r2.norm.neg.bw \</span><br><span class="line">    --direction r &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 也可以用deeptools的bamCoverage来生成bigwig</span></span><br><span class="line"><span class="comment"># for file in $wkdir/6.read2bam/*.bam</span></span><br><span class="line"><span class="comment"># do</span></span><br><span class="line"><span class="comment"># 	samtools index $file</span></span><br><span class="line"><span class="comment"># 	bamCoverage -b mergedbam/$prefix.r2.bam \</span></span><br><span class="line"><span class="comment"># 		-o bigwigfile/$prefix.r2.norm.pos.bw \</span></span><br><span class="line"><span class="comment"># 		-p 2 --binSize 10 --smoothLength 100 \</span></span><br><span class="line"><span class="comment"># 		--normalizeUsing CPM --filterRNAstrand forward &amp;</span></span><br><span class="line"><span class="comment"># 	bamCoverage -b mergedbam/$prefix.r2.bam \</span></span><br><span class="line"><span class="comment"># 		-o bigwigfile/$prefix.r2.norm.neg.bw \</span></span><br><span class="line"><span class="comment"># 		-p 2 --binSize 10 --smoothLength 100 \</span></span><br><span class="line"><span class="comment"># 		--normalizeUsing CPM --filterRNAstrand reverse &amp;</span></span><br><span class="line"><span class="comment"># done</span></span><br><span class="line"><span class="comment"># 如果想把负链为负数，则可以用bigwigtowig结合awk运算，再转回bigwig</span></span><br><span class="line"></span><br><span class="line"><span class="comment">############## clipper生成peak #############</span></span><br><span class="line"><span class="comment"># clipper依赖特定的scikit-learn，单独给它建立一个conda env，但可以用软链接链到py2环境</span></span><br><span class="line"><span class="comment"># 必须为sorted bam以及index过的</span></span><br><span class="line"><span class="comment"># hg38会报错，但GRCh38_v29则不会，无解</span></span><br><span class="line"><span class="comment"># clipper很慢，用nohup或screen防止中断</span></span><br><span class="line"><span class="comment"># -g ENSG00000166579 此处用一个基因作为示例，可以快速得到结果</span></span><br><span class="line">mkdir -p <span class="variable">$wkdir</span>/8.peakcalling/</span><br><span class="line"><span class="keyword">for</span> rep <span class="keyword">in</span> rep1 rep2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  clipper -b <span class="variable">$wkdir</span>/6.read2bam/RBFOX2.CLIP.<span class="variable">$rep</span>.srt.bam \</span><br><span class="line">    -s GRCh38_v29 \</span><br><span class="line">    -o <span class="variable">$wkdir</span>/8.peakcalling/RBFOX2.<span class="variable">$rep</span>.peaks.bed \</span><br><span class="line">    --processors <span class="variable">$ncores</span> -g ENSG00000166579</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="comment"># -s 支持的物种名称</span></span><br><span class="line"><span class="comment"># ll /public/workspace/shaojf/anaconda3/envs/clipper3/lib/python3.7/site-packages/clipper/data/</span></span><br><span class="line"><span class="comment"># ll /public/workspace/shaojf/anaconda3/envs/clipper3/lib/python3.7/site-packages/clipper/data/regions</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># alternative peak calling with MACS2</span></span><br><span class="line"><span class="comment"># -f BAMPE 不能加，为什么呢（思考一下？）</span></span><br><span class="line"><span class="comment"># 以r2.bam为输入会报错，无法读入read，只能用初始的比对结果</span></span><br><span class="line">mkdir -p <span class="variable">$wkdir</span>/8.peakcalling/MACS2</span><br><span class="line"><span class="keyword">for</span> rep <span class="keyword">in</span> rep1 rep2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  macs2 callpeak -g hs --keep-dup all \</span><br><span class="line">    -t <span class="variable">$wkdir</span>/5.mergedbam/RBFOX2.CLIP.<span class="variable">$rep</span>.bam \</span><br><span class="line">    -c <span class="variable">$wkdir</span>/4.rmDupbam/RBFOX2.INPUT.NIL.rmDup.bam \</span><br><span class="line">    -n <span class="variable">$wkdir</span>/8.peakcalling/MACS2/RBFOX2.<span class="variable">$rep</span>.macs2 &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="comment">############## pureclip生成peak #############</span></span><br><span class="line"><span class="comment"># pureclip [OPTIONS] &lt;-i BAM FILE&gt; &lt;-bai BAI FILE&gt; &lt;-g GENOME FILE&gt; &lt;-o OUTPUT BED FILE&gt;</span></span><br><span class="line"><span class="comment"># 超级占内存！！！</span></span><br><span class="line"><span class="comment"># 会死机！！！</span></span><br><span class="line"><span class="comment"># 运行之前top一下看看当前正在运行的命令</span></span><br><span class="line"><span class="comment"># 切勿在server上运行</span></span><br><span class="line"><span class="comment"># -iv, --inter STR</span></span><br><span class="line"><span class="comment">#           Genomic chromosomes to learn HMM parameters, e.g. &#x27;chr1;chr2;chr3&#x27;. Contigs have to be in the same order as</span></span><br><span class="line"><span class="comment">#           in BAM file. Useful to reduce runtime and memory consumption. Default: all contigs from reference file are</span></span><br><span class="line"><span class="comment">#           used (useful when applying to transcript-wise alignments or poor data).</span></span><br><span class="line">mkdir -p <span class="variable">$wkdir</span>/8.peakcalling/pureclip.peaks</span><br><span class="line"><span class="keyword">for</span> rep <span class="keyword">in</span> rep1 rep2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  pureclip -i <span class="variable">$wkdir</span>/6.read2bam/RBFOX2.CLIP.<span class="variable">$rep</span>.srt.bam \</span><br><span class="line">    -bai <span class="variable">$wkdir</span>/6.read2bam/RBFOX2.CLIP.<span class="variable">$rep</span>.srt.bam.bai \</span><br><span class="line">    -ibam <span class="variable">$wkdir</span>/6.read2bam/RBFOX2.INPUT.NIL.rmDup.srt.bam \</span><br><span class="line">    -ibai <span class="variable">$wkdir</span>/6.read2bam/RBFOX2.INPUT.NIL.rmDup.srt.bam.bai \</span><br><span class="line">    -g <span class="variable">$genomefa</span> \</span><br><span class="line">    -o <span class="variable">$wkdir</span>/8.peakcalling/pureclip.peaks/RBFOX2.CLIP.<span class="variable">$rep</span>.called_crosslinksites.bed \</span><br><span class="line">    -or <span class="variable">$wkdir</span>/8.peakcalling/pureclip.peaks/RBFOX2.CLIP.<span class="variable">$rep</span>.called_bindingregions.bed \</span><br><span class="line">    -nt <span class="variable">$ncores</span> -iv <span class="string">&#x27;chr1;chr10;chr20&#x27;</span></span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="comment">############## Input normalization #############</span></span><br><span class="line">mkdir -p <span class="variable">$wkdir</span>/9.normalization</span><br><span class="line"><span class="keyword">for</span> rep <span class="keyword">in</span> rep1 rep2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  samtools view -cF 4 <span class="variable">$wkdir</span>/6.read2bam/RBFOX2.CLIP.<span class="variable">$rep</span>.srt.bam \</span><br><span class="line">    &gt; <span class="variable">$wkdir</span>/9.normalization/RBFOX2.CLIP.<span class="variable">$rep</span>.mapped_readnum.txt &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line">samtools view -cF 4 <span class="variable">$wkdir</span>/6.read2bam/RBFOX2.INPUT.NIL.rmDup.srt.bam \</span><br><span class="line">    &gt; <span class="variable">$wkdir</span>/9.normalization/RBFOX2.INPUT.NIL.mapped_readnum.txt &amp;</span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> rep <span class="keyword">in</span> rep1 rep2</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">  overlap_peakfi_with_bam.pl \</span><br><span class="line">    <span class="variable">$wkdir</span>/6.read2bam/RBFOX2.CLIP.<span class="variable">$rep</span>.srt.bam \</span><br><span class="line">    <span class="variable">$wkdir</span>/6.read2bam/RBFOX2.INPUT.NIL.rmDup.srt.bam \</span><br><span class="line">    <span class="variable">$wkdir</span>/8.peakcalling/RBFOX2.<span class="variable">$rep</span>.peaks.bed \</span><br><span class="line">    <span class="variable">$wkdir</span>/9.normalization/RBFOX2.CLIP.<span class="variable">$rep</span>.mapped_readnum.txt \</span><br><span class="line">  <span class="variable">$wkdir</span>/9.normalization/RBFOX2.INPUT.NIL.mapped_readnum.txt \</span><br><span class="line">    <span class="variable">$wkdir</span>/9.normalization/RBFOX2.<span class="variable">$rep</span>.peaks.normed.bed &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">wait</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># MEME-chip不大适合RNA peak的分析，但也只有这种方法</span></span><br></pre></td></tr></table></figure></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="mailto:undefined">yilingzhang</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="http://yilingzhang.gitee.io/2021/03/31/eCLIP-Analysis/">http://yilingzhang.gitee.io/2021/03/31/eCLIP-Analysis/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/eCLIP/">eCLIP</a></div><div class="post_share"><div class="social-share" data-image="/img/cover/eCLIP.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/social-share.js/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/social-share.js/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2021/04/01/Metagenomics/"><img class="prev-cover" src="/img/cover/Metagenomics.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">Metagenomics Analysis</div></div></a></div><div class="next-post pull-right"><a href="/2021/03/30/WES-Project/"><img class="next-cover" src="/img/cover/WES.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">WES Project</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="card-info-avatar is-center"><img class="avatar-img" src="/img/selfie.JPG" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/><div class="author-info__name">yilingzhang</div><div class="author-info__description">Every Day is My Day</div></div><div class="card-info-data"><div class="card-info-data-item is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">21</div></a></div><div class="card-info-data-item is-center"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">11</div></a></div><div class="card-info-data-item is-center"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">12</div></a></div></div><div class="card-info-social-icons is-center"><a class="social-icon" href="mailto:jocelynzhang1119@icloud.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a><a class="social-icon" href="https://blog.csdn.net/weixin_43893431" target="_blank" title="CSDN"><i class="fas fa-paper-plane"></i></a></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%88%86%E6%9E%90%E8%B7%AF%E7%BA%BF"><span class="toc-number">1.</span> <span class="toc-text">分析路线</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%B7%A5%E5%85%B7"><span class="toc-number">2.</span> <span class="toc-text">工具</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#fastqc"><span class="toc-number">2.1.</span> <span class="toc-text">fastqc</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#STAR"><span class="toc-number">2.2.</span> <span class="toc-text">STAR</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Pureclip"><span class="toc-number">2.3.</span> <span class="toc-text">Pureclip</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#MACS2"><span class="toc-number">2.4.</span> <span class="toc-text">MACS2</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#UMI-Tools"><span class="toc-number">2.5.</span> <span class="toc-text">UMI_Tools</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Code"><span class="toc-number">3.</span> <span class="toc-text">Code</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2021/04/08/scATAC-seq-Analysis/" title="scATAC-seq Analysis"><img src="/img/cover/scATAC.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scATAC-seq Analysis"/></a><div class="content"><a class="title" href="/2021/04/08/scATAC-seq-Analysis/" title="scATAC-seq Analysis">scATAC-seq Analysis</a><time datetime="2021-04-08T11:01:25.000Z" title="Created 2021-04-08 19:01:25">2021-04-08</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/04/06/TOEFL-writing-05/" title="TOEFL-writing-05"><img src="/img/cover/TOEFL-writing-04.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="TOEFL-writing-05"/></a><div class="content"><a class="title" href="/2021/04/06/TOEFL-writing-05/" title="TOEFL-writing-05">TOEFL-writing-05</a><time datetime="2021-04-06T00:48:39.000Z" title="Created 2021-04-06 08:48:39">2021-04-06</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/04/04/conda-mamba-Installation/" title="conda/mamba Installation"><img src="/img/cover/conda.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="conda/mamba Installation"/></a><div class="content"><a class="title" href="/2021/04/04/conda-mamba-Installation/" title="conda/mamba Installation">conda/mamba Installation</a><time datetime="2021-04-04T03:42:27.000Z" title="Created 2021-04-04 11:42:27">2021-04-04</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/04/01/Metagenomics/" title="Metagenomics Analysis"><img src="/img/cover/Metagenomics.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Metagenomics Analysis"/></a><div class="content"><a class="title" href="/2021/04/01/Metagenomics/" title="Metagenomics Analysis">Metagenomics Analysis</a><time datetime="2021-04-01T11:31:45.000Z" title="Created 2021-04-01 19:31:45">2021-04-01</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/03/31/eCLIP-Analysis/" title="eCLIP Analysis"><img src="/img/cover/eCLIP.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="eCLIP Analysis"/></a><div class="content"><a class="title" href="/2021/03/31/eCLIP-Analysis/" title="eCLIP Analysis">eCLIP Analysis</a><time datetime="2021-03-31T00:19:05.000Z" title="Created 2021-03-31 08:19:05">2021-03-31</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2021 By yilingzhang</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"></div><script defer="defer" id="ribbon" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/canvas-ribbon.min.js" size="150" alpha="0.6" zIndex="-1" mobile="false" data-click="true"></script><script id="click-heart" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/click-heart.min.js" async="async" mobile="false"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>