<!DOCTYPE html><html lang="en | zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1"><title>scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv | Jocelyn's Blog</title><meta name="keywords" content="scRNA-seq"><meta name="author" content="yilingzhang"><meta name="copyright" content="yilingzhang"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="拟时分析拟时分析本质上是根据每个细胞的基因表达情况找出关键基因的表达模式来对细胞进行排序，反推出细胞发育轨迹。用于细胞分化的研究(胚胎发育，细胞重编程，癌细胞分化…)细胞在外界刺激和细胞发育过程中，细胞的功能会发生变化。这些功能的变化是靠蛋白表达的调控即转录组信息的变化实现的(一些基因表达上调一些基因表达下调)。细胞在发育过程中不是完全同步的，在一次取样的过程中可以取到不同发育状况的细胞。通过单细">
<meta property="og:type" content="article">
<meta property="og:title" content="scRNAseq-05 Monocle3 &amp; GO&#x2F;KEGG &amp; TCGASurv">
<meta property="og:url" content="https://hiiiyilingzhang.github.io/2021/05/24/scRNAseq-05-Monocle3-GO-KEGG-TCGASurv/index.html">
<meta property="og:site_name" content="Jocelyn&#39;s Blog">
<meta property="og:description" content="拟时分析拟时分析本质上是根据每个细胞的基因表达情况找出关键基因的表达模式来对细胞进行排序，反推出细胞发育轨迹。用于细胞分化的研究(胚胎发育，细胞重编程，癌细胞分化…)细胞在外界刺激和细胞发育过程中，细胞的功能会发生变化。这些功能的变化是靠蛋白表达的调控即转录组信息的变化实现的(一些基因表达上调一些基因表达下调)。细胞在发育过程中不是完全同步的，在一次取样的过程中可以取到不同发育状况的细胞。通过单细">
<meta property="og:locale">
<meta property="og:image" content="https://hiiiyilingzhang.github.io/img/cover/scRNA.png">
<meta property="article:published_time" content="2021-05-24T04:51:09.000Z">
<meta property="article:modified_time" content="2021-05-24T05:35:33.558Z">
<meta property="article:author" content="yilingzhang">
<meta property="article:tag" content="scRNA-seq">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://hiiiyilingzhang.github.io/img/cover/scRNA.png"><link rel="shortcut icon" href="/img/love_32px.png"><link rel="canonical" href="https://hiiiyilingzhang.github.io/2021/05/24/scRNAseq-05-Monocle3-GO-KEGG-TCGASurv/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    jQuery: 'https://cdn.jsdelivr.net/npm/jquery@latest/dist/jquery.min.js',
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/js/jquery.justifiedGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/css/justifiedGallery.min.css'
    },
    fancybox: {
      js: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.js',
      css: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isanchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = { 
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2021-05-24 13:35:33'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    })(window)</script><meta name="generator" content="Hexo 5.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="author-avatar"><img class="avatar-img" src="/img/selfie.JPG" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data"><div class="data-item is-center"><div class="data-item-link"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">39</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">17</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">14</div></a></div></div></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/img/cover/scRNA.png')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Jocelyn's Blog</a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2021-05-24T04:51:09.000Z" title="Created 2021-05-24 12:51:09">2021-05-24</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2021-05-24T05:35:33.558Z" title="Updated 2021-05-24 13:35:33">2021-05-24</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/Single-Cell/">Single Cell</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">Word count:</span><span class="word-count">4.2k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">Reading time:</span><span>16min</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><p><strong>拟时分析</strong><br>拟时分析本质上是根据每个细胞的基因表达情况找出关键基因的表达模式来对细胞进行排序，反推出细胞发育轨迹。用于细胞分化的研究(胚胎发育，细胞重编程，癌细胞分化…)<br>细胞在外界刺激和细胞发育过程中，细胞的功能会发生变化。这些功能的变化是靠蛋白表达的调控即<strong>转录组信息</strong>的变化实现的(一些基因表达上调一些基因表达下调)。<br>细胞在发育过程中不是完全同步的，在一次取样的过程中可以取到不同发育状况的细胞。通过单细胞测序得到不同时期细胞的转录组信息，可以推断出这些细胞之前的先后顺序，从而反推出细胞发育轨迹。</p>
<ul>
<li>RNA Velovity</li>
<li>增殖相关基因表达量（自行定义根节点）</li>
</ul>
<h1 id="Monocle3"><a href="#Monocle3" class="headerlink" title="Monocle3"></a>Monocle3</h1><ul>
<li>从UMAP图识别发育轨迹，可以继承Seurat的质控、批次校正和降维分析结果，实现“一张图”展现细胞的聚类、鉴定和轨迹分析结果。</li>
<li>自动对UMAP图分区（partition），可以选择多个起点，轨迹分析算法的逻辑更符合生物学现实。</li>
<li>官网：<a target="_blank" rel="noopener" href="https://cole-trapnell-lab.github.io/monocle3/">https://cole-trapnell-lab.github.io/monocle3/</a></li>
</ul>
<h2 id="步骤"><a href="#步骤" class="headerlink" title="步骤"></a>步骤</h2><p>数据导入→数据预处理→降维→聚类→(细胞类群之间的比较)→拟时分析</p>
<h3 id="数据导入"><a href="#数据导入" class="headerlink" title="数据导入"></a>数据导入</h3><ul>
<li>可导入Cellranger结果</li>
<li>导入非Cellranger的结果，需要准备三个input文件：<br>  1.expression_matrix：<br>  2.cell metadata：细胞属性的数据框，每一行对应一个细胞，每一列对应不同的细胞属性<br>  3.gene_metadata：记录了基因属性的数据框,每一行对应一个基因，每一列对应不同的基因属性</li>
<li>从Seurat结果中导入数据<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##1.读取seurat产生的RDS文件</span></span><br><span class="line">pbmc&lt;-read RDS(＂seurat_rds_for_test.rds＂)</span><br><span class="line"><span class="comment">##或者可以用seurat自带的pbmc_small进行测试</span></span><br><span class="line">library(Seurat)</span><br><span class="line">pbmc&lt;-pbmc_small</span><br><span class="line"><span class="comment">##2.拆分成3个文件</span></span><br><span class="line">Expression_data&lt;-as(as.matrix(pbmc@assays$RNA@counts),＂sparse Matrix＂)</span><br><span class="line">Cell_data-&gt;pbmc@meta.data</span><br><span class="line">Gene_data&lt;-data.frame(gene_short_name=row.names(data), row.names=row.names(data))</span><br><span class="line"><span class="comment">##3.生成monocle 3对象</span></span><br><span class="line">cds&lt;- new_cell_data_set(expression_matrix=Expression_data，</span><br><span class="line">            cell_metadata=Cell_data，</span><br><span class="line">            gene_metadata=Gene_data)</span><br></pre></td></tr></table></figure>
<h3 id="数据预处理"><a href="#数据预处理" class="headerlink" title="数据预处理"></a>数据预处理</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 后面的例子采用官网教程数据</span></span><br><span class="line">expression_matrix &lt;- readRDS(url(<span class="string">&quot;http://staff.washington.edu/hpliner/data/packer_embryo_expression.rds&quot;</span>))</span><br><span class="line">cell_metadata &lt;- readRDS(url(<span class="string">&quot;http://staff.washington.edu/hpliner/data/packer_embryo_colData.rds&quot;</span>))</span><br><span class="line">gene_annotation &lt;- readRDS(url(<span class="string">&quot;http://staff.washington.edu/hpliner/data/packer_embryo_rowData.rds&quot;</span>))</span><br><span class="line"></span><br><span class="line">cds &lt;- new_cell_data_set(expression_matrix,</span><br><span class="line">                         cell_metadata = cell_metadata,</span><br><span class="line">                         gene_metadata = gene_annotation)</span><br><span class="line">cds = preprocess_cds(cds, norm_method=＂<span class="built_in">log</span>“, method=＂PCA“, num_dim=<span class="number">100</span>)</span><br></pre></td></tr></table></figure></li>
<li>标准化：默认对表达矩阵的raw reads进行log处理(减小数据的离散程度)</li>
<li>预降维：方法有PCA， LSI。默认为PCA降维， 一般RNAseq采用PCA降维即可</li>
<li>numdim：PCA维度，默认为50，可以根据细胞数进行调整</li>
</ul>
<h3 id="降维及可视化"><a href="#降维及可视化" class="headerlink" title="降维及可视化"></a>降维及可视化</h3><p>进行t-SNE/UMAP降维(如果细胞数较多，比如几万个细胞，推荐UMAP降维，减少运算时间)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">cds&lt;-reduce_dimension(cds)</span><br><span class="line">plot_cells(cds, label_groups_by_cluster=<span class="literal">FALSE</span>)</span><br><span class="line"></span><br><span class="line">ciliated_genes &lt;- <span class="built_in">c</span>(<span class="string">&quot;che-1&quot;</span>,</span><br><span class="line">                    <span class="string">&quot;hlh-17&quot;</span>,</span><br><span class="line">                    <span class="string">&quot;nhr-6&quot;</span>,</span><br><span class="line">                    <span class="string">&quot;dmd-6&quot;</span>,</span><br><span class="line">                    <span class="string">&quot;ceh-36&quot;</span>,</span><br><span class="line">                    <span class="string">&quot;ham-1&quot;</span>)</span><br><span class="line"></span><br><span class="line">plot_cells(cds,</span><br><span class="line">           genes=ciliated_genes,</span><br><span class="line">           label_cell_groups=<span class="literal">FALSE</span>,</span><br><span class="line">           show_trajectory_graph=<span class="literal">FALSE</span>)</span><br></pre></td></tr></table></figure>
<h3 id="聚类分析"><a href="#聚类分析" class="headerlink" title="聚类分析"></a>聚类分析</h3><p>目的是为了找出那些细胞在一个发育轨迹</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">cds &lt;- cluster_cells(cds)</span><br><span class="line">plot_cells(cds, color_cells_by = <span class="string">&quot;partition&quot;</span>)</span><br></pre></td></tr></table></figure>
<h3 id="拟时分析"><a href="#拟时分析" class="headerlink" title="拟时分析"></a>拟时分析</h3><p>在上一步分群的基础上，找出每个群内的细胞发育轨迹(无监督，没有基于生物学背景知识)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cds &lt;- learn_graph(cds)</span><br><span class="line">plot_cells(cds,</span><br><span class="line">           color_cells_by = <span class="string">&quot;cell.type&quot;</span>,</span><br><span class="line">           label_groups_by_cluster=<span class="literal">FALSE</span>,</span><br><span class="line">           label_leaves=<span class="literal">FALSE</span>,</span><br><span class="line">           label_branch_points=<span class="literal">FALSE</span>)</span><br></pre></td></tr></table></figure>
<h3 id="确定轨迹的先后顺序"><a href="#确定轨迹的先后顺序" class="headerlink" title="确定轨迹的先后顺序"></a>确定轨迹的先后顺序</h3><p>根据颜色，时间越长越往后，时间=0，代表根</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">get_earliest_principal_node</span><br><span class="line">&lt;-<span class="keyword">function</span>(cds， time_bin=<span class="string">&quot;130-170&quot;</span>) &#123;</span><br><span class="line">cell_ids&lt;-which(col Data(cds) [, <span class="string">&quot;embryo.time.bin&quot;</span>] ==time_bin)</span><br><span class="line">closest_vertex&lt;-</span><br><span class="line">cds@principal_graph_aux[[<span class="string">&quot;UMAP&quot;</span>] ] $pr_graph_cell_proj_closest_vertex</span><br><span class="line">closest_vertex &lt;-as.matrix(closest_vertex[col <span class="built_in">names</span>(cds), ] )</span><br><span class="line">root_pr_nodes&lt;-</span><br><span class="line">igraph::V(principal_graph(cds) [[<span class="string">&quot;UMAP&quot;</span>] ] ) $name[<span class="built_in">as.numeric</span>(<span class="built_in">names</span></span><br><span class="line">(which.max(table(closest_vertex[cell_ids, ] ) ) ) ) ]</span><br><span class="line">root_pr_nodes</span><br><span class="line">&#125;</span><br><span class="line">cds&lt;-order_cells(cds， root_pr_nodes=get_earliest_principal_node(cds))</span><br></pre></td></tr></table></figure>
<p>手动(每一个细胞轨迹均需要指定一个根节点)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">cds &lt;- order_cells(cds)</span><br><span class="line"></span><br><span class="line">plot_cells(cds,</span><br><span class="line">           color_cells_by = <span class="string">&quot;pseudotime&quot;</span>,</span><br><span class="line">           label_cell_groups=<span class="literal">FALSE</span>,</span><br><span class="line">           label_leaves=<span class="literal">FALSE</span>,</span><br><span class="line">           label_branch_points=<span class="literal">FALSE</span>,</span><br><span class="line">           graph_label_size=<span class="number">1.5</span>)</span><br></pre></td></tr></table></figure>
<h2 id="练习"><a href="#练习" class="headerlink" title="练习"></a>练习</h2><ul>
<li>数据：/public/workspace/lincs/lab7/singlecell/monocle3/下GSE139324文件夹，从头分析。还有很多RDS文件，可以拷贝过去，从中间步骤开始做，注意：count.rds和azimuth_pred.tsv需要同时拷贝过去</li>
<li>代码： /public/workspace/lincs/lab7/singlecell/monocle3/下monocle3-new.R</li>
<li>环境：在你们的conda环境中ln -s /public/workspace/lincs/.conda/envs/monocle3py37 。此例子包含前面harmony等多个包内容。</li>
<li>azimuth_pred.tsv来源于Seurat官网的细胞注释，推出了在线细胞类型鉴定服务，可以准确鉴定pbmc细胞，<a target="_blank" rel="noopener" href="http://azimuth.satijalab.org/app/azimuth">http://azimuth.satijalab.org/app/azimuth</a> ，上传count.rds，下载得到azimuth_pred.tsv</li>
</ul>
<h3 id="Installation"><a href="#Installation" class="headerlink" title="Installation"></a>Installation</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#  install Bioconductor</span></span><br><span class="line"><span class="keyword">if</span> (!requireNamespace(<span class="string">&quot;BiocManager&quot;</span>, quietly = <span class="literal">TRUE</span>))</span><br><span class="line">install.packages(<span class="string">&quot;BiocManager&quot;</span>)</span><br><span class="line">BiocManager::install(version = <span class="string">&quot;3.10&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#  install a few Bioconductor dependencies that aren&#x27;t automatically installed:</span></span><br><span class="line">BiocManager::install(<span class="built_in">c</span>(<span class="string">&#x27;BiocGenerics&#x27;</span>, <span class="string">&#x27;DelayedArray&#x27;</span>, <span class="string">&#x27;DelayedMatrixStats&#x27;</span>,<span class="string">&#x27;limma&#x27;</span>, <span class="string">&#x27;S4Vectors&#x27;</span>, <span class="string">&#x27;SingleCellExperiment&#x27;</span>, <span class="string">&#x27;SummarizedExperiment&#x27;</span>, <span class="string">&#x27;batchelor&#x27;</span>, <span class="string">&#x27;Matrix.utils&#x27;</span>))</span><br><span class="line"></span><br><span class="line"><span class="comment">#  install monocle3 through the cole-trapnell-lab GitHub, execute:</span></span><br><span class="line">install.packages(<span class="string">&quot;devtools&quot;</span>)</span><br><span class="line">devtools::install_github(<span class="string">&#x27;cole-trapnell-lab/leidenbase&#x27;</span>)</span><br><span class="line">devtools::install_github(<span class="string">&#x27;cole-trapnell-lab/monocle3&#x27;</span>)</span><br></pre></td></tr></table></figure>
<h3 id="Run-Monocle3"><a href="#Run-Monocle3" class="headerlink" title="Run Monocle3"></a>Run Monocle3</h3><p><a href="/RmdOutput/monocle3.html">Rmd Output of Monocle3</a></p>
<h1 id="Monocle2"><a href="#Monocle2" class="headerlink" title="Monocle2"></a>Monocle2</h1><ul>
<li><p>Monocle进行伪时间分析的核心技术是一种机器学习算法——反向图形嵌入 (Reversed Graph Embedding)。</p>
</li>
<li><p>它分析的前提需要一张展现细胞转录特征相似性关系的图，Monocle2使用DDTree降维图，Monocle3使用UMAP降维图。</p>
</li>
<li><p>Monocle的机器学习算法可以依据上述降维图形，学习描述细胞如何从一种状态过渡到另一种状态的轨迹。</p>
</li>
<li><p>Monocle假设轨迹是树状结构，一端是“根”，另一端是“叶”。一个细胞在生物过程的开始，从根开始沿着主干进行，直到它到达第一个分支。然后，该细胞必须选择一条路径，并沿着树移动越来越远，直到它到达一片叶子。一个细胞的假时间值是它返回根所需的距离。</p>
</li>
<li><p>降维方面monocle与seurat的过程大同小异，首先进行数据标准化，其次选择部分基因代表细胞转录特征 ，最后选用适当的算法降维。</p>
</li>
<li><p>完成数据导入和预处理后，就可以考虑选择哪些基因代表细胞的发育特征，Monocle官网教程提供了4个选择方法：</p>
<blockquote>
<p>选择发育差异表达基因<br>选择clusters差异表达基因<br>选择离散程度高的基因<br>自定义发育marker基因</p>
</blockquote>
</li>
<li><p>前三种都是无监督分析方法，细胞发育轨迹生成完全不受人工干预；最后一种是半监督分析方法，可以使用先验知识辅助分析。第一种方法要求实验设计有不同的时间点，对起点和终点的样本做基因表达差异分析，挑选显著差异的基因进行后续分析。对于没有时序设计的实验样本，可以使用第2、3种方法挑选基因。第2种方法要先对细胞降维聚类，然后用clusters之间差异表达的基因开展后续分析。</p>
</li>
<li><p>Monocle有一套自己的降维聚类方法，与seurat的方法大同小异，很多教程直接使用seurat的差异分析结果。第3种方法使用离散程度高的基因开展分析，seurat有挑选高变基因的方法，monocle也有自己选择的算法。本案例数据不具备使用第1、4种方法的条件，因此这里只演示2、3种方法的使用。</p>
</li>
</ul>
<h2 id="拟时相关基因聚类热图"><a href="#拟时相关基因聚类热图" class="headerlink" title="拟时相关基因聚类热图"></a>拟时相关基因聚类热图</h2><p>Monocle中differentialGeneTest()函数可以按条件进行差异分析，将相关参数设为<code>fullModelFormulaStr = “~sm.ns(Pseudotime)</code>时，可以找到与拟时相关的差异基因。我们可以按一定的条件筛选基因后进行差异分析，全部基因都输入会耗费比较长的时间。建议使用cluster差异基因或高变基因输入函数计算。分析结果主要依据qval区分差异的显著性，筛选之后可以用<code>plot_pseudotime_heatmap</code>函数绘制成热图。</p>
<h2 id="BEAM分析"><a href="#BEAM分析" class="headerlink" title="BEAM分析"></a>BEAM分析</h2><p>单细胞轨迹中通常包括分支，它们的出现是因为细胞的表达模式不同。当细胞做出命运选择时，或者遗传、化学或环境扰动时，就会表现出不同的基因表达模式。BEAM(Branched expression analysis modeling)是一种统计方法，用于寻找以依赖于分支的方式调控的基因。</p>
<h2 id="练习-1"><a href="#练习-1" class="headerlink" title="练习"></a>练习</h2><p>数据：/public/workspace/lincs/lab7/singlecell/monocle2/下scRNAsub.rds和diff_genes_wilcox.csv。 pseudotime文件夹不用拷贝，为输出结果结果。<br>代码： /public/workspace/lincs/lab7/singlecell/monocle2/下monocle2.R<br>环境：在你们的conda环境中ln -s /public/workspace/lincs/.conda/envs/monocle2。</p>
<h1 id="GO-KEGG富集分析"><a href="#GO-KEGG富集分析" class="headerlink" title="GO/KEGG富集分析"></a>GO/KEGG富集分析</h1><p>数据在/public/workspace/lincs/lab7/singlecell/cluster/下，<br>环境在/public/workspace/lincs/.conda/envs/singlecellcluter/ 代码如下</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br></pre></td><td class="code"><pre><span class="line">rm(<span class="built_in">list</span>=ls())</span><br><span class="line">library(Seurat)</span><br><span class="line">library(tidyverse)</span><br><span class="line">library(patchwork)</span><br><span class="line">library(monocle)</span><br><span class="line">library(clusterProfiler)</span><br><span class="line">library(org.Hs.eg.db)</span><br><span class="line">setwd(<span class="string">&quot;&quot;</span>)</span><br><span class="line">dir.create(<span class="string">&quot;enrich&quot;</span>)</span><br><span class="line">scRNA &lt;- readRDS(<span class="string">&quot;scRNA.rds&quot;</span>)</span><br><span class="line">mycds &lt;- readRDS(<span class="string">&quot;mycds.rds&quot;</span>)</span><br><span class="line"><span class="comment">#比较cluster0和cluster1的差异表达基因</span></span><br><span class="line">dge.cluster &lt;- FindMarkers(scRNA,ident.1 = <span class="number">0</span>,ident.2 = <span class="number">1</span>)</span><br><span class="line">sig_dge.cluster &lt;- subset(dge.cluster, p_val_adj&lt;<span class="number">0.01</span>&amp;<span class="built_in">abs</span>(avg_log2FC)&gt;<span class="number">1</span>)</span><br><span class="line"><span class="comment">#比较B_cell和T_cells的差异表达基因</span></span><br><span class="line">dge.celltype &lt;- FindMarkers(scRNA, ident.1 = <span class="string">&#x27;B_cell&#x27;</span>, ident.2 = <span class="string">&#x27;T_cells&#x27;</span>, group.by = <span class="string">&#x27;celltype&#x27;</span>)</span><br><span class="line">sig_dge.celltype &lt;- subset(dge.celltype, p_val_adj&lt;<span class="number">0.01</span>&amp;<span class="built_in">abs</span>(avg_log2FC)&gt;<span class="number">1</span>)</span><br><span class="line"><span class="comment">#比较拟时State1和State3的差异表达基因</span></span><br><span class="line">p_data &lt;- subset(pData(mycds),select=<span class="string">&#x27;State&#x27;</span>)</span><br><span class="line">scRNAsub &lt;- subset(scRNA, cells=row.names(p_data))</span><br><span class="line">scRNAsub &lt;- AddMetaData(scRNAsub,p_data,col.name = <span class="string">&#x27;State&#x27;</span>)</span><br><span class="line"></span><br><span class="line">dge.State &lt;- FindMarkers(scRNAsub, ident.1 = <span class="number">1</span>, ident.2 = <span class="number">3</span>, group.by = <span class="string">&#x27;State&#x27;</span>)</span><br><span class="line">sig_dge.State &lt;- subset(dge.State, p_val_adj&lt;<span class="number">0.01</span>&amp;<span class="built_in">abs</span>(avg_log2FC)&gt;<span class="number">1</span>)</span><br><span class="line"><span class="comment">#差异基因GO富集分析</span></span><br><span class="line">ego_ALL &lt;- enrichGO(gene          = row.names(sig_dge.celltype),</span><br><span class="line">                   <span class="comment">#universe     = row.names(dge.celltype),</span></span><br><span class="line">                   OrgDb         = <span class="string">&#x27;org.Hs.eg.db&#x27;</span>,</span><br><span class="line">                   keyType       = <span class="string">&#x27;SYMBOL&#x27;</span>,</span><br><span class="line">                   ont           = <span class="string">&quot;ALL&quot;</span>,</span><br><span class="line">                   pAdjustMethod = <span class="string">&quot;BH&quot;</span>,</span><br><span class="line">                   pvalueCutoff  = <span class="number">0.01</span>,</span><br><span class="line">                   qvalueCutoff  = <span class="number">0.05</span>)</span><br><span class="line">ego_all &lt;- data.frame(ego_ALL)</span><br><span class="line">write.csv(ego_all,<span class="string">&#x27;enrich/enrichGO.csv&#x27;</span>)           </span><br><span class="line">ego_CC &lt;- enrichGO(gene          = row.names(sig_dge.celltype),</span><br><span class="line">                   <span class="comment">#universe     = row.names(dge.celltype),</span></span><br><span class="line">                   OrgDb         = <span class="string">&#x27;org.Hs.eg.db&#x27;</span>,</span><br><span class="line">                   keyType       = <span class="string">&#x27;SYMBOL&#x27;</span>,</span><br><span class="line">                   ont           = <span class="string">&quot;CC&quot;</span>,</span><br><span class="line">                   pAdjustMethod = <span class="string">&quot;BH&quot;</span>,</span><br><span class="line">                   pvalueCutoff  = <span class="number">0.01</span>,</span><br><span class="line">                   qvalueCutoff  = <span class="number">0.05</span>)</span><br><span class="line">ego_MF &lt;- enrichGO(gene          = row.names(sig_dge.celltype),</span><br><span class="line">                   <span class="comment">#universe     = row.names(dge.celltype),</span></span><br><span class="line">                   OrgDb         = <span class="string">&#x27;org.Hs.eg.db&#x27;</span>,</span><br><span class="line">                   keyType       = <span class="string">&#x27;SYMBOL&#x27;</span>,</span><br><span class="line">                   ont           = <span class="string">&quot;MF&quot;</span>,</span><br><span class="line">                   pAdjustMethod = <span class="string">&quot;BH&quot;</span>,</span><br><span class="line">                   pvalueCutoff  = <span class="number">0.01</span>,</span><br><span class="line">                   qvalueCutoff  = <span class="number">0.05</span>)</span><br><span class="line"></span><br><span class="line">ego_BP &lt;- enrichGO(gene          = row.names(sig_dge.celltype),</span><br><span class="line">                   <span class="comment">#universe     = row.names(dge.celltype),</span></span><br><span class="line">                   OrgDb         = <span class="string">&#x27;org.Hs.eg.db&#x27;</span>,</span><br><span class="line">                   keyType       = <span class="string">&#x27;SYMBOL&#x27;</span>,</span><br><span class="line">                   ont           = <span class="string">&quot;BP&quot;</span>,</span><br><span class="line">                   pAdjustMethod = <span class="string">&quot;BH&quot;</span>,</span><br><span class="line">                   pvalueCutoff  = <span class="number">0.01</span>,</span><br><span class="line">                   qvalueCutoff  = <span class="number">0.05</span>) </span><br><span class="line">ego_CC@result$Description &lt;- substring(ego_CC@result$Description,<span class="number">1</span>,<span class="number">70</span>)</span><br><span class="line">ego_MF@result$Description &lt;- substring(ego_MF@result$Description,<span class="number">1</span>,<span class="number">70</span>)</span><br><span class="line">ego_BP@result$Description &lt;- substring(ego_BP@result$Description,<span class="number">1</span>,<span class="number">70</span>)</span><br><span class="line">p_BP &lt;- barplot(ego_BP,showCategory = <span class="number">10</span>) + ggtitle(<span class="string">&quot;barplot for Biological process&quot;</span>)</span><br><span class="line">p_CC &lt;- barplot(ego_CC,showCategory = <span class="number">10</span>) + ggtitle(<span class="string">&quot;barplot for Cellular component&quot;</span>)</span><br><span class="line">p_MF &lt;- barplot(ego_MF,showCategory = <span class="number">10</span>) + ggtitle(<span class="string">&quot;barplot for Molecular function&quot;</span>)</span><br><span class="line">plotc &lt;- p_BP/p_CC/p_MF</span><br><span class="line">ggsave(<span class="string">&#x27;enrich/enrichGO.png&#x27;</span>, plotc, width = <span class="number">12</span>,height = <span class="number">10</span>)</span><br><span class="line"></span><br><span class="line">genelist &lt;- bitr(row.names(sig_dge.celltype), fromType=<span class="string">&quot;SYMBOL&quot;</span>,</span><br><span class="line">                           toType=<span class="string">&quot;ENTREZID&quot;</span>, OrgDb=<span class="string">&#x27;org.Hs.eg.db&#x27;</span>)</span><br><span class="line">genelist &lt;- pull(genelist,ENTREZID)               </span><br><span class="line">ekegg &lt;- enrichKEGG(gene = genelist, organism = <span class="string">&#x27;hsa&#x27;</span>)</span><br><span class="line">p1 &lt;- barplot(ekegg, showCategory=<span class="number">20</span>)</span><br><span class="line">p2 &lt;- dotplot(ekegg, showCategory=<span class="number">20</span>)</span><br><span class="line">plotc = p1/p2</span><br><span class="line">ggsave(<span class="string">&quot;enrich/enrichKEGG.png&quot;</span>, plot = plotc, width = <span class="number">12</span>, height = <span class="number">10</span>)</span><br></pre></td></tr></table></figure>
<h1 id="GSEA-GSVA分析"><a href="#GSEA-GSVA分析" class="headerlink" title="GSEA/GSVA分析"></a>GSEA/GSVA分析</h1><h2 id="基因集的概念"><a href="#基因集的概念" class="headerlink" title="基因集的概念"></a>基因集的概念</h2><p>GSEA全称Gene Set Enrichment Analysis，GSVA全称Gene Set Variation Analysis，它们都是基于基因集开展的分析，因此我们先要了解基因集的定义。基因集顾名思义就是一些基因的集合，任何一些基因放在一起都可以叫做基因集，但是我们用来分析的基因集要求有一定的生物学意义。最常见的基因集数据库如GO与KEGG，它们一个按照基因本体论将基因分门别类，一个按照代谢通路将相关基因集合在一起。除此之外我们还可以按转录因子调控网络、共表达网络、定义生物状态的marker基因列表等把基因集合成有一定生物学意义的基因集。</p>
<h2 id="MSigDB基因集数据库"><a href="#MSigDB基因集数据库" class="headerlink" title="MSigDB基因集数据库"></a>MSigDB基因集数据库</h2><p>GSEA是由Broad研究所开发的一种富集方法，他们在提出该方法的同时还提供了一个基因集数据库——MSigdb。它从位置，功能，代谢途径，靶标结合等多种角度出发，构建出了许多的基因集合，Broad研究所将他们构建的基因集合保存在MSigDB，官网地址如下：<a target="_blank" rel="noopener" href="http://software.broadinstitute.org/gsea/msigdb/index.jsp">http://software.broadinstitute.org/gsea/msigdb/index.jsp</a></p>
<h2 id="GSVA"><a href="#GSVA" class="headerlink" title="GSVA"></a>GSVA</h2><ul>
<li>GSEA虽然是一种强大的富集分析工具，但是它的应用场景通常局限于Case/Control的实验设计。对于表型（分组）复杂的大样本量研究，例如TCGA和单细胞图谱这样的项目，分析起来就非常困难。</li>
<li>目的：对单细胞亚群进行功能注释后，如何求每群细胞某条pathway的gene set score</li>
<li>GSVA不需要预先进行样本之间的差异分析，它依据表达矩阵就可以计算每个样本中特定基因集的变异分数。</li>
<li>简单的说，输入以基因为行的表达矩阵和基因集数据库给GSVA，它就输出以基因集名称为行的变异分数矩阵，如下图所示：</li>
</ul>
<center><img src="/img/Figures/scRNA/GSVA.png" alt="GSVA" style="zoom:70%;" /></center>

<h3 id="GSEA练习"><a href="#GSEA练习" class="headerlink" title="GSEA练习"></a>GSEA练习</h3><p>数据和代码在/public/workspace/lincs/lab7/singlecell/GSEAGSVA/下，代码为GSEA.R<br>环境: ln -s /public/workspace/lincs/.conda/envs/gseagsva</p>
<ol>
<li>提取HNC01TIL和Tonsil2的T细胞做富集分析</li>
<li>写入表达数据gct文件</li>
<li>分组cls文件</li>
<li>运行GSEA #gs.db参数对应的是基因集数据文件，这里只对kegg收录的基因集做了分析<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">GSEA::GSEA(<span class="string">&quot;GSEA/input/expr.gct&quot;</span>, <span class="string">&quot;GSEA/input/group.cls&quot;</span>,gs.db=“./c2.cp.kegg.v7.1.symbols.gmt<span class="string">&quot;, output.directory=&quot;</span>GSEA/output/<span class="string">&quot;)</span></span><br></pre></td></tr></table></figure>
<h3 id="GSVA练习"><a href="#GSVA练习" class="headerlink" title="GSVA练习"></a>GSVA练习</h3>数据和代码在/public/workspace/lincs/lab7/singlecell/GSEAGSVA/下，代码为GSVA.R<br>环境在/public/workspace/lincs/.conda/envs/gseagsva</li>
</ol>
<p>在pbmc这个seurat object 的assays中存储了三种基因表达数据，保存在PBMC_pc10_res0.6.rds里</p>
<ol>
<li>pbmc@assays$RNA@counts : raw count</li>
<li>pbmc@assays$RNA@data : Normalized data</li>
<li>pbmc@assays$RNA@scale.data: scaled之后的data做GSVA分析时，使用normalized data.</li>
</ol>
<ul>
<li>步骤一：GSVA 分析和差异表达分析<br>选择感兴趣的两个细胞亚群（cluster），用GSVA算出pathway的表达量；</li>
<li>步骤二：根据已经得到的分组信息，挑选出需要的通路，画图</li>
</ul>
<p>需要注意的代码：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">genesets &lt;- getGmt(<span class="string">&quot;h.all.v7.0.symbols.gmt&quot;</span>)</span><br><span class="line"><span class="comment">#找到感兴趣的细胞组别，两者间进行比较这里以cluster6和cluster8为例</span></span><br><span class="line">subcluster = <span class="built_in">c</span>(<span class="number">6</span>, <span class="number">8</span>)</span><br><span class="line">sub &lt;- subset(pbmc, idents = subcluster)</span><br><span class="line">df.data &lt;- GetAssayData(object = sub, slot = <span class="string">&quot;data&quot;</span>)</span><br></pre></td></tr></table></figure>
<ul>
<li>结果：两个cluster之间的差别</li>
</ul>
<h1 id="联合TCGA生存分析"><a href="#联合TCGA生存分析" class="headerlink" title="联合TCGA生存分析"></a>联合TCGA生存分析</h1><ul>
<li>通过scRNA分析得到新的marker基因，gene signatures，以及解释研究结论的基因集和代谢通路，一般都会用公共数据库的bulk RNA数据验证一下。</li>
<li>常见的验证方式是：单细胞分析筛选的多个基因用TCGA/GEO数据做生存分析验证。</li>
</ul>
<p>数据和代码在/public/workspace/lincs/lab7/singlecell/tcgasurv/下，代码为tcgasurv.R, 数据为Data文件夹<br>环境在/public/workspace/lincs/.conda/envs/tcgasurv</p>
<p>注意的代码：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">TFHset = <span class="built_in">list</span>(<span class="built_in">c</span>(<span class="string">&#x27;PDCD1&#x27;</span>, <span class="string">&#x27;CXCL13&#x27;</span>, <span class="string">&#x27;TIGIT&#x27;</span>, <span class="string">&#x27;TOX&#x27;</span>, <span class="string">&#x27;MAF&#x27;</span>, <span class="string">&#x27;CXCR5&#x27;</span>, <span class="string">&#x27;CD40LG&#x27;</span>, <span class="string">&#x27;IL6ST&#x27;</span>))</span><br><span class="line"><span class="built_in">names</span>(TFHset) &lt;- <span class="string">&#x27;TFHset&#x27;</span></span><br><span class="line">ES &lt;- gsva(TCGAexpr, TFHset)</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br></pre></td><td class="code"><pre><span class="line">library(survival)</span><br><span class="line">library(survminer)</span><br><span class="line">library(tidyverse)</span><br><span class="line">library(GSVA)</span><br><span class="line"><span class="comment">#设置合适的路径，含有Data</span></span><br><span class="line">setwd()</span><br><span class="line">dir.create(<span class="string">&quot;Survival&quot;</span>)</span><br><span class="line"><span class="comment">#不要着急，读入有点慢</span></span><br><span class="line">TCGAexpr &lt;- read.delim(<span class="string">&quot;Data/TCGA-HNSC.htseq_fpkm.tsv&quot;</span>, check.names=<span class="built_in">F</span>)</span><br><span class="line">id.annota &lt;- read.delim(<span class="string">&quot;Data/gencode.v22.annotation.gene.probeMap&quot;</span>, check.names=<span class="built_in">F</span>)</span><br><span class="line">id.annota &lt;- id.annota[, <span class="number">1</span>:<span class="number">2</span>]</span><br><span class="line">TCGAexpr &lt;- merge(id.annota, TCGAexpr, by.x=<span class="string">&quot;id&quot;</span>, by.y=<span class="string">&quot;Ensembl_ID&quot;</span>)</span><br><span class="line">TCGAexpr &lt;- aggregate(TCGAexpr[,-<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">2</span>)], <span class="built_in">list</span>(TCGAexpr$gene), FUN=<span class="built_in">sum</span>)</span><br><span class="line">TCGAexpr &lt;- column_to_rownames(TCGAexpr,<span class="string">&quot;Group.1&quot;</span>)</span><br><span class="line"><span class="comment">#过滤表达量过低的基因，减少分析数据量</span></span><br><span class="line">TCGAexpr &lt;- TCGAexpr[rowSums(TCGAexpr)&gt;<span class="number">10</span>, ] %&gt;% as.matrix()</span><br><span class="line">TCGAclin &lt;- read.delim(<span class="string">&quot;Data/TCGA-HNSC.survival.tsv&quot;</span>, check.names = <span class="built_in">F</span>)</span><br><span class="line">TCGAclin &lt;- TCGAclin[<span class="built_in">as.character</span>(TCGAclin$sample) %in% colnames(TCGAexpr),]</span><br><span class="line">TCGAclin &lt;- TCGAclin[,-<span class="number">3</span>]</span><br><span class="line"><span class="comment">#GSVA分析</span></span><br><span class="line"><span class="comment">#原文富集分析用的200个基因找不到具体名称，这用了作者特异强调的8个上调基因代替</span></span><br><span class="line">TFHset = <span class="built_in">list</span>(<span class="built_in">c</span>(<span class="string">&#x27;PDCD1&#x27;</span>, <span class="string">&#x27;CXCL13&#x27;</span>, <span class="string">&#x27;TIGIT&#x27;</span>, <span class="string">&#x27;TOX&#x27;</span>, <span class="string">&#x27;MAF&#x27;</span>, <span class="string">&#x27;CXCR5&#x27;</span>, <span class="string">&#x27;CD40LG&#x27;</span>, <span class="string">&#x27;IL6ST&#x27;</span>))</span><br><span class="line"><span class="built_in">names</span>(TFHset) &lt;- <span class="string">&#x27;TFHset&#x27;</span></span><br><span class="line">ES &lt;- gsva(TCGAexpr, TFHset)</span><br><span class="line"></span><br><span class="line"><span class="comment">#生存曲线</span></span><br><span class="line">ES &lt;- data.frame(t(ES), check.names=<span class="built_in">F</span>)</span><br><span class="line">TCGAclin &lt;- merge(TCGAclin, ES, by.x=<span class="string">&quot;sample&quot;</span>, by.y=<span class="number">0</span>)</span><br><span class="line">TCGAclin$TFH.set = ifelse(TCGAclin$TFHset&gt;median(TCGAclin$TFHset),<span class="string">&#x27;high&#x27;</span>,<span class="string">&#x27;low&#x27;</span>)</span><br><span class="line">png(<span class="string">&#x27;Survival/survival.png&#x27;</span>, <span class="number">800</span>, <span class="number">600</span>)</span><br><span class="line">ggsurvplot(survfit(Surv(OS.time, OS)~TFH.set, data=TCGAclin), conf.int=<span class="built_in">F</span>, pval=<span class="literal">TRUE</span>)</span><br><span class="line">dev.off()</span><br><span class="line">write.csv(TCGAclin, <span class="string">&#x27;Survival/TCGAclin.csv&#x27;</span>, row.names = <span class="built_in">F</span>)</span><br><span class="line"><span class="comment">#Cox回归分析</span></span><br><span class="line"><span class="comment">#上述Kaplan-Meier方法只能针对分类变量做单因素生存分析，因此我将GSVA分析的结果转换成了只有high和low的二分类变量。如果不想转换可以用Cox比例风险回归模型分析：</span></span><br><span class="line">res.cox &lt;- coxph(Surv(OS.time, OS)~TFHset, data=TCGAclin)</span><br><span class="line">res.cox</span><br><span class="line"><span class="comment">#Cox结果解读：</span></span><br><span class="line"><span class="comment">#z值：Wald统计值，它对应于每个回归系数与其标准误差的比率（z = coef / se（coef））。</span></span><br><span class="line"><span class="comment">#coef：回归系数，正数表示危险（死亡风险）较高，负数代表危险较低。</span></span><br><span class="line"><span class="comment">#exp(coef)：在此代表危险比(HR，hazard ratio)。</span></span><br></pre></td></tr></table></figure></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="mailto:undefined">yilingzhang</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="https://hiiiyilingzhang.github.io/2021/05/24/scRNAseq-05-Monocle3-GO-KEGG-TCGASurv/">https://hiiiyilingzhang.github.io/2021/05/24/scRNAseq-05-Monocle3-GO-KEGG-TCGASurv/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/scRNA-seq/">scRNA-seq</a></div><div class="post_share"><div class="social-share" data-image="/img/cover/scRNA.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/social-share.js/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/social-share.js/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2021/05/24/Tumor-immune-infiltration-analysis/"><img class="prev-cover" src="/img/cover/Immue.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">Tumor immune infiltration analysis</div></div></a></div><div class="next-post pull-right"><a href="/2021/05/18/TCGA-data-mining/"><img class="next-cover" src="/img/cover/TCGA.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">TCGA data mining</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span> Related Articles</span></div><div class="relatedPosts-list"><div><a href="/2021/04/26/scRNAseq-01-Read-Data-and-Aggregation/" title="scRNAseq-01 Read Data and Aggregation"><img class="cover" src="/img/cover/scRNA.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2021-04-26</div><div class="title">scRNAseq-01 Read Data and Aggregation</div></div></a></div><div><a href="/2021/04/26/scRNAseq-02-Annotation/" title="scRNAseq-02 Annotation"><img class="cover" src="/img/cover/scRNA.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2021-04-26</div><div class="title">scRNAseq-02 Annotation</div></div></a></div><div><a href="/2021/05/10/scRNAseq-03-Doublet-velocyto-inferCNV/" title="scRNAseq-03 Doublet & velocyto & inferCNV"><img class="cover" src="/img/cover/scRNA.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2021-05-10</div><div class="title">scRNAseq-03 Doublet & velocyto & inferCNV</div></div></a></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="card-info-avatar is-center"><img class="avatar-img" src="/img/selfie.JPG" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/><div class="author-info__name">yilingzhang</div><div class="author-info__description">Every Day is My Day</div></div><div class="card-info-data"><div class="card-info-data-item is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">39</div></a></div><div class="card-info-data-item is-center"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">17</div></a></div><div class="card-info-data-item is-center"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">14</div></a></div></div><div class="card-info-social-icons is-center"><a class="social-icon" href="mailto:jocelynzhang1119@icloud.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a><a class="social-icon" href="https://blog.csdn.net/weixin_43893431" target="_blank" title="CSDN"><i class="fas fa-paper-plane"></i></a></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#Monocle3"><span class="toc-number">1.</span> <span class="toc-text">Monocle3</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%AD%A5%E9%AA%A4"><span class="toc-number">1.1.</span> <span class="toc-text">步骤</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E5%AF%BC%E5%85%A5"><span class="toc-number">1.1.1.</span> <span class="toc-text">数据导入</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86"><span class="toc-number">1.1.2.</span> <span class="toc-text">数据预处理</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%99%8D%E7%BB%B4%E5%8F%8A%E5%8F%AF%E8%A7%86%E5%8C%96"><span class="toc-number">1.1.3.</span> <span class="toc-text">降维及可视化</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%81%9A%E7%B1%BB%E5%88%86%E6%9E%90"><span class="toc-number">1.1.4.</span> <span class="toc-text">聚类分析</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%8B%9F%E6%97%B6%E5%88%86%E6%9E%90"><span class="toc-number">1.1.5.</span> <span class="toc-text">拟时分析</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%A1%AE%E5%AE%9A%E8%BD%A8%E8%BF%B9%E7%9A%84%E5%85%88%E5%90%8E%E9%A1%BA%E5%BA%8F"><span class="toc-number">1.1.6.</span> <span class="toc-text">确定轨迹的先后顺序</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%BB%83%E4%B9%A0"><span class="toc-number">1.2.</span> <span class="toc-text">练习</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#Installation"><span class="toc-number">1.2.1.</span> <span class="toc-text">Installation</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Run-Monocle3"><span class="toc-number">1.2.2.</span> <span class="toc-text">Run Monocle3</span></a></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Monocle2"><span class="toc-number">2.</span> <span class="toc-text">Monocle2</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%8B%9F%E6%97%B6%E7%9B%B8%E5%85%B3%E5%9F%BA%E5%9B%A0%E8%81%9A%E7%B1%BB%E7%83%AD%E5%9B%BE"><span class="toc-number">2.1.</span> <span class="toc-text">拟时相关基因聚类热图</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#BEAM%E5%88%86%E6%9E%90"><span class="toc-number">2.2.</span> <span class="toc-text">BEAM分析</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%BB%83%E4%B9%A0-1"><span class="toc-number">2.3.</span> <span class="toc-text">练习</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#GO-KEGG%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90"><span class="toc-number">3.</span> <span class="toc-text">GO&#x2F;KEGG富集分析</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#GSEA-GSVA%E5%88%86%E6%9E%90"><span class="toc-number">4.</span> <span class="toc-text">GSEA&#x2F;GSVA分析</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%9F%BA%E5%9B%A0%E9%9B%86%E7%9A%84%E6%A6%82%E5%BF%B5"><span class="toc-number">4.1.</span> <span class="toc-text">基因集的概念</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#MSigDB%E5%9F%BA%E5%9B%A0%E9%9B%86%E6%95%B0%E6%8D%AE%E5%BA%93"><span class="toc-number">4.2.</span> <span class="toc-text">MSigDB基因集数据库</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#GSVA"><span class="toc-number">4.3.</span> <span class="toc-text">GSVA</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#GSEA%E7%BB%83%E4%B9%A0"><span class="toc-number">4.3.1.</span> <span class="toc-text">GSEA练习</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#GSVA%E7%BB%83%E4%B9%A0"><span class="toc-number">4.3.2.</span> <span class="toc-text">GSVA练习</span></a></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%81%94%E5%90%88TCGA%E7%94%9F%E5%AD%98%E5%88%86%E6%9E%90"><span class="toc-number">5.</span> <span class="toc-text">联合TCGA生存分析</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2021/09/06/All-I-Learn-About-Circulating-Cell-free-DNA-cfDNA/" title="All I Learn About Circulating Cell-free DNA(cfDNA)"><img src="/img/cover/cfDNA.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="All I Learn About Circulating Cell-free DNA(cfDNA)"/></a><div class="content"><a class="title" href="/2021/09/06/All-I-Learn-About-Circulating-Cell-free-DNA-cfDNA/" title="All I Learn About Circulating Cell-free DNA(cfDNA)">All I Learn About Circulating Cell-free DNA(cfDNA)</a><time datetime="2021-09-06T08:31:28.000Z" title="Created 2021-09-06 16:31:28">2021-09-06</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/08/31/CITE-seq-theory-and-analysis/" title="CITE-seq: theory and analysis"><img src="/img/cover/CITE.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="CITE-seq: theory and analysis"/></a><div class="content"><a class="title" href="/2021/08/31/CITE-seq-theory-and-analysis/" title="CITE-seq: theory and analysis">CITE-seq: theory and analysis</a><time datetime="2021-08-31T13:09:19.000Z" title="Created 2021-08-31 21:09:19">2021-08-31</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/07/23/scRNA-seq-Project-LIHC-microenvironment/" title="scRNA-seq Project -- LIHC microenvironment"><img src="/img/cover/scRNA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scRNA-seq Project -- LIHC microenvironment"/></a><div class="content"><a class="title" href="/2021/07/23/scRNA-seq-Project-LIHC-microenvironment/" title="scRNA-seq Project -- LIHC microenvironment">scRNA-seq Project -- LIHC microenvironment</a><time datetime="2021-07-23T14:09:24.000Z" title="Created 2021-07-23 22:09:24">2021-07-23</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/07/18/Get-Start-With-Bioinfo-Pipeline/" title="Get Start With Bioinfo Pipeline"><img src="/img/cover/pipeline.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Get Start With Bioinfo Pipeline"/></a><div class="content"><a class="title" href="/2021/07/18/Get-Start-With-Bioinfo-Pipeline/" title="Get Start With Bioinfo Pipeline">Get Start With Bioinfo Pipeline</a><time datetime="2021-07-18T11:59:13.000Z" title="Created 2021-07-18 19:59:13">2021-07-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/24/Tumor-immune-infiltration-analysis/" title="Tumor immune infiltration analysis"><img src="/img/cover/Immue.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Tumor immune infiltration analysis"/></a><div class="content"><a class="title" href="/2021/05/24/Tumor-immune-infiltration-analysis/" title="Tumor immune infiltration analysis">Tumor immune infiltration analysis</a><time datetime="2021-05-24T10:51:35.000Z" title="Created 2021-05-24 18:51:35">2021-05-24</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2021 By yilingzhang</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"></div><script defer="defer" id="ribbon" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/canvas-ribbon.min.js" size="150" alpha="0.6" zIndex="-1" mobile="false" data-click="true"></script><script id="click-heart" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/click-heart.min.js" async="async" mobile="false"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>