<!DOCTYPE html><html lang="en | zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1"><title>TCGA data mining | Jocelyn's Blog</title><meta name="keywords" content="TCGA"><meta name="author" content="yilingzhang"><meta name="copyright" content="yilingzhang"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="Basic info前言34种癌型，数据类型：  Genome Transcriptome Proteome Epigenome Clinic  数据产生过程 分析pipeline以及由GDACs产生的数据结果通过DCC保存到研究社区（community）中。  Barcode   TCGA data processingData download 网页下载（最靠谱） 命令行下载（可能会断，一定要">
<meta property="og:type" content="article">
<meta property="og:title" content="TCGA data mining">
<meta property="og:url" content="http://yilingzhang.gitee.io/2021/05/18/TCGA-data-mining/index.html">
<meta property="og:site_name" content="Jocelyn&#39;s Blog">
<meta property="og:description" content="Basic info前言34种癌型，数据类型：  Genome Transcriptome Proteome Epigenome Clinic  数据产生过程 分析pipeline以及由GDACs产生的数据结果通过DCC保存到研究社区（community）中。  Barcode   TCGA data processingData download 网页下载（最靠谱） 命令行下载（可能会断，一定要">
<meta property="og:locale">
<meta property="og:image" content="http://yilingzhang.gitee.io/img/cover/TCGA.png">
<meta property="article:published_time" content="2021-05-18T12:01:08.000Z">
<meta property="article:modified_time" content="2021-05-18T12:17:02.227Z">
<meta property="article:author" content="yilingzhang">
<meta property="article:tag" content="TCGA">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://yilingzhang.gitee.io/img/cover/TCGA.png"><link rel="shortcut icon" href="/img/love_32px.png"><link rel="canonical" href="http://yilingzhang.gitee.io/2021/05/18/TCGA-data-mining/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    jQuery: 'https://cdn.jsdelivr.net/npm/jquery@latest/dist/jquery.min.js',
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/js/jquery.justifiedGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/justifiedGallery/dist/css/justifiedGallery.min.css'
    },
    fancybox: {
      js: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.js',
      css: 'https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@latest/dist/jquery.fancybox.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isanchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = { 
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2021-05-18 20:17:02'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    })(window)</script><meta name="generator" content="Hexo 5.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="author-avatar"><img class="avatar-img" src="/img/selfie.JPG" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data"><div class="data-item is-center"><div class="data-item-link"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">38</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">16</div></a></div></div><div class="data-item is-center"><div class="data-item-link"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">14</div></a></div></div></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/img/cover/TCGA.png')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Jocelyn's Blog</a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down expand"></i></a><ul class="menus_item_child"><li><a class="site-page" target="_blank" rel="noopener" href="https://music.apple.com/cn/playlist/fav/pl.u-yZyVEjrFYkEpKjk"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page" target="_blank" rel="noopener" href="https://ohmyphotography.lofter.com/"><i class="fa-fw fas fa-video"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">TCGA data mining</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2021-05-18T12:01:08.000Z" title="Created 2021-05-18 20:01:08">2021-05-18</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2021-05-18T12:17:02.227Z" title="Updated 2021-05-18 20:17:02">2021-05-18</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/Database/">Database</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">Word count:</span><span class="word-count">1.8k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">Reading time:</span><span>8min</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1 id="Basic-info"><a href="#Basic-info" class="headerlink" title="Basic info"></a>Basic info</h1><h2 id="前言"><a href="#前言" class="headerlink" title="前言"></a>前言</h2><p>34种癌型，数据类型：</p>
<ul>
<li>Genome</li>
<li>Transcriptome</li>
<li>Proteome</li>
<li>Epigenome</li>
<li>Clinic</li>
</ul>
<h2 id="数据产生过程"><a href="#数据产生过程" class="headerlink" title="数据产生过程"></a>数据产生过程</h2><center><img src="/img/Figures/TCGA/dataproduce.png" alt="dataproduce" style="zoom:70%;" /></center>
分析pipeline以及由GDACs产生的数据结果通过DCC保存到研究社区（community）中。

<h2 id="Barcode"><a href="#Barcode" class="headerlink" title="Barcode"></a>Barcode</h2><center><img src="/img/Figures/TCGA/barcode.png" alt="barcode" style="zoom:60%;" /></center>
<center><img src="/img/Figures/TCGA/barcode.png" alt="barcode" style="zoom:60%;" /></center>

<h1 id="TCGA-data-processing"><a href="#TCGA-data-processing" class="headerlink" title="TCGA data processing"></a>TCGA data processing</h1><h2 id="Data-download"><a href="#Data-download" class="headerlink" title="Data download"></a>Data download</h2><ul>
<li>网页下载（最靠谱）</li>
<li>命令行下载（可能会断，一定要校验）</li>
<li>R包下载（有问题，经过清洗）</li>
</ul>
<p>下载文件：</p>
<ul>
<li>Metadata（json文件，meta文件中含有样品号的TCGA barcode，而需要根据文件名转换成样本号）</li>
<li>Manifest, Cart（命令行下载需要这两个文件）</li>
</ul>
<p>下载数据类型选择：</p>
<ol>
<li>htseq.counts是RNA-Seq原始reads文件，利用该数据通过不同的算法得到表达值（如果要DeSeq处理，下载这个）</li>
<li>FPKM.txt.gz是存储每个样品所有测序基因的FPKM值（一般下这个）</li>
<li>FPKM-UQ.txt.gz是存储每个样品所有测序基因的FPKM-UQ（upper quantile）值，即调整后FPKM值（用reads值3/4分位值替代原reads值）。</li>
</ol>
<h2 id="表达谱数据整合"><a href="#表达谱数据整合" class="headerlink" title="表达谱数据整合"></a>表达谱数据整合</h2><p>存在少量的多对一情况—&gt;变成矩阵<br>数据整理的结果:生成mRNAmatrix.txt文件<br><img src="C:\Users\lenovo\Desktop\matrix.png" alt="matrix"></p>
<ul>
<li>行名：Ensembl ID</li>
<li>列名为：Barcode，样本号；</li>
<li>第四部分：1-9为肿瘤，10以上为正常，判断正常及肿瘤的代码，常用的R函数 <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">group_list &lt;- ifelse(<span class="built_in">as.numeric</span>(str_sub(colnames(expr),<span class="number">14</span>,<span class="number">15</span>))&lt;<span class="number">10</span>,<span class="string">&quot;tumor&quot;</span>,<span class="string">&quot;normal&quot;</span>) </span><br></pre></td></tr></table></figure>
处理时需要注意，需要切割，取前面的整数</li>
</ul>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#解析meta文件，目的是为了拿到 样本名-barcode对应关系,</span></span><br><span class="line"><span class="comment">#注意json文件的位置</span></span><br><span class="line">metadata &lt;- jsonlite::fromJSON(<span class="string">&quot;data/1.rawdata/metadata.cart.2021-05-16.json&quot;</span>)</span><br><span class="line">library(dplyr)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 转换的信息是两列filename和associate dentities</span></span><br><span class="line"><span class="comment"># 样本名称在associated_entities,列表中里面包括了 entity_id, case_id, entity_submitter_id, entity_type这四个项目</span></span><br><span class="line">metadata_id &lt;- metadata %&gt;% dplyr::select(<span class="built_in">c</span>(file_name,associated_entities)) </span><br><span class="line"><span class="comment"># metadata$associatedentities[1][[1]]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 把filename和 associated_entities中的entity_submitter_id提取出来，做成数据框批量对应转换</span></span><br><span class="line"><span class="comment"># substr(s, first, last)</span></span><br><span class="line"><span class="comment"># nchar(metadata_id$file_name[i])-3 去掉.gz</span></span><br><span class="line">naid_df &lt;- data.frame()</span><br><span class="line"><span class="keyword">for</span> (i <span class="keyword">in</span> <span class="number">1</span>:<span class="built_in">dim</span>(metadata_id)[<span class="number">1</span>])&#123;</span><br><span class="line">  naid_df[i,<span class="number">1</span>] &lt;- substr(metadata_id$file_name[i],<span class="number">1</span>,nchar(metadata_id$file_name[i])-<span class="number">3</span>)</span><br><span class="line">  naid_df[i,<span class="number">2</span>] &lt;- metadata_id$associated_entities[i][[<span class="number">1</span>]]$entity_submitter_id</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment"># 提取前12个</span></span><br><span class="line">library(stringi)</span><br><span class="line">library(stringr)</span><br><span class="line"><span class="keyword">for</span> (i <span class="keyword">in</span> <span class="number">1</span>:<span class="built_in">dim</span>(naid_df)[<span class="number">1</span>])&#123;</span><br><span class="line">  naid_df$barcode[i]&lt;- str_sub(naid_df[i,<span class="number">2</span>],start=<span class="number">1</span>,end=<span class="number">12</span>)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment"># 提取出文件夹中的txt.gz</span></span><br><span class="line"><span class="comment"># 这里面有个技巧，先读文件夹中的内容，再新建文件夹“files”,要不然会报错</span></span><br><span class="line">dir1 &lt;- dir(<span class="string">&quot;data/1.rawdata/files&quot;</span>)</span><br><span class="line">dir.create(<span class="string">&quot;data/1.rawdata/pureFile&quot;</span>) </span><br><span class="line"><span class="keyword">for</span> (dirname <span class="keyword">in</span> dir1)&#123;  </span><br><span class="line">  file &lt;- list.files(paste0(<span class="string">&quot;data/1.rawdata/files/&quot;</span>,dirname),pattern = <span class="string">&quot;*.FPKM.txt.gz&quot;</span>)  <span class="comment">#找到对应文件夹中的内容，pattern可以是正则表达式</span></span><br><span class="line">  file.copy(paste0(<span class="string">&quot;data/1.rawdata/files/&quot;</span>,dirname,<span class="string">&quot;/&quot;</span>,file),<span class="string">&quot;data/1.rawdata/pureFile/&quot;</span>)  <span class="comment">#复制内容到新的文件夹</span></span><br><span class="line">&#125;</span><br><span class="line">rm(dir1,dirname)</span><br><span class="line"></span><br><span class="line"><span class="comment">## gunzip</span></span><br><span class="line">files1&lt;-dir(<span class="string">&quot;data/1.rawdata/pureFile/&quot;</span>)</span><br><span class="line">library(R.utils)</span><br><span class="line"><span class="keyword">for</span> (i <span class="keyword">in</span> <span class="number">1</span>:<span class="built_in">length</span>(files1))&#123;</span><br><span class="line">  gunzip(paste0(<span class="string">&quot;data/1.rawdata/pureFile/&quot;</span>,files1[i]),remove=<span class="literal">TRUE</span>)  <span class="comment">## 解压时原来的文件删掉</span></span><br><span class="line">&#125;</span><br><span class="line">rm(files1,file,i)</span><br><span class="line"></span><br><span class="line"><span class="comment">#合并</span></span><br><span class="line">library(data.table)</span><br><span class="line"></span><br><span class="line">nameList &lt;- list.files(<span class="string">&quot;data/1.rawdata/pureFile/&quot;</span>,pattern = <span class="string">&quot;*.FPKM.txt&quot;</span>) <span class="comment">## 注意pattern 参数，获取该目录下所有的结尾为FPKM.txt</span></span><br><span class="line">location &lt;- which(naid_df==nameList[<span class="number">1</span>],arr.ind = <span class="literal">TRUE</span>)  <span class="comment">## which函数有一个已知value返回坐标的功能</span></span><br><span class="line">TCGA_id &lt;- <span class="built_in">as.character</span>(naid_df[location[<span class="number">1</span>],<span class="number">2</span>]) <span class="comment">## 通过坐标，获取TCGA_id</span></span><br><span class="line"></span><br><span class="line">expr_df&lt;- fread(paste0(<span class="string">&quot;data/1.rawdata/pureFile/&quot;</span>,nameList[<span class="number">1</span>])) <span class="comment">## 读入第一个文件，保存为data.frame</span></span><br><span class="line"><span class="built_in">names</span>(expr_df) &lt;- <span class="built_in">c</span>(<span class="string">&quot;gene_id&quot;</span>,TCGA_id)  <span class="comment">## 给刚才数据库命名</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> (i <span class="keyword">in</span> <span class="number">2</span>:<span class="built_in">length</span>(nameList))&#123;</span><br><span class="line">  location &lt;- which(naid_df==nameList[i],arr.ind = <span class="literal">TRUE</span>)</span><br><span class="line">  TCGA_id &lt;- <span class="built_in">as.character</span>(naid_df[location[<span class="number">1</span>],<span class="number">2</span>])</span><br><span class="line">  dfnew &lt;- fread(paste0(<span class="string">&quot;data/1.rawdata/pureFile/&quot;</span>,nameList[i]))</span><br><span class="line">  <span class="built_in">names</span>(dfnew) &lt;- <span class="built_in">c</span>(<span class="string">&quot;gene_id&quot;</span>,TCGA_id)</span><br><span class="line">  expr_df &lt;- inner_join(expr_df,dfnew,by=<span class="string">&quot;gene_id&quot;</span>)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">mRNAmatrix &lt;-expr_df</span><br><span class="line">save(mRNAmatrix,file = <span class="string">&quot;data/2.ExprData/mRNAmatrix.Rda&quot;</span>)</span><br><span class="line">write.table(mRNAmatrix,file = <span class="string">&quot;data/2.ExprData/mRNAmatrix.txt&quot;</span>,<span class="built_in">quote</span> = <span class="built_in">F</span>,row.names = <span class="built_in">F</span>,sep = <span class="string">&quot;\t&quot;</span>)</span><br></pre></td></tr></table></figure>
<h2 id="id转换"><a href="#id转换" class="headerlink" title="id转换"></a>id转换</h2><p>输入文件：mRNAmatrix.txt（前面产生）<br>                  GTF文件（下载）</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">library(dplyr)</span><br><span class="line">library(tidyr)</span><br><span class="line">library(rtracklayer)</span><br><span class="line"></span><br><span class="line"><span class="comment"># GTF的下载从官方网站，ftp://ftp.ensembl.org/pub/current_gtf/homo_sapiens，</span></span><br><span class="line"><span class="comment"># 现在已经是GRCh38（100）版本了，此版本中的ID号肯定和矩阵文件中的不一样，应为当初人家比对不是100版本</span></span><br><span class="line">gtf1 &lt;- rtracklayer::import(<span class="string">&#x27;data/Reference/Homo_sapiens.GRCh38.100.gtf&#x27;</span>)</span><br><span class="line">gtf_df &lt;- as.data.frame(gtf1)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 选择两列，有2908224行</span></span><br><span class="line">gtf_df1&lt;-select(gtf_df,gene_id,gene_name)</span><br><span class="line"><span class="comment"># 去除重复行，有60683行，但里面还有多个ID对应一个基因的行</span></span><br><span class="line">gtf_df&lt;-distinct(gtf_df1,gene_id,gene_name)</span><br><span class="line"></span><br><span class="line">save(gtf_df,file=<span class="string">&quot;data/Reference/GTF_selected.Rdata&quot;</span>)</span><br><span class="line">rm(gtf_df1,gtf_df2,gtf_df,gtf1)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 进行注释替换</span></span><br><span class="line">mRNAmatrix&lt;-read.table(<span class="string">&quot;data/2.ExprData/mRNAmatrix.txt&quot;</span>,sep = <span class="string">&quot;\t&quot;</span>,stringsAsFactors = <span class="built_in">F</span>,header = <span class="built_in">T</span>)</span><br><span class="line">load(<span class="string">&quot;../../CancerGenomics/TCGA/tcga-classproj/data/Reference/GTF_selected.Rdata&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 表达矩阵中的ensemble号中含有小数点， 而GTF文件中没有,调整一下，先以“.”分列，在去掉小数点后的列</span></span><br><span class="line"><span class="comment"># 此行看下名字，不一定需要colnames(mRNAmatrix)&lt;-”gene_id” </span></span><br><span class="line">mRNAmatrix_nopoint &lt;- mRNAmatrix %&gt;% tidyr::separate(gene_id,into = <span class="built_in">c</span>(<span class="string">&quot;gene_id&quot;</span>,<span class="string">&quot;drop&quot;</span>),sep=<span class="string">&quot;\\.&quot;</span>) %&gt;% dplyr::select(-drop)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 采用左连接，左边的全部保留</span></span><br><span class="line">mRNA_symbol&lt;-left_join(x=mRNAmatrix_nopoint,y=gtf_df,by=<span class="string">&quot;gene_id&quot;</span>)</span><br><span class="line"><span class="comment"># 删除没能注释的行</span></span><br><span class="line">mRNA_symbol&lt;-na.omit(mRNA_symbol)</span><br><span class="line"><span class="comment"># 构建新的表达矩阵，第一列为gene symbol</span></span><br><span class="line">mRNA_symbol2&lt;-cbind(gene_name=mRNA_symbol$gene_name,mRNA_symbol[,-<span class="built_in">c</span>(<span class="number">1</span>,ncol(mRNA_symbol))])</span><br><span class="line"></span><br><span class="line">save(mRNA_symbol2,file=<span class="string">&quot;data/3.Symbol/mRNA_symbol.Rdata&quot;</span>)</span><br><span class="line">write.table(mRNA_symbol2,file = <span class="string">&quot;data/3.Symbol/symbol.txt&quot;</span>,<span class="built_in">quote</span> = <span class="built_in">F</span>,row.names = <span class="built_in">F</span>,sep = <span class="string">&quot;\t&quot;</span>)</span><br></pre></td></tr></table></figure>
<h2 id="TME-scoring"><a href="#TME-scoring" class="headerlink" title="TME scoring"></a>TME scoring</h2><p>肿瘤微环境是肿瘤细胞所处的细胞环境，其组成包括细胞外基质、可溶性分子和肿瘤基质细胞。肿瘤微环境一旦形成，众多免疫细胞，如T细胞、髓源抑制性细胞，巨噬细胞等，都被趋化至此，构成肿瘤的微环境。在肿瘤微环境中，免疫细胞和基质细胞是两种主要类型的非肿瘤组分，并且已被提出对于肿瘤的诊断和预后评估是有价值的。</p>
<p><strong>ESTIMATE</strong> (Estimation of STromal and Immune cells in MAlignant Tumor tissues using Expression data) is a tool for predicting tumor purity, and the presence of infiltrating stromal/immune cells in tumor tissues using gene expression data. ESTIMATE algorithm is based on single sample Gene Set Enrichment Analysis and generates three scores:</p>
<ul>
<li>stromal score (that captures the presence of stroma in tumor tissue).</li>
<li>immune score (that represents the infiltration of immune cells in tumor tissue).</li>
<li>estimate score (that infers tumor purity).</li>
</ul>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#library(utils)</span></span><br><span class="line"><span class="comment">#rforge &lt;- &quot;http://r-forge.r-project.org&quot;</span></span><br><span class="line"><span class="comment"># install.packages(&quot;estimate&quot;, repos=&quot;http://R-Forge.R-project.org&quot;)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))</span></span><br><span class="line"><span class="comment">#    install.packages(&quot;BiocManager&quot;)</span></span><br><span class="line"><span class="comment">#BiocManager::install(&quot;limma&quot;, version = &quot;3.8&quot;)</span></span><br><span class="line"></span><br><span class="line">library(limma)</span><br><span class="line">library(estimate)</span><br><span class="line"></span><br><span class="line"><span class="comment">#读取文件</span></span><br><span class="line">rt &lt;- read.table(<span class="string">&quot;data/3.Symbol/symbol.txt&quot;</span>,sep=<span class="string">&quot;\t&quot;</span>,header=<span class="built_in">T</span>,check.names=<span class="built_in">F</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#如果一个基因占了多行，取均值</span></span><br><span class="line"><span class="comment">## 去除.1这种</span></span><br><span class="line">rt$gene_name_fixed &lt;- <span class="literal">NA</span></span><br><span class="line"><span class="keyword">for</span> (i <span class="keyword">in</span> <span class="number">1</span>:nrow(rt)) &#123;</span><br><span class="line">  rt[i,<span class="string">&quot;gene_name_fixed&quot;</span>] &lt;- strsplit(<span class="built_in">as.character</span>(rt[i,<span class="string">&quot;gene_name&quot;</span>]),<span class="string">&quot;.&quot;</span>,fixed = <span class="built_in">T</span>)[[<span class="number">1</span>]][<span class="number">1</span>]</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">rt_fixed &lt;- aggregate(rt,by = <span class="built_in">list</span>(gene_name_fixed=rt$gene_name_fixed), FUN=mean, drop = <span class="literal">FALSE</span>)</span><br><span class="line">rownames(rt_fixed) &lt;- rt_fixed[,<span class="number">1</span>]</span><br><span class="line">rt_fixed &lt;- rt_fixed[,-<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">2</span>,<span class="number">410</span>)]</span><br><span class="line">rt_fixed &lt;- rt_fixed[rowMeans(rt_fixed)&gt;<span class="number">0</span>,]</span><br><span class="line"></span><br><span class="line"><span class="comment">#输出整理后的矩阵文件</span></span><br><span class="line">write.table(rt_fixed,file=<span class="string">&quot;data/3.Symbol/uniq.symbol.txt&quot;</span>,sep=<span class="string">&quot;\t&quot;</span>,<span class="built_in">quote</span>=<span class="built_in">F</span>,col.names=<span class="built_in">F</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#运行estimate包</span></span><br><span class="line">filterCommonGenes(input.f=<span class="string">&quot;data/3.Symbol/uniq.symbol.txt&quot;</span>, </span><br><span class="line">                  output.f=<span class="string">&quot;data/4.ESTIMATE/commonGenes.gct&quot;</span>, </span><br><span class="line">                  id=<span class="string">&quot;GeneSymbol&quot;</span>)</span><br><span class="line"></span><br><span class="line">estimateScore(input.ds = <span class="string">&quot;data/4.ESTIMATE/commonGenes.gct&quot;</span>,</span><br><span class="line">              output.ds=<span class="string">&quot;data/4.ESTIMATE/estimateScore.gct&quot;</span>, </span><br><span class="line">              platform=<span class="string">&quot;illumina&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#输出每个样品的打分</span></span><br><span class="line">scores=read.table(<span class="string">&quot;data/4.ESTIMATE/estimateScore.gct&quot;</span>,skip = <span class="number">2</span>,header = <span class="built_in">T</span>)</span><br><span class="line">rownames(scores)=scores[,<span class="number">1</span>]</span><br><span class="line">scores=t(scores[,<span class="number">3</span>:ncol(scores)])</span><br><span class="line">rownames(scores)=gsub(<span class="string">&quot;\\.&quot;</span>,<span class="string">&quot;\\-&quot;</span>,rownames(scores))</span><br><span class="line">out=rbind(ID=colnames(scores),scores)</span><br><span class="line">write.table(out,file=<span class="string">&quot;data/4.ESTIMATE/scores.txt&quot;</span>,sep=<span class="string">&quot;\t&quot;</span>,<span class="built_in">quote</span>=<span class="built_in">F</span>,col.names=<span class="built_in">F</span>)</span><br></pre></td></tr></table></figure></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="mailto:undefined">yilingzhang</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="http://yilingzhang.gitee.io/2021/05/18/TCGA-data-mining/">http://yilingzhang.gitee.io/2021/05/18/TCGA-data-mining/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/TCGA/">TCGA</a></div><div class="post_share"><div class="social-share" data-image="/img/cover/TCGA.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/social-share.js/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/social-share.js/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2021/05/24/scRNAseq-05-Monocle3-GO-KEGG-TCGASurv/"><img class="prev-cover" src="/img/cover/scRNA.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv</div></div></a></div><div class="next-post pull-right"><a href="/2021/05/18/scRNAseq-04-Cell-cell-communication/"><img class="next-cover" src="/img/cover/scRNA.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">scRNAseq-04 Cell-cell communication</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="card-info-avatar is-center"><img class="avatar-img" src="/img/selfie.JPG" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/><div class="author-info__name">yilingzhang</div><div class="author-info__description">Every Day is My Day</div></div><div class="card-info-data"><div class="card-info-data-item is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">38</div></a></div><div class="card-info-data-item is-center"><a href="/tags/"><div class="headline">Tags</div><div class="length-num">16</div></a></div><div class="card-info-data-item is-center"><a href="/categories/"><div class="headline">Categories</div><div class="length-num">14</div></a></div></div><div class="card-info-social-icons is-center"><a class="social-icon" href="mailto:jocelynzhang1119@icloud.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a><a class="social-icon" href="https://blog.csdn.net/weixin_43893431" target="_blank" title="CSDN"><i class="fas fa-paper-plane"></i></a></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#Basic-info"><span class="toc-number">1.</span> <span class="toc-text">Basic info</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%89%8D%E8%A8%80"><span class="toc-number">1.1.</span> <span class="toc-text">前言</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E4%BA%A7%E7%94%9F%E8%BF%87%E7%A8%8B"><span class="toc-number">1.2.</span> <span class="toc-text">数据产生过程</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Barcode"><span class="toc-number">1.3.</span> <span class="toc-text">Barcode</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#TCGA-data-processing"><span class="toc-number">2.</span> <span class="toc-text">TCGA data processing</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#Data-download"><span class="toc-number">2.1.</span> <span class="toc-text">Data download</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%A1%A8%E8%BE%BE%E8%B0%B1%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88"><span class="toc-number">2.2.</span> <span class="toc-text">表达谱数据整合</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#id%E8%BD%AC%E6%8D%A2"><span class="toc-number">2.3.</span> <span class="toc-text">id转换</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#TME-scoring"><span class="toc-number">2.4.</span> <span class="toc-text">TME scoring</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2021/07/18/Get-Start-With-Bioinfo-Pipeline/" title="Get Start With Bioinfo Pipeline"><img src="/img/banner/paofu.JPG" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Get Start With Bioinfo Pipeline"/></a><div class="content"><a class="title" href="/2021/07/18/Get-Start-With-Bioinfo-Pipeline/" title="Get Start With Bioinfo Pipeline">Get Start With Bioinfo Pipeline</a><time datetime="2021-07-18T11:59:13.000Z" title="Created 2021-07-18 19:59:13">2021-07-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/24/Tumor-immune-infiltration-analysis/" title="Tumor immune infiltration analysis"><img src="/img/cover/Immue.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Tumor immune infiltration analysis"/></a><div class="content"><a class="title" href="/2021/05/24/Tumor-immune-infiltration-analysis/" title="Tumor immune infiltration analysis">Tumor immune infiltration analysis</a><time datetime="2021-05-24T10:51:35.000Z" title="Created 2021-05-24 18:51:35">2021-05-24</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/24/scRNAseq-05-Monocle3-GO-KEGG-TCGASurv/" title="scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv"><img src="/img/cover/scRNA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv"/></a><div class="content"><a class="title" href="/2021/05/24/scRNAseq-05-Monocle3-GO-KEGG-TCGASurv/" title="scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv">scRNAseq-05 Monocle3 &amp; GO/KEGG &amp; TCGASurv</a><time datetime="2021-05-24T04:51:09.000Z" title="Created 2021-05-24 12:51:09">2021-05-24</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/18/TCGA-data-mining/" title="TCGA data mining"><img src="/img/cover/TCGA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="TCGA data mining"/></a><div class="content"><a class="title" href="/2021/05/18/TCGA-data-mining/" title="TCGA data mining">TCGA data mining</a><time datetime="2021-05-18T12:01:08.000Z" title="Created 2021-05-18 20:01:08">2021-05-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/05/18/scRNAseq-04-Cell-cell-communication/" title="scRNAseq-04 Cell-cell communication"><img src="/img/cover/scRNA.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="scRNAseq-04 Cell-cell communication"/></a><div class="content"><a class="title" href="/2021/05/18/scRNAseq-04-Cell-cell-communication/" title="scRNAseq-04 Cell-cell communication">scRNAseq-04 Cell-cell communication</a><time datetime="2021-05-18T11:59:21.000Z" title="Created 2021-05-18 19:59:21">2021-05-18</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2021 By yilingzhang</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"></div><script defer="defer" id="ribbon" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/canvas-ribbon.min.js" size="150" alpha="0.6" zIndex="-1" mobile="false" data-click="true"></script><script id="click-heart" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1/dist/click-heart.min.js" async="async" mobile="false"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>